{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify generic genes and pathways\n",
    "\n",
    "Studies have found that some genes are more likely to be differentially expressed even across a wide range of experimental designs. These generic genes and subsequent pathways are not necessarily specific to the biological process being studied but instead represent a more systematic change. \n",
    "\n",
    "We have developed an approach, outlined below, to automatically identify these generic genes and pathways. We have validated this simulation approach can identify generic genes and pathways in the analysis notebooks: [human_general_analysis](../human_general_analysis/) and [human_cancer_analysis](../human_cancer_analysis/). Here\n",
    "\n",
    "This notebook applies this approach to identify generic genes and pathways in the pseudomonas compendium. \n",
    "\n",
    "**Steps to identify generic genes:**\n",
    "1. Simulates N gene expression experiments using [ponyo](https://github.com/ajlee21/ponyo)\n",
    "2. Perform DE analysis to get association statistics for each gene\n",
    "\n",
    "In this case the DE analysis is based on the experimental design of the template experiment, described in the previous [notebook](1_process_pseudomonas_data.ipynb). \n",
    "The template experiment is [GEOD-33245](https://www.ebi.ac.uk/arrayexpress/experiments/E-GEOD-33245/?s_sortby=col_8&s_sortorder=ascending), which contains multiple different comparisons including WT vs *crc* mutants, WT vs *cbr* mutants in different conditions. So the DE analysis is comparing WT vs mutant.\n",
    "\n",
    "3. For each gene, aggregate statistics across all simulated experiments \n",
    "4. Rank genes based on this aggregated statistic\n",
    "\n",
    "**Steps to identify generic gene sets (pathways):**\n",
    "1. Using the same simulated experiments from above, perform GSEA analysis. This analysis will determine whether the genes contained in a gene set are clustered towards the beginning or the end of the ranked list of genes, where genes are ranked by log fold change, indicating a correlation with change in expression.\n",
    "2. For each gene set (pathway), aggregate statistics across all simulated experiments\n",
    "3. Rank gene sets based on this aggregated statistic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "\n",
    "from rpy2.robjects import pandas2ri\n",
    "pandas2ri.activate()\n",
    "\n",
    "from ponyo import utils, simulate_expression_data\n",
    "from generic_expression_patterns_modules import process, stats, ranking\n",
    "\n",
    "np.random.seed(123)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(),\"../\"))\n",
    "\n",
    "config_filename = os.path.abspath(os.path.join(base_dir,\n",
    "                                           \"configs\",\n",
    "                                           \"config_pseudomonas_33245.tsv\"))\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/sklearn/base.py:315: UserWarning: Trying to unpickle estimator MinMaxScaler from version 0.23.2 when using version 0.24.0. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  UserWarning)\n"
     ]
    }
   ],
   "source": [
    "# Load params\n",
    "local_dir = params[\"local_dir\"]\n",
    "dataset_name = params['dataset_name']\n",
    "NN_architecture = params['NN_architecture']\n",
    "num_runs = params['num_simulated']\n",
    "project_id = params['project_id']\n",
    "metadata_col_id = params['metadata_colname']\n",
    "raw_template_filename = params['raw_template_filename']\n",
    "processed_template_filename = params['processed_template_filename']\n",
    "normalized_compendium_filename = params['normalized_compendium_filename']\n",
    "scaler_filename = params['scaler_filename']\n",
    "col_to_rank_genes = params['rank_genes_by']\n",
    "col_to_rank_pathways = params['rank_pathways_by']\n",
    "statistic = params['gsea_statistic']\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "sample_id_metadata_filename = os.path.join(\n",
    "    base_dir,\n",
    "    dataset_name,\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    f\"{project_id}_process_samples.tsv\")\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "metadata_filename = os.path.join(\n",
    "    base_dir,\n",
    "    dataset_name,\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    f\"{project_id}_groups.tsv\"\n",
    ")\n",
    "\n",
    "# Load pickled file\n",
    "scaler = pickle.load(open(scaler_filename, \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output files\n",
    "gene_summary_filename = os.path.join(\n",
    "    base_dir, \n",
    "    dataset_name, \n",
    "    f\"generic_gene_summary_{project_id}_test.tsv\"\n",
    ")\n",
    "pathway_summary_filename = os.path.join(\n",
    "    base_dir, \n",
    "    dataset_name, \n",
    "    f\"generic_pathway_summary_{project_id}_test.tsv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulate experiments using selected template experiment\n",
    "Workflow:\n",
    "\n",
    "1. Get the gene expression data for the selected template experiment\n",
    "2. Encode this experiment into a latent space using the trained VAE model\n",
    "3. Linearly shift the encoded template experiment in the latent space\n",
    "4. Decode the samples. This results in a new experiment\n",
    "5. Repeat steps 1-4 to get multiple simulated experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate multiple experiments\n",
    "# This step creates the following files in \"<local_dir>/pseudo_experiment/\" directory:           \n",
    "#   - selected_simulated_data_SRP012656_<n>.txt\n",
    "#   - selected_simulated_encoded_data_SRP012656_<n>.txt\n",
    "#   - template_normalized_data_SRP012656_test.txt\n",
    "# in which \"<n>\" is an integer in the range of [0, num_runs-1] \n",
    "os.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)\n",
    "for run_id in range(num_runs):\n",
    "    simulate_expression_data.shift_template_experiment(\n",
    "        normalized_compendium_filename,\n",
    "        project_id,\n",
    "        metadata_col_id,\n",
    "        NN_architecture,\n",
    "        dataset_name,\n",
    "        scaler,\n",
    "        local_dir,\n",
    "        base_dir,\n",
    "        run_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process template and simulated data\n",
    "\n",
    "* Remove samples not required for comparison. \n",
    "* Make sure ordering of samples matches metadata for proper comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(sample_id_metadata_filename):\n",
    "    sample_id_metadata_filename = None\n",
    "\n",
    "stats.process_samples_for_limma(\n",
    "    raw_template_filename,\n",
    "    metadata_filename,\n",
    "    processed_template_filename,\n",
    "    sample_id_metadata_filename,\n",
    ")\n",
    "\n",
    "for i in range(num_runs):\n",
    "    simulated_filename = os.path.join(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment\",\n",
    "        f\"selected_simulated_data_{project_id}_{i}.txt\"\n",
    "    )\n",
    "    stats.process_samples_for_limma(\n",
    "        simulated_filename,\n",
    "        metadata_filename,\n",
    "        None,\n",
    "        sample_id_metadata_filename,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quick check\n",
    "template_data = pd.read_csv(\n",
    "    processed_template_filename, \n",
    "    header=0,\n",
    "    index_col=0,\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "\n",
    "assert(template_data.shape[0] == 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: S4Vectors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: stats4\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocGenerics\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: parallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: IRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomicRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomeInfoDb\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: SummarizedExperiment\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Biobase\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: DelayedArray\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: matrixStats\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocParallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] 15\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i processed_template_filename -i local_dir -i base_dir\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "get_DE_stats_limma(metadata_filename,\n",
    "                   project_id, \n",
    "                   processed_template_filename,\n",
    "                   \"template\",\n",
    "                   local_dir,\n",
    "                   \"real\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i metadata_filename -i project_id -i base_dir -i local_dir -i num_runs -o num_sign_DEGs_simulated\n",
    "\n",
    "source(paste0(base_dir,'/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "num_sign_DEGs_simulated <- c()\n",
    "\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir, \n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\", \n",
    "        i,\n",
    "        \".txt\",\n",
    "        sep=\"\"\n",
    "    )\n",
    "    \n",
    "    run_output <- get_DE_stats_limma(\n",
    "        metadata_filename,\n",
    "        project_id, \n",
    "        simulated_data_filename,\n",
    "        \"simulated\",\n",
    "        local_dir,\n",
    "        i\n",
    "    )\n",
    "    num_sign_DEGs_simulated <- c(num_sign_DEGs_simulated, run_output)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rank genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir,\n",
    "    \"DE_stats\",\n",
    "    f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (1 levels on the left, 2 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(logFC) (Real)</th>\n",
       "      <th>logFC (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean abs(logFC) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA4726</th>\n",
       "      <td>PA4726</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>5549.0</td>\n",
       "      <td>4.530917</td>\n",
       "      <td>-4.530917</td>\n",
       "      <td>0.846422</td>\n",
       "      <td>4845.5</td>\n",
       "      <td>0.16708</td>\n",
       "      <td>0.101184</td>\n",
       "      <td>25</td>\n",
       "      <td>43.127565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA1123</th>\n",
       "      <td>PA1123</td>\n",
       "      <td>0.002578</td>\n",
       "      <td>5537.0</td>\n",
       "      <td>2.175900</td>\n",
       "      <td>-2.175900</td>\n",
       "      <td>0.682483</td>\n",
       "      <td>4391.5</td>\n",
       "      <td>0.16190</td>\n",
       "      <td>0.106530</td>\n",
       "      <td>25</td>\n",
       "      <td>18.905485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA4620</th>\n",
       "      <td>PA4620</td>\n",
       "      <td>0.001107</td>\n",
       "      <td>5512.0</td>\n",
       "      <td>1.611311</td>\n",
       "      <td>1.611311</td>\n",
       "      <td>0.842271</td>\n",
       "      <td>2220.5</td>\n",
       "      <td>0.11220</td>\n",
       "      <td>0.092593</td>\n",
       "      <td>25</td>\n",
       "      <td>16.190288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA1217</th>\n",
       "      <td>PA1217</td>\n",
       "      <td>0.001039</td>\n",
       "      <td>5531.0</td>\n",
       "      <td>1.941753</td>\n",
       "      <td>1.941753</td>\n",
       "      <td>0.850699</td>\n",
       "      <td>1612.0</td>\n",
       "      <td>0.11886</td>\n",
       "      <td>0.120587</td>\n",
       "      <td>25</td>\n",
       "      <td>15.116766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA1216</th>\n",
       "      <td>PA1216</td>\n",
       "      <td>0.000223</td>\n",
       "      <td>5545.0</td>\n",
       "      <td>2.600594</td>\n",
       "      <td>2.600594</td>\n",
       "      <td>0.832988</td>\n",
       "      <td>3823.5</td>\n",
       "      <td>0.19604</td>\n",
       "      <td>0.159747</td>\n",
       "      <td>25</td>\n",
       "      <td>15.052271</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene ID  Adj P-value (Real)  Rank (Real)  abs(logFC) (Real)  \\\n",
       "PA4726  PA4726            0.000010       5549.0           4.530917   \n",
       "PA1123  PA1123            0.002578       5537.0           2.175900   \n",
       "PA4620  PA4620            0.001107       5512.0           1.611311   \n",
       "PA1217  PA1217            0.001039       5531.0           1.941753   \n",
       "PA1216  PA1216            0.000223       5545.0           2.600594   \n",
       "\n",
       "        logFC (Real)  Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "PA4726     -4.530917                        0.846422            4845.5   \n",
       "PA1123     -2.175900                        0.682483            4391.5   \n",
       "PA4620      1.611311                        0.842271            2220.5   \n",
       "PA1217      1.941753                        0.850699            1612.0   \n",
       "PA1216      2.600594                        0.832988            3823.5   \n",
       "\n",
       "        Mean abs(logFC) (simulated)  Std deviation (simulated)  \\\n",
       "PA4726                      0.16708                   0.101184   \n",
       "PA1123                      0.16190                   0.106530   \n",
       "PA4620                      0.11220                   0.092593   \n",
       "PA1217                      0.11886                   0.120587   \n",
       "PA1216                      0.19604                   0.159747   \n",
       "\n",
       "        Number of experiments (simulated)    Z score  \n",
       "PA4726                                 25  43.127565  \n",
       "PA1123                                 25  18.905485  \n",
       "PA4620                                 25  16.190288  \n",
       "PA1217                                 25  15.116766  \n",
       "PA1216                                 25  15.052271  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    'gene',\n",
    "    params\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(logFC) (Real)</th>\n",
       "      <th>logFC (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean abs(logFC) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "      <th>Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA4726</th>\n",
       "      <td>PA4726</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>5549.0</td>\n",
       "      <td>4.530917</td>\n",
       "      <td>-4.530917</td>\n",
       "      <td>0.846422</td>\n",
       "      <td>4845.5</td>\n",
       "      <td>0.16708</td>\n",
       "      <td>0.101184</td>\n",
       "      <td>25</td>\n",
       "      <td>43.127565</td>\n",
       "      <td>cbrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA1123</th>\n",
       "      <td>PA1123</td>\n",
       "      <td>0.002578</td>\n",
       "      <td>5537.0</td>\n",
       "      <td>2.175900</td>\n",
       "      <td>-2.175900</td>\n",
       "      <td>0.682483</td>\n",
       "      <td>4391.5</td>\n",
       "      <td>0.16190</td>\n",
       "      <td>0.106530</td>\n",
       "      <td>25</td>\n",
       "      <td>18.905485</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA4620</th>\n",
       "      <td>PA4620</td>\n",
       "      <td>0.001107</td>\n",
       "      <td>5512.0</td>\n",
       "      <td>1.611311</td>\n",
       "      <td>1.611311</td>\n",
       "      <td>0.842271</td>\n",
       "      <td>2220.5</td>\n",
       "      <td>0.11220</td>\n",
       "      <td>0.092593</td>\n",
       "      <td>25</td>\n",
       "      <td>16.190288</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA1217</th>\n",
       "      <td>PA1217</td>\n",
       "      <td>0.001039</td>\n",
       "      <td>5531.0</td>\n",
       "      <td>1.941753</td>\n",
       "      <td>1.941753</td>\n",
       "      <td>0.850699</td>\n",
       "      <td>1612.0</td>\n",
       "      <td>0.11886</td>\n",
       "      <td>0.120587</td>\n",
       "      <td>25</td>\n",
       "      <td>15.116766</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA1216</th>\n",
       "      <td>PA1216</td>\n",
       "      <td>0.000223</td>\n",
       "      <td>5545.0</td>\n",
       "      <td>2.600594</td>\n",
       "      <td>2.600594</td>\n",
       "      <td>0.832988</td>\n",
       "      <td>3823.5</td>\n",
       "      <td>0.19604</td>\n",
       "      <td>0.159747</td>\n",
       "      <td>25</td>\n",
       "      <td>15.052271</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene ID  Adj P-value (Real)  Rank (Real)  abs(logFC) (Real)  \\\n",
       "PA4726  PA4726            0.000010       5549.0           4.530917   \n",
       "PA1123  PA1123            0.002578       5537.0           2.175900   \n",
       "PA4620  PA4620            0.001107       5512.0           1.611311   \n",
       "PA1217  PA1217            0.001039       5531.0           1.941753   \n",
       "PA1216  PA1216            0.000223       5545.0           2.600594   \n",
       "\n",
       "        logFC (Real)  Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "PA4726     -4.530917                        0.846422            4845.5   \n",
       "PA1123     -2.175900                        0.682483            4391.5   \n",
       "PA4620      1.611311                        0.842271            2220.5   \n",
       "PA1217      1.941753                        0.850699            1612.0   \n",
       "PA1216      2.600594                        0.832988            3823.5   \n",
       "\n",
       "        Mean abs(logFC) (simulated)  Std deviation (simulated)  \\\n",
       "PA4726                      0.16708                   0.101184   \n",
       "PA1123                      0.16190                   0.106530   \n",
       "PA4620                      0.11220                   0.092593   \n",
       "PA1217                      0.11886                   0.120587   \n",
       "PA1216                      0.19604                   0.159747   \n",
       "\n",
       "        Number of experiments (simulated)    Z score Gene Name  \n",
       "PA4726                                 25  43.127565      cbrB  \n",
       "PA1123                                 25  18.905485       NaN  \n",
       "PA4620                                 25  16.190288       NaN  \n",
       "PA1217                                 25  15.116766       NaN  \n",
       "PA1216                                 25  15.052271       NaN  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add gene name as column to summary dataframe\n",
    "summary_gene_ranks = ranking.add_pseudomonas_gene_name_col(summary_gene_ranks, base_dir)\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene ID                              False\n",
       "Adj P-value (Real)                   False\n",
       "Rank (Real)                          False\n",
       "abs(logFC) (Real)                    False\n",
       "logFC (Real)                         False\n",
       "Median adj p-value (simulated)       False\n",
       "Rank (simulated)                     False\n",
       "Mean abs(logFC) (simulated)          False\n",
       "Std deviation (simulated)            False\n",
       "Number of experiments (simulated)    False\n",
       "Z score                              False\n",
       "Gene Name                             True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if there is an NaN values, there should not be\n",
    "summary_gene_ranks.isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create `gene_summary_filename`\n",
    "summary_gene_ranks.to_csv(gene_summary_filename, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GSEA \n",
    "**Goal:** To detect modest but coordinated changes in prespecified sets of related genes (i.e. those genes in the same pathway or share the same GO term).\n",
    "\n",
    "1. Ranks all genes based using DE association statistics. \n",
    "2. An enrichment score (ES) is defined as the maximum distance from the middle of the ranked list. Thus, the enrichment score indicates whether the genes contained in a gene set are clustered towards the beginning or the end of the ranked list (indicating a correlation with change in expression). \n",
    "3. Estimate the statistical significance of the ES by a phenotypic-based permutation test in order to produce a null distribution for the ES( i.e. scores based on permuted phenotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create \"<local_dir>/GSEA_stats/\" subdirectory\n",
    "os.makedirs(os.path.join(local_dir, \"GSEA_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load pathway data\n",
    "adage_kegg_DB_filename = params['pathway_DB_filename']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need to format data into tab-delimited matrix\n",
    "# with columns= KEGG pathway name, description, gene ids\n",
    "# Each gene ids is tab separated\n",
    "adage_kegg_DB_processed_filename = os.path.join(\n",
    "    base_dir,\n",
    "    dataset_name,\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    \"adage_kegg_DB_process_filename.gmt\"\n",
    ")\n",
    "stats.format_pseudomonas_pathway_DB(adage_kegg_DB_filename, local_dir, adage_kegg_DB_processed_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Rcpp\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py:191: FutureWarning: from_items is deprecated. Please use DataFrame.from_dict(dict(items), ...) instead. DataFrame.from_dict(OrderedDict(items)) may be used to preserve the key order.\n",
      "  res = PandasDataFrame.from_items(items)\n"
     ]
    }
   ],
   "source": [
    "%%R -i base_dir -i template_DE_stats_filename -i adage_kegg_DB_processed_filename -i statistic -o template_enriched_pathways\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/GSEA_analysis.R'))\n",
    "template_enriched_pathways <- find_enriched_pathways(template_DE_stats_filename, adage_kegg_DB_processed_filename, statistic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(169, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway</th>\n",
       "      <th>pval</th>\n",
       "      <th>padj</th>\n",
       "      <th>ES</th>\n",
       "      <th>NES</th>\n",
       "      <th>nMoreExtreme</th>\n",
       "      <th>size</th>\n",
       "      <th>leadingEdge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>KEGG-Module-M00178: Ribosome, bacteria</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.001157</td>\n",
       "      <td>0.914364</td>\n",
       "      <td>6.117751</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>KEGG-Pathway-pae00010: Glycolysis / Gluconeoge...</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>0.001157</td>\n",
       "      <td>0.451513</td>\n",
       "      <td>2.578504</td>\n",
       "      <td>0.0</td>\n",
       "      <td>35</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>KEGG-Pathway-pae00680: Methane metabolism - Ps...</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.001157</td>\n",
       "      <td>0.550329</td>\n",
       "      <td>2.716780</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>KEGG-Pathway-pae00250: Alanine, aspartate and ...</td>\n",
       "      <td>0.000121</td>\n",
       "      <td>0.001157</td>\n",
       "      <td>0.471680</td>\n",
       "      <td>2.661697</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynth...</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.001157</td>\n",
       "      <td>0.669514</td>\n",
       "      <td>3.410769</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               pathway      pval      padj  \\\n",
       "168             KEGG-Module-M00178: Ribosome, bacteria  0.000113  0.001157   \n",
       "91   KEGG-Pathway-pae00010: Glycolysis / Gluconeoge...  0.000120  0.001157   \n",
       "74   KEGG-Pathway-pae00680: Methane metabolism - Ps...  0.000128  0.001157   \n",
       "70   KEGG-Pathway-pae00250: Alanine, aspartate and ...  0.000121  0.001157   \n",
       "61   KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynth...  0.000127  0.001157   \n",
       "\n",
       "           ES       NES  nMoreExtreme  size leadingEdge  \n",
       "168  0.914364  6.117751           0.0    55        None  \n",
       "91   0.451513  2.578504           0.0    35        None  \n",
       "74   0.550329  2.716780           0.0    24        None  \n",
       "70   0.471680  2.661697           0.0    34        None  \n",
       "61   0.669514  3.410769           0.0    26        None  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(template_enriched_pathways.shape)\n",
    "template_enriched_pathways[template_enriched_pathways['padj'] < 0.05].sort_values(by='padj').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i project_id -i local_dir -i adage_kegg_DB_processed_filename -i num_runs -i statistic\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/GSEA_analysis.R'))\n",
    "\n",
    "# New files created: \"<local_dir>/GSEA_stats/GSEA_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)) {\n",
    "    simulated_DE_stats_filename <- paste(local_dir, \n",
    "                                     \"DE_stats/DE_stats_simulated_data_\", \n",
    "                                     project_id,\n",
    "                                     \"_\", \n",
    "                                     i,\n",
    "                                     \".txt\",\n",
    "                                     sep = \"\")\n",
    "    \n",
    "    out_filename <- paste(local_dir, \n",
    "                     \"GSEA_stats/GSEA_stats_simulated_data_\",\n",
    "                     project_id,\n",
    "                     \"_\",\n",
    "                     i,\n",
    "                     \".txt\", \n",
    "                     sep = \"\")\n",
    "    \n",
    "    enriched_pathways <- find_enriched_pathways(simulated_DE_stats_filename, adage_kegg_DB_processed_filename, statistic) \n",
    "    \n",
    "    # Remove column with leading edge since its causing parsing issues\n",
    "    write.table(as.data.frame(enriched_pathways[1:7]), file = out_filename, row.names = F, sep = \"\\t\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rank pathways "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_type = \"GSEA\"\n",
    "template_GSEA_stats_filename = os.path.join(\n",
    "    local_dir,\n",
    "    \"GSEA_stats\",\n",
    "    f\"GSEA_stats_template_data_{project_id}_real.txt\"    \n",
    ")\n",
    "template_GSEA_stats, simulated_GSEA_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_GSEA_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_pathways,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pathway summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>padj (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean padj (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathway</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00178: Ribosome, bacteria</th>\n",
       "      <td>KEGG-Module-M00178: Ribosome, bacteria</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>0.001238</td>\n",
       "      <td>161.5</td>\n",
       "      <td>0.001307</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.223324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynthesis - Pseudomonas aeruginosa PAO1</th>\n",
       "      <td>KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynth...</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>0.001238</td>\n",
       "      <td>161.5</td>\n",
       "      <td>0.001307</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.223324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00011: Citrate cycle, second carbon oxidation, 2-oxoglutarate =&gt; oxaloacetate</th>\n",
       "      <td>KEGG-Module-M00011: Citrate cycle, second carb...</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>0.001238</td>\n",
       "      <td>161.5</td>\n",
       "      <td>0.001345</td>\n",
       "      <td>0.000405</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.236316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00335: Sec (secretion) system</th>\n",
       "      <td>KEGG-Module-M00335: Sec (secretion) system</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>0.001238</td>\n",
       "      <td>161.5</td>\n",
       "      <td>0.001599</td>\n",
       "      <td>0.001568</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.223061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00009: Citrate cycle (TCA cycle, Krebs cycle)</th>\n",
       "      <td>KEGG-Module-M00009: Citrate cycle (TCA cycle, ...</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>0.001238</td>\n",
       "      <td>161.5</td>\n",
       "      <td>0.001307</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.223324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                           Pathway ID  \\\n",
       "pathway                                                                                                 \n",
       "KEGG-Module-M00178: Ribosome, bacteria                         KEGG-Module-M00178: Ribosome, bacteria   \n",
       "KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynthe...  KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynth...   \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo...  KEGG-Module-M00011: Citrate cycle, second carb...   \n",
       "KEGG-Module-M00335: Sec (secretion) system                 KEGG-Module-M00335: Sec (secretion) system   \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K...  KEGG-Module-M00009: Citrate cycle (TCA cycle, ...   \n",
       "\n",
       "                                                    Adj P-value (Real)  \\\n",
       "pathway                                                                  \n",
       "KEGG-Module-M00178: Ribosome, bacteria                        0.001249   \n",
       "KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynthe...            0.001249   \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo...            0.001249   \n",
       "KEGG-Module-M00335: Sec (secretion) system                    0.001249   \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K...            0.001249   \n",
       "\n",
       "                                                    Rank (Real)  padj (Real)  \\\n",
       "pathway                                                                        \n",
       "KEGG-Module-M00178: Ribosome, bacteria                    159.0     0.001249   \n",
       "KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynthe...        159.0     0.001249   \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo...        159.0     0.001249   \n",
       "KEGG-Module-M00335: Sec (secretion) system                159.0     0.001249   \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K...        159.0     0.001249   \n",
       "\n",
       "                                                    Median adj p-value (simulated)  \\\n",
       "pathway                                                                              \n",
       "KEGG-Module-M00178: Ribosome, bacteria                                    0.001238   \n",
       "KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynthe...                        0.001238   \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo...                        0.001238   \n",
       "KEGG-Module-M00335: Sec (secretion) system                                0.001238   \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K...                        0.001238   \n",
       "\n",
       "                                                    Rank (simulated)  \\\n",
       "pathway                                                                \n",
       "KEGG-Module-M00178: Ribosome, bacteria                         161.5   \n",
       "KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynthe...             161.5   \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo...             161.5   \n",
       "KEGG-Module-M00335: Sec (secretion) system                     161.5   \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K...             161.5   \n",
       "\n",
       "                                                    Mean padj (simulated)  \\\n",
       "pathway                                                                     \n",
       "KEGG-Module-M00178: Ribosome, bacteria                           0.001307   \n",
       "KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynthe...               0.001307   \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo...               0.001345   \n",
       "KEGG-Module-M00335: Sec (secretion) system                       0.001599   \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K...               0.001307   \n",
       "\n",
       "                                                    Std deviation (simulated)  \\\n",
       "pathway                                                                         \n",
       "KEGG-Module-M00178: Ribosome, bacteria                               0.000257   \n",
       "KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynthe...                   0.000257   \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo...                   0.000405   \n",
       "KEGG-Module-M00335: Sec (secretion) system                           0.001568   \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K...                   0.000257   \n",
       "\n",
       "                                                    Number of experiments (simulated)  \\\n",
       "pathway                                                                                 \n",
       "KEGG-Module-M00178: Ribosome, bacteria                                             25   \n",
       "KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynthe...                                 25   \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo...                                 25   \n",
       "KEGG-Module-M00335: Sec (secretion) system                                         25   \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K...                                 25   \n",
       "\n",
       "                                                     Z score  \n",
       "pathway                                                       \n",
       "KEGG-Module-M00178: Ribosome, bacteria             -0.223324  \n",
       "KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynthe... -0.223324  \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo... -0.236316  \n",
       "KEGG-Module-M00335: Sec (secretion) system         -0.223061  \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K... -0.223324  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create intermediate file: \"<local_dir>/gene_summary_table_<col_to_rank_pathways>.tsv\"\n",
    "summary_pathway_ranks = ranking.generate_summary_table(\n",
    "    template_GSEA_stats_filename,\n",
    "    template_GSEA_stats,\n",
    "    simulated_GSEA_summary_stats,\n",
    "    col_to_rank_pathways,\n",
    "    local_dir,\n",
    "    'pathway',\n",
    "    params\n",
    ")\n",
    "\n",
    "summary_pathway_ranks.sort_values(by='Rank (simulated)', ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create `pathway_summary_filename`\n",
    "summary_pathway_ranks.to_csv(pathway_summary_filename, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Quick check:**\n",
    "\n",
    "Looks like Ribosomes are found to be significantly differentiated in the template and across 25 simulated experiments. So this pathway looks to be generic, which is consistent with findings from [Crow. et. al.](https://www.pnas.org/content/116/13/6491.abstract), using Human data. This is also consistent with [paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2679180/) which observed that Ribosome proteins have many different functions in *S. cerevisiae*, which might explain their variability across a wide range of contexts.\n",
    "\n",
    "* Need to think about why it might make sense that Ribosomes being generic in Humans would also imply that they are generic in P. aeruginosa."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
