{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify generic genes and pathways\n",
    "\n",
    "Studies have found that some genes are more likely to be differentially expressed even across a wide range of experimental designs. These generic genes and subsequent pathways are not necessarily specific to the biological process being studied but instead represent a more systematic change. \n",
    "\n",
    "We have developed an approach, outlined below, to automatically identify these generic genes and pathways. We have validated this simulation approach can identify generic genes and pathways in the analysis notebooks: [human_general_analysis](../human_general_analysis/) and [human_cancer_analysis](../human_cancer_analysis/). Here\n",
    "\n",
    "This notebook applies this approach to identify generic genes and pathways in the pseudomonas compendium. \n",
    "\n",
    "**Steps to identify generic genes:**\n",
    "1. Simulates N gene expression experiments using [ponyo](https://github.com/ajlee21/ponyo)\n",
    "2. Perform DE analysis to get association statistics for each gene\n",
    "\n",
    "In this case the DE analysis is based on the experimental design of the template experiment, described in the previous [notebook](1_process_pseudomonas_data.ipynb). \n",
    "The template experiment is [GEOD-33245](https://www.ebi.ac.uk/arrayexpress/experiments/E-GEOD-33245/?s_sortby=col_8&s_sortorder=ascending), which contains multiple different comparisons including WT vs *crc* mutants, WT vs *cbr* mutants in different conditions. So the DE analysis is comparing WT vs mutant.\n",
    "\n",
    "3. For each gene, aggregate statistics across all simulated experiments \n",
    "4. Rank genes based on this aggregated statistic\n",
    "\n",
    "**Steps to identify generic gene sets (pathways):**\n",
    "1. Using the same simulated experiments from above, perform GSEA analysis. This analysis will determine whether the genes contained in a gene set are clustered towards the beginning or the end of the ranked list of genes, where genes are ranked by log fold change, indicating a correlation with change in expression.\n",
    "2. For each gene set (pathway), aggregate statistics across all simulated experiments\n",
    "3. Rank gene sets based on this aggregated statistic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_qint8 = np.dtype([(\"qint8\", np.int8, 1)])\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_quint8 = np.dtype([(\"quint8\", np.uint8, 1)])\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_qint16 = np.dtype([(\"qint16\", np.int16, 1)])\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:529: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_quint16 = np.dtype([(\"quint16\", np.uint16, 1)])\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:530: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  _np_qint32 = np.dtype([(\"qint32\", np.int32, 1)])\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:535: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
      "  np_resource = np.dtype([(\"resource\", np.ubyte, 1)])\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "from rpy2.robjects import pandas2ri\n",
    "pandas2ri.activate()\n",
    "\n",
    "from ponyo import utils, simulate_expression_data\n",
    "from generic_expression_patterns_modules import calc, process\n",
    "\n",
    "np.random.seed(123)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(),\"../\"))\n",
    "\n",
    "config_file = os.path.abspath(os.path.join(base_dir,\n",
    "                                           \"configs\",\n",
    "                                           \"config_pseudomonas_33245.tsv\"))\n",
    "params = utils.read_config(config_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load params\n",
    "local_dir = params[\"local_dir\"]\n",
    "dataset_name = params['dataset_name']\n",
    "NN_architecture = params['NN_architecture']\n",
    "num_runs = params['num_simulated']\n",
    "project_id = params['project_id']\n",
    "metadata_col_id = params['metadata_colname']\n",
    "processed_template_filename = params['processed_template_filename']\n",
    "normalized_compendium_filename = params['normalized_compendium_filename']\n",
    "scaler_filename = params['scaler_filename']\n",
    "col_to_rank_genes = params['rank_genes_by']\n",
    "col_to_rank_pathways = params['rank_pathways_by']\n",
    "statistic = params['gsea_statistic']\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "sample_id_metadata_filename = os.path.join(\n",
    "    base_dir,\n",
    "    dataset_name,\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    f\"{project_id}_process_samples.tsv\")\n",
    "\n",
    "# Load pickled file\n",
    "scaler = pickle.load(open(scaler_filename, \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output files\n",
    "gene_summary_filename = os.path.join(\n",
    "    base_dir, \n",
    "    dataset_name, \n",
    "    f\"generic_gene_summary_{project_id}.tsv\"\n",
    ")\n",
    "pathway_summary_filename = os.path.join(\n",
    "    base_dir, \n",
    "    dataset_name, \n",
    "    f\"generic_pathway_summary_{project_id}.tsv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulate experiments using selected template experiment\n",
    "Workflow:\n",
    "\n",
    "1. Get the gene expression data for the selected template experiment\n",
    "2. Encode this experiment into a latent space using the trained VAE model\n",
    "3. Linearly shift the encoded template experiment in the latent space\n",
    "4. Decode the samples. This results in a new experiment\n",
    "5. Repeat steps 1-4 to get multiple simulated experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'# Simulate multiple experiments\\n# This step creates the following files in \"<local_dir>/pseudo_experiment/\" directory:           \\n#   - selected_simulated_data_SRP012656_<n>.txt\\n#   - selected_simulated_encoded_data_SRP012656_<n>.txt\\n#   - template_normalized_data_SRP012656_test.txt\\n# in which \"<n>\" is an integer in the range of [0, num_runs-1] \\nos.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)\\nfor run_id in range(num_runs):\\n    simulate_expression_data.shift_template_experiment(\\n        normalized_compendium_filename,\\n        project_id,\\n        metadata_col_id,\\n        NN_architecture,\\n        dataset_name,\\n        scaler,\\n        local_dir,\\n        base_dir,\\n        run_id)'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"# Simulate multiple experiments\n",
    "# This step creates the following files in \"<local_dir>/pseudo_experiment/\" directory:           \n",
    "#   - selected_simulated_data_SRP012656_<n>.txt\n",
    "#   - selected_simulated_encoded_data_SRP012656_<n>.txt\n",
    "#   - template_normalized_data_SRP012656_test.txt\n",
    "# in which \"<n>\" is an integer in the range of [0, num_runs-1] \n",
    "os.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)\n",
    "for run_id in range(num_runs):\n",
    "    simulate_expression_data.shift_template_experiment(\n",
    "        normalized_compendium_filename,\n",
    "        project_id,\n",
    "        metadata_col_id,\n",
    "        NN_architecture,\n",
    "        dataset_name,\n",
    "        scaler,\n",
    "        local_dir,\n",
    "        base_dir,\n",
    "        run_id)\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'# This step modifies the following files:\\n# \"<local_dir>/pseudo_experiments/selected_simulated_data_SRP012656_<n>.txt\"\\nif os.path.exists(sample_id_metadata_filename):\\n    # Read in metadata\\n    metadata = pd.read_csv(sample_id_metadata_filename, sep=\\'\\t\\', header=0, index_col=0)\\n    \\n    # Get samples to be dropped\\n    sample_ids_to_drop = list(metadata[metadata[\"processing\"] == \"drop\"].index)\\n\\n    process.subset_samples(\\n        sample_ids_to_drop,\\n        num_runs,\\n        local_dir,\\n        project_id\\n    )'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"# This step modifies the following files:\n",
    "# \"<local_dir>/pseudo_experiments/selected_simulated_data_SRP012656_<n>.txt\"\n",
    "if os.path.exists(sample_id_metadata_filename):\n",
    "    # Read in metadata\n",
    "    metadata = pd.read_csv(sample_id_metadata_filename, sep='\\t', header=0, index_col=0)\n",
    "    \n",
    "    # Get samples to be dropped\n",
    "    sample_ids_to_drop = list(metadata[metadata[\"processing\"] == \"drop\"].index)\n",
    "\n",
    "    process.subset_samples(\n",
    "        sample_ids_to_drop,\n",
    "        num_runs,\n",
    "        local_dir,\n",
    "        project_id\n",
    "    )\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata file with grouping assignments for samples\n",
    "metadata_filename = os.path.join(\n",
    "    base_dir,\n",
    "    dataset_name,\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    f\"{project_id}_groups.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "# Check whether ordering of sample ids is consistent between gene expression data and metadata\n",
    "process.compare_and_reorder_samples(processed_template_filename, metadata_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Error in paste0(base_dir, \"/generic_expression_patterns_modules/DE_analysis.R\") : \n",
      "  object 'base_dir' not found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Error in paste0(base_dir, \"/generic_expression_patterns_modules/DE_analysis.R\") : \n",
      "  object 'base_dir' not found\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i processed_template_filename -i local_dir\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "get_DE_stats_limma(metadata_filename,\n",
    "                   project_id, \n",
    "                   processed_template_filename,\n",
    "                   \"template\",\n",
    "                   local_dir,\n",
    "                   \"real\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "# Check whether ordering of sample ids is consistent between gene expression data and metadata\n",
    "for i in range(num_runs):\n",
    "    simulated_data_filename = os.path.join(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment\",\n",
    "        f\"selected_simulated_data_{project_id}_{i}.txt\")\n",
    "        \n",
    "    process.compare_and_reorder_samples(simulated_data_filename, metadata_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: S4Vectors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: stats4\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocGenerics\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: parallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: IRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomicRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomeInfoDb\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: SummarizedExperiment\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Biobase\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: DelayedArray\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: matrixStats\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocParallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i base_dir -i local_dir -i num_runs -o num_sign_DEGs_simulated\n",
    "\n",
    "source(paste0(base_dir,'/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "num_sign_DEGs_simulated <- c()\n",
    "\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir, \n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\", \n",
    "        i,\n",
    "        \".txt\",\n",
    "        sep=\"\"\n",
    "    )\n",
    "    \n",
    "    run_output <- get_DE_stats_limma(\n",
    "        metadata_filename,\n",
    "        project_id, \n",
    "        simulated_data_filename,\n",
    "        \"simulated\",\n",
    "        local_dir,\n",
    "        i\n",
    "    )\n",
    "    num_sign_DEGs_simulated <- c(num_sign_DEGs_simulated, run_output)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rank genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(138725, 7)\n"
     ]
    }
   ],
   "source": [
    "# Concatenate simulated experiments\n",
    "simulated_DE_stats_all = process.concat_simulated_data(local_dir, num_runs, project_id, 'DE')\n",
    "\n",
    "print(simulated_DE_stats_all.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take absolute value of logFC and t statistic\n",
    "simulated_DE_stats_all = process.abs_value_stats(simulated_DE_stats_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aggregate statistics across all simulated experiments\n",
    "simulated_DE_summary_stats = calc.aggregate_stats(\n",
    "    col_to_rank_genes,\n",
    "    simulated_DE_stats_all,\n",
    "    'DE'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load association statistics for template experiment\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir,\n",
    "    \"DE_stats\",\n",
    "    \"DE_stats_template_data_\"+project_id+\"_real.txt\")\n",
    "\n",
    "template_DE_stats = pd.read_csv(\n",
    "    template_DE_stats_filename,\n",
    "    header=0,\n",
    "    sep='\\t',\n",
    "    index_col=0)\n",
    "\n",
    "# Take absolute value of logFC and t statistic\n",
    "template_DE_stats = process.abs_value_stats(template_DE_stats)\n",
    "\n",
    "# Rank genes in template experiment\n",
    "template_DE_stats = calc.rank_genes_or_pathways(\n",
    "    col_to_rank_genes,\n",
    "    template_DE_stats,\n",
    "    True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rank genes in simulated experiments\n",
    "simulated_DE_summary_stats = calc.rank_genes_or_pathways(\n",
    "    col_to_rank_genes,\n",
    "    simulated_DE_summary_stats,\n",
    "    False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5549, 13)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (1 levels on the left, 2 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>Test statistic (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean test statistic (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>abs(Z score)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA3600</th>\n",
       "      <td>PA3600</td>\n",
       "      <td>7.068633e-08</td>\n",
       "      <td>5549.0</td>\n",
       "      <td>6.137566</td>\n",
       "      <td>0.658483</td>\n",
       "      <td>5386.0</td>\n",
       "      <td>0.298250</td>\n",
       "      <td>0.181704</td>\n",
       "      <td>6</td>\n",
       "      <td>32.136455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA4470</th>\n",
       "      <td>PA4470</td>\n",
       "      <td>1.176030e-08</td>\n",
       "      <td>5548.0</td>\n",
       "      <td>6.016552</td>\n",
       "      <td>0.715044</td>\n",
       "      <td>4864.5</td>\n",
       "      <td>0.206083</td>\n",
       "      <td>0.145293</td>\n",
       "      <td>6</td>\n",
       "      <td>39.991398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA3601</th>\n",
       "      <td>PA3601</td>\n",
       "      <td>1.100335e-08</td>\n",
       "      <td>5547.0</td>\n",
       "      <td>5.836628</td>\n",
       "      <td>0.678337</td>\n",
       "      <td>4229.0</td>\n",
       "      <td>0.171667</td>\n",
       "      <td>0.100604</td>\n",
       "      <td>6</td>\n",
       "      <td>56.309776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA2426</th>\n",
       "      <td>PA2426</td>\n",
       "      <td>3.642444e-08</td>\n",
       "      <td>5546.0</td>\n",
       "      <td>5.533709</td>\n",
       "      <td>0.623670</td>\n",
       "      <td>5355.0</td>\n",
       "      <td>0.377667</td>\n",
       "      <td>0.264477</td>\n",
       "      <td>6</td>\n",
       "      <td>19.495214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA4469</th>\n",
       "      <td>PA4469</td>\n",
       "      <td>3.642444e-08</td>\n",
       "      <td>5545.0</td>\n",
       "      <td>5.423924</td>\n",
       "      <td>0.710824</td>\n",
       "      <td>3832.0</td>\n",
       "      <td>0.252917</td>\n",
       "      <td>0.277097</td>\n",
       "      <td>6</td>\n",
       "      <td>18.661361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene ID  Adj P-value (Real)  Rank (Real)  Test statistic (Real)  \\\n",
       "PA3600  PA3600        7.068633e-08       5549.0               6.137566   \n",
       "PA4470  PA4470        1.176030e-08       5548.0               6.016552   \n",
       "PA3601  PA3601        1.100335e-08       5547.0               5.836628   \n",
       "PA2426  PA2426        3.642444e-08       5546.0               5.533709   \n",
       "PA4469  PA4469        3.642444e-08       5545.0               5.423924   \n",
       "\n",
       "        Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "PA3600                        0.658483            5386.0   \n",
       "PA4470                        0.715044            4864.5   \n",
       "PA3601                        0.678337            4229.0   \n",
       "PA2426                        0.623670            5355.0   \n",
       "PA4469                        0.710824            3832.0   \n",
       "\n",
       "        Mean test statistic (simulated)  Std deviation (simulated)  \\\n",
       "PA3600                         0.298250                   0.181704   \n",
       "PA4470                         0.206083                   0.145293   \n",
       "PA3601                         0.171667                   0.100604   \n",
       "PA2426                         0.377667                   0.264477   \n",
       "PA4469                         0.252917                   0.277097   \n",
       "\n",
       "        Number of experiments (simulated)  abs(Z score)  \n",
       "PA3600                                  6     32.136455  \n",
       "PA4470                                  6     39.991398  \n",
       "PA3601                                  6     56.309776  \n",
       "PA2426                                  6     19.495214  \n",
       "PA4469                                  6     18.661361  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks = process.generate_summary_table(\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir\n",
    ")\n",
    "\n",
    "summary_gene_ranks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add gene name as column to summary dataframe\n",
    "summary_gene_ranks = process.add_pseudomonas_gene_name_col(summary_gene_ranks, base_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create `gene_summary_filename`\n",
    "summary_gene_ranks.to_csv(gene_summary_filename, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GSEA \n",
    "**Goal:** To detect modest but coordinated changes in prespecified sets of related genes (i.e. those genes in the same pathway or share the same GO term).\n",
    "\n",
    "1. Ranks all genes based using DE association statistics. In this case we used the p-value scores to rank genes. logFC returned error -- need to look into this.\n",
    "2. An enrichment score (ES) is defined as the maximum distance from the middle of the ranked list. Thus, the enrichment score indicates whether the genes contained in a gene set are clustered towards the beginning or the end of the ranked list (indicating a correlation with change in expression). \n",
    "3. Estimate the statistical significance of the ES by a phenotypic-based permutation test in order to produce a null distribution for the ES( i.e. scores based on permuted phenotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load pathway data\n",
    "adage_kegg_DB_filename = \"https://raw.githubusercontent.com/greenelab/adage/master/Node_interpretation/pseudomonas_KEGG_terms.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KEGG-Pathway-pae00072: Synthesis and degradati...</td>\n",
       "      <td>10</td>\n",
       "      <td>PA2553;PA2000;PA2011;PA1999;PA2001;PA3925;PA17...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KEGG-Pathway-pae00071: Fatty acid degradation ...</td>\n",
       "      <td>32</td>\n",
       "      <td>PA5427;PA1821;PA2553;PA1737;PA1027;PA3014;PA25...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KEGG-Pathway-pae00903: Limonene and pinene deg...</td>\n",
       "      <td>9</td>\n",
       "      <td>PA1821;PA1737;PA1027;PA3014;PA3426;PA4899;PA24...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KEGG-Pathway-pae00380: Tryptophan metabolism -...</td>\n",
       "      <td>27</td>\n",
       "      <td>PA1821;PA3366;PA2080;PA0421;PA2553;PA2081;PA17...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KEGG-Pathway-pae00900: Terpenoid backbone bios...</td>\n",
       "      <td>16</td>\n",
       "      <td>PA2553;PA4044;PA2001;PA3925;PA3803;PA4043;PA17...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   0   1  \\\n",
       "0  KEGG-Pathway-pae00072: Synthesis and degradati...  10   \n",
       "1  KEGG-Pathway-pae00071: Fatty acid degradation ...  32   \n",
       "2  KEGG-Pathway-pae00903: Limonene and pinene deg...   9   \n",
       "3  KEGG-Pathway-pae00380: Tryptophan metabolism -...  27   \n",
       "4  KEGG-Pathway-pae00900: Terpenoid backbone bios...  16   \n",
       "\n",
       "                                                   2  \n",
       "0  PA2553;PA2000;PA2011;PA1999;PA2001;PA3925;PA17...  \n",
       "1  PA5427;PA1821;PA2553;PA1737;PA1027;PA3014;PA25...  \n",
       "2  PA1821;PA1737;PA1027;PA3014;PA3426;PA4899;PA24...  \n",
       "3  PA1821;PA3366;PA2080;PA0421;PA2553;PA2081;PA17...  \n",
       "4  PA2553;PA4044;PA2001;PA3925;PA3803;PA4043;PA17...  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adage_kegg_DB = pd.read_csv(adage_kegg_DB_filename, sep=\"\\t\", header=None)\n",
    "adage_kegg_DB.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KEGG-Pathway-pae00072: Synthesis and degradati...</td>\n",
       "      <td>PA2553;PA2000;PA2011;PA1999;PA2001;PA3925;PA17...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KEGG-Pathway-pae00071: Fatty acid degradation ...</td>\n",
       "      <td>PA5427;PA1821;PA2553;PA1737;PA1027;PA3014;PA25...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KEGG-Pathway-pae00903: Limonene and pinene deg...</td>\n",
       "      <td>PA1821;PA1737;PA1027;PA3014;PA3426;PA4899;PA24...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KEGG-Pathway-pae00380: Tryptophan metabolism -...</td>\n",
       "      <td>PA1821;PA3366;PA2080;PA0421;PA2553;PA2081;PA17...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KEGG-Pathway-pae00900: Terpenoid backbone bios...</td>\n",
       "      <td>PA2553;PA4044;PA2001;PA3925;PA3803;PA4043;PA17...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   0  \\\n",
       "0  KEGG-Pathway-pae00072: Synthesis and degradati...   \n",
       "1  KEGG-Pathway-pae00071: Fatty acid degradation ...   \n",
       "2  KEGG-Pathway-pae00903: Limonene and pinene deg...   \n",
       "3  KEGG-Pathway-pae00380: Tryptophan metabolism -...   \n",
       "4  KEGG-Pathway-pae00900: Terpenoid backbone bios...   \n",
       "\n",
       "                                                   2  \n",
       "0  PA2553;PA2000;PA2011;PA1999;PA2001;PA3925;PA17...  \n",
       "1  PA5427;PA1821;PA2553;PA1737;PA1027;PA3014;PA25...  \n",
       "2  PA1821;PA1737;PA1027;PA3014;PA3426;PA4899;PA24...  \n",
       "3  PA1821;PA3366;PA2080;PA0421;PA2553;PA2081;PA17...  \n",
       "4  PA2553;PA4044;PA2001;PA3925;PA3803;PA4043;PA17...  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adage_kegg_DB.drop(columns=[1], inplace=True)\n",
    "adage_kegg_DB.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "adage_kegg_DB[2] = adage_kegg_DB[2].str.split(\";\").str.join(\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KEGG-Pathway-pae00072: Synthesis and degradati...</td>\n",
       "      <td>PA2553\\tPA2000\\tPA2011\\tPA1999\\tPA2001\\tPA3925...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KEGG-Pathway-pae00071: Fatty acid degradation ...</td>\n",
       "      <td>PA5427\\tPA1821\\tPA2553\\tPA1737\\tPA1027\\tPA3014...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KEGG-Pathway-pae00903: Limonene and pinene deg...</td>\n",
       "      <td>PA1821\\tPA1737\\tPA1027\\tPA3014\\tPA3426\\tPA4899...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KEGG-Pathway-pae00380: Tryptophan metabolism -...</td>\n",
       "      <td>PA1821\\tPA3366\\tPA2080\\tPA0421\\tPA2553\\tPA2081...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KEGG-Pathway-pae00900: Terpenoid backbone bios...</td>\n",
       "      <td>PA2553\\tPA4044\\tPA2001\\tPA3925\\tPA3803\\tPA4043...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   0  \\\n",
       "0  KEGG-Pathway-pae00072: Synthesis and degradati...   \n",
       "1  KEGG-Pathway-pae00071: Fatty acid degradation ...   \n",
       "2  KEGG-Pathway-pae00903: Limonene and pinene deg...   \n",
       "3  KEGG-Pathway-pae00380: Tryptophan metabolism -...   \n",
       "4  KEGG-Pathway-pae00900: Terpenoid backbone bios...   \n",
       "\n",
       "                                                   2  \n",
       "0  PA2553\\tPA2000\\tPA2011\\tPA1999\\tPA2001\\tPA3925...  \n",
       "1  PA5427\\tPA1821\\tPA2553\\tPA1737\\tPA1027\\tPA3014...  \n",
       "2  PA1821\\tPA1737\\tPA1027\\tPA3014\\tPA3426\\tPA4899...  \n",
       "3  PA1821\\tPA3366\\tPA2080\\tPA0421\\tPA2553\\tPA2081...  \n",
       "4  PA2553\\tPA4044\\tPA2001\\tPA3925\\tPA3803\\tPA4043...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adage_kegg_DB.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(169, 2)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adage_kegg_DB.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "adage_kegg_DB.to_csv(\"adage_kegg_DB_tmp_filename.gmt\", \n",
    "                     quoting=csv.QUOTE_NONE,\n",
    "                     escapechar=\"\\\\\",\n",
    "                     index=False,\n",
    "                     header=False,\n",
    "                     sep=\"\\t\"\n",
    "                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"adage_kegg_DB_tmp_filename.gmt\", \"r\") as f:\n",
    "    temp = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"adage_kegg_DB_process_filename.gmt\", \"w\") as of:\n",
    "    of.write(temp.replace(\"\\\\\", \"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "adage_kegg_DB_processed_filename = \"adage_kegg_DB_process_filename.gmt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need to format data into tab-delimited matrix\n",
    "# with columns= KEGG pathway name, description, gene ids\n",
    "# Each gene ids is tab separated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Rcpp\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py:191: FutureWarning: from_items is deprecated. Please use DataFrame.from_dict(dict(items), ...) instead. DataFrame.from_dict(OrderedDict(items)) may be used to preserve the key order.\n",
      "  res = PandasDataFrame.from_items(items)\n"
     ]
    }
   ],
   "source": [
    "%%R -i base_dir -i template_DE_stats_filename -i adage_kegg_DB_processed_filename -i statistic -o template_enriched_pathways\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/GSEA_analysis.R'))\n",
    "template_enriched_pathways <- find_enriched_pathways(template_DE_stats_filename, adage_kegg_DB_processed_filename, statistic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(169, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway</th>\n",
       "      <th>pval</th>\n",
       "      <th>padj</th>\n",
       "      <th>ES</th>\n",
       "      <th>NES</th>\n",
       "      <th>nMoreExtreme</th>\n",
       "      <th>size</th>\n",
       "      <th>leadingEdge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>KEGG-Module-M00178: Ribosome, bacteria</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.900619</td>\n",
       "      <td>6.085628</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>KEGG-Module-M00335: Sec (secretion) system</td>\n",
       "      <td>0.000146</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.730965</td>\n",
       "      <td>2.579776</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>KEGG-Pathway-pae00250: Alanine, aspartate and ...</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.460515</td>\n",
       "      <td>2.615437</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>KEGG-Pathway-pae00400: Phenylalanine, tyrosine...</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.483272</td>\n",
       "      <td>2.557964</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynth...</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.638701</td>\n",
       "      <td>3.293676</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>KEGG-Pathway-pae00240: Pyrimidine metabolism -...</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.513021</td>\n",
       "      <td>3.180929</td>\n",
       "      <td>0.0</td>\n",
       "      <td>43</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>KEGG-Pathway-pae03010: Ribosome - Pseudomonas ...</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.900619</td>\n",
       "      <td>6.085628</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>KEGG-Pathway-pae00330: Arginine and proline me...</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.350134</td>\n",
       "      <td>2.550105</td>\n",
       "      <td>0.0</td>\n",
       "      <td>69</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>KEGG-Module-M00009: Citrate cycle (TCA cycle, ...</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.655828</td>\n",
       "      <td>3.176966</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>KEGG-Pathway-pae00230: Purine metabolism - Pse...</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.405511</td>\n",
       "      <td>3.040558</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>KEGG-Module-M00011: Citrate cycle, second carb...</td>\n",
       "      <td>0.000135</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.642079</td>\n",
       "      <td>2.762956</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>KEGG-Pathway-pae01210: 2-Oxocarboxylic acid me...</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.470734</td>\n",
       "      <td>2.587238</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>KEGG-Pathway-pae03018: RNA degradation - Pseud...</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.691998</td>\n",
       "      <td>2.818566</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>KEGG-Pathway-pae00020: Citrate cycle (TCA cycl...</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.544917</td>\n",
       "      <td>3.064743</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>KEGG-Module-M00360: Aminoacyl-tRNA biosynthesi...</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.690190</td>\n",
       "      <td>3.282871</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>KEGG-Module-M00144: NADH:quinone oxidoreductas...</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.655344</td>\n",
       "      <td>2.588272</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>KEGG-Pathway-pae03060: Protein export - Pseudo...</td>\n",
       "      <td>0.000135</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.666837</td>\n",
       "      <td>2.869491</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>KEGG-Pathway-pae00270: Cysteine and methionine...</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.465093</td>\n",
       "      <td>2.588585</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>KEGG-Pathway-pae00190: Oxidative phosphorylati...</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.567832</td>\n",
       "      <td>3.787781</td>\n",
       "      <td>0.0</td>\n",
       "      <td>53</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>KEGG-Module-M00157: F-type ATPase, prokaryotes...</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.933959</td>\n",
       "      <td>2.852426</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>KEGG-Module-M00156: Cytochrome c oxidase, cbb3...</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.823172</td>\n",
       "      <td>2.514070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>KEGG-Pathway-pae00010: Glycolysis / Gluconeoge...</td>\n",
       "      <td>0.000242</td>\n",
       "      <td>0.001788</td>\n",
       "      <td>0.413157</td>\n",
       "      <td>2.375335</td>\n",
       "      <td>1.0</td>\n",
       "      <td>35</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>KEGG-Pathway-pae02040: Flagellar assembly - Ps...</td>\n",
       "      <td>0.000243</td>\n",
       "      <td>0.001788</td>\n",
       "      <td>0.425094</td>\n",
       "      <td>2.414266</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>KEGG-Module-M00002: Glycolysis, core module in...</td>\n",
       "      <td>0.000458</td>\n",
       "      <td>0.003223</td>\n",
       "      <td>0.751750</td>\n",
       "      <td>2.295939</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>KEGG-Pathway-pae00540: Lipopolysaccharide bios...</td>\n",
       "      <td>0.000795</td>\n",
       "      <td>0.005373</td>\n",
       "      <td>0.483895</td>\n",
       "      <td>2.211730</td>\n",
       "      <td>5.0</td>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>KEGG-Pathway-pae00620: Pyruvate metabolism - P...</td>\n",
       "      <td>0.000892</td>\n",
       "      <td>0.005620</td>\n",
       "      <td>0.325672</td>\n",
       "      <td>2.241867</td>\n",
       "      <td>7.0</td>\n",
       "      <td>58</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>KEGG-Pathway-pae00680: Methane metabolism - Ps...</td>\n",
       "      <td>0.000900</td>\n",
       "      <td>0.005620</td>\n",
       "      <td>0.446158</td>\n",
       "      <td>2.225449</td>\n",
       "      <td>6.0</td>\n",
       "      <td>24</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>KEGG-Pathway-pae01220: Degradation of aromatic...</td>\n",
       "      <td>0.000964</td>\n",
       "      <td>0.005620</td>\n",
       "      <td>-0.345706</td>\n",
       "      <td>-2.264434</td>\n",
       "      <td>1.0</td>\n",
       "      <td>26</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>KEGG-Module-M00010: Citrate cycle, first carbo...</td>\n",
       "      <td>0.000947</td>\n",
       "      <td>0.005620</td>\n",
       "      <td>0.779454</td>\n",
       "      <td>2.079944</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>KEGG-Pathway-pae00450: Selenocompound metaboli...</td>\n",
       "      <td>0.001033</td>\n",
       "      <td>0.005716</td>\n",
       "      <td>0.615162</td>\n",
       "      <td>2.080158</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>KEGG-Pathway-pae00550: Peptidoglycan biosynthe...</td>\n",
       "      <td>0.003312</td>\n",
       "      <td>0.013992</td>\n",
       "      <td>0.447560</td>\n",
       "      <td>2.045655</td>\n",
       "      <td>24.0</td>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>KEGG-Module-M00115: NAD biosynthesis, aspartat...</td>\n",
       "      <td>0.003422</td>\n",
       "      <td>0.014107</td>\n",
       "      <td>0.791704</td>\n",
       "      <td>1.920905</td>\n",
       "      <td>20.0</td>\n",
       "      <td>4</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>KEGG-Module-M00616: Sulfate-sulfur assimilation</td>\n",
       "      <td>0.003737</td>\n",
       "      <td>0.015038</td>\n",
       "      <td>0.537725</td>\n",
       "      <td>1.978550</td>\n",
       "      <td>25.0</td>\n",
       "      <td>11</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>KEGG-Module-M00048: Inosine monophosphate bios...</td>\n",
       "      <td>0.003958</td>\n",
       "      <td>0.015556</td>\n",
       "      <td>0.534604</td>\n",
       "      <td>2.043684</td>\n",
       "      <td>27.0</td>\n",
       "      <td>12</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>KEGG-Module-M00034: Methionine salvage pathway</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.017422</td>\n",
       "      <td>0.529611</td>\n",
       "      <td>1.948695</td>\n",
       "      <td>31.0</td>\n",
       "      <td>11</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>KEGG-Module-M00452: CusS-CusR (copper toleranc...</td>\n",
       "      <td>0.004639</td>\n",
       "      <td>0.017422</td>\n",
       "      <td>-0.617106</td>\n",
       "      <td>-2.073694</td>\n",
       "      <td>15.0</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>KEGG-Module-M00050: Guanine ribonucleotide bio...</td>\n",
       "      <td>0.005680</td>\n",
       "      <td>0.020868</td>\n",
       "      <td>0.706300</td>\n",
       "      <td>1.884735</td>\n",
       "      <td>35.0</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>KEGG-Pathway-pae00770: Pantothenate and CoA bi...</td>\n",
       "      <td>0.005884</td>\n",
       "      <td>0.021157</td>\n",
       "      <td>0.418544</td>\n",
       "      <td>1.953852</td>\n",
       "      <td>44.0</td>\n",
       "      <td>20</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>KEGG-Module-M00331: Type II general secretion ...</td>\n",
       "      <td>0.006568</td>\n",
       "      <td>0.022855</td>\n",
       "      <td>-0.275396</td>\n",
       "      <td>-2.006574</td>\n",
       "      <td>11.0</td>\n",
       "      <td>32</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>KEGG-Module-M00018: Threonine biosynthesis, as...</td>\n",
       "      <td>0.006627</td>\n",
       "      <td>0.022855</td>\n",
       "      <td>0.700036</td>\n",
       "      <td>1.868021</td>\n",
       "      <td>41.0</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>KEGG-Pathway-pae00030: Pentose phosphate pathw...</td>\n",
       "      <td>0.006814</td>\n",
       "      <td>0.023032</td>\n",
       "      <td>0.383921</td>\n",
       "      <td>1.915012</td>\n",
       "      <td>52.0</td>\n",
       "      <td>24</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>KEGG-Pathway-pae03440: Homologous recombinatio...</td>\n",
       "      <td>0.006974</td>\n",
       "      <td>0.023110</td>\n",
       "      <td>0.397155</td>\n",
       "      <td>1.923901</td>\n",
       "      <td>53.0</td>\n",
       "      <td>22</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>KEGG-Module-M00210: Putative ABC transport system</td>\n",
       "      <td>0.007731</td>\n",
       "      <td>0.025126</td>\n",
       "      <td>0.691739</td>\n",
       "      <td>1.845880</td>\n",
       "      <td>48.0</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>KEGG-Module-M00188: NitT/TauT family transport...</td>\n",
       "      <td>0.008880</td>\n",
       "      <td>0.028315</td>\n",
       "      <td>-0.441202</td>\n",
       "      <td>-1.943573</td>\n",
       "      <td>25.0</td>\n",
       "      <td>12</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>KEGG-Module-M00119: Pantothenate biosynthesis,...</td>\n",
       "      <td>0.009282</td>\n",
       "      <td>0.029050</td>\n",
       "      <td>0.635857</td>\n",
       "      <td>1.830504</td>\n",
       "      <td>59.0</td>\n",
       "      <td>6</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>KEGG-Pathway-pae00521: Streptomycin biosynthes...</td>\n",
       "      <td>0.009765</td>\n",
       "      <td>0.029846</td>\n",
       "      <td>0.592833</td>\n",
       "      <td>1.810585</td>\n",
       "      <td>63.0</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>KEGG-Pathway-pae00364: Fluorobenzoate degradat...</td>\n",
       "      <td>0.009890</td>\n",
       "      <td>0.029846</td>\n",
       "      <td>-0.616623</td>\n",
       "      <td>-1.918883</td>\n",
       "      <td>34.0</td>\n",
       "      <td>6</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>KEGG-Module-M00019: Valine/isoleucine biosynth...</td>\n",
       "      <td>0.010186</td>\n",
       "      <td>0.029850</td>\n",
       "      <td>0.566841</td>\n",
       "      <td>1.826503</td>\n",
       "      <td>67.0</td>\n",
       "      <td>8</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>KEGG-Pathway-pae00500: Starch and sucrose meta...</td>\n",
       "      <td>0.010244</td>\n",
       "      <td>0.029850</td>\n",
       "      <td>-0.361307</td>\n",
       "      <td>-1.904232</td>\n",
       "      <td>25.0</td>\n",
       "      <td>17</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>KEGG-Pathway-pae03430: Mismatch repair - Pseud...</td>\n",
       "      <td>0.013264</td>\n",
       "      <td>0.037993</td>\n",
       "      <td>0.414025</td>\n",
       "      <td>1.824144</td>\n",
       "      <td>98.0</td>\n",
       "      <td>17</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>KEGG-Pathway-pae00740: Riboflavin metabolism -...</td>\n",
       "      <td>0.013631</td>\n",
       "      <td>0.038394</td>\n",
       "      <td>0.554303</td>\n",
       "      <td>1.786102</td>\n",
       "      <td>90.0</td>\n",
       "      <td>8</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>KEGG-Pathway-pae00440: Phosphonate and phosphi...</td>\n",
       "      <td>0.015482</td>\n",
       "      <td>0.042200</td>\n",
       "      <td>-0.457069</td>\n",
       "      <td>-1.834462</td>\n",
       "      <td>48.0</td>\n",
       "      <td>10</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>KEGG-Module-M00083: Fatty acid biosynthesis, e...</td>\n",
       "      <td>0.015281</td>\n",
       "      <td>0.042200</td>\n",
       "      <td>0.419523</td>\n",
       "      <td>1.805267</td>\n",
       "      <td>112.0</td>\n",
       "      <td>16</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KEGG-Pathway-pae00900: Terpenoid backbone bios...</td>\n",
       "      <td>0.017251</td>\n",
       "      <td>0.046276</td>\n",
       "      <td>0.424001</td>\n",
       "      <td>1.781474</td>\n",
       "      <td>125.0</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>KEGG-Pathway-pae00627: Aminobenzoate degradati...</td>\n",
       "      <td>0.018883</td>\n",
       "      <td>0.047291</td>\n",
       "      <td>-0.327231</td>\n",
       "      <td>-1.773231</td>\n",
       "      <td>46.0</td>\n",
       "      <td>18</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>KEGG-Pathway-pae00640: Propanoate metabolism -...</td>\n",
       "      <td>0.018931</td>\n",
       "      <td>0.047291</td>\n",
       "      <td>0.296915</td>\n",
       "      <td>1.762832</td>\n",
       "      <td>158.0</td>\n",
       "      <td>38</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>KEGG-Module-M00176: Assimilatory sulfate reduc...</td>\n",
       "      <td>0.019028</td>\n",
       "      <td>0.047291</td>\n",
       "      <td>0.603242</td>\n",
       "      <td>1.736611</td>\n",
       "      <td>122.0</td>\n",
       "      <td>6</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>KEGG-Module-M00016: Lysine biosynthesis, succi...</td>\n",
       "      <td>0.017917</td>\n",
       "      <td>0.047291</td>\n",
       "      <td>0.448249</td>\n",
       "      <td>1.770352</td>\n",
       "      <td>127.0</td>\n",
       "      <td>13</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>KEGG-Pathway-pae00061: Fatty acid biosynthesis...</td>\n",
       "      <td>0.018985</td>\n",
       "      <td>0.047291</td>\n",
       "      <td>0.364891</td>\n",
       "      <td>1.767606</td>\n",
       "      <td>146.0</td>\n",
       "      <td>22</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>KEGG-Pathway-pae00460: Cyanoamino acid metabol...</td>\n",
       "      <td>0.020038</td>\n",
       "      <td>0.049079</td>\n",
       "      <td>0.488604</td>\n",
       "      <td>1.724415</td>\n",
       "      <td>136.0</td>\n",
       "      <td>10</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               pathway      pval      padj  \\\n",
       "168             KEGG-Module-M00178: Ribosome, bacteria  0.000113  0.001228   \n",
       "113         KEGG-Module-M00335: Sec (secretion) system  0.000146  0.001228   \n",
       "70   KEGG-Pathway-pae00250: Alanine, aspartate and ...  0.000122  0.001228   \n",
       "68   KEGG-Pathway-pae00400: Phenylalanine, tyrosine...  0.000125  0.001228   \n",
       "61   KEGG-Pathway-pae00970: Aminoacyl-tRNA biosynth...  0.000126  0.001228   \n",
       "59   KEGG-Pathway-pae00240: Pyrimidine metabolism -...  0.000117  0.001228   \n",
       "50   KEGG-Pathway-pae03010: Ribosome - Pseudomonas ...  0.000113  0.001228   \n",
       "48   KEGG-Pathway-pae00330: Arginine and proline me...  0.000109  0.001228   \n",
       "145  KEGG-Module-M00009: Citrate cycle (TCA cycle, ...  0.000129  0.001228   \n",
       "46   KEGG-Pathway-pae00230: Purine metabolism - Pse...  0.000108  0.001228   \n",
       "151  KEGG-Module-M00011: Citrate cycle, second carb...  0.000135  0.001228   \n",
       "40   KEGG-Pathway-pae01210: 2-Oxocarboxylic acid me...  0.000123  0.001228   \n",
       "49   KEGG-Pathway-pae03018: RNA degradation - Pseud...  0.000139  0.001228   \n",
       "37   KEGG-Pathway-pae00020: Citrate cycle (TCA cycl...  0.000122  0.001228   \n",
       "162  KEGG-Module-M00360: Aminoacyl-tRNA biosynthesi...  0.000130  0.001228   \n",
       "158  KEGG-Module-M00144: NADH:quinone oxidoreductas...  0.000140  0.001228   \n",
       "38   KEGG-Pathway-pae03060: Protein export - Pseudo...  0.000135  0.001228   \n",
       "24   KEGG-Pathway-pae00270: Cysteine and methionine...  0.000122  0.001228   \n",
       "27   KEGG-Pathway-pae00190: Oxidative phosphorylati...  0.000114  0.001228   \n",
       "155  KEGG-Module-M00157: F-type ATPase, prokaryotes...  0.000153  0.001228   \n",
       "154  KEGG-Module-M00156: Cytochrome c oxidase, cbb3...  0.000153  0.001228   \n",
       "91   KEGG-Pathway-pae00010: Glycolysis / Gluconeoge...  0.000242  0.001788   \n",
       "15   KEGG-Pathway-pae02040: Flagellar assembly - Ps...  0.000243  0.001788   \n",
       "142  KEGG-Module-M00002: Glycolysis, core module in...  0.000458  0.003223   \n",
       "90   KEGG-Pathway-pae00540: Lipopolysaccharide bios...  0.000795  0.005373   \n",
       "51   KEGG-Pathway-pae00620: Pyruvate metabolism - P...  0.000892  0.005620   \n",
       "74   KEGG-Pathway-pae00680: Methane metabolism - Ps...  0.000900  0.005620   \n",
       "75   KEGG-Pathway-pae01220: Degradation of aromatic...  0.000964  0.005620   \n",
       "152  KEGG-Module-M00010: Citrate cycle, first carbo...  0.000947  0.005620   \n",
       "33   KEGG-Pathway-pae00450: Selenocompound metaboli...  0.001033  0.005716   \n",
       "..                                                 ...       ...       ...   \n",
       "79   KEGG-Pathway-pae00550: Peptidoglycan biosynthe...  0.003312  0.013992   \n",
       "111  KEGG-Module-M00115: NAD biosynthesis, aspartat...  0.003422  0.014107   \n",
       "102    KEGG-Module-M00616: Sulfate-sulfur assimilation  0.003737  0.015038   \n",
       "98   KEGG-Module-M00048: Inosine monophosphate bios...  0.003958  0.015556   \n",
       "130     KEGG-Module-M00034: Methionine salvage pathway  0.004600  0.017422   \n",
       "106  KEGG-Module-M00452: CusS-CusR (copper toleranc...  0.004639  0.017422   \n",
       "110  KEGG-Module-M00050: Guanine ribonucleotide bio...  0.005680  0.020868   \n",
       "43   KEGG-Pathway-pae00770: Pantothenate and CoA bi...  0.005884  0.021157   \n",
       "115  KEGG-Module-M00331: Type II general secretion ...  0.006568  0.022855   \n",
       "150  KEGG-Module-M00018: Threonine biosynthesis, as...  0.006627  0.022855   \n",
       "47   KEGG-Pathway-pae00030: Pentose phosphate pathw...  0.006814  0.023032   \n",
       "23   KEGG-Pathway-pae03440: Homologous recombinatio...  0.006974  0.023110   \n",
       "95   KEGG-Module-M00210: Putative ABC transport system  0.007731  0.025126   \n",
       "104  KEGG-Module-M00188: NitT/TauT family transport...  0.008880  0.028315   \n",
       "143  KEGG-Module-M00119: Pantothenate biosynthesis,...  0.009282  0.029050   \n",
       "28   KEGG-Pathway-pae00521: Streptomycin biosynthes...  0.009765  0.029846   \n",
       "9    KEGG-Pathway-pae00364: Fluorobenzoate degradat...  0.009890  0.029846   \n",
       "149  KEGG-Module-M00019: Valine/isoleucine biosynth...  0.010186  0.029850   \n",
       "42   KEGG-Pathway-pae00500: Starch and sucrose meta...  0.010244  0.029850   \n",
       "80   KEGG-Pathway-pae03430: Mismatch repair - Pseud...  0.013264  0.037993   \n",
       "54   KEGG-Pathway-pae00740: Riboflavin metabolism -...  0.013631  0.038394   \n",
       "18   KEGG-Pathway-pae00440: Phosphonate and phosphi...  0.015482  0.042200   \n",
       "128  KEGG-Module-M00083: Fatty acid biosynthesis, e...  0.015281  0.042200   \n",
       "4    KEGG-Pathway-pae00900: Terpenoid backbone bios...  0.017251  0.046276   \n",
       "52   KEGG-Pathway-pae00627: Aminobenzoate degradati...  0.018883  0.047291   \n",
       "76   KEGG-Pathway-pae00640: Propanoate metabolism -...  0.018931  0.047291   \n",
       "165  KEGG-Module-M00176: Assimilatory sulfate reduc...  0.019028  0.047291   \n",
       "153  KEGG-Module-M00016: Lysine biosynthesis, succi...  0.017917  0.047291   \n",
       "83   KEGG-Pathway-pae00061: Fatty acid biosynthesis...  0.018985  0.047291   \n",
       "88   KEGG-Pathway-pae00460: Cyanoamino acid metabol...  0.020038  0.049079   \n",
       "\n",
       "           ES       NES  nMoreExtreme  size leadingEdge  \n",
       "168  0.900619  6.085628           0.0    55        None  \n",
       "113  0.730965  2.579776           0.0    10        None  \n",
       "70   0.460515  2.615437           0.0    34        None  \n",
       "68   0.483272  2.557964           0.0    28        None  \n",
       "61   0.638701  3.293676           0.0    26        None  \n",
       "59   0.513021  3.180929           0.0    43        None  \n",
       "50   0.900619  6.085628           0.0    55        None  \n",
       "48   0.350134  2.550105           0.0    69        None  \n",
       "145  0.655828  3.176966           0.0    22        None  \n",
       "46   0.405511  3.040558           0.0    76        None  \n",
       "151  0.642079  2.762956           0.0    16        None  \n",
       "40   0.470734  2.587238           0.0    31        None  \n",
       "49   0.691998  2.818566           0.0    14        None  \n",
       "37   0.544917  3.064743           0.0    33        None  \n",
       "162  0.690190  3.282871           0.0    21        None  \n",
       "158  0.655344  2.588272           0.0    13        None  \n",
       "38   0.666837  2.869491           0.0    16        None  \n",
       "24   0.465093  2.588585           0.0    32        None  \n",
       "27   0.567832  3.787781           0.0    53        None  \n",
       "155  0.933959  2.852426           0.0     7        None  \n",
       "154  0.823172  2.514070           0.0     7        None  \n",
       "91   0.413157  2.375335           1.0    35        None  \n",
       "15   0.425094  2.414266           1.0    34        None  \n",
       "142  0.751750  2.295939           2.0     7        None  \n",
       "90   0.483895  2.211730           5.0    19        None  \n",
       "51   0.325672  2.241867           7.0    58        None  \n",
       "74   0.446158  2.225449           6.0    24        None  \n",
       "75  -0.345706 -2.264434           1.0    26        None  \n",
       "152  0.779454  2.079944           5.0     5        None  \n",
       "33   0.615162  2.080158           6.0     9        None  \n",
       "..        ...       ...           ...   ...         ...  \n",
       "79   0.447560  2.045655          24.0    19        None  \n",
       "111  0.791704  1.920905          20.0     4        None  \n",
       "102  0.537725  1.978550          25.0    11        None  \n",
       "98   0.534604  2.043684          27.0    12        None  \n",
       "130  0.529611  1.948695          31.0    11        None  \n",
       "106 -0.617106 -2.073694          15.0     7        None  \n",
       "110  0.706300  1.884735          35.0     5        None  \n",
       "43   0.418544  1.953852          44.0    20        None  \n",
       "115 -0.275396 -2.006574          11.0    32        None  \n",
       "150  0.700036  1.868021          41.0     5        None  \n",
       "47   0.383921  1.915012          52.0    24        None  \n",
       "23   0.397155  1.923901          53.0    22        None  \n",
       "95   0.691739  1.845880          48.0     5        None  \n",
       "104 -0.441202 -1.943573          25.0    12        None  \n",
       "143  0.635857  1.830504          59.0     6        None  \n",
       "28   0.592833  1.810585          63.0     7        None  \n",
       "9   -0.616623 -1.918883          34.0     6        None  \n",
       "149  0.566841  1.826503          67.0     8        None  \n",
       "42  -0.361307 -1.904232          25.0    17        None  \n",
       "80   0.414025  1.824144          98.0    17        None  \n",
       "54   0.554303  1.786102          90.0     8        None  \n",
       "18  -0.457069 -1.834462          48.0    10        None  \n",
       "128  0.419523  1.805267         112.0    16        None  \n",
       "4    0.424001  1.781474         125.0    15        None  \n",
       "52  -0.327231 -1.773231          46.0    18        None  \n",
       "76   0.296915  1.762832         158.0    38        None  \n",
       "165  0.603242  1.736611         122.0     6        None  \n",
       "153  0.448249  1.770352         127.0    13        None  \n",
       "83   0.364891  1.767606         146.0    22        None  \n",
       "88   0.488604  1.724415         136.0    10        None  \n",
       "\n",
       "[69 rows x 8 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(template_enriched_pathways.shape)\n",
    "template_enriched_pathways[template_enriched_pathways['padj'] < 0.05].sort_values(by='padj')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create \"<local_dir>/GSEA_stats/\" subdirectory\n",
    "os.makedirs(os.path.join(local_dir, \"GSEA_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i project_id -i local_dir -i adage_kegg_DB_processed_filename -i num_runs -i statistic\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/GSEA_analysis.R'))\n",
    "\n",
    "# New files created: \"<local_dir>/GSEA_stats/GSEA_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)) {\n",
    "    simulated_DE_stats_file <- paste(local_dir, \n",
    "                                     \"DE_stats/DE_stats_simulated_data_\", \n",
    "                                     project_id,\n",
    "                                     \"_\", \n",
    "                                     i,\n",
    "                                     \".txt\",\n",
    "                                     sep = \"\")\n",
    "    \n",
    "    out_file <- paste(local_dir, \n",
    "                     \"GSEA_stats/GSEA_stats_simulated_data_\",\n",
    "                     project_id,\n",
    "                     \"_\",\n",
    "                     i,\n",
    "                     \".txt\", \n",
    "                     sep = \"\")\n",
    "    \n",
    "    enriched_pathways <- find_enriched_pathways(simulated_DE_stats_file, adage_kegg_DB_processed_filename, statistic) \n",
    "    \n",
    "    # Remove column with leading edge since its causing parsing issues\n",
    "    write.table(as.data.frame(enriched_pathways[1:7]), file = out_file, row.names = F, sep = \"\\t\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rank pathways "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4225, 6)\n"
     ]
    }
   ],
   "source": [
    "# Concatenate simulated experiments\n",
    "simulated_GSEA_stats_all = process.concat_simulated_data(local_dir, num_runs, project_id, 'GSEA')\n",
    "simulated_GSEA_stats_all.set_index('pathway', inplace=True)\n",
    "print(simulated_GSEA_stats_all.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>median</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathway</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00002: Glycolysis, core module involving three-carbon compounds</th>\n",
       "      <td>0.003091</td>\n",
       "      <td>0.008388</td>\n",
       "      <td>0.011080</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00008: Entner-Doudoroff pathway, glucose-6P =&gt; glyceraldehyde-3P + pyruvate</th>\n",
       "      <td>0.049850</td>\n",
       "      <td>0.076868</td>\n",
       "      <td>0.074792</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00009: Citrate cycle (TCA cycle, Krebs cycle)</th>\n",
       "      <td>0.001348</td>\n",
       "      <td>0.001369</td>\n",
       "      <td>0.000370</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00010: Citrate cycle, first carbon oxidation, oxaloacetate =&gt; 2-oxoglutarate</th>\n",
       "      <td>0.001865</td>\n",
       "      <td>0.003223</td>\n",
       "      <td>0.002802</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00011: Citrate cycle, second carbon oxidation, 2-oxoglutarate =&gt; oxaloacetate</th>\n",
       "      <td>0.001348</td>\n",
       "      <td>0.001434</td>\n",
       "      <td>0.000546</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        padj            \\\n",
       "                                                      median      mean   \n",
       "pathway                                                                  \n",
       "KEGG-Module-M00002: Glycolysis, core module inv...  0.003091  0.008388   \n",
       "KEGG-Module-M00008: Entner-Doudoroff pathway, g...  0.049850  0.076868   \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K...  0.001348  0.001369   \n",
       "KEGG-Module-M00010: Citrate cycle, first carbon...  0.001865  0.003223   \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo...  0.001348  0.001434   \n",
       "\n",
       "                                                                    \n",
       "                                                         std count  \n",
       "pathway                                                             \n",
       "KEGG-Module-M00002: Glycolysis, core module inv...  0.011080    25  \n",
       "KEGG-Module-M00008: Entner-Doudoroff pathway, g...  0.074792    25  \n",
       "KEGG-Module-M00009: Citrate cycle (TCA cycle, K...  0.000370    25  \n",
       "KEGG-Module-M00010: Citrate cycle, first carbon...  0.002802    25  \n",
       "KEGG-Module-M00011: Citrate cycle, second carbo...  0.000546    25  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aggregate statistics across all simulated experiments\n",
    "simulated_GSEA_summary_stats = calc.aggregate_stats(\n",
    "    col_to_rank_pathways,\n",
    "    simulated_GSEA_stats_all,\n",
    "    'GSEA'\n",
    ")\n",
    "\n",
    "simulated_GSEA_summary_stats.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load association statistics for template experiment\n",
    "template_GSEA_stats = template_enriched_pathways.iloc[:, :-1]\n",
    "template_GSEA_stats.set_index('pathway', inplace=True)\n",
    "\n",
    "template_GSEA_stats.head()\n",
    "\n",
    "# Rank genes in template experiment\n",
    "template_GSEA_stats = calc.rank_genes_or_pathways(\n",
    "    col_to_rank_pathways,\n",
    "    template_GSEA_stats,\n",
    "    True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rank genes in simulated experiments\n",
    "simulated_GSEA_summary_stats = calc.rank_genes_or_pathways(\n",
    "    col_to_rank_pathways,\n",
    "    simulated_GSEA_summary_stats,\n",
    "    False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pathway summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(169, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>Test statistic (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean test statistic (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>abs(Z score)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathway</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00178: Ribosome, bacteria</th>\n",
       "      <td>KEGG-Module-M00178: Ribosome, bacteria</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>162.5</td>\n",
       "      <td>0.001369</td>\n",
       "      <td>0.000370</td>\n",
       "      <td>25</td>\n",
       "      <td>0.379764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Pathway-pae03060: Protein export - Pseudomonas aeruginosa PAO1</th>\n",
       "      <td>KEGG-Pathway-pae03060: Protein export - Pseudo...</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>162.5</td>\n",
       "      <td>0.001447</td>\n",
       "      <td>0.000507</td>\n",
       "      <td>25</td>\n",
       "      <td>0.432351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Pathway-pae00020: Citrate cycle (TCA cycle) - Pseudomonas aeruginosa PAO1</th>\n",
       "      <td>KEGG-Pathway-pae00020: Citrate cycle (TCA cycl...</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>162.5</td>\n",
       "      <td>0.001369</td>\n",
       "      <td>0.000370</td>\n",
       "      <td>25</td>\n",
       "      <td>0.379764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00156: Cytochrome c oxidase, cbb3-type</th>\n",
       "      <td>KEGG-Module-M00156: Cytochrome c oxidase, cbb3...</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.003548</td>\n",
       "      <td>143.0</td>\n",
       "      <td>0.059462</td>\n",
       "      <td>0.173213</td>\n",
       "      <td>25</td>\n",
       "      <td>0.336202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG-Module-M00157: F-type ATPase, prokaryotes and chloroplasts</th>\n",
       "      <td>KEGG-Module-M00157: F-type ATPase, prokaryotes...</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>162.5</td>\n",
       "      <td>0.001369</td>\n",
       "      <td>0.000371</td>\n",
       "      <td>25</td>\n",
       "      <td>0.381636</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                              Gene ID  \\\n",
       "pathway                                                                                                 \n",
       "KEGG-Module-M00178: Ribosome, bacteria                         KEGG-Module-M00178: Ribosome, bacteria   \n",
       "KEGG-Pathway-pae03060: Protein export - Pseudom...  KEGG-Pathway-pae03060: Protein export - Pseudo...   \n",
       "KEGG-Pathway-pae00020: Citrate cycle (TCA cycle...  KEGG-Pathway-pae00020: Citrate cycle (TCA cycl...   \n",
       "KEGG-Module-M00156: Cytochrome c oxidase, cbb3-...  KEGG-Module-M00156: Cytochrome c oxidase, cbb3...   \n",
       "KEGG-Module-M00157: F-type ATPase, prokaryotes ...  KEGG-Module-M00157: F-type ATPase, prokaryotes...   \n",
       "\n",
       "                                                    Adj P-value (Real)  \\\n",
       "pathway                                                                  \n",
       "KEGG-Module-M00178: Ribosome, bacteria                        0.001228   \n",
       "KEGG-Pathway-pae03060: Protein export - Pseudom...            0.001228   \n",
       "KEGG-Pathway-pae00020: Citrate cycle (TCA cycle...            0.001228   \n",
       "KEGG-Module-M00156: Cytochrome c oxidase, cbb3-...            0.001228   \n",
       "KEGG-Module-M00157: F-type ATPase, prokaryotes ...            0.001228   \n",
       "\n",
       "                                                    Rank (Real)  \\\n",
       "pathway                                                           \n",
       "KEGG-Module-M00178: Ribosome, bacteria                    159.0   \n",
       "KEGG-Pathway-pae03060: Protein export - Pseudom...        159.0   \n",
       "KEGG-Pathway-pae00020: Citrate cycle (TCA cycle...        159.0   \n",
       "KEGG-Module-M00156: Cytochrome c oxidase, cbb3-...        159.0   \n",
       "KEGG-Module-M00157: F-type ATPase, prokaryotes ...        159.0   \n",
       "\n",
       "                                                    Test statistic (Real)  \\\n",
       "pathway                                                                     \n",
       "KEGG-Module-M00178: Ribosome, bacteria                           0.001228   \n",
       "KEGG-Pathway-pae03060: Protein export - Pseudom...               0.001228   \n",
       "KEGG-Pathway-pae00020: Citrate cycle (TCA cycle...               0.001228   \n",
       "KEGG-Module-M00156: Cytochrome c oxidase, cbb3-...               0.001228   \n",
       "KEGG-Module-M00157: F-type ATPase, prokaryotes ...               0.001228   \n",
       "\n",
       "                                                    Median adj p-value (simulated)  \\\n",
       "pathway                                                                              \n",
       "KEGG-Module-M00178: Ribosome, bacteria                                    0.001348   \n",
       "KEGG-Pathway-pae03060: Protein export - Pseudom...                        0.001348   \n",
       "KEGG-Pathway-pae00020: Citrate cycle (TCA cycle...                        0.001348   \n",
       "KEGG-Module-M00156: Cytochrome c oxidase, cbb3-...                        0.003548   \n",
       "KEGG-Module-M00157: F-type ATPase, prokaryotes ...                        0.001348   \n",
       "\n",
       "                                                    Rank (simulated)  \\\n",
       "pathway                                                                \n",
       "KEGG-Module-M00178: Ribosome, bacteria                         162.5   \n",
       "KEGG-Pathway-pae03060: Protein export - Pseudom...             162.5   \n",
       "KEGG-Pathway-pae00020: Citrate cycle (TCA cycle...             162.5   \n",
       "KEGG-Module-M00156: Cytochrome c oxidase, cbb3-...             143.0   \n",
       "KEGG-Module-M00157: F-type ATPase, prokaryotes ...             162.5   \n",
       "\n",
       "                                                    Mean test statistic (simulated)  \\\n",
       "pathway                                                                               \n",
       "KEGG-Module-M00178: Ribosome, bacteria                                     0.001369   \n",
       "KEGG-Pathway-pae03060: Protein export - Pseudom...                         0.001447   \n",
       "KEGG-Pathway-pae00020: Citrate cycle (TCA cycle...                         0.001369   \n",
       "KEGG-Module-M00156: Cytochrome c oxidase, cbb3-...                         0.059462   \n",
       "KEGG-Module-M00157: F-type ATPase, prokaryotes ...                         0.001369   \n",
       "\n",
       "                                                    Std deviation (simulated)  \\\n",
       "pathway                                                                         \n",
       "KEGG-Module-M00178: Ribosome, bacteria                               0.000370   \n",
       "KEGG-Pathway-pae03060: Protein export - Pseudom...                   0.000507   \n",
       "KEGG-Pathway-pae00020: Citrate cycle (TCA cycle...                   0.000370   \n",
       "KEGG-Module-M00156: Cytochrome c oxidase, cbb3-...                   0.173213   \n",
       "KEGG-Module-M00157: F-type ATPase, prokaryotes ...                   0.000371   \n",
       "\n",
       "                                                    Number of experiments (simulated)  \\\n",
       "pathway                                                                                 \n",
       "KEGG-Module-M00178: Ribosome, bacteria                                             25   \n",
       "KEGG-Pathway-pae03060: Protein export - Pseudom...                                 25   \n",
       "KEGG-Pathway-pae00020: Citrate cycle (TCA cycle...                                 25   \n",
       "KEGG-Module-M00156: Cytochrome c oxidase, cbb3-...                                 25   \n",
       "KEGG-Module-M00157: F-type ATPase, prokaryotes ...                                 25   \n",
       "\n",
       "                                                    abs(Z score)  \n",
       "pathway                                                           \n",
       "KEGG-Module-M00178: Ribosome, bacteria                  0.379764  \n",
       "KEGG-Pathway-pae03060: Protein export - Pseudom...      0.432351  \n",
       "KEGG-Pathway-pae00020: Citrate cycle (TCA cycle...      0.379764  \n",
       "KEGG-Module-M00156: Cytochrome c oxidase, cbb3-...      0.336202  \n",
       "KEGG-Module-M00157: F-type ATPase, prokaryotes ...      0.381636  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create intermediate file: \"<local_dir>/gene_summary_table_<col_to_rank_pathways>.tsv\"\n",
    "summary_pathway_ranks = process.generate_summary_table(\n",
    "    template_GSEA_stats,\n",
    "    simulated_GSEA_summary_stats,\n",
    "    col_to_rank_pathways,\n",
    "    local_dir\n",
    ")\n",
    "\n",
    "summary_pathway_ranks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create `pathway_summary_filename`\n",
    "summary_pathway_ranks.to_csv(pathway_summary_filename, sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
