{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify generic genes and pathways\n",
    "\n",
    "This notebook allows users to find common and specific genes in their experiment of interest using their newly trained VAE model.\n",
    "\n",
    "This notebook will generate a `generic_gene_summary_<experiment id>.tsv` file that contains z-scores per gene that indicates how specific a gene is the experiment in question."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "import scipy.stats as ss\n",
    "from keras.models import load_model\n",
    "from rpy2.robjects import pandas2ri\n",
    "from ponyo import utils, simulate_expression_data\n",
    "from generic_expression_patterns_modules import process, stats, ranking\n",
    "\n",
    "pandas2ri.activate()\n",
    "\n",
    "np.random.seed(123)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User inputs needed\n",
    "\n",
    "User needs to define the following in the [config file](../configs/config_new_model_experiment.tsv):\n",
    "\n",
    "1. Directory on your local machine to store intermediate and output data files generated (`local_dir`). Make sure to end with `\\`.\n",
    "2. Template experiment (`raw_template_filename`). This is the experiment you are interested in studying.\n",
    "3. Training compendium used to train VAE, including unnormalized gene mapped version (`mapped_compendium_filename`) and normalized version (`normalized_compendium_filename`). These files should have been generated from the previous notebook.\n",
    "4. Scaler transform (`scaler_filename`) used to normalize the training compendium. This can be found in the `data/` directory within the analysis folder.\n",
    "5. Directory (`vae_model_dir`) containing trained VAE model (.h5 files) from the previous notebook.\n",
    "6. Size of latent dimension (`latent_dim`).\n",
    "7. File that maps experiment ids to the associated sample ids (`experiment_to_sample_filename`)\n",
    "8. The delimiter used in the 'experiment_to_sample_filename' file (`metadata_delimiter`)\n",
    "9. The column header/name that contains the experiment ids (`experiment_id_colname`)\n",
    "10. Experiment id (`project_id`) to label newly create simulated experiments.\n",
    "11. The column header/name in the metadatathat contains the sample ids (`sample_id_colname`)\n",
    "12. The number of experiments to simulate (`num_simulated`)\n",
    "13. Differential expression method to use: either 'limma' or 'desesq' (`DE_method`)\n",
    "14. Minimum average read count to filter data by (`count_threshold`)\n",
    "15. Name of column header (`rank_genes_by`) from DE association statistic results. This column will be use to rank genes. Select \"logFC\", \"P.Value\", \"adj.P.Val\", \"t\" if using Limma. Select \"log2FoldChange\", \"pvalue\", \"padj\" if using DESeq.\n",
    "16. `DE_logFC_name` is either \"logFC\" (Limma) or \"log2FoldChange\" (DESeq). This is used for plotting volcano plots.\n",
    "17. `DE_pvalue_name` is either \"adj.P.Val\" (Limma) or \"padj\" (DESeq). This is used for plotting volcano plots.\n",
    "\n",
    "The remaining parameters within the `config` file specify filenames that are intermediate data files that will be generated when SOPHIE runs.\n",
    "\n",
    "The user also needs to provide metadata files:\n",
    "\n",
    "These files should be located in `data/metadata/` directory.\n",
    "1. `<experiment id>_process_samples.tsv` contains 2 columns (sample ids, label that indicates if the sample is kept or removed). See [example](data/metadata/cis-gem-par-KU1919_process_samples.tsv). **Note: This file is not required if the user wishes to use all the samples in the template experiment file.**\n",
    "2. `<experiment id>_groups.tsv` contains 2 columns: sample ids, group label to perform DE analysis. See [example](data/metadata/cis-gem-par-KU1919_groups.tsv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = os.path.abspath(\n",
    "    os.path.join(base_dir, \"configs\", \"config_new_model_experiment.tsv\")\n",
    ")\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load params\n",
    "local_dir = params[\"local_dir\"]\n",
    "dataset_name = params[\"dataset_name\"]\n",
    "NN_architecture = params[\"NN_architecture\"]\n",
    "latent_dim = params[\"latent_dim\"]\n",
    "num_runs = params[\"num_simulated\"]\n",
    "metadata_simulate_filename = params[\"experiment_to_sample_filename\"]\n",
    "metadata_delimiter = params[\"metadata_delimiter\"]\n",
    "experiment_id_colname = params[\"experiment_id_colname\"]\n",
    "project_id = params[\"project_id\"]\n",
    "metadata_col_id = params[\"sample_id_colname\"]\n",
    "raw_template_filename = params[\"raw_template_filename\"]\n",
    "processed_template_filename = params[\"processed_template_filename\"]\n",
    "normalized_compendium_filename = params[\"normalized_compendium_filename\"]\n",
    "scaler_filename = params[\"scaler_filename\"]\n",
    "method = params[\"DE_method\"]\n",
    "col_to_rank_genes = params[\"rank_genes_by\"]\n",
    "logFC_name = params[\"DE_logFC_name\"]\n",
    "pvalue_name = params[\"DE_pvalue_name\"]\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "sample_id_metadata_filename = os.path.join(\n",
    "    base_dir, dataset_name, \"data\", \"metadata\", f\"{project_id}_process_samples.tsv\"\n",
    ")\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "metadata_filename = os.path.join(\n",
    "    base_dir, dataset_name, \"data\", \"metadata\", f\"{project_id}_groups.tsv\"\n",
    ")\n",
    "\n",
    "# Load pickled file\n",
    "scaler = pickle.load(open(scaler_filename, \"rb\"))\n",
    "\n",
    "# Percentile threshold to identify generic genes\n",
    "percentile_threshold = 80.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output files\n",
    "gene_summary_filename = os.path.join(\n",
    "    base_dir, dataset_name, f\"generic_gene_summary_{project_id}.tsv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate experiments using selected template experiment\n",
    "Workflow:\n",
    "\n",
    "1. Get the gene expression data for the selected template experiment\n",
    "2. Encode this experiment into a latent space using the trained VAE model\n",
    "3. Linearly shift the encoded template experiment in the latent space\n",
    "4. Decode the samples. This results in a new experiment\n",
    "5. Repeat steps 1-4 to get multiple simulated experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate multiple experiments\n",
    "# This step creates the following files in \"<local_dir>/pseudo_experiment/\" directory:\n",
    "#   - selected_simulated_data_SRP012656_<n>.txt\n",
    "#   - selected_simulated_encoded_data_SRP012656_<n>.txt\n",
    "#   - template_normalized_data_SRP012656_test.txt\n",
    "# in which \"<n>\" is an integer in the range of [0, num_runs-1]\n",
    "os.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)\n",
    "simulate_expression_data.shift_template_experiment(\n",
    "    normalized_compendium_filename,\n",
    "    NN_architecture,\n",
    "    latent_dim,\n",
    "    dataset_name,\n",
    "    scaler,\n",
    "    metadata_simulate_filename,\n",
    "    metadata_delimiter,\n",
    "    experiment_id_colname,\n",
    "    metadata_col_id,\n",
    "    project_id,\n",
    "    local_dir,\n",
    "    base_dir,\n",
    "    num_runs,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "simulated_filename = os.path.join(\n",
    "    local_dir, \"pseudo_experiment\", f\"selected_simulated_data_{project_id}_1.txt\"\n",
    ")\n",
    "\n",
    "test = pd.read_csv(simulated_filename, sep=\"\\t\", index_col=0, header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PA14_55610</th>\n",
       "      <th>PA14_55600</th>\n",
       "      <th>PA14_55590</th>\n",
       "      <th>PA14_55580</th>\n",
       "      <th>PA14_55570</th>\n",
       "      <th>PA14_55560</th>\n",
       "      <th>PA14_55550</th>\n",
       "      <th>PA14_55540</th>\n",
       "      <th>PA14_55530</th>\n",
       "      <th>PA14_55520</th>\n",
       "      <th>...</th>\n",
       "      <th>PA14_19205</th>\n",
       "      <th>PA14_17675</th>\n",
       "      <th>PA14_67975</th>\n",
       "      <th>PA14_36345</th>\n",
       "      <th>PA14_43405</th>\n",
       "      <th>PA14_38825</th>\n",
       "      <th>PA14_24245</th>\n",
       "      <th>PA14_28895</th>\n",
       "      <th>PA14_55117</th>\n",
       "      <th>PA14_59845</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRX1740204</th>\n",
       "      <td>111.777</td>\n",
       "      <td>50.537</td>\n",
       "      <td>10.080</td>\n",
       "      <td>19.057</td>\n",
       "      <td>21.094</td>\n",
       "      <td>19.976</td>\n",
       "      <td>17.336</td>\n",
       "      <td>18.764</td>\n",
       "      <td>16.352</td>\n",
       "      <td>4.021</td>\n",
       "      <td>...</td>\n",
       "      <td>44.504</td>\n",
       "      <td>162.106</td>\n",
       "      <td>633.776</td>\n",
       "      <td>144.577</td>\n",
       "      <td>0.171</td>\n",
       "      <td>29.336</td>\n",
       "      <td>137.091</td>\n",
       "      <td>683.270</td>\n",
       "      <td>155.982</td>\n",
       "      <td>209.853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740205</th>\n",
       "      <td>103.666</td>\n",
       "      <td>51.088</td>\n",
       "      <td>10.955</td>\n",
       "      <td>26.069</td>\n",
       "      <td>16.420</td>\n",
       "      <td>16.012</td>\n",
       "      <td>12.438</td>\n",
       "      <td>15.369</td>\n",
       "      <td>10.092</td>\n",
       "      <td>2.846</td>\n",
       "      <td>...</td>\n",
       "      <td>41.855</td>\n",
       "      <td>160.967</td>\n",
       "      <td>799.159</td>\n",
       "      <td>103.454</td>\n",
       "      <td>0.135</td>\n",
       "      <td>18.419</td>\n",
       "      <td>116.171</td>\n",
       "      <td>550.327</td>\n",
       "      <td>152.655</td>\n",
       "      <td>87.486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740206</th>\n",
       "      <td>102.711</td>\n",
       "      <td>49.355</td>\n",
       "      <td>9.203</td>\n",
       "      <td>15.811</td>\n",
       "      <td>18.975</td>\n",
       "      <td>19.545</td>\n",
       "      <td>14.986</td>\n",
       "      <td>18.000</td>\n",
       "      <td>18.599</td>\n",
       "      <td>4.126</td>\n",
       "      <td>...</td>\n",
       "      <td>48.051</td>\n",
       "      <td>176.501</td>\n",
       "      <td>683.735</td>\n",
       "      <td>174.970</td>\n",
       "      <td>0.236</td>\n",
       "      <td>43.901</td>\n",
       "      <td>134.203</td>\n",
       "      <td>682.758</td>\n",
       "      <td>140.616</td>\n",
       "      <td>211.644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740207</th>\n",
       "      <td>123.578</td>\n",
       "      <td>25.647</td>\n",
       "      <td>10.035</td>\n",
       "      <td>12.350</td>\n",
       "      <td>37.381</td>\n",
       "      <td>10.266</td>\n",
       "      <td>13.395</td>\n",
       "      <td>21.082</td>\n",
       "      <td>54.470</td>\n",
       "      <td>3.969</td>\n",
       "      <td>...</td>\n",
       "      <td>37.466</td>\n",
       "      <td>64.821</td>\n",
       "      <td>314.995</td>\n",
       "      <td>250.590</td>\n",
       "      <td>0.176</td>\n",
       "      <td>38.048</td>\n",
       "      <td>202.545</td>\n",
       "      <td>432.848</td>\n",
       "      <td>146.042</td>\n",
       "      <td>801.247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740208</th>\n",
       "      <td>127.446</td>\n",
       "      <td>41.818</td>\n",
       "      <td>11.924</td>\n",
       "      <td>28.756</td>\n",
       "      <td>43.823</td>\n",
       "      <td>21.700</td>\n",
       "      <td>17.510</td>\n",
       "      <td>27.530</td>\n",
       "      <td>37.405</td>\n",
       "      <td>6.179</td>\n",
       "      <td>...</td>\n",
       "      <td>44.970</td>\n",
       "      <td>102.648</td>\n",
       "      <td>465.611</td>\n",
       "      <td>257.010</td>\n",
       "      <td>0.473</td>\n",
       "      <td>56.969</td>\n",
       "      <td>189.951</td>\n",
       "      <td>570.901</td>\n",
       "      <td>139.930</td>\n",
       "      <td>400.680</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 5891 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PA14_55610  PA14_55600  PA14_55590  PA14_55580  PA14_55570  \\\n",
       "SRX1740204     111.777      50.537      10.080      19.057      21.094   \n",
       "SRX1740205     103.666      51.088      10.955      26.069      16.420   \n",
       "SRX1740206     102.711      49.355       9.203      15.811      18.975   \n",
       "SRX1740207     123.578      25.647      10.035      12.350      37.381   \n",
       "SRX1740208     127.446      41.818      11.924      28.756      43.823   \n",
       "\n",
       "            PA14_55560  PA14_55550  PA14_55540  PA14_55530  PA14_55520  \\\n",
       "SRX1740204      19.976      17.336      18.764      16.352       4.021   \n",
       "SRX1740205      16.012      12.438      15.369      10.092       2.846   \n",
       "SRX1740206      19.545      14.986      18.000      18.599       4.126   \n",
       "SRX1740207      10.266      13.395      21.082      54.470       3.969   \n",
       "SRX1740208      21.700      17.510      27.530      37.405       6.179   \n",
       "\n",
       "               ...      PA14_19205  PA14_17675  PA14_67975  PA14_36345  \\\n",
       "SRX1740204     ...          44.504     162.106     633.776     144.577   \n",
       "SRX1740205     ...          41.855     160.967     799.159     103.454   \n",
       "SRX1740206     ...          48.051     176.501     683.735     174.970   \n",
       "SRX1740207     ...          37.466      64.821     314.995     250.590   \n",
       "SRX1740208     ...          44.970     102.648     465.611     257.010   \n",
       "\n",
       "            PA14_43405  PA14_38825  PA14_24245  PA14_28895  PA14_55117  \\\n",
       "SRX1740204       0.171      29.336     137.091     683.270     155.982   \n",
       "SRX1740205       0.135      18.419     116.171     550.327     152.655   \n",
       "SRX1740206       0.236      43.901     134.203     682.758     140.616   \n",
       "SRX1740207       0.176      38.048     202.545     432.848     146.042   \n",
       "SRX1740208       0.473      56.969     189.951     570.901     139.930   \n",
       "\n",
       "            PA14_59845  \n",
       "SRX1740204     209.853  \n",
       "SRX1740205      87.486  \n",
       "SRX1740206     211.644  \n",
       "SRX1740207     801.247  \n",
       "SRX1740208     400.680  \n",
       "\n",
       "[5 rows x 5891 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process template and simulated data\n",
    "\n",
    "* Remove samples not required for comparison.\n",
    "* Make sure ordering of samples matches metadata for proper comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(sample_id_metadata_filename):\n",
    "    sample_id_metadata_filename = None\n",
    "\n",
    "if method == \"deseq\":\n",
    "    stats.process_samples_for_DESeq(\n",
    "        mapped_template_filename,\n",
    "        metadata_filename,\n",
    "        processed_template_filename,\n",
    "        count_threshold,\n",
    "        sample_id_metadata_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        out_simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}_processed.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_DESeq(\n",
    "            simulated_filename,\n",
    "            metadata_filename,\n",
    "            out_simulated_filename,\n",
    "            count_threshold,\n",
    "            sample_id_metadata_filename,\n",
    "        )\n",
    "else:\n",
    "    stats.process_samples_for_limma(\n",
    "        mapped_template_filename,\n",
    "        metadata_filename,\n",
    "        processed_template_filename,\n",
    "        sample_id_metadata_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_limma(\n",
    "            simulated_filename,\n",
    "            metadata_filename,\n",
    "            None,\n",
    "            sample_id_metadata_filename,\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i metadata_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        metadata_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        metadata_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i metadata_filename -i project_id -i base_dir -i local_dir -i num_runs -i method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# Files created: \"<local_dir>/DE_stats/DE_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\",\n",
    "        i,\n",
    "        \"_processed.txt\",\n",
    "        sep = \"\"\n",
    "    )\n",
    "    if (method == \"deseq\"){\n",
    "        get_DE_stats_DESeq(\n",
    "            metadata_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "    }\n",
    "    else {\n",
    "        get_DE_stats_limma(\n",
    "            metadata_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rank genes\n",
    "Genes are ranked by their \"generic-ness\" - how frequently these genes are changed across the simulated experiments using user-specific test statistic (i.e. log2 fold change)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (2 levels on the left, 1 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    "    logFC_name,\n",
    "    pvalue_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary table\n",
    "\n",
    "* Gene ID: Gene identifier (hgnc symbols for human data or PA number for *P. aeruginosa* data)\n",
    "* (Real): Statistics for template experiment\n",
    "* (Simulated): Statistics across simulated experiments\n",
    "* Number of experiments: Number of simulated experiments\n",
    "* Z-score: High z-score indicates that gene is more changed in template compared to the null set of simulated experiments (high z-score = highly specific to template experiment)\n",
    "\n",
    "\n",
    "**Note:**\n",
    "* If using DESeq, genes with NaN in only the `Adj P-value (Real)` column are those genes flagged because of the `cooksCutoff` parameter. The cook's distance as a diagnostic to tell if a single sample has a count which has a disproportionate impact on the log fold change and p-values. These genes are flagged with an NA in the pvalue and padj columns of the result table.\n",
    "\n",
    "* If using DESeq with count threshold, some genes may not be present in all simulated experiments (i.e. the `Number of experiments (simulated)` will not equal the number of simulated experiments you specified in the beginning. This pre-filtering will lead to some genes found in few simulated experiments and so the background/null set for that gene is not robust. Thus, the user should sort by both z-score and number of experiments to identify specific expressed genes.\n",
    "\n",
    "* If using DESeq without count threshold, some genes may still not be present in all simulated experiments (i.e. the `Number of experiments (simulated)`  will not equal the number of simulated experiments you specified in the beginning. If the gene is 0 expressed across all samples and thus automatically given an NA in `log fold change, adjusted p-value` columns. Thus, the user should sort by both z-score and number of experiments to identify specific expressed genes.\n",
    "\n",
    "For more information you can read [DESeq FAQs](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA14_53300</th>\n",
       "      <td>PA14_53300</td>\n",
       "      <td>3.277490e-44</td>\n",
       "      <td>5890.0</td>\n",
       "      <td>3.820634</td>\n",
       "      <td>-3.820634</td>\n",
       "      <td>0.615595</td>\n",
       "      <td>4244.0</td>\n",
       "      <td>72.037351</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.249700</td>\n",
       "      <td>0.194925</td>\n",
       "      <td>25</td>\n",
       "      <td>18.319543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_21530</th>\n",
       "      <td>PA14_21530</td>\n",
       "      <td>1.617143e-28</td>\n",
       "      <td>5887.0</td>\n",
       "      <td>3.273077</td>\n",
       "      <td>-3.273077</td>\n",
       "      <td>0.757436</td>\n",
       "      <td>4379.0</td>\n",
       "      <td>74.329372</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.222476</td>\n",
       "      <td>0.172918</td>\n",
       "      <td>25</td>\n",
       "      <td>17.641922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_66480</th>\n",
       "      <td>PA14_66480</td>\n",
       "      <td>5.056304e-07</td>\n",
       "      <td>5700.0</td>\n",
       "      <td>1.299171</td>\n",
       "      <td>1.299171</td>\n",
       "      <td>0.646066</td>\n",
       "      <td>1318.0</td>\n",
       "      <td>22.359932</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.127432</td>\n",
       "      <td>0.066487</td>\n",
       "      <td>25</td>\n",
       "      <td>17.623670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_12590</th>\n",
       "      <td>PA14_12590</td>\n",
       "      <td>2.669583e-01</td>\n",
       "      <td>5791.0</td>\n",
       "      <td>1.517689</td>\n",
       "      <td>1.517689</td>\n",
       "      <td>0.949002</td>\n",
       "      <td>1069.0</td>\n",
       "      <td>18.132428</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.115920</td>\n",
       "      <td>0.080223</td>\n",
       "      <td>25</td>\n",
       "      <td>17.473402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_10500</th>\n",
       "      <td>PA14_10500</td>\n",
       "      <td>2.400867e-30</td>\n",
       "      <td>5884.0</td>\n",
       "      <td>3.173688</td>\n",
       "      <td>-3.173688</td>\n",
       "      <td>0.286293</td>\n",
       "      <td>4426.0</td>\n",
       "      <td>75.127334</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.254018</td>\n",
       "      <td>0.173930</td>\n",
       "      <td>25</td>\n",
       "      <td>16.786449</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "PA14_53300  PA14_53300        3.277490e-44       5890.0   \n",
       "PA14_21530  PA14_21530        1.617143e-28       5887.0   \n",
       "PA14_66480  PA14_66480        5.056304e-07       5700.0   \n",
       "PA14_12590  PA14_12590        2.669583e-01       5791.0   \n",
       "PA14_10500  PA14_10500        2.400867e-30       5884.0   \n",
       "\n",
       "            abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "PA14_53300                    3.820634              -3.820634   \n",
       "PA14_21530                    3.273077              -3.273077   \n",
       "PA14_66480                    1.299171               1.299171   \n",
       "PA14_12590                    1.517689               1.517689   \n",
       "PA14_10500                    3.173688              -3.173688   \n",
       "\n",
       "            Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "PA14_53300                        0.615595            4244.0   \n",
       "PA14_21530                        0.757436            4379.0   \n",
       "PA14_66480                        0.646066            1318.0   \n",
       "PA14_12590                        0.949002            1069.0   \n",
       "PA14_10500                        0.286293            4426.0   \n",
       "\n",
       "            Percentile (simulated)  Percent DE (simulated)  \\\n",
       "PA14_53300               72.037351                     0.0   \n",
       "PA14_21530               74.329372                     0.0   \n",
       "PA14_66480               22.359932                     0.0   \n",
       "PA14_12590               18.132428                     0.0   \n",
       "PA14_10500               75.127334                     0.0   \n",
       "\n",
       "            Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "PA14_53300                              0.249700                   0.194925   \n",
       "PA14_21530                              0.222476                   0.172918   \n",
       "PA14_66480                              0.127432                   0.066487   \n",
       "PA14_12590                              0.115920                   0.080223   \n",
       "PA14_10500                              0.254018                   0.173930   \n",
       "\n",
       "            Number of experiments (simulated)    Z score  \n",
       "PA14_53300                                 25  18.319543  \n",
       "PA14_21530                                 25  17.641922  \n",
       "PA14_66480                                 25  17.623670  \n",
       "PA14_12590                                 25  17.473402  \n",
       "PA14_10500                                 25  16.786449  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    \"gene\",\n",
    "    params,\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene ID                                 False\n",
       "Adj P-value (Real)                       True\n",
       "Rank (Real)                              True\n",
       "abs(log2FoldChange) (Real)               True\n",
       "log2FoldChange (Real)                    True\n",
       "Median adj p-value (simulated)          False\n",
       "Rank (simulated)                        False\n",
       "Percentile (simulated)                  False\n",
       "Percent DE (simulated)                  False\n",
       "Mean abs(log2FoldChange) (simulated)    False\n",
       "Std deviation (simulated)               False\n",
       "Number of experiments (simulated)       False\n",
       "Z score                                  True\n",
       "Gene Name                                True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if there is an NaN values, there should not be\n",
    "summary_gene_ranks.isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create `gene_summary_filename`\n",
    "summary_gene_ranks.to_csv(gene_summary_filename, sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "encoding": "# -*- coding: utf-8 -*-",
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression_new] *",
   "language": "python",
   "name": "conda-env-generic_expression_new-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
