{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apply enrichment method\n",
    "\n",
    "This notebook plugs in other gene set enrichment methods to demonstrate that our method, SOPHIE, can be inserted into different pipelines and work with other methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "\n",
    "from rpy2.robjects import pandas2ri\n",
    "pandas2ri.activate()\n",
    "\n",
    "from ponyo import utils\n",
    "from generic_expression_patterns_modules import ranking\n",
    "\n",
    "np.random.seed(123)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = os.path.abspath(\n",
    "    os.path.join(base_dir, \"configs\", \"config_human_general.tsv\")\n",
    ")\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load params\n",
    "local_dir = params[\"local_dir\"]\n",
    "project_id = params[\"project_id\"]\n",
    "statistic = params[\"gsea_statistic\"]\n",
    "hallmark_DB_filename = params[\"pathway_DB_filename\"]\n",
    "num_runs = params[\"num_simulated\"]\n",
    "dataset_name = params[\"dataset_name\"]\n",
    "\n",
    "# Select enrichment method\n",
    "# enrichment_method = [\"GSEA\", GSVA\", \"ROAST\", \"CAMERA\", \"ORA\"]\n",
    "# If enrichment_method == \"GSEA\" then use \"padj\" to rank\n",
    "# If enrichment_method == \"GSVA\" then use \"ES\" to rank\n",
    "# If enrichment_method == \"ROAST\" or \"CAMERA\" then use \"FDR\" to rank\n",
    "# If using \"ORA\" then use \"p.adjust\" to rank\n",
    "enrichment_method = \"ROAST\"\n",
    "col_to_rank_pathways = params[\"rank_pathways_by\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load DE stats directory\n",
    "DE_stats_dir = os.path.join(local_dir, \"DE_stats\")\n",
    "\n",
    "# Template experiment gene expression\n",
    "template_expression_filename = os.path.join(base_dir, dataset_name, params[\"processed_template_filename\"])\n",
    "\n",
    "# Template experiment DE stats\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    DE_stats_dir,\n",
    "    f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "# Metadata file with sample grouping to define comparison\n",
    "metadata_filename = os.path.join(\n",
    "    base_dir,\n",
    "    dataset_name,\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    f\"{project_id}_groups.tsv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output file\n",
    "pathway_summary_filename = os.path.join(\n",
    "    base_dir, \n",
    "    dataset_name, \n",
    "    f\"generic_pathway_summary_{project_id}_{enrichment_method}.tsv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enrichment methods\n",
    "These methods are important because they permit differential expression questions to be posed in terms of ensembles of genes representing pathways or other biologically interpretable processes. \n",
    "\n",
    "There are two main groups of methods: ‘self-contained’ gene set tests examine a set of genes in their own right without reference to other genes in the genome whereas ‘competitive’ gene set tests compare genes in the test set relative to all other genes. Self-contained tests are of interest for assessing the relevance of an individual biological process to the experiment at hand, whereas the competitive tests focus more on distinguishing the most important biological processes from those that are less important. Competitive tests are overwhelmingly more commonly used in the genomic literature. (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3458527/)\n",
    "\n",
    "* [ROAST](https://pubmed.ncbi.nlm.nih.gov/20610611/) (rotation gene set tests) performs a focused gene set testing, in which interest focuses on a few gene sets as opposed to a large dataset. (available in limma).\n",
    "  * Self contained gene set test\n",
    "  * Instead of permutations they use rotations (i.e. fractional permutation) in order to allow for more complex experimental designs than binary experiments (i.e. time-course, more than 2 groups)\n",
    "  * Also esimates correlation between genes\n",
    "  * Best power to detect modest fold changes with minority of genes changed\n",
    "* [CAMERA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3458527/) (Correlation Adjusted MEan RAnk gene set test) is based on the idea of estimating the variance inflation factor associated with inter-gene correlation, and incorporating this into parametric or rank-based test procedures. (available in limma) \n",
    "  * Competitive gene set test\n",
    "  * Performs the same rank-based test procedure as GSEA, but also estimates the correlation between genes, instead of treating genes as independent\n",
    "  * Recall GSEA: 1) Rank all genes using DE association statistics. 2) An enrichment score (ES) is defined as the maximum distance from the middle of the ranked list. Thus, the enrichment score indicates whether the genes contained in a gene set are clustered towards the beginning or the end of the ranked list (indicating a correlation with change in expression). 3) Estimate the statistical significance of the ES by a phenotypic-based permutation (permute samples assigned to label) test in order to produce a null distribution for the ES (i.e. scores based on permuted phenotype)\n",
    "  * Appropriate for small and large fold changes\n",
    "* [GSVA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3618321/) (Gene Set Variation Analysis) calculates sample-wise gene set enrichment scores as a function of genes inside and outside the gene set. This method is well-suited for assessing gene set variation across a dichotomous phenotype. (biocontuctor package GSVA) \n",
    "  * Competitive gene set test\n",
    "  * Estimates variation of gene set enrichment over the samples independently of any class label\n",
    "* [ORA](https://www.rdocumentation.org/packages/clusterProfiler/versions/3.0.4/topics/enricher) (over-representation analysis) uses the hypergeometric test to determine if there a significant over-representation of pathway in the selected set of DEGs. Here we're using clusterProfiler library but there are multiple options for this analysis. See [slide 6](https://docs.google.com/presentation/d/1t4rK7UiLAeIKIzeRJK-YzspNUfGM-8nuRCcevh2lx34/edit?usp=sharing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create \"<local_dir>/GSEA_stats/\" subdirectory\n",
    "os.makedirs(os.path.join(local_dir, \"GSA_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load pathway data\n",
    "hallmark_DB_filename = params[\"pathway_DB_filename\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Apply enrichment to template experiment**\n",
    "\n",
    "See supplementary tables: https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbz158/5722384"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Rcpp\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘edgeR’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:DEFormats’:\n",
      "\n",
      "    DGEList\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: S4Vectors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: stats4\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocGenerics\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: parallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: IRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomicRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomeInfoDb\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: SummarizedExperiment\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Biobase\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: DelayedArray\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: matrixStats\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocParallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Registered S3 method overwritten by 'enrichplot':\n",
      "  method               from\n",
      "  fortify.enrichResult DOSE\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: clusterProfiler v3.14.0  For help: https://guangchuangyu.github.io/software/clusterProfiler\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "Guangchuang Yu, Li-Gen Wang, Yanyan Han, Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology. 2012, 16(5):284-287.\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:DelayedArray’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py:191: FutureWarning: from_items is deprecated. Please use DataFrame.from_dict(dict(items), ...) instead. DataFrame.from_dict(OrderedDict(items)) may be used to preserve the key order.\n",
      "  res = PandasDataFrame.from_items(items)\n"
     ]
    }
   ],
   "source": [
    "%%R -i base_dir -i local_dir -i project_id -i template_expression_filename -i hallmark_DB_filename -i metadata_filename -i enrichment_method -o template_enriched_pathways\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/other_enrichment_methods.R'))\n",
    "\n",
    "out_filename <- paste(local_dir, \n",
    "                      \"GSA_stats/\",\n",
    "                      enrichment_method,\n",
    "                      \"_stats_template_data_\",\n",
    "                      project_id,\n",
    "                      \"_real.txt\", \n",
    "                      sep = \"\")\n",
    "\n",
    "if (enrichment_method == \"GSVA\"){\n",
    "    \n",
    "    template_enriched_pathways <- find_enriched_pathways_GSVA(\n",
    "        template_expression_filename,\n",
    "        hallmark_DB_filename\n",
    "    )\n",
    "}\n",
    "else if (enrichment_method == \"ROAST\"){\n",
    "    \n",
    "    template_enriched_pathways <- find_enriched_pathways_ROAST(\n",
    "        template_expression_filename,\n",
    "        metadata_filename,\n",
    "        hallmark_DB_filename\n",
    "    )\n",
    "}\n",
    "else if (enrichment_method == \"CAMERA\"){\n",
    "    \n",
    "    template_enriched_pathways <- find_enriched_pathways_CAMERA(\n",
    "        template_expression_filename,\n",
    "        metadata_filename,\n",
    "        hallmark_DB_filename\n",
    "    )\n",
    "}\n",
    "else if (enrichment_method == \"ORA\"){\n",
    "    \n",
    "    template_enriched_pathways <- find_enriched_pathways_ORA(\n",
    "        template_expression_filename,\n",
    "        metadata_filename, \n",
    "        hallmark_DB_filename\n",
    "    )\n",
    "}\n",
    "write.table(as.data.frame(template_enriched_pathways), file = out_filename, row.names = F, sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(50, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NGenes</th>\n",
       "      <th>PropDown</th>\n",
       "      <th>PropUp</th>\n",
       "      <th>Direction</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>PValue.Mixed</th>\n",
       "      <th>FDR.Mixed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>193</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>189</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>181</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>180</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>138</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>89</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>55</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>54</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>32</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>185</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.994595</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>184</td>\n",
       "      <td>0.005435</td>\n",
       "      <td>0.994565</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>137</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.992701</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>104</td>\n",
       "      <td>0.009615</td>\n",
       "      <td>0.990385</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>93</td>\n",
       "      <td>0.010753</td>\n",
       "      <td>0.989247</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>185</td>\n",
       "      <td>0.005405</td>\n",
       "      <td>0.983784</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>197</td>\n",
       "      <td>0.015228</td>\n",
       "      <td>0.979695</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>193</td>\n",
       "      <td>0.005181</td>\n",
       "      <td>0.979275</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>188</td>\n",
       "      <td>0.015957</td>\n",
       "      <td>0.978723</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>94</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>0.978723</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>45</td>\n",
       "      <td>0.022222</td>\n",
       "      <td>0.977778</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>187</td>\n",
       "      <td>0.021390</td>\n",
       "      <td>0.973262</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>67</td>\n",
       "      <td>0.014925</td>\n",
       "      <td>0.970149</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>191</td>\n",
       "      <td>0.020942</td>\n",
       "      <td>0.963351</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>156</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.961538</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>193</td>\n",
       "      <td>0.025907</td>\n",
       "      <td>0.958549</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>196</td>\n",
       "      <td>0.040816</td>\n",
       "      <td>0.943878</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>35</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.942857</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>33</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.939394</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>148</td>\n",
       "      <td>0.047297</td>\n",
       "      <td>0.932432</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>103</td>\n",
       "      <td>0.067961</td>\n",
       "      <td>0.932039</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>85</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.929412</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>42</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>150</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.926667</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>188</td>\n",
       "      <td>0.047872</td>\n",
       "      <td>0.925532</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>191</td>\n",
       "      <td>0.057592</td>\n",
       "      <td>0.921466</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>192</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>99</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>43</td>\n",
       "      <td>0.069767</td>\n",
       "      <td>0.906977</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>197</td>\n",
       "      <td>0.055838</td>\n",
       "      <td>0.903553</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>195</td>\n",
       "      <td>0.071795</td>\n",
       "      <td>0.902564</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>181</td>\n",
       "      <td>0.082873</td>\n",
       "      <td>0.895028</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>191</td>\n",
       "      <td>0.068063</td>\n",
       "      <td>0.890052</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>189</td>\n",
       "      <td>0.116402</td>\n",
       "      <td>0.878307</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>192</td>\n",
       "      <td>0.135417</td>\n",
       "      <td>0.854167</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>195</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.835897</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>110</td>\n",
       "      <td>0.154545</td>\n",
       "      <td>0.818182</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>133</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.781955</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>130</td>\n",
       "      <td>0.338462</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>185</td>\n",
       "      <td>0.362162</td>\n",
       "      <td>0.545946</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>37</td>\n",
       "      <td>0.459459</td>\n",
       "      <td>0.459459</td>\n",
       "      <td>Up</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    NGenes  PropDown    PropUp Direction  PValue     FDR  PValue.Mixed  \\\n",
       "0      193  0.000000  1.000000        Up  0.0001  0.0001        0.0001   \n",
       "1      189  0.000000  1.000000        Up  0.0001  0.0001        0.0001   \n",
       "2      181  0.000000  1.000000        Up  0.0001  0.0001        0.0001   \n",
       "3      180  0.000000  1.000000        Up  0.0001  0.0001        0.0001   \n",
       "4      138  0.000000  1.000000        Up  0.0001  0.0001        0.0001   \n",
       "5       89  0.000000  1.000000        Up  0.0001  0.0001        0.0001   \n",
       "6       55  0.000000  1.000000        Up  0.0001  0.0001        0.0001   \n",
       "7       54  0.000000  1.000000        Up  0.0001  0.0001        0.0001   \n",
       "8       32  0.000000  1.000000        Up  0.0001  0.0001        0.0001   \n",
       "9      185  0.000000  0.994595        Up  0.0001  0.0001        0.0001   \n",
       "10     184  0.005435  0.994565        Up  0.0001  0.0001        0.0001   \n",
       "11     137  0.000000  0.992701        Up  0.0001  0.0001        0.0001   \n",
       "12     104  0.009615  0.990385        Up  0.0001  0.0001        0.0001   \n",
       "13      93  0.010753  0.989247        Up  0.0001  0.0001        0.0001   \n",
       "14     185  0.005405  0.983784        Up  0.0001  0.0001        0.0001   \n",
       "15     197  0.015228  0.979695        Up  0.0001  0.0001        0.0001   \n",
       "16     193  0.005181  0.979275        Up  0.0001  0.0001        0.0001   \n",
       "17     188  0.015957  0.978723        Up  0.0001  0.0001        0.0001   \n",
       "18      94  0.021277  0.978723        Up  0.0001  0.0001        0.0001   \n",
       "19      45  0.022222  0.977778        Up  0.0001  0.0001        0.0001   \n",
       "20     187  0.021390  0.973262        Up  0.0001  0.0001        0.0001   \n",
       "21      67  0.014925  0.970149        Up  0.0001  0.0001        0.0001   \n",
       "22     191  0.020942  0.963351        Up  0.0001  0.0001        0.0001   \n",
       "23     156  0.025641  0.961538        Up  0.0001  0.0001        0.0001   \n",
       "24     193  0.025907  0.958549        Up  0.0001  0.0001        0.0001   \n",
       "25     196  0.040816  0.943878        Up  0.0001  0.0001        0.0001   \n",
       "26      35  0.028571  0.942857        Up  0.0001  0.0001        0.0001   \n",
       "27      33  0.030303  0.939394        Up  0.0001  0.0001        0.0001   \n",
       "28     148  0.047297  0.932432        Up  0.0001  0.0001        0.0001   \n",
       "29     103  0.067961  0.932039        Up  0.0001  0.0001        0.0001   \n",
       "30      85  0.058824  0.929412        Up  0.0001  0.0001        0.0001   \n",
       "31      42  0.047619  0.928571        Up  0.0001  0.0001        0.0001   \n",
       "32     150  0.066667  0.926667        Up  0.0001  0.0001        0.0001   \n",
       "33     188  0.047872  0.925532        Up  0.0001  0.0001        0.0001   \n",
       "34     191  0.057592  0.921466        Up  0.0001  0.0001        0.0001   \n",
       "35     192  0.062500  0.916667        Up  0.0001  0.0001        0.0001   \n",
       "36      99  0.060606  0.909091        Up  0.0001  0.0001        0.0001   \n",
       "37      43  0.069767  0.906977        Up  0.0001  0.0001        0.0001   \n",
       "38     197  0.055838  0.903553        Up  0.0001  0.0001        0.0001   \n",
       "39     195  0.071795  0.902564        Up  0.0001  0.0001        0.0001   \n",
       "40     181  0.082873  0.895028        Up  0.0001  0.0001        0.0001   \n",
       "41     191  0.068063  0.890052        Up  0.0001  0.0001        0.0001   \n",
       "42     189  0.116402  0.878307        Up  0.0001  0.0001        0.0001   \n",
       "43     192  0.135417  0.854167        Up  0.0001  0.0001        0.0001   \n",
       "44     195  0.133333  0.835897        Up  0.0001  0.0001        0.0001   \n",
       "45     110  0.154545  0.818182        Up  0.0001  0.0001        0.0001   \n",
       "46     133  0.142857  0.781955        Up  0.0001  0.0001        0.0001   \n",
       "47     130  0.338462  0.615385        Up  0.0001  0.0001        0.0001   \n",
       "48     185  0.362162  0.545946        Up  0.0001  0.0001        0.0001   \n",
       "49      37  0.459459  0.459459        Up  0.0001  0.0001        0.0001   \n",
       "\n",
       "    FDR.Mixed  \n",
       "0      0.0001  \n",
       "1      0.0001  \n",
       "2      0.0001  \n",
       "3      0.0001  \n",
       "4      0.0001  \n",
       "5      0.0001  \n",
       "6      0.0001  \n",
       "7      0.0001  \n",
       "8      0.0001  \n",
       "9      0.0001  \n",
       "10     0.0001  \n",
       "11     0.0001  \n",
       "12     0.0001  \n",
       "13     0.0001  \n",
       "14     0.0001  \n",
       "15     0.0001  \n",
       "16     0.0001  \n",
       "17     0.0001  \n",
       "18     0.0001  \n",
       "19     0.0001  \n",
       "20     0.0001  \n",
       "21     0.0001  \n",
       "22     0.0001  \n",
       "23     0.0001  \n",
       "24     0.0001  \n",
       "25     0.0001  \n",
       "26     0.0001  \n",
       "27     0.0001  \n",
       "28     0.0001  \n",
       "29     0.0001  \n",
       "30     0.0001  \n",
       "31     0.0001  \n",
       "32     0.0001  \n",
       "33     0.0001  \n",
       "34     0.0001  \n",
       "35     0.0001  \n",
       "36     0.0001  \n",
       "37     0.0001  \n",
       "38     0.0001  \n",
       "39     0.0001  \n",
       "40     0.0001  \n",
       "41     0.0001  \n",
       "42     0.0001  \n",
       "43     0.0001  \n",
       "44     0.0001  \n",
       "45     0.0001  \n",
       "46     0.0001  \n",
       "47     0.0001  \n",
       "48     0.0001  \n",
       "49     0.0001  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Quick check\n",
    "print(template_enriched_pathways.shape)\n",
    "template_enriched_pathways"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Apply enrichment to simulated experiments**\n",
    "\n",
    "Note: GSA takes a while to run (each simulated experiment took ~2-3 minutes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in ROAST\"\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i project_id -i local_dir -i hallmark_DB_filename -i metadata_filename -i num_runs -i base_dir -i enrichment_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/other_enrichment_methods.R'))\n",
    "\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_expression_filename <- paste(local_dir, \n",
    "                                           \"pseudo_experiment/selected_simulated_data_\",\n",
    "                                           project_id,\n",
    "                                           \"_\", \n",
    "                                           i,\n",
    "                                           \"_processed.txt\",\n",
    "                                           sep = \"\")\n",
    "\n",
    "    out_filename <- paste(local_dir,\n",
    "                          \"GSA_stats/\",\n",
    "                          enrichment_method,\n",
    "                          \"_stats_simulated_data_\",\n",
    "                          project_id,\n",
    "                          \"_\",\n",
    "                          i,\n",
    "                          \".txt\", \n",
    "                          sep = \"\")\n",
    "    \n",
    "    if (enrichment_method == \"GSVA\"){\n",
    "        enriched_pathways <- find_enriched_pathways_GSVA(\n",
    "            simulated_expression_filename, \n",
    "            hallmark_DB_filename\n",
    "        ) \n",
    "        write.table(as.data.frame(enriched_pathways), file = out_filename, row.names = F, sep = \"\\t\")\n",
    "        print(\"in GSVA\")\n",
    "    }\n",
    "    else if (enrichment_method == \"ROAST\"){\n",
    "        enriched_pathways <- find_enriched_pathways_ROAST(\n",
    "            simulated_expression_filename,\n",
    "            metadata_filename,\n",
    "            hallmark_DB_filename\n",
    "        ) \n",
    "        write.table(as.data.frame(enriched_pathways), file = out_filename, row.names = F, sep = \"\\t\")\n",
    "        print(\"in ROAST\")\n",
    "    }\n",
    "    else if (enrichment_method == \"CAMERA\"){\n",
    "        enriched_pathways <- find_enriched_pathways_CAMERA(\n",
    "            simulated_expression_filename,\n",
    "            metadata_filename, \n",
    "            hallmark_DB_filename\n",
    "        ) \n",
    "        write.table(as.data.frame(enriched_pathways), file = out_filename, row.names = F, sep = \"\\t\")\n",
    "        print(\"in CAMERA\")\n",
    "    }\n",
    "    else if (enrichment_method == \"ORA\"){\n",
    "        enriched_pathways <- find_enriched_pathways_ORA(\n",
    "            simulated_expression_filename,\n",
    "            metadata_filename, \n",
    "            hallmark_DB_filename\n",
    "        ) \n",
    "        write.table(as.data.frame(enriched_pathways), file = out_filename, row.names = F, sep = \"\\t\")\n",
    "        print(\"in ORA\")\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format enrichment output\n",
    "\n",
    "Each method yields a different output format so we will need to format the data before we can rank and summarize it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Read 50 records\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Read 7424 items\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "12345678910111213141516171819202122232425262728293031323334353637383940414243444546474849501\n",
       "2\n",
       "3\n",
       "4\n",
       "5\n",
       "6\n",
       "7\n",
       "8\n",
       "9\n",
       "10\n",
       "11\n",
       "12\n",
       "13\n",
       "14\n",
       "15\n",
       "16\n",
       "17\n",
       "18\n",
       "19\n",
       "20\n",
       "21\n",
       "22\n",
       "23\n",
       "24\n",
       "25\n",
       "26\n",
       "27\n",
       "28\n",
       "29\n",
       "30\n",
       "31\n",
       "32\n",
       "33\n",
       "34\n",
       "35\n",
       "36\n",
       "37\n",
       "38\n",
       "39\n",
       "40\n",
       "41\n",
       "42\n",
       "43\n",
       "44\n",
       "45\n",
       "46\n",
       "47\n",
       "48\n",
       "49\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i hallmark_DB_filename -o hallmark_DB_names\n",
    "library(\"GSA\")\n",
    "\n",
    "hallmark_DB <- GSA.read.gmt(hallmark_DB_filename)\n",
    "\n",
    "hallmark_DB_names <- as.data.frame(hallmark_DB$geneset.names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ranking.format_enrichment_output(\n",
    "    local_dir, \n",
    "    project_id, \n",
    "    enrichment_method, \n",
    "    hallmark_DB_names,\n",
    "    num_runs\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rank pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "analysis_type = \"GSA\"\n",
    "\n",
    "template_GSEA_stats_filename = os.path.join(\n",
    "    local_dir,\n",
    "    \"GSA_stats\",\n",
    "    f\"{enrichment_method}_stats_template_data_{project_id}_real.txt\"    \n",
    ")\n",
    "template_GSEA_stats, simulated_GSEA_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_GSEA_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_pathways,\n",
    "    enrichment_method\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pathway summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (1 levels on the left, 2 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>FDR (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean FDR (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathway</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HALLMARK_SPERMATOGENESIS</th>\n",
       "      <td>HALLMARK_SPERMATOGENESIS</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>25.5</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>25.5</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY</th>\n",
       "      <td>HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>25.5</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>25.5</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_TGF_BETA_SIGNALING</th>\n",
       "      <td>HALLMARK_TGF_BETA_SIGNALING</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>25.5</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>25.5</td>\n",
       "      <td>0.000197</td>\n",
       "      <td>0.000425</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.227416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_UV_RESPONSE_UP</th>\n",
       "      <td>HALLMARK_UV_RESPONSE_UP</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>25.5</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>25.5</td>\n",
       "      <td>0.041571</td>\n",
       "      <td>0.162964</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.254482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_TNFA_SIGNALING_VIA_NFKB</th>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>25.5</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>25.5</td>\n",
       "      <td>0.000959</td>\n",
       "      <td>0.003070</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.279695</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        Pathway ID  \\\n",
       "pathway                                                                              \n",
       "HALLMARK_SPERMATOGENESIS                                  HALLMARK_SPERMATOGENESIS   \n",
       "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY  HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY   \n",
       "HALLMARK_TGF_BETA_SIGNALING                            HALLMARK_TGF_BETA_SIGNALING   \n",
       "HALLMARK_UV_RESPONSE_UP                                    HALLMARK_UV_RESPONSE_UP   \n",
       "HALLMARK_TNFA_SIGNALING_VIA_NFKB                  HALLMARK_TNFA_SIGNALING_VIA_NFKB   \n",
       "\n",
       "                                          Adj P-value (Real)  Rank (Real)  \\\n",
       "pathway                                                                     \n",
       "HALLMARK_SPERMATOGENESIS                              0.0001         25.5   \n",
       "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY              0.0001         25.5   \n",
       "HALLMARK_TGF_BETA_SIGNALING                           0.0001         25.5   \n",
       "HALLMARK_UV_RESPONSE_UP                               0.0001         25.5   \n",
       "HALLMARK_TNFA_SIGNALING_VIA_NFKB                      0.0001         25.5   \n",
       "\n",
       "                                          FDR (Real)  \\\n",
       "pathway                                                \n",
       "HALLMARK_SPERMATOGENESIS                      0.0001   \n",
       "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY      0.0001   \n",
       "HALLMARK_TGF_BETA_SIGNALING                   0.0001   \n",
       "HALLMARK_UV_RESPONSE_UP                       0.0001   \n",
       "HALLMARK_TNFA_SIGNALING_VIA_NFKB              0.0001   \n",
       "\n",
       "                                          Median adj p-value (simulated)  \\\n",
       "pathway                                                                    \n",
       "HALLMARK_SPERMATOGENESIS                                          0.0001   \n",
       "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY                          0.0001   \n",
       "HALLMARK_TGF_BETA_SIGNALING                                       0.0001   \n",
       "HALLMARK_UV_RESPONSE_UP                                           0.0001   \n",
       "HALLMARK_TNFA_SIGNALING_VIA_NFKB                                  0.0001   \n",
       "\n",
       "                                          Rank (simulated)  \\\n",
       "pathway                                                      \n",
       "HALLMARK_SPERMATOGENESIS                              25.5   \n",
       "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY              25.5   \n",
       "HALLMARK_TGF_BETA_SIGNALING                           25.5   \n",
       "HALLMARK_UV_RESPONSE_UP                               25.5   \n",
       "HALLMARK_TNFA_SIGNALING_VIA_NFKB                      25.5   \n",
       "\n",
       "                                          Mean FDR (simulated)  \\\n",
       "pathway                                                          \n",
       "HALLMARK_SPERMATOGENESIS                              0.000112   \n",
       "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY              0.000104   \n",
       "HALLMARK_TGF_BETA_SIGNALING                           0.000197   \n",
       "HALLMARK_UV_RESPONSE_UP                               0.041571   \n",
       "HALLMARK_TNFA_SIGNALING_VIA_NFKB                      0.000959   \n",
       "\n",
       "                                          Std deviation (simulated)  \\\n",
       "pathway                                                               \n",
       "HALLMARK_SPERMATOGENESIS                                   0.000060   \n",
       "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY                   0.000020   \n",
       "HALLMARK_TGF_BETA_SIGNALING                                0.000425   \n",
       "HALLMARK_UV_RESPONSE_UP                                    0.162964   \n",
       "HALLMARK_TNFA_SIGNALING_VIA_NFKB                           0.003070   \n",
       "\n",
       "                                          Number of experiments (simulated)  \\\n",
       "pathway                                                                       \n",
       "HALLMARK_SPERMATOGENESIS                                                 25   \n",
       "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY                                 25   \n",
       "HALLMARK_TGF_BETA_SIGNALING                                              25   \n",
       "HALLMARK_UV_RESPONSE_UP                                                  25   \n",
       "HALLMARK_TNFA_SIGNALING_VIA_NFKB                                         25   \n",
       "\n",
       "                                           Z score  \n",
       "pathway                                             \n",
       "HALLMARK_SPERMATOGENESIS                 -0.200000  \n",
       "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY -0.200000  \n",
       "HALLMARK_TGF_BETA_SIGNALING              -0.227416  \n",
       "HALLMARK_UV_RESPONSE_UP                  -0.254482  \n",
       "HALLMARK_TNFA_SIGNALING_VIA_NFKB         -0.279695  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create intermediate file: \"<local_dir>/gene_summary_table_<col_to_rank_pathways>.tsv\"\n",
    "summary_pathway_ranks = ranking.generate_summary_table(\n",
    "    template_GSEA_stats_filename,\n",
    "    template_GSEA_stats,\n",
    "    simulated_GSEA_summary_stats,\n",
    "    col_to_rank_pathways,\n",
    "    local_dir,\n",
    "    'pathway',\n",
    "    params\n",
    ")\n",
    "\n",
    "summary_pathway_ranks.sort_values(by=\"Z score\", ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create `pathway_summary_filename`\n",
    "summary_pathway_ranks.to_csv(pathway_summary_filename, sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
