{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apply enrichment method\n",
    "\n",
    "This notebook aims to show that SOPHIE can be easily extended to use multiple different enrichment methods. The enrichment methods were selected from [Geistlinger et. al. paper](https://academic.oup.com/bib/article/22/1/545/5722384). We selected a subset of the methods in the paper based on those that could 1) be applied to both RNA-seq and array data and 2)those that covered a wide range of statistical performances.\n",
    "\n",
    "The methods we selected were (description of the methods are below):\n",
    "* GSEA\n",
    "* GSVA\n",
    "* CAMERA\n",
    "* ORA\n",
    "\n",
    "These methods have a range of different characteristics:\n",
    "* Runtime: GSEA has the longest runtime compared to CAMERA which has the shortest, GSVA and ORA are in the middle\n",
    "* Type of method: GSVA is self-contained vs the others are competative\n",
    "* Number of gene sets found to be statistically signficant: GSVA found the most (median ~70%), CAMERA found ~25% and ORA and GSEA found very few\n",
    "\n",
    "**Implementation note:** \n",
    "\n",
    "This notebook assumes that `2_identify_generic_genes_pathways.ipynb` has already been run on the dataset of interest. `2_identify_generic_genes_pathways.ipynb` notebook needs to be run to generate simulated data as well as get results for GSEA analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "\n",
    "from rpy2.robjects import pandas2ri\n",
    "pandas2ri.activate()\n",
    "\n",
    "from ponyo import utils\n",
    "from generic_expression_patterns_modules import ranking\n",
    "\n",
    "np.random.seed(123)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = os.path.abspath(\n",
    "    os.path.join(base_dir, \"configs\", \"config_human_cancer.tsv\")\n",
    ")\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load params\n",
    "local_dir = params[\"local_dir\"]\n",
    "project_id = params[\"project_id\"]\n",
    "statistic = params[\"gsea_statistic\"]\n",
    "hallmark_DB_filename = params[\"pathway_DB_filename\"]\n",
    "num_runs = params[\"num_simulated\"]\n",
    "dataset_name = params[\"dataset_name\"]\n",
    "logFC_name = params['DE_logFC_name']\n",
    "pvalue_name = params['DE_pvalue_name']\n",
    "\n",
    "# Select enrichment method\n",
    "\n",
    "# enrichment_method = [\"GSVA\", \"CAMERA\", \"ORA\"]\n",
    "# If enrichment_method == \"GSVA\" then use \"ES\" to rank\n",
    "# If enrichment_method == \"CAMERA\" then use \"FDR\" to rank\n",
    "# If using \"ORA\" then use \"p.adjust\" to rank\n",
    "enrichment_method = \"CAMERA\"\n",
    "col_to_rank_pathways = params[\"rank_pathways_by\"]\n",
    "\n",
    "# Select platform technology\n",
    "\n",
    "# platform = [\"array\", \"rnaseq\"]\n",
    "platform = \"array\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load DE stats directory\n",
    "DE_stats_dir = os.path.join(local_dir, \"DE_stats\")\n",
    "\n",
    "# Template experiment gene expression\n",
    "template_expression_filename = os.path.join(base_dir, dataset_name, params[\"processed_template_filename\"])\n",
    "\n",
    "# Template experiment DE stats\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    DE_stats_dir,\n",
    "    f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "# Metadata file with sample grouping to define comparison\n",
    "metadata_filename = os.path.join(\n",
    "    base_dir,\n",
    "    dataset_name,\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    f\"{project_id}_groups.tsv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output file\n",
    "pathway_summary_filename = os.path.join(\n",
    "    base_dir, \n",
    "    dataset_name, \n",
    "    f\"generic_pathway_summary_{project_id}_{enrichment_method}.tsv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enrichment methods\n",
    "These methods are important because they permit differential expression questions to be posed in terms of ensembles of genes representing pathways or other biologically interpretable processes. \n",
    "\n",
    "There are two main groups of methods: ‘self-contained’ gene set tests examine a set of genes in their own right without reference to other genes in the genome whereas ‘competitive’ gene set tests compare genes in the test set relative to all other genes. Self-contained tests are of interest for assessing the relevance of an individual biological process to the experiment at hand, whereas the competitive tests focus more on distinguishing the most important biological processes from those that are less important. Competitive tests are overwhelmingly more commonly used in the genomic literature. (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3458527/)\n",
    "\n",
    "* [CAMERA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3458527/) (Correlation Adjusted MEan RAnk gene set test) is based on the idea of estimating the variance inflation factor associated with inter-gene correlation, and incorporating this into parametric or rank-based test procedures. (available in limma) \n",
    "  * Competitive gene set test\n",
    "  * Performs the same rank-based test procedure as GSEA, but also estimates the correlation between genes, instead of treating genes as independent\n",
    "  * Recall GSEA: 1) Rank all genes using DE association statistics. 2) An enrichment score (ES) is defined as the maximum distance from the middle of the ranked list. Thus, the enrichment score indicates whether the genes contained in a gene set are clustered towards the beginning or the end of the ranked list (indicating a correlation with change in expression). 3) Estimate the statistical significance of the ES by a phenotypic-based permutation (permute samples assigned to label) test in order to produce a null distribution for the ES (i.e. scores based on permuted phenotype)\n",
    "  * Appropriate for small and large fold changes\n",
    "* [GSVA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3618321/) (Gene Set Variation Analysis) calculates sample-wise gene set enrichment scores as a function of genes inside and outside the gene set. This method is well-suited for assessing gene set variation across a dichotomous phenotype. (biocontuctor package GSVA) \n",
    "  * Self-contained gene set test\n",
    "  * Estimates variation of gene set enrichment over the samples independently of any class label\n",
    "* [ORA](https://www.rdocumentation.org/packages/clusterProfiler/versions/3.0.4/topics/enricher) (over-representation analysis) uses the hypergeometric test to determine if there a significant over-representation of pathway in the selected set of DEGs. Here we're using clusterProfiler library but there are multiple options for this analysis. See [slide 6](https://docs.google.com/presentation/d/1t4rK7UiLAeIKIzeRJK-YzspNUfGM-8nuRCcevh2lx34/edit?usp=sharing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create \"<local_dir>/GSEA_stats/\" subdirectory\n",
    "os.makedirs(os.path.join(local_dir, \"GSA_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load pathway data\n",
    "hallmark_DB_filename = params[\"pathway_DB_filename\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Apply enrichment to template experiment**\n",
    "\n",
    "See supplementary tables: https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbz158/5722384"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Rcpp\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘edgeR’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:DEFormats’:\n",
      "\n",
      "    DGEList\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: S4Vectors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: stats4\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocGenerics\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: parallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: IRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomicRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomeInfoDb\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: SummarizedExperiment\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Biobase\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: DelayedArray\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: matrixStats\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocParallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Registered S3 method overwritten by 'enrichplot':\n",
      "  method               from\n",
      "  fortify.enrichResult DOSE\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: clusterProfiler v3.14.0  For help: https://guangchuangyu.github.io/software/clusterProfiler\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "Guangchuang Yu, Li-Gen Wang, Yanyan Han, Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology. 2012, 16(5):284-287.\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:DelayedArray’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py:191: FutureWarning: from_items is deprecated. Please use DataFrame.from_dict(dict(items), ...) instead. DataFrame.from_dict(OrderedDict(items)) may be used to preserve the key order.\n",
      "  res = PandasDataFrame.from_items(items)\n"
     ]
    }
   ],
   "source": [
    "%%R -i base_dir -i local_dir -i project_id -i template_expression_filename -i hallmark_DB_filename -i metadata_filename -i enrichment_method -i platform -o template_enriched_pathways\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/other_enrichment_methods.R'))\n",
    "\n",
    "out_filename <- paste(local_dir, \n",
    "                      \"GSA_stats/\",\n",
    "                      enrichment_method,\n",
    "                      \"_stats_template_data_\",\n",
    "                      project_id,\n",
    "                      \"_real.txt\", \n",
    "                      sep = \"\")\n",
    "\n",
    "if (enrichment_method == \"GSVA\"){\n",
    "    \n",
    "    template_enriched_pathways <- find_enriched_pathways_GSVA(\n",
    "        template_expression_filename,\n",
    "        hallmark_DB_filename,\n",
    "        platform\n",
    "    )\n",
    "}\n",
    "else if (enrichment_method == \"CAMERA\"){\n",
    "    \n",
    "    template_enriched_pathways <- find_enriched_pathways_CAMERA(\n",
    "        template_expression_filename,\n",
    "        metadata_filename,\n",
    "        hallmark_DB_filename,\n",
    "        platform\n",
    "    )\n",
    "}\n",
    "else if (enrichment_method == \"ORA\"){\n",
    "    \n",
    "    template_enriched_pathways <- find_enriched_pathways_ORA(\n",
    "        template_expression_filename,\n",
    "        metadata_filename, \n",
    "        hallmark_DB_filename,\n",
    "        platform\n",
    "    )\n",
    "}\n",
    "write.table(as.data.frame(template_enriched_pathways), file = out_filename, row.names = F, sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(50, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NGenes</th>\n",
       "      <th>Direction</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>65.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>1.661477e-09</td>\n",
       "      <td>8.307384e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>57.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>1.880893e-07</td>\n",
       "      <td>3.459403e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>39.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>2.075642e-07</td>\n",
       "      <td>3.459403e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>58.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>3.276393e-06</td>\n",
       "      <td>4.095491e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>52.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>4.157648e-06</td>\n",
       "      <td>4.157648e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>29.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>9.921172e-05</td>\n",
       "      <td>8.267644e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>34.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>6.762263e-04</td>\n",
       "      <td>4.830188e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>38.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>9.447601e-04</td>\n",
       "      <td>5.904751e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>50.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>3.912528e-03</td>\n",
       "      <td>2.173627e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>52.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>6.784222e-03</td>\n",
       "      <td>3.351343e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>51.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>7.372954e-03</td>\n",
       "      <td>3.351343e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>42.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>9.739344e-03</td>\n",
       "      <td>3.851647e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>67.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>1.001428e-02</td>\n",
       "      <td>3.851647e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>70.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>2.521718e-02</td>\n",
       "      <td>9.006137e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>58.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>3.226412e-02</td>\n",
       "      <td>1.017446e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>64.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>3.255828e-02</td>\n",
       "      <td>1.017446e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>65.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>4.070835e-02</td>\n",
       "      <td>1.197305e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>13.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>4.381047e-02</td>\n",
       "      <td>1.216957e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>36.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>7.612387e-02</td>\n",
       "      <td>1.882608e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>35.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>7.827957e-02</td>\n",
       "      <td>1.882608e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>45.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>8.120377e-02</td>\n",
       "      <td>1.882608e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>9.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>8.283474e-02</td>\n",
       "      <td>1.882608e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>9.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>1.322123e-01</td>\n",
       "      <td>2.874181e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>60.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>1.444788e-01</td>\n",
       "      <td>2.917683e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>9.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>1.458841e-01</td>\n",
       "      <td>2.917683e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>31.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>1.824760e-01</td>\n",
       "      <td>3.509153e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>25.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>2.948625e-01</td>\n",
       "      <td>5.460417e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>60.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>3.066144e-01</td>\n",
       "      <td>5.475257e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>73.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>3.418136e-01</td>\n",
       "      <td>5.784389e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>11.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>3.470633e-01</td>\n",
       "      <td>5.784389e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>24.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>4.140235e-01</td>\n",
       "      <td>6.405517e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>68.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>4.186177e-01</td>\n",
       "      <td>6.405517e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>37.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>4.227641e-01</td>\n",
       "      <td>6.405517e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>41.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>4.580793e-01</td>\n",
       "      <td>6.593388e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>29.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>4.615371e-01</td>\n",
       "      <td>6.593388e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>70.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>5.196691e-01</td>\n",
       "      <td>7.217626e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>52.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>5.620326e-01</td>\n",
       "      <td>7.438143e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>14.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>5.652989e-01</td>\n",
       "      <td>7.438143e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>32.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>6.067129e-01</td>\n",
       "      <td>7.741911e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>14.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>6.193529e-01</td>\n",
       "      <td>7.741911e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>68.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>6.580144e-01</td>\n",
       "      <td>8.024566e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>41.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>6.880175e-01</td>\n",
       "      <td>8.189454e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>37.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>7.042930e-01</td>\n",
       "      <td>8.189454e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>81.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>7.542946e-01</td>\n",
       "      <td>8.571529e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>54.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>7.858568e-01</td>\n",
       "      <td>8.731742e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>9.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>8.398861e-01</td>\n",
       "      <td>9.129196e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>51.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>8.896653e-01</td>\n",
       "      <td>9.357933e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>10.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>8.983616e-01</td>\n",
       "      <td>9.357933e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>53.0</td>\n",
       "      <td>Up</td>\n",
       "      <td>9.532961e-01</td>\n",
       "      <td>9.727511e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>64.0</td>\n",
       "      <td>Down</td>\n",
       "      <td>9.739308e-01</td>\n",
       "      <td>9.739308e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    NGenes Direction        PValue           FDR\n",
       "0     65.0        Up  1.661477e-09  8.307384e-08\n",
       "1     57.0        Up  1.880893e-07  3.459403e-06\n",
       "2     39.0        Up  2.075642e-07  3.459403e-06\n",
       "3     58.0        Up  3.276393e-06  4.095491e-05\n",
       "4     52.0        Up  4.157648e-06  4.157648e-05\n",
       "5     29.0        Up  9.921172e-05  8.267644e-04\n",
       "6     34.0        Up  6.762263e-04  4.830188e-03\n",
       "7     38.0        Up  9.447601e-04  5.904751e-03\n",
       "8     50.0        Up  3.912528e-03  2.173627e-02\n",
       "9     52.0        Up  6.784222e-03  3.351343e-02\n",
       "10    51.0        Up  7.372954e-03  3.351343e-02\n",
       "11    42.0        Up  9.739344e-03  3.851647e-02\n",
       "12    67.0        Up  1.001428e-02  3.851647e-02\n",
       "13    70.0        Up  2.521718e-02  9.006137e-02\n",
       "14    58.0      Down  3.226412e-02  1.017446e-01\n",
       "15    64.0        Up  3.255828e-02  1.017446e-01\n",
       "16    65.0        Up  4.070835e-02  1.197305e-01\n",
       "17    13.0        Up  4.381047e-02  1.216957e-01\n",
       "18    36.0        Up  7.612387e-02  1.882608e-01\n",
       "19    35.0      Down  7.827957e-02  1.882608e-01\n",
       "20    45.0        Up  8.120377e-02  1.882608e-01\n",
       "21     9.0        Up  8.283474e-02  1.882608e-01\n",
       "22     9.0        Up  1.322123e-01  2.874181e-01\n",
       "23    60.0        Up  1.444788e-01  2.917683e-01\n",
       "24     9.0      Down  1.458841e-01  2.917683e-01\n",
       "25    31.0        Up  1.824760e-01  3.509153e-01\n",
       "26    25.0        Up  2.948625e-01  5.460417e-01\n",
       "27    60.0        Up  3.066144e-01  5.475257e-01\n",
       "28    73.0      Down  3.418136e-01  5.784389e-01\n",
       "29    11.0        Up  3.470633e-01  5.784389e-01\n",
       "30    24.0        Up  4.140235e-01  6.405517e-01\n",
       "31    68.0      Down  4.186177e-01  6.405517e-01\n",
       "32    37.0        Up  4.227641e-01  6.405517e-01\n",
       "33    41.0      Down  4.580793e-01  6.593388e-01\n",
       "34    29.0        Up  4.615371e-01  6.593388e-01\n",
       "35    70.0        Up  5.196691e-01  7.217626e-01\n",
       "36    52.0        Up  5.620326e-01  7.438143e-01\n",
       "37    14.0        Up  5.652989e-01  7.438143e-01\n",
       "38    32.0        Up  6.067129e-01  7.741911e-01\n",
       "39    14.0        Up  6.193529e-01  7.741911e-01\n",
       "40    68.0      Down  6.580144e-01  8.024566e-01\n",
       "41    41.0      Down  6.880175e-01  8.189454e-01\n",
       "42    37.0      Down  7.042930e-01  8.189454e-01\n",
       "43    81.0      Down  7.542946e-01  8.571529e-01\n",
       "44    54.0      Down  7.858568e-01  8.731742e-01\n",
       "45     9.0        Up  8.398861e-01  9.129196e-01\n",
       "46    51.0      Down  8.896653e-01  9.357933e-01\n",
       "47    10.0      Down  8.983616e-01  9.357933e-01\n",
       "48    53.0        Up  9.532961e-01  9.727511e-01\n",
       "49    64.0      Down  9.739308e-01  9.739308e-01"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Quick check\n",
    "print(template_enriched_pathways.shape)\n",
    "template_enriched_pathways"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Apply enrichment to simulated experiments**\n",
    "\n",
    "Note: GSA takes a while to run (each simulated experiment took ~2-3 minutes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"in CAMERA\"\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i project_id -i local_dir -i hallmark_DB_filename -i metadata_filename -i num_runs -i base_dir -i enrichment_method -i platform\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/other_enrichment_methods.R'))\n",
    "\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_expression_filename <- paste(local_dir, \n",
    "                                       \"pseudo_experiment/selected_simulated_data_\",\n",
    "                                       project_id,\n",
    "                                       \"_\", \n",
    "                                       i,\n",
    "                                       \"_processed.txt\",\n",
    "                                       sep = \"\")\n",
    "\n",
    "    out_filename <- paste(local_dir,\n",
    "                          \"GSA_stats/\",\n",
    "                          enrichment_method,\n",
    "                          \"_stats_simulated_data_\",\n",
    "                          project_id,\n",
    "                          \"_\",\n",
    "                          i,\n",
    "                          \".txt\", \n",
    "                          sep = \"\")\n",
    "    \n",
    "    if (enrichment_method == \"GSVA\"){\n",
    "        enriched_pathways <- find_enriched_pathways_GSVA(\n",
    "            simulated_expression_filename, \n",
    "            hallmark_DB_filename,\n",
    "            platform\n",
    "        ) \n",
    "        write.table(as.data.frame(enriched_pathways), file = out_filename, row.names = F, sep = \"\\t\")\n",
    "        print(\"in GSVA\")\n",
    "    }\n",
    "    else if (enrichment_method == \"CAMERA\"){\n",
    "        enriched_pathways <- find_enriched_pathways_CAMERA(\n",
    "            simulated_expression_filename,\n",
    "            metadata_filename, \n",
    "            hallmark_DB_filename,\n",
    "            platform\n",
    "        ) \n",
    "        write.table(as.data.frame(enriched_pathways), file = out_filename, row.names = F, sep = \"\\t\")\n",
    "        print(\"in CAMERA\")\n",
    "    }\n",
    "    else if (enrichment_method == \"ORA\"){\n",
    "        enriched_pathways <- find_enriched_pathways_ORA(\n",
    "            simulated_expression_filename,\n",
    "            metadata_filename, \n",
    "            hallmark_DB_filename,\n",
    "            platform\n",
    "        ) \n",
    "        write.table(as.data.frame(enriched_pathways), file = out_filename, row.names = F, sep = \"\\t\")\n",
    "        print(\"in ORA\")\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format enrichment output\n",
    "\n",
    "Each method yields a different output format so we will need to format the data before we can rank and summarize it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Read 50 records\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Read 7424 items\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "12345678910111213141516171819202122232425262728293031323334353637383940414243444546474849501\n",
       "2\n",
       "3\n",
       "4\n",
       "5\n",
       "6\n",
       "7\n",
       "8\n",
       "9\n",
       "10\n",
       "11\n",
       "12\n",
       "13\n",
       "14\n",
       "15\n",
       "16\n",
       "17\n",
       "18\n",
       "19\n",
       "20\n",
       "21\n",
       "22\n",
       "23\n",
       "24\n",
       "25\n",
       "26\n",
       "27\n",
       "28\n",
       "29\n",
       "30\n",
       "31\n",
       "32\n",
       "33\n",
       "34\n",
       "35\n",
       "36\n",
       "37\n",
       "38\n",
       "39\n",
       "40\n",
       "41\n",
       "42\n",
       "43\n",
       "44\n",
       "45\n",
       "46\n",
       "47\n",
       "48\n",
       "49\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i hallmark_DB_filename -o hallmark_DB_names\n",
    "library(\"GSA\")\n",
    "\n",
    "hallmark_DB <- GSA.read.gmt(hallmark_DB_filename)\n",
    "\n",
    "hallmark_DB_names <- as.data.frame(hallmark_DB$geneset.names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ranking.format_enrichment_output(\n",
    "    local_dir, \n",
    "    project_id, \n",
    "    enrichment_method, \n",
    "    hallmark_DB_names,\n",
    "    num_runs\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rank pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "analysis_type = \"GSA\"\n",
    "\n",
    "template_GSEA_stats_filename = os.path.join(\n",
    "    local_dir,\n",
    "    \"GSA_stats\",\n",
    "    f\"{enrichment_method}_stats_template_data_{project_id}_real.txt\"    \n",
    ")\n",
    "template_GSEA_stats, simulated_GSEA_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_GSEA_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_pathways,\n",
    "    logFC_name,\n",
    "    pvalue_name,\n",
    "    enrichment_method\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pathway summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (1 levels on the left, 2 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>FDR (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Mean FDR (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathway</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HALLMARK_WNT_BETA_CATENIN_SIGNALING</th>\n",
       "      <td>HALLMARK_WNT_BETA_CATENIN_SIGNALING</td>\n",
       "      <td>0.972751</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.972751</td>\n",
       "      <td>0.947583</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.040816</td>\n",
       "      <td>0.920829</td>\n",
       "      <td>0.072060</td>\n",
       "      <td>25</td>\n",
       "      <td>0.720536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_UV_RESPONSE_DN</th>\n",
       "      <td>HALLMARK_UV_RESPONSE_DN</td>\n",
       "      <td>0.935793</td>\n",
       "      <td>3.5</td>\n",
       "      <td>0.935793</td>\n",
       "      <td>0.848518</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.122449</td>\n",
       "      <td>0.861494</td>\n",
       "      <td>0.113685</td>\n",
       "      <td>25</td>\n",
       "      <td>0.653556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_OXIDATIVE_PHOSPHORYLATION</th>\n",
       "      <td>HALLMARK_OXIDATIVE_PHOSPHORYLATION</td>\n",
       "      <td>0.721763</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.721763</td>\n",
       "      <td>0.531132</td>\n",
       "      <td>15.0</td>\n",
       "      <td>28.571429</td>\n",
       "      <td>0.534451</td>\n",
       "      <td>0.293512</td>\n",
       "      <td>25</td>\n",
       "      <td>0.638174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_P53_PATHWAY</th>\n",
       "      <td>HALLMARK_P53_PATHWAY</td>\n",
       "      <td>0.743814</td>\n",
       "      <td>13.5</td>\n",
       "      <td>0.743814</td>\n",
       "      <td>0.550823</td>\n",
       "      <td>14.0</td>\n",
       "      <td>26.530612</td>\n",
       "      <td>0.561691</td>\n",
       "      <td>0.287317</td>\n",
       "      <td>25</td>\n",
       "      <td>0.633877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HALLMARK_PEROXISOME</th>\n",
       "      <td>HALLMARK_PEROXISOME</td>\n",
       "      <td>0.774191</td>\n",
       "      <td>11.5</td>\n",
       "      <td>0.774191</td>\n",
       "      <td>0.599094</td>\n",
       "      <td>12.0</td>\n",
       "      <td>22.448980</td>\n",
       "      <td>0.606046</td>\n",
       "      <td>0.271027</td>\n",
       "      <td>25</td>\n",
       "      <td>0.620398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                              Pathway ID  \\\n",
       "pathway                                                                    \n",
       "HALLMARK_WNT_BETA_CATENIN_SIGNALING  HALLMARK_WNT_BETA_CATENIN_SIGNALING   \n",
       "HALLMARK_UV_RESPONSE_DN                          HALLMARK_UV_RESPONSE_DN   \n",
       "HALLMARK_OXIDATIVE_PHOSPHORYLATION    HALLMARK_OXIDATIVE_PHOSPHORYLATION   \n",
       "HALLMARK_P53_PATHWAY                                HALLMARK_P53_PATHWAY   \n",
       "HALLMARK_PEROXISOME                                  HALLMARK_PEROXISOME   \n",
       "\n",
       "                                     Adj P-value (Real)  Rank (Real)  \\\n",
       "pathway                                                                \n",
       "HALLMARK_WNT_BETA_CATENIN_SIGNALING            0.972751          2.0   \n",
       "HALLMARK_UV_RESPONSE_DN                        0.935793          3.5   \n",
       "HALLMARK_OXIDATIVE_PHOSPHORYLATION             0.721763         15.0   \n",
       "HALLMARK_P53_PATHWAY                           0.743814         13.5   \n",
       "HALLMARK_PEROXISOME                            0.774191         11.5   \n",
       "\n",
       "                                     FDR (Real)  \\\n",
       "pathway                                           \n",
       "HALLMARK_WNT_BETA_CATENIN_SIGNALING    0.972751   \n",
       "HALLMARK_UV_RESPONSE_DN                0.935793   \n",
       "HALLMARK_OXIDATIVE_PHOSPHORYLATION     0.721763   \n",
       "HALLMARK_P53_PATHWAY                   0.743814   \n",
       "HALLMARK_PEROXISOME                    0.774191   \n",
       "\n",
       "                                     Median adj p-value (simulated)  \\\n",
       "pathway                                                               \n",
       "HALLMARK_WNT_BETA_CATENIN_SIGNALING                        0.947583   \n",
       "HALLMARK_UV_RESPONSE_DN                                    0.848518   \n",
       "HALLMARK_OXIDATIVE_PHOSPHORYLATION                         0.531132   \n",
       "HALLMARK_P53_PATHWAY                                       0.550823   \n",
       "HALLMARK_PEROXISOME                                        0.599094   \n",
       "\n",
       "                                     Rank (simulated)  Percentile (simulated)  \\\n",
       "pathway                                                                         \n",
       "HALLMARK_WNT_BETA_CATENIN_SIGNALING               2.0                2.040816   \n",
       "HALLMARK_UV_RESPONSE_DN                           4.0                6.122449   \n",
       "HALLMARK_OXIDATIVE_PHOSPHORYLATION               15.0               28.571429   \n",
       "HALLMARK_P53_PATHWAY                             14.0               26.530612   \n",
       "HALLMARK_PEROXISOME                              12.0               22.448980   \n",
       "\n",
       "                                     Mean FDR (simulated)  \\\n",
       "pathway                                                     \n",
       "HALLMARK_WNT_BETA_CATENIN_SIGNALING              0.920829   \n",
       "HALLMARK_UV_RESPONSE_DN                          0.861494   \n",
       "HALLMARK_OXIDATIVE_PHOSPHORYLATION               0.534451   \n",
       "HALLMARK_P53_PATHWAY                             0.561691   \n",
       "HALLMARK_PEROXISOME                              0.606046   \n",
       "\n",
       "                                     Std deviation (simulated)  \\\n",
       "pathway                                                          \n",
       "HALLMARK_WNT_BETA_CATENIN_SIGNALING                   0.072060   \n",
       "HALLMARK_UV_RESPONSE_DN                               0.113685   \n",
       "HALLMARK_OXIDATIVE_PHOSPHORYLATION                    0.293512   \n",
       "HALLMARK_P53_PATHWAY                                  0.287317   \n",
       "HALLMARK_PEROXISOME                                   0.271027   \n",
       "\n",
       "                                     Number of experiments (simulated)  \\\n",
       "pathway                                                                  \n",
       "HALLMARK_WNT_BETA_CATENIN_SIGNALING                                 25   \n",
       "HALLMARK_UV_RESPONSE_DN                                             25   \n",
       "HALLMARK_OXIDATIVE_PHOSPHORYLATION                                  25   \n",
       "HALLMARK_P53_PATHWAY                                                25   \n",
       "HALLMARK_PEROXISOME                                                 25   \n",
       "\n",
       "                                      Z score  \n",
       "pathway                                        \n",
       "HALLMARK_WNT_BETA_CATENIN_SIGNALING  0.720536  \n",
       "HALLMARK_UV_RESPONSE_DN              0.653556  \n",
       "HALLMARK_OXIDATIVE_PHOSPHORYLATION   0.638174  \n",
       "HALLMARK_P53_PATHWAY                 0.633877  \n",
       "HALLMARK_PEROXISOME                  0.620398  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create intermediate file: \"<local_dir>/gene_summary_table_<col_to_rank_pathways>.tsv\"\n",
    "summary_pathway_ranks = ranking.generate_summary_table(\n",
    "    template_GSEA_stats_filename,\n",
    "    template_GSEA_stats,\n",
    "    simulated_GSEA_summary_stats,\n",
    "    col_to_rank_pathways,\n",
    "    local_dir,\n",
    "    'pathway',\n",
    "    params\n",
    ")\n",
    "\n",
    "summary_pathway_ranks.sort_values(by=\"Z score\", ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create `pathway_summary_filename`\n",
    "summary_pathway_ranks.to_csv(pathway_summary_filename, sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
