{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process recount2 data\n",
    "This notebook does the following:\n",
    "\n",
    "1. Selects template experiment\n",
    "2. Downloads subset of recount2 data, including the template experiment (subset of random experiments + 1 template experiment)\n",
    "3. Train VAE on subset of recount2 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import rpy2\n",
    "import seaborn as sns\n",
    "from sklearn import preprocessing\n",
    "import pickle\n",
    "\n",
    "from ponyo import utils, train_vae_modules\n",
    "from generic_expression_patterns_modules import process, calc\n",
    "\n",
    "np.random.seed(123)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(),\"../\"))\n",
    "\n",
    "config_file = os.path.abspath(os.path.join(base_dir,\n",
    "                                           \"config_human.tsv\"))\n",
    "params = utils.read_config(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select template experiment\n",
    "\n",
    "We manually selected bioproject [SRP012656](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37764), which contains primary non-small cell lung adenocarcinoma tumors and adjacent normal tissues of 6 never-smoker Korean female patients with 2 replicates each."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load params\n",
    "local_dir = params[\"local_dir\"]\n",
    "dataset_name = params['dataset_name']\n",
    "NN_architecture = params['NN_architecture']\n",
    "project_id = params['project_id']\n",
    "num_recount2_experiments = params['num_recount2_experiments']\n",
    "template_data_file = params['template_data_file']\n",
    "original_compendium_file = params['compendium_data_file']\n",
    "normalized_data_file = params['normalized_compendium_data_file']\n",
    "shared_genes_file = params['shared_genes_file']\n",
    "scaler_file = params['scaler_transform_file']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download subset of recount2 to use as a compendium\n",
    "The compendium will be composed of random experiments + the selected template experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "# Select 59\n",
    "# Select a\n",
    "# Run one time\n",
    "#if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "#    install.packages(\"BiocManager\")\n",
    "#BiocManager::install(\"derfinder\")\n",
    "#BiocManager::install(\"recount\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: SummarizedExperiment\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomicRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: stats4\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocGenerics\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: parallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: S4Vectors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: IRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomeInfoDb\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Biobase\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: DelayedArray\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: matrixStats\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocParallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Setting options('download.file.method.GEOquery'='auto')\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Setting options('GEOquery.inmemory.gpl'=FALSE)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library('recount')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: 2020-07-09 15:32:35 downloading the metadata to /tmp/RtmpxYcHjG/metadata_clean_sra.Rdata\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: trying URL 'https://github.com/leekgroup/recount-website/blob/master/metadata/metadata_clean_sra.Rdata?raw=true'\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Content type 'application/octet-stream'\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning:  length 2531337 bytes (2.4 MB)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: =\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: downloaded 2.4 MB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] \"SRP012651\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP014574\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP000542\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP016059\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP001540\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP034626\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP062873\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041956\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP046254\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP044956\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP045269\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP039357\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP049238\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP021130\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP018571\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP056201\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP060708\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP023262\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP037775\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041531\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP034953\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP010041\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP038143\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP057248\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP048811\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP031507\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"DRP002667\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP015943\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP050138\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP052896\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP060650\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP002278\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP034547\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP056041\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP018284\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP017464\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP064624\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP010961\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP031849\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP042597\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP017262\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP015138\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP056637\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP061255\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP033466\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP034606\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP063840\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP039598\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP034009\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP035255\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP008496\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP042228\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP062956\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP066889\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP044809\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP011000\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP002621\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP045911\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP018010\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP004043\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP018234\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP055809\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP009282\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP024674\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP064475\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP032540\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP010065\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP007881\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP007481\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP051000\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP001304\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP013473\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP004682\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP008930\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP044766\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP016130\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041943\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP056293\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP028301\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP026454\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP014428\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP018377\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP064270\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP040418\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP043085\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP026600\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP044593\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP011008\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041538\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP000573\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP015904\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"DRP001150\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP003933\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP065445\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP026648\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP012557\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP047097\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP045673\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041494\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP034163\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP037718\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP017105\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP035665\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP007819\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP050074\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP064820\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP045876\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP021459\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP007022\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041825\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP043081\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041044\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP051544\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP051368\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP014842\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP058714\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP003815\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP057613\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP037982\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP065202\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP063978\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041469\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP010930\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP057244\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP040647\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041210\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP043470\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP048993\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP004637\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP018779\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP009276\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP026013\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP008445\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP002306\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP022029\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP061462\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP029592\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP047126\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP034543\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP059643\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP033098\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP005242\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP064457\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP032165\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP061685\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041840\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP066009\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP018883\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP014790\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP009992\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP021134\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP056437\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP028118\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP010500\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP049484\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP010136\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP013389\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP038964\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP009251\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP061455\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP045711\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP061971\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP028738\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP055512\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP063860\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP045204\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP060598\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP048562\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP035679\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP000992\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP029953\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP035862\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP008992\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP014320\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP059379\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP055438\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP018241\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP041387\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP058375\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP063838\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP001558\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP060715\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP021916\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP065848\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP067221\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP015640\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP050551\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP064143\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP009861\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP064652\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP014540\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP034698\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP039552\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP032953\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP039338\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP043166\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP040679\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP033078\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP016629\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"ERP013422\"\n",
       "Loading objects:\n",
       "  rse_gene\n",
       "[1] \"SRP012656\"\n",
       "Loading objects:\n",
       "  rse_gene\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i project_id -i num_recount2_experiments -i local_dir -i base_dir\n",
    "\n",
    "source('../generic_expression_patterns_modules/download_recount2_data.R')\n",
    "\n",
    "get_recount2_compendium(project_id, num_recount2_experiments, local_dir, base_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download expression data for selected project id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Loading objects:\n",
       "  rse_gene\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i project_id -i local_dir\n",
    "\n",
    "source('../generic_expression_patterns_modules/download_recount2_data.R')\n",
    "\n",
    "get_recount2_template_experiment(project_id, local_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset genes and convert gene names\n",
    "For our downstream analysis we will be comparing our set of differentially expression genes against the set found in [Crow et. al. publication](https://www.pnas.org/content/pnas/116/13/6491.full.pdf), therefore we will limit our genes to include only those genes shared between our starting set of genes and those in publication. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Order</th>\n",
       "      <th>Gene_EntrezID</th>\n",
       "      <th>N_HitLists</th>\n",
       "      <th>DE_Prior_Rank</th>\n",
       "      <th>Gene_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>7503</td>\n",
       "      <td>79</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>8653</td>\n",
       "      <td>64</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>DDX3Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>9086</td>\n",
       "      <td>62</td>\n",
       "      <td>0.999870</td>\n",
       "      <td>EIF1AY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>8284</td>\n",
       "      <td>52</td>\n",
       "      <td>0.999870</td>\n",
       "      <td>KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>8287</td>\n",
       "      <td>46</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>USP9Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene_Order  Gene_EntrezID  N_HitLists  DE_Prior_Rank Gene_Name\n",
       "0           1           7503          79       1.000000      XIST\n",
       "1           2           8653          64       0.999948     DDX3Y\n",
       "2           3           9086          62       0.999870    EIF1AY\n",
       "3           4           8284          52       0.999870     KDM5D\n",
       "4           5           8287          46       0.999791     USP9Y"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get generic genes identified by Crow et. al.\n",
    "DE_prior_file = \"https://raw.githubusercontent.com/maggiecrow/DEprior/master/DE_Prior.txt\"\n",
    "\n",
    "DE_prior = pd.read_csv(DE_prior_file,\n",
    "                       header=0,\n",
    "                       sep=\"\\t\")\n",
    "\n",
    "DE_prior.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of published generic genes\n",
    "published_generic_genes = list(DE_prior['Gene_Name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of our genes\n",
    "\n",
    "# Read template data\n",
    "template_data = pd.read_csv(\n",
    "    template_data_file,\n",
    "    header=0,\n",
    "    sep='\\t',\n",
    "    index_col=0)\n",
    "\n",
    "our_gene_ids = list(template_data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# File mapping ensembl ids to hgnc symbols\n",
    "gene_id_file = os.path.join(\n",
    "    local_dir,\n",
    "    \"ensembl_hgnc_mapping.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "suppressWarnings(library(\"biomaRt\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R -i template_data_file -i gene_id_file\n",
    "\n",
    "# Get mapping between ensembl gene ids (ours) to HGNC gene symbols (published)\n",
    "\n",
    "source('../generic_expression_patterns_modules/process_names.R')\n",
    "\n",
    "if (file.exists(gene_id_file) == FALSE){\n",
    "    gene_id_mapping <- get_ensembl_symbol_mapping(template_data_file, gene_id_file)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(57210, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000003137</td>\n",
       "      <td>CYP26B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000003249</td>\n",
       "      <td>DBNDD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000004799</td>\n",
       "      <td>PDK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000006062</td>\n",
       "      <td>MAP3K14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id hgnc_symbol\n",
       "1  ENSG00000002330         BAD\n",
       "2  ENSG00000003137     CYP26B1\n",
       "3  ENSG00000003249      DBNDD1\n",
       "4  ENSG00000004799        PDK4\n",
       "5  ENSG00000006062     MAP3K14"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read gene id mapping\n",
    "gene_id_mapping = pd.read_csv(\n",
    "        gene_id_file,\n",
    "        header=0,\n",
    "        sep='\\t',\n",
    "        index_col=0)\n",
    "\n",
    "print(gene_id_mapping.shape)\n",
    "gene_id_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58037, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_version</th>\n",
       "      <th>ensembl_parsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ensembl_version   ensembl_parsed\n",
       "0  ENSG00000000003.14  ENSG00000000003\n",
       "1   ENSG00000000005.5  ENSG00000000005\n",
       "2  ENSG00000000419.12  ENSG00000000419\n",
       "3  ENSG00000000457.13  ENSG00000000457\n",
       "4  ENSG00000000460.16  ENSG00000000460"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get mapping between ensembl ids with and without version numbers\n",
    "# Expressiond data uses ensembl gene ids with version number \n",
    "ensembl_gene_ids = pd.DataFrame(data={'ensembl_version': our_gene_ids,\n",
    "                                      'ensembl_parsed': [gene_id.split('.')[0] for gene_id in our_gene_ids]})\n",
    "\n",
    "print(ensembl_gene_ids.shape)\n",
    "ensembl_gene_ids.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58129, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>ensembl_parsed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_version</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000002330.13</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003137.8</th>\n",
       "      <td>ENSG00000003137</td>\n",
       "      <td>CYP26B1</td>\n",
       "      <td>ENSG00000003137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003249.13</th>\n",
       "      <td>ENSG00000003249</td>\n",
       "      <td>DBNDD1</td>\n",
       "      <td>ENSG00000003249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000004799.7</th>\n",
       "      <td>ENSG00000004799</td>\n",
       "      <td>PDK4</td>\n",
       "      <td>ENSG00000004799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000006062.13</th>\n",
       "      <td>ENSG00000006062</td>\n",
       "      <td>MAP3K14</td>\n",
       "      <td>ENSG00000006062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ensembl_gene_id hgnc_symbol   ensembl_parsed\n",
       "ensembl_version                                                 \n",
       "ENSG00000002330.13  ENSG00000002330         BAD  ENSG00000002330\n",
       "ENSG00000003137.8   ENSG00000003137     CYP26B1  ENSG00000003137\n",
       "ENSG00000003249.13  ENSG00000003249      DBNDD1  ENSG00000003249\n",
       "ENSG00000004799.7   ENSG00000004799        PDK4  ENSG00000004799\n",
       "ENSG00000006062.13  ENSG00000006062     MAP3K14  ENSG00000006062"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map ensembl ids with version number to gene_id_mapping_df\n",
    "gene_id_mapping = pd.merge(gene_id_mapping, \n",
    "                           ensembl_gene_ids, \n",
    "                           left_on='ensembl_gene_id',\n",
    "                           right_on='ensembl_parsed', \n",
    "                           how='outer')\n",
    "\n",
    "print(gene_id_mapping.shape)\n",
    "gene_id_mapping.set_index('ensembl_version', inplace=True)\n",
    "gene_id_mapping.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since this experiment contains both RNA-seq and smRNA-seq samples which are in different ranges so we will drop smRNA samples so that samples are within the same range. The analysis identifying these two subsets of samples can be found in this [notebook](0_explore_input_data.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/frame.py:3781: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  return super(DataFrame, self).rename(**kwargs)\n"
     ]
    }
   ],
   "source": [
    "# Replace ensembl ids with gene symbols\n",
    "template_data = process.replace_ensembl_ids(template_data,\n",
    "                                            gene_id_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>NFYA</th>\n",
       "      <th>...</th>\n",
       "      <th>MIR6830</th>\n",
       "      <th>MIR548AH</th>\n",
       "      <th>MYOCOS</th>\n",
       "      <th>MIR522</th>\n",
       "      <th>MIR6715B</th>\n",
       "      <th>MIR3116-2</th>\n",
       "      <th>MIR3202-2</th>\n",
       "      <th>HSFX3</th>\n",
       "      <th>VINAC1P</th>\n",
       "      <th>MIR4481</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493937</th>\n",
       "      <td>3.800154</td>\n",
       "      <td>0.018190</td>\n",
       "      <td>9.128459</td>\n",
       "      <td>1.585803</td>\n",
       "      <td>0.713952</td>\n",
       "      <td>84.990542</td>\n",
       "      <td>19.943002</td>\n",
       "      <td>12.065586</td>\n",
       "      <td>2.498822</td>\n",
       "      <td>7.713002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.135649</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.588513</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493938</th>\n",
       "      <td>3.798473</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.157669</td>\n",
       "      <td>1.456017</td>\n",
       "      <td>0.738967</td>\n",
       "      <td>85.788323</td>\n",
       "      <td>19.621252</td>\n",
       "      <td>11.738855</td>\n",
       "      <td>2.435308</td>\n",
       "      <td>7.125808</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005417</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.127873</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.452511</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493939</th>\n",
       "      <td>20.222134</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.522072</td>\n",
       "      <td>3.944919</td>\n",
       "      <td>2.512265</td>\n",
       "      <td>8.072948</td>\n",
       "      <td>47.246078</td>\n",
       "      <td>18.866957</td>\n",
       "      <td>2.443230</td>\n",
       "      <td>12.143670</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011112</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.446833</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493940</th>\n",
       "      <td>19.988329</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17.140304</td>\n",
       "      <td>3.977217</td>\n",
       "      <td>2.378849</td>\n",
       "      <td>7.599225</td>\n",
       "      <td>46.594458</td>\n",
       "      <td>18.921190</td>\n",
       "      <td>2.448276</td>\n",
       "      <td>12.048960</td>\n",
       "      <td>...</td>\n",
       "      <td>0.147412</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.037886</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.07756</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.462295</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.168159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493941</th>\n",
       "      <td>3.239231</td>\n",
       "      <td>0.019398</td>\n",
       "      <td>9.069316</td>\n",
       "      <td>2.181582</td>\n",
       "      <td>1.008155</td>\n",
       "      <td>17.675885</td>\n",
       "      <td>19.371137</td>\n",
       "      <td>11.759756</td>\n",
       "      <td>1.970725</td>\n",
       "      <td>6.740705</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002860</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156220</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.070955</td>\n",
       "      <td>0.739403</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37911 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TSPAN6      TNMD       DPM1     SCYL3  C1orf112        FGR  \\\n",
       "SRR493937   3.800154  0.018190   9.128459  1.585803  0.713952  84.990542   \n",
       "SRR493938   3.798473  0.000000   9.157669  1.456017  0.738967  85.788323   \n",
       "SRR493939  20.222134  0.000000  15.522072  3.944919  2.512265   8.072948   \n",
       "SRR493940  19.988329  0.000000  17.140304  3.977217  2.378849   7.599225   \n",
       "SRR493941   3.239231  0.019398   9.069316  2.181582  1.008155  17.675885   \n",
       "\n",
       "                 CFH      FUCA2      GCLC       NFYA    ...      MIR6830  \\\n",
       "SRR493937  19.943002  12.065586  2.498822   7.713002    ...     0.000000   \n",
       "SRR493938  19.621252  11.738855  2.435308   7.125808    ...     0.005417   \n",
       "SRR493939  47.246078  18.866957  2.443230  12.143670    ...     0.000000   \n",
       "SRR493940  46.594458  18.921190  2.448276  12.048960    ...     0.147412   \n",
       "SRR493941  19.371137  11.759756  1.970725   6.740705    ...     0.002860   \n",
       "\n",
       "           MIR548AH    MYOCOS  MIR522  MIR6715B  MIR3116-2  MIR3202-2  \\\n",
       "SRR493937       0.0  0.135649     0.0       0.0    0.00000   0.000000   \n",
       "SRR493938       0.0  0.127873     0.0       0.0    0.00000   0.000000   \n",
       "SRR493939       0.0  0.011112     0.0       0.0    0.00000   0.000000   \n",
       "SRR493940       0.0  0.037886     0.0       0.0    0.07756   0.000000   \n",
       "SRR493941       0.0  0.156220     0.0       0.0    0.00000   0.070955   \n",
       "\n",
       "              HSFX3  VINAC1P   MIR4481  \n",
       "SRR493937  0.588513      0.0  0.000000  \n",
       "SRR493938  0.452511      0.0  0.000000  \n",
       "SRR493939  0.446833      0.0  0.000000  \n",
       "SRR493940  0.462295      0.0  0.168159  \n",
       "SRR493941  0.739403      0.0  0.000000  \n",
       "\n",
       "[5 rows x 37911 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17785\n"
     ]
    }
   ],
   "source": [
    "# Get intersection of gene lists\n",
    "our_gene_ids_hgnc = template_data.columns\n",
    "shared_genes_hgnc = list(set(our_gene_ids_hgnc).intersection(published_generic_genes))\n",
    "print(len(shared_genes_hgnc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save shared genes\n",
    "outfile = open(shared_genes_file,'wb')\n",
    "pickle.dump(shared_genes_hgnc,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop smRNA samples so that samples are within the same range\n",
    "smRNA_samples = [\"SRR493961\",\n",
    "                 \"SRR493962\",\n",
    "                 \"SRR493963\",\n",
    "                 \"SRR493964\",\n",
    "                 \"SRR493965\",\n",
    "                 \"SRR493966\",\n",
    "                 \"SRR493967\",\n",
    "                 \"SRR493968\",\n",
    "                 \"SRR493969\",\n",
    "                 \"SRR493970\",\n",
    "                 \"SRR493971\",\n",
    "                 \"SRR493972\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop samples\n",
    "template_data = template_data.drop(smRNA_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24, 17788)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRMT10C</th>\n",
       "      <th>RUFY3</th>\n",
       "      <th>CNOT1</th>\n",
       "      <th>SLC35F1</th>\n",
       "      <th>RBM26</th>\n",
       "      <th>IHH</th>\n",
       "      <th>GPR20</th>\n",
       "      <th>ADRB3</th>\n",
       "      <th>MFSD10</th>\n",
       "      <th>DCAF8</th>\n",
       "      <th>...</th>\n",
       "      <th>MRPL3</th>\n",
       "      <th>CEP152</th>\n",
       "      <th>C1QTNF7</th>\n",
       "      <th>RARA</th>\n",
       "      <th>HBD</th>\n",
       "      <th>LINC01526</th>\n",
       "      <th>MFSD14A</th>\n",
       "      <th>TFAP2B</th>\n",
       "      <th>NDUFA4</th>\n",
       "      <th>SULT2B1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493937</th>\n",
       "      <td>4.665383</td>\n",
       "      <td>4.081109</td>\n",
       "      <td>16.805435</td>\n",
       "      <td>0.105384</td>\n",
       "      <td>3.345764</td>\n",
       "      <td>0.596137</td>\n",
       "      <td>0.700501</td>\n",
       "      <td>0.035386</td>\n",
       "      <td>22.347079</td>\n",
       "      <td>13.608744</td>\n",
       "      <td>...</td>\n",
       "      <td>9.813451</td>\n",
       "      <td>0.446482</td>\n",
       "      <td>1.483983</td>\n",
       "      <td>27.105052</td>\n",
       "      <td>0.388193</td>\n",
       "      <td>0.056681</td>\n",
       "      <td>16.617661</td>\n",
       "      <td>0.009494</td>\n",
       "      <td>10.581429</td>\n",
       "      <td>1.526560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493938</th>\n",
       "      <td>4.796513</td>\n",
       "      <td>4.234347</td>\n",
       "      <td>16.808594</td>\n",
       "      <td>0.195406</td>\n",
       "      <td>3.056731</td>\n",
       "      <td>0.471109</td>\n",
       "      <td>0.699908</td>\n",
       "      <td>0.031939</td>\n",
       "      <td>22.376378</td>\n",
       "      <td>14.316637</td>\n",
       "      <td>...</td>\n",
       "      <td>9.553095</td>\n",
       "      <td>0.462525</td>\n",
       "      <td>1.294170</td>\n",
       "      <td>26.411033</td>\n",
       "      <td>0.433801</td>\n",
       "      <td>0.113388</td>\n",
       "      <td>16.912483</td>\n",
       "      <td>0.004779</td>\n",
       "      <td>10.738915</td>\n",
       "      <td>1.518376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493939</th>\n",
       "      <td>3.284528</td>\n",
       "      <td>4.539443</td>\n",
       "      <td>15.327375</td>\n",
       "      <td>0.035567</td>\n",
       "      <td>3.547286</td>\n",
       "      <td>0.227920</td>\n",
       "      <td>0.152627</td>\n",
       "      <td>0.038542</td>\n",
       "      <td>57.659052</td>\n",
       "      <td>21.526116</td>\n",
       "      <td>...</td>\n",
       "      <td>12.157567</td>\n",
       "      <td>0.737306</td>\n",
       "      <td>1.450699</td>\n",
       "      <td>11.622323</td>\n",
       "      <td>0.265947</td>\n",
       "      <td>0.034443</td>\n",
       "      <td>13.683262</td>\n",
       "      <td>0.536620</td>\n",
       "      <td>17.572848</td>\n",
       "      <td>0.803435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493940</th>\n",
       "      <td>3.812965</td>\n",
       "      <td>4.346555</td>\n",
       "      <td>15.367022</td>\n",
       "      <td>0.053466</td>\n",
       "      <td>3.388571</td>\n",
       "      <td>0.286209</td>\n",
       "      <td>0.371584</td>\n",
       "      <td>0.011037</td>\n",
       "      <td>56.854383</td>\n",
       "      <td>22.026572</td>\n",
       "      <td>...</td>\n",
       "      <td>13.030322</td>\n",
       "      <td>0.738708</td>\n",
       "      <td>1.447023</td>\n",
       "      <td>11.706894</td>\n",
       "      <td>0.316107</td>\n",
       "      <td>0.034396</td>\n",
       "      <td>14.278460</td>\n",
       "      <td>0.435270</td>\n",
       "      <td>18.259412</td>\n",
       "      <td>0.924137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493941</th>\n",
       "      <td>2.733277</td>\n",
       "      <td>4.228729</td>\n",
       "      <td>16.461116</td>\n",
       "      <td>0.051296</td>\n",
       "      <td>4.198360</td>\n",
       "      <td>2.105660</td>\n",
       "      <td>2.876334</td>\n",
       "      <td>0.076718</td>\n",
       "      <td>33.461660</td>\n",
       "      <td>16.452750</td>\n",
       "      <td>...</td>\n",
       "      <td>8.254008</td>\n",
       "      <td>0.740339</td>\n",
       "      <td>4.223101</td>\n",
       "      <td>20.432419</td>\n",
       "      <td>0.098966</td>\n",
       "      <td>0.054669</td>\n",
       "      <td>15.831326</td>\n",
       "      <td>0.075468</td>\n",
       "      <td>12.470698</td>\n",
       "      <td>6.449160</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17788 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            TRMT10C     RUFY3      CNOT1   SLC35F1     RBM26       IHH  \\\n",
       "SRR493937  4.665383  4.081109  16.805435  0.105384  3.345764  0.596137   \n",
       "SRR493938  4.796513  4.234347  16.808594  0.195406  3.056731  0.471109   \n",
       "SRR493939  3.284528  4.539443  15.327375  0.035567  3.547286  0.227920   \n",
       "SRR493940  3.812965  4.346555  15.367022  0.053466  3.388571  0.286209   \n",
       "SRR493941  2.733277  4.228729  16.461116  0.051296  4.198360  2.105660   \n",
       "\n",
       "              GPR20     ADRB3     MFSD10      DCAF8    ...         MRPL3  \\\n",
       "SRR493937  0.700501  0.035386  22.347079  13.608744    ...      9.813451   \n",
       "SRR493938  0.699908  0.031939  22.376378  14.316637    ...      9.553095   \n",
       "SRR493939  0.152627  0.038542  57.659052  21.526116    ...     12.157567   \n",
       "SRR493940  0.371584  0.011037  56.854383  22.026572    ...     13.030322   \n",
       "SRR493941  2.876334  0.076718  33.461660  16.452750    ...      8.254008   \n",
       "\n",
       "             CEP152   C1QTNF7       RARA       HBD  LINC01526    MFSD14A  \\\n",
       "SRR493937  0.446482  1.483983  27.105052  0.388193   0.056681  16.617661   \n",
       "SRR493938  0.462525  1.294170  26.411033  0.433801   0.113388  16.912483   \n",
       "SRR493939  0.737306  1.450699  11.622323  0.265947   0.034443  13.683262   \n",
       "SRR493940  0.738708  1.447023  11.706894  0.316107   0.034396  14.278460   \n",
       "SRR493941  0.740339  4.223101  20.432419  0.098966   0.054669  15.831326   \n",
       "\n",
       "             TFAP2B     NDUFA4   SULT2B1  \n",
       "SRR493937  0.009494  10.581429  1.526560  \n",
       "SRR493938  0.004779  10.738915  1.518376  \n",
       "SRR493939  0.536620  17.572848  0.803435  \n",
       "SRR493940  0.435270  18.259412  0.924137  \n",
       "SRR493941  0.075468  12.470698  6.449160  \n",
       "\n",
       "[5 rows x 17788 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop genes\n",
    "template_data = template_data[shared_genes_hgnc]\n",
    "\n",
    "print(template_data.shape)\n",
    "template_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    }
   ],
   "source": [
    "print(len(template_data.columns) - len(shared_genes_hgnc))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Note:* There is a difference in the number of `shared_genes_hgnc` and genes in the template experiment because 3 genes have 2 different ensembl gene ids have map to the same hgnc symbol (one forward, one reverse)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalize compendium "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3388, 58037)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <th>ENSG00000000938.12</th>\n",
       "      <th>ENSG00000000971.15</th>\n",
       "      <th>ENSG00000001036.13</th>\n",
       "      <th>ENSG00000001084.10</th>\n",
       "      <th>ENSG00000001167.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000283690.1</th>\n",
       "      <th>ENSG00000283691.1</th>\n",
       "      <th>ENSG00000283692.1</th>\n",
       "      <th>ENSG00000283693.1</th>\n",
       "      <th>ENSG00000283694.1</th>\n",
       "      <th>ENSG00000283695.1</th>\n",
       "      <th>ENSG00000283696.1</th>\n",
       "      <th>ENSG00000283697.1</th>\n",
       "      <th>ENSG00000283698.1</th>\n",
       "      <th>ENSG00000283699.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493816</th>\n",
       "      <td>2.893395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.298572</td>\n",
       "      <td>1.552609</td>\n",
       "      <td>1.880483</td>\n",
       "      <td>0.264440</td>\n",
       "      <td>1.632884</td>\n",
       "      <td>21.194289</td>\n",
       "      <td>10.014925</td>\n",
       "      <td>4.762246</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.540156</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.314956</td>\n",
       "      <td>0.358752</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493817</th>\n",
       "      <td>1.923686</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.401872</td>\n",
       "      <td>1.245607</td>\n",
       "      <td>1.529450</td>\n",
       "      <td>0.408554</td>\n",
       "      <td>0.071068</td>\n",
       "      <td>16.599504</td>\n",
       "      <td>11.798680</td>\n",
       "      <td>4.036652</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.188427</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.307766</td>\n",
       "      <td>0.512091</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR547975</th>\n",
       "      <td>1.920405</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.610590</td>\n",
       "      <td>1.630213</td>\n",
       "      <td>3.371858</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.919197</td>\n",
       "      <td>64.923769</td>\n",
       "      <td>1.306540</td>\n",
       "      <td>17.485584</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.242607</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.080863</td>\n",
       "      <td>0.460912</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR547973</th>\n",
       "      <td>6.344573</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.921543</td>\n",
       "      <td>3.250220</td>\n",
       "      <td>3.201103</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17.733091</td>\n",
       "      <td>10.638572</td>\n",
       "      <td>4.659644</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.359655</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.118196</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.943308</td>\n",
       "      <td>0.254015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR547968</th>\n",
       "      <td>1.282905</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.224764</td>\n",
       "      <td>2.086982</td>\n",
       "      <td>2.213324</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>80.792929</td>\n",
       "      <td>7.692393</td>\n",
       "      <td>13.100063</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.039251</td>\n",
       "      <td>0.658290</td>\n",
       "      <td>0.794068</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 58037 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           ENSG00000000003.14  ENSG00000000005.5  ENSG00000000419.12  \\\n",
       "SRR493816            2.893395                0.0           18.298572   \n",
       "SRR493817            1.923686                0.0           16.401872   \n",
       "SRR547975            1.920405                0.0           42.610590   \n",
       "SRR547973            6.344573                0.0           27.921543   \n",
       "SRR547968            1.282905                0.0           16.224764   \n",
       "\n",
       "           ENSG00000000457.13  ENSG00000000460.16  ENSG00000000938.12  \\\n",
       "SRR493816            1.552609            1.880483            0.264440   \n",
       "SRR493817            1.245607            1.529450            0.408554   \n",
       "SRR547975            1.630213            3.371858            0.000000   \n",
       "SRR547973            3.250220            3.201103            0.000000   \n",
       "SRR547968            2.086982            2.213324            0.000000   \n",
       "\n",
       "           ENSG00000000971.15  ENSG00000001036.13  ENSG00000001084.10  \\\n",
       "SRR493816            1.632884           21.194289           10.014925   \n",
       "SRR493817            0.071068           16.599504           11.798680   \n",
       "SRR547975            2.919197           64.923769            1.306540   \n",
       "SRR547973            0.000000           17.733091           10.638572   \n",
       "SRR547968            0.000000           80.792929            7.692393   \n",
       "\n",
       "           ENSG00000001167.14        ...          ENSG00000283690.1  \\\n",
       "SRR493816            4.762246        ...                        0.0   \n",
       "SRR493817            4.036652        ...                        0.0   \n",
       "SRR547975           17.485584        ...                        0.0   \n",
       "SRR547973            4.659644        ...                        0.0   \n",
       "SRR547968           13.100063        ...                        0.0   \n",
       "\n",
       "           ENSG00000283691.1  ENSG00000283692.1  ENSG00000283693.1  \\\n",
       "SRR493816           0.540156                0.0                0.0   \n",
       "SRR493817           0.188427                0.0                0.0   \n",
       "SRR547975           2.242607                0.0                0.0   \n",
       "SRR547973           0.359655                0.0                0.0   \n",
       "SRR547968           0.000000                0.0                0.0   \n",
       "\n",
       "           ENSG00000283694.1  ENSG00000283695.1  ENSG00000283696.1  \\\n",
       "SRR493816           0.000000           0.000000           0.314956   \n",
       "SRR493817           0.000000           0.000000           0.307766   \n",
       "SRR547975           0.000000           0.000000           0.080863   \n",
       "SRR547973           0.118196           0.000000           0.943308   \n",
       "SRR547968           0.000000           0.039251           0.658290   \n",
       "\n",
       "           ENSG00000283697.1  ENSG00000283698.1  ENSG00000283699.1  \n",
       "SRR493816           0.358752                0.0                0.0  \n",
       "SRR493817           0.512091                0.0                0.0  \n",
       "SRR547975           0.460912                0.0                0.0  \n",
       "SRR547973           0.254015                0.0                0.0  \n",
       "SRR547968           0.794068                0.0                0.0  \n",
       "\n",
       "[5 rows x 58037 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read data\n",
    "original_compendium = pd.read_table(\n",
    "    original_compendium_file,\n",
    "    header=0,\n",
    "    sep='\\t',\n",
    "    index_col=0)\n",
    "\n",
    "print(original_compendium.shape)\n",
    "original_compendium.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace ensembl ids with gene symbols\n",
    "original_compendium = process.replace_ensembl_ids(original_compendium,\n",
    "                                                gene_id_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRMT10C</th>\n",
       "      <th>RUFY3</th>\n",
       "      <th>CNOT1</th>\n",
       "      <th>SLC35F1</th>\n",
       "      <th>RBM26</th>\n",
       "      <th>IHH</th>\n",
       "      <th>GPR20</th>\n",
       "      <th>ADRB3</th>\n",
       "      <th>MFSD10</th>\n",
       "      <th>DCAF8</th>\n",
       "      <th>...</th>\n",
       "      <th>MRPL3</th>\n",
       "      <th>CEP152</th>\n",
       "      <th>C1QTNF7</th>\n",
       "      <th>RARA</th>\n",
       "      <th>HBD</th>\n",
       "      <th>LINC01526</th>\n",
       "      <th>MFSD14A</th>\n",
       "      <th>TFAP2B</th>\n",
       "      <th>NDUFA4</th>\n",
       "      <th>SULT2B1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493816</th>\n",
       "      <td>3.240427</td>\n",
       "      <td>1.487418</td>\n",
       "      <td>22.294177</td>\n",
       "      <td>0.013186</td>\n",
       "      <td>2.094926</td>\n",
       "      <td>0.126898</td>\n",
       "      <td>0.210860</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.114229</td>\n",
       "      <td>13.971167</td>\n",
       "      <td>...</td>\n",
       "      <td>15.714002</td>\n",
       "      <td>0.695362</td>\n",
       "      <td>0.013531</td>\n",
       "      <td>5.337377</td>\n",
       "      <td>0.000301</td>\n",
       "      <td>0.066481</td>\n",
       "      <td>9.192732</td>\n",
       "      <td>0.001145</td>\n",
       "      <td>13.334783</td>\n",
       "      <td>0.618180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493817</th>\n",
       "      <td>2.009355</td>\n",
       "      <td>1.204978</td>\n",
       "      <td>24.093032</td>\n",
       "      <td>0.010005</td>\n",
       "      <td>1.367029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.164727</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.579942</td>\n",
       "      <td>13.048861</td>\n",
       "      <td>...</td>\n",
       "      <td>10.849816</td>\n",
       "      <td>0.476430</td>\n",
       "      <td>0.001260</td>\n",
       "      <td>6.103372</td>\n",
       "      <td>0.012302</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.885370</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.365625</td>\n",
       "      <td>1.538041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR547975</th>\n",
       "      <td>10.485186</td>\n",
       "      <td>0.377008</td>\n",
       "      <td>53.087109</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.279572</td>\n",
       "      <td>0.020207</td>\n",
       "      <td>0.191273</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.922340</td>\n",
       "      <td>13.850461</td>\n",
       "      <td>...</td>\n",
       "      <td>51.864493</td>\n",
       "      <td>3.202080</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.681785</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>22.189305</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>27.861512</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR547973</th>\n",
       "      <td>5.378727</td>\n",
       "      <td>2.146560</td>\n",
       "      <td>69.606976</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.245030</td>\n",
       "      <td>2.929866</td>\n",
       "      <td>0.009603</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.734198</td>\n",
       "      <td>12.349436</td>\n",
       "      <td>...</td>\n",
       "      <td>42.563711</td>\n",
       "      <td>2.007285</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.435816</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.012944</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.785109</td>\n",
       "      <td>3.258487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR547968</th>\n",
       "      <td>11.799357</td>\n",
       "      <td>0.088415</td>\n",
       "      <td>28.885471</td>\n",
       "      <td>0.003233</td>\n",
       "      <td>5.417210</td>\n",
       "      <td>39.274797</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.232692</td>\n",
       "      <td>11.469034</td>\n",
       "      <td>...</td>\n",
       "      <td>61.760986</td>\n",
       "      <td>4.224323</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.368031</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>19.540065</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>61.917697</td>\n",
       "      <td>3.690493</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17788 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             TRMT10C     RUFY3      CNOT1   SLC35F1     RBM26        IHH  \\\n",
       "SRR493816   3.240427  1.487418  22.294177  0.013186  2.094926   0.126898   \n",
       "SRR493817   2.009355  1.204978  24.093032  0.010005  1.367029   0.000000   \n",
       "SRR547975  10.485186  0.377008  53.087109  0.000000  8.279572   0.020207   \n",
       "SRR547973   5.378727  2.146560  69.606976  0.000000  4.245030   2.929866   \n",
       "SRR547968  11.799357  0.088415  28.885471  0.003233  5.417210  39.274797   \n",
       "\n",
       "              GPR20  ADRB3     MFSD10      DCAF8    ...         MRPL3  \\\n",
       "SRR493816  0.210860    0.0  27.114229  13.971167    ...     15.714002   \n",
       "SRR493817  0.164727    0.0  21.579942  13.048861    ...     10.849816   \n",
       "SRR547975  0.191273    0.0  10.922340  13.850461    ...     51.864493   \n",
       "SRR547973  0.009603    0.0  10.734198  12.349436    ...     42.563711   \n",
       "SRR547968  0.000000    0.0  15.232692  11.469034    ...     61.760986   \n",
       "\n",
       "             CEP152   C1QTNF7      RARA       HBD  LINC01526    MFSD14A  \\\n",
       "SRR493816  0.695362  0.013531  5.337377  0.000301   0.066481   9.192732   \n",
       "SRR493817  0.476430  0.001260  6.103372  0.012302   0.000000   8.885370   \n",
       "SRR547975  3.202080  0.000000  3.681785  0.000000   0.000000  22.189305   \n",
       "SRR547973  2.007285  0.000000  4.435816  0.000000   0.000000   8.012944   \n",
       "SRR547968  4.224323  0.000000  6.368031  0.000000   0.000000  19.540065   \n",
       "\n",
       "             TFAP2B     NDUFA4   SULT2B1  \n",
       "SRR493816  0.001145  13.334783  0.618180  \n",
       "SRR493817  0.000000  11.365625  1.538041  \n",
       "SRR547975  0.000000  27.861512  0.000000  \n",
       "SRR547973  0.000000  15.785109  3.258487  \n",
       "SRR547968  0.000000  61.917697  3.690493  \n",
       "\n",
       "[5 rows x 17788 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop genes\n",
    "original_compendium = original_compendium[shared_genes_hgnc]\n",
    "\n",
    "original_compendium.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3388, 17788)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRMT10C</th>\n",
       "      <th>RUFY3</th>\n",
       "      <th>CNOT1</th>\n",
       "      <th>SLC35F1</th>\n",
       "      <th>RBM26</th>\n",
       "      <th>IHH</th>\n",
       "      <th>GPR20</th>\n",
       "      <th>ADRB3</th>\n",
       "      <th>MFSD10</th>\n",
       "      <th>DCAF8</th>\n",
       "      <th>...</th>\n",
       "      <th>MRPL3</th>\n",
       "      <th>CEP152</th>\n",
       "      <th>C1QTNF7</th>\n",
       "      <th>RARA</th>\n",
       "      <th>HBD</th>\n",
       "      <th>LINC01526</th>\n",
       "      <th>MFSD14A</th>\n",
       "      <th>TFAP2B</th>\n",
       "      <th>NDUFA4</th>\n",
       "      <th>SULT2B1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493816</th>\n",
       "      <td>0.008236</td>\n",
       "      <td>0.008567</td>\n",
       "      <td>0.107553</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.030474</td>\n",
       "      <td>0.002778</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.229998</td>\n",
       "      <td>0.146681</td>\n",
       "      <td>...</td>\n",
       "      <td>0.052023</td>\n",
       "      <td>0.013885</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>0.023876</td>\n",
       "      <td>2.364617e-08</td>\n",
       "      <td>0.006778</td>\n",
       "      <td>0.054133</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.037151</td>\n",
       "      <td>0.008155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493817</th>\n",
       "      <td>0.005107</td>\n",
       "      <td>0.006941</td>\n",
       "      <td>0.116232</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.019886</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.183053</td>\n",
       "      <td>0.136998</td>\n",
       "      <td>...</td>\n",
       "      <td>0.035919</td>\n",
       "      <td>0.009513</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.027302</td>\n",
       "      <td>9.650961e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.052323</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.031665</td>\n",
       "      <td>0.020289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR547975</th>\n",
       "      <td>0.026650</td>\n",
       "      <td>0.002172</td>\n",
       "      <td>0.256107</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.120439</td>\n",
       "      <td>0.000442</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.092649</td>\n",
       "      <td>0.145414</td>\n",
       "      <td>...</td>\n",
       "      <td>0.171702</td>\n",
       "      <td>0.063938</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016470</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.130666</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.077624</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR547973</th>\n",
       "      <td>0.013671</td>\n",
       "      <td>0.012364</td>\n",
       "      <td>0.335804</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.061750</td>\n",
       "      <td>0.064150</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.091053</td>\n",
       "      <td>0.129655</td>\n",
       "      <td>...</td>\n",
       "      <td>0.140911</td>\n",
       "      <td>0.040081</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019843</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.047186</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.043978</td>\n",
       "      <td>0.042983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR547968</th>\n",
       "      <td>0.029991</td>\n",
       "      <td>0.000509</td>\n",
       "      <td>0.139352</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.078802</td>\n",
       "      <td>0.859928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.129212</td>\n",
       "      <td>0.120411</td>\n",
       "      <td>...</td>\n",
       "      <td>0.204466</td>\n",
       "      <td>0.084350</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028486</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115065</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.172506</td>\n",
       "      <td>0.048682</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17788 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            TRMT10C     RUFY3     CNOT1   SLC35F1     RBM26       IHH  \\\n",
       "SRR493816  0.008236  0.008567  0.107553  0.000022  0.030474  0.002778   \n",
       "SRR493817  0.005107  0.006941  0.116232  0.000016  0.019886  0.000000   \n",
       "SRR547975  0.026650  0.002172  0.256107  0.000000  0.120439  0.000442   \n",
       "SRR547973  0.013671  0.012364  0.335804  0.000000  0.061750  0.064150   \n",
       "SRR547968  0.029991  0.000509  0.139352  0.000005  0.078802  0.859928   \n",
       "\n",
       "              GPR20  ADRB3    MFSD10     DCAF8    ...        MRPL3    CEP152  \\\n",
       "SRR493816  0.000114    0.0  0.229998  0.146681    ...     0.052023  0.013885   \n",
       "SRR493817  0.000089    0.0  0.183053  0.136998    ...     0.035919  0.009513   \n",
       "SRR547975  0.000103    0.0  0.092649  0.145414    ...     0.171702  0.063938   \n",
       "SRR547973  0.000005    0.0  0.091053  0.129655    ...     0.140911  0.040081   \n",
       "SRR547968  0.000000    0.0  0.129212  0.120411    ...     0.204466  0.084350   \n",
       "\n",
       "            C1QTNF7      RARA           HBD  LINC01526   MFSD14A    TFAP2B  \\\n",
       "SRR493816  0.000248  0.023876  2.364617e-08   0.006778  0.054133  0.000009   \n",
       "SRR493817  0.000023  0.027302  9.650961e-07   0.000000  0.052323  0.000000   \n",
       "SRR547975  0.000000  0.016470  0.000000e+00   0.000000  0.130666  0.000000   \n",
       "SRR547973  0.000000  0.019843  0.000000e+00   0.000000  0.047186  0.000000   \n",
       "SRR547968  0.000000  0.028486  0.000000e+00   0.000000  0.115065  0.000000   \n",
       "\n",
       "             NDUFA4   SULT2B1  \n",
       "SRR493816  0.037151  0.008155  \n",
       "SRR493817  0.031665  0.020289  \n",
       "SRR547975  0.077624  0.000000  \n",
       "SRR547973  0.043978  0.042983  \n",
       "SRR547968  0.172506  0.048682  \n",
       "\n",
       "[5 rows x 17788 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 0-1 normalize per gene\n",
    "scaler = preprocessing.MinMaxScaler()\n",
    "original_data_scaled = scaler.fit_transform(original_compendium)\n",
    "original_data_scaled_df = pd.DataFrame(original_data_scaled,\n",
    "                                columns=original_compendium.columns,\n",
    "                                index=original_compendium.index)\n",
    "\n",
    "print(original_data_scaled_df.shape)\n",
    "original_data_scaled_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "template_data.to_csv(\n",
    "    template_data_file, float_format='%.5f', sep='\\t')\n",
    "\n",
    "original_compendium.to_csv(\n",
    "    original_compendium_file, float_format='%.3f', sep='\\t')\n",
    "\n",
    "original_data_scaled_df.to_csv(\n",
    "    normalized_data_file, float_format='%.3f', sep='\\t')\n",
    "\n",
    "# Save scaler transform\n",
    "outfile = open(scaler_file,'wb')\n",
    "pickle.dump(scaler,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train VAE \n",
    "Performed exploratory analysis of compendium data [here](../explore_data/viz_recount2_compendium.ipynb) to help interpret loss curve."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup directories\n",
    "# Create VAE directories\n",
    "output_dirs = [os.path.join(base_dir, dataset_name, \"models\"),\n",
    "               os.path.join(base_dir, dataset_name, \"logs\")]\n",
    "\n",
    "# Check if analysis output directory exist otherwise create\n",
    "for each_dir in output_dirs:\n",
    "    if os.path.exists(each_dir) == False:\n",
    "        print('creating new directory: {}'.format(each_dir))\n",
    "        os.makedirs(each_dir, exist_ok=True)\n",
    "\n",
    "# Check if NN architecture directory exist otherwise create\n",
    "for each_dir in output_dirs:\n",
    "    new_dir = os.path.join(each_dir, NN_architecture)\n",
    "    if os.path.exists(new_dir) == False:\n",
    "        print('creating new directory: {}'.format(new_dir))\n",
    "        os.makedirs(new_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Train VAE on new compendium data\n",
    "train_vae_modules.train_vae(config_file,\n",
    "                   normalized_data_file)"
   ]
  }
 ],
 "metadata": {
  "finalized": {
   "timestamp": 1592246126078,
   "trusted": true
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
