{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process recount2 data\n",
    "This notebook does the following:\n",
    "\n",
    "1. Selects template experiment\n",
    "2. Downloads subset of recount2 data, including the template experiment (50 random experiments + 1 template experiment)\n",
    "3. Train VAE on subset of recount2 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import rpy2\n",
    "import seaborn as sns\n",
    "from sklearn import preprocessing\n",
    "import pickle\n",
    "\n",
    "sys.path.append(\"../\")\n",
    "from ponyo import generate_template_data, utils, pipeline\n",
    "from functions import process, calc\n",
    "\n",
    "from numpy.random import seed\n",
    "random_state = 123\n",
    "seed(random_state)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(),\"../\"))\n",
    "\n",
    "config_file = os.path.abspath(os.path.join(base_dir,\n",
    "                                           \"config_human.tsv\"))\n",
    "params = utils.read_config(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select template experiment\n",
    "\n",
    "We manually selected bioproject [SRP012656](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37764), which contains primary non-small cell lung adenocarcinoma tumors and adjacent normal tissues of 6 never-smoker Korean female patients with 2 replicates each."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load params\n",
    "local_dir = params[\"local_dir\"]\n",
    "dataset_name = params['dataset_name']\n",
    "NN_architecture = params['NN_architecture']\n",
    "project_id = params['project_id']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download subset of recount2 to use as a compendium\n",
    "The compendium will be composed of random experiments + the selected template experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "# Select 59\n",
    "# Select a\n",
    "# Run one time\n",
    "#if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "#    install.packages(\"BiocManager\")\n",
    "#BiocManager::install(\"derfinder\")\n",
    "#BiocManager::install(\"recount\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library('recount')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#%%R -i project_id -i local_dir\n",
    "\n",
    "#source('../functions/download_recount2_data.R')\n",
    "\n",
    "#get_recount2_compendium(project_id, local_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download expression data for selected project id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Loading objects:\n",
       "  rse_gene\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i project_id -i local_dir\n",
    "\n",
    "source('../functions/download_recount2_data.R')\n",
    "\n",
    "get_recount2_template_experiment(project_id, local_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset genes\n",
    "For our downstream analysis we will be comparing our set of differentially expression genes against the set found in [Crow et. al. publication](https://www.pnas.org/content/pnas/116/13/6491.full.pdf), therefore we will limit our genes to include only those genes shared between our starting set of genes and those in publication. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Order</th>\n",
       "      <th>Gene_EntrezID</th>\n",
       "      <th>N_HitLists</th>\n",
       "      <th>DE_Prior_Rank</th>\n",
       "      <th>Gene_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>7503</td>\n",
       "      <td>79</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>8653</td>\n",
       "      <td>64</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>DDX3Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>9086</td>\n",
       "      <td>62</td>\n",
       "      <td>0.999870</td>\n",
       "      <td>EIF1AY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>8284</td>\n",
       "      <td>52</td>\n",
       "      <td>0.999870</td>\n",
       "      <td>KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>8287</td>\n",
       "      <td>46</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>USP9Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene_Order  Gene_EntrezID  N_HitLists  DE_Prior_Rank Gene_Name\n",
       "0           1           7503          79       1.000000      XIST\n",
       "1           2           8653          64       0.999948     DDX3Y\n",
       "2           3           9086          62       0.999870    EIF1AY\n",
       "3           4           8284          52       0.999870     KDM5D\n",
       "4           5           8287          46       0.999791     USP9Y"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get generic genes identified by Crow et. al.\n",
    "DE_prior_file = \"https://raw.githubusercontent.com/maggiecrow/DEprior/master/DE_Prior.txt\"\n",
    "\n",
    "DE_prior = pd.read_csv(DE_prior_file,\n",
    "                       header=0,\n",
    "                       sep=\"\\t\")\n",
    "\n",
    "DE_prior.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of published generic genes\n",
    "published_generic_genes = list(DE_prior['Gene_Name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of our genes\n",
    "# Load real template experiment\n",
    "template_data_file = params['template_data_file']\n",
    "\n",
    "# Read template data\n",
    "template_data = pd.read_csv(\n",
    "    template_data_file,\n",
    "    header=0,\n",
    "    sep='\\t',\n",
    "    index_col=0)\n",
    "\n",
    "our_gene_ids = list(template_data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# File mapping ensembl ids to hgnc symbols\n",
    "gene_id_file = os.path.join(\n",
    "    local_dir,\n",
    "    \"ensembl_hgnc_mapping.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "suppressWarnings(library(\"biomaRt\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R -i template_data_file -i gene_id_file\n",
    "\n",
    "# Get mapping between ensembl gene ids (ours) to HGNC gene symbols (published)\n",
    "\n",
    "source('../functions/process_names.R')\n",
    "\n",
    "if (file.exists(gene_id_file) == FALSE){\n",
    "    gene_id_mapping <- get_ensembl_symbol_mapping(template_data_file, gene_id_file)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(57210, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000003137</td>\n",
       "      <td>CYP26B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000003249</td>\n",
       "      <td>DBNDD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000004799</td>\n",
       "      <td>PDK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000006062</td>\n",
       "      <td>MAP3K14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id hgnc_symbol\n",
       "1  ENSG00000002330         BAD\n",
       "2  ENSG00000003137     CYP26B1\n",
       "3  ENSG00000003249      DBNDD1\n",
       "4  ENSG00000004799        PDK4\n",
       "5  ENSG00000006062     MAP3K14"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read gene id mapping\n",
    "gene_id_mapping = pd.read_csv(\n",
    "        gene_id_file,\n",
    "        header=0,\n",
    "        sep='\\t',\n",
    "        index_col=0)\n",
    "\n",
    "print(gene_id_mapping.shape)\n",
    "gene_id_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58037, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_version</th>\n",
       "      <th>ensembl_parsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ensembl_version   ensembl_parsed\n",
       "0  ENSG00000000003.14  ENSG00000000003\n",
       "1   ENSG00000000005.5  ENSG00000000005\n",
       "2  ENSG00000000419.12  ENSG00000000419\n",
       "3  ENSG00000000457.13  ENSG00000000457\n",
       "4  ENSG00000000460.16  ENSG00000000460"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get mapping between ensembl ids with and without version numbers\n",
    "# Expressiond data uses ensembl gene ids with version number \n",
    "ensembl_gene_ids = pd.DataFrame(data={'ensembl_version': our_gene_ids,\n",
    "                                      'ensembl_parsed': [gene_id.split('.')[0] for gene_id in our_gene_ids]})\n",
    "\n",
    "print(ensembl_gene_ids.shape)\n",
    "ensembl_gene_ids.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58129, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>ensembl_parsed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_version</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000002330.13</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003137.8</th>\n",
       "      <td>ENSG00000003137</td>\n",
       "      <td>CYP26B1</td>\n",
       "      <td>ENSG00000003137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003249.13</th>\n",
       "      <td>ENSG00000003249</td>\n",
       "      <td>DBNDD1</td>\n",
       "      <td>ENSG00000003249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000004799.7</th>\n",
       "      <td>ENSG00000004799</td>\n",
       "      <td>PDK4</td>\n",
       "      <td>ENSG00000004799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000006062.13</th>\n",
       "      <td>ENSG00000006062</td>\n",
       "      <td>MAP3K14</td>\n",
       "      <td>ENSG00000006062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ensembl_gene_id hgnc_symbol   ensembl_parsed\n",
       "ensembl_version                                                 \n",
       "ENSG00000002330.13  ENSG00000002330         BAD  ENSG00000002330\n",
       "ENSG00000003137.8   ENSG00000003137     CYP26B1  ENSG00000003137\n",
       "ENSG00000003249.13  ENSG00000003249      DBNDD1  ENSG00000003249\n",
       "ENSG00000004799.7   ENSG00000004799        PDK4  ENSG00000004799\n",
       "ENSG00000006062.13  ENSG00000006062     MAP3K14  ENSG00000006062"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map ensembl ids with version number to gene_id_mapping_df\n",
    "gene_id_mapping = pd.merge(gene_id_mapping, \n",
    "                           ensembl_gene_ids, \n",
    "                           left_on='ensembl_gene_id',\n",
    "                           right_on='ensembl_parsed', \n",
    "                           how='outer')\n",
    "\n",
    "print(gene_id_mapping.shape)\n",
    "gene_id_mapping.set_index('ensembl_version', inplace=True)\n",
    "gene_id_mapping.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since this experiment contains both RNA-seq and smRNA-seq samples which are in different ranges so we will drop smRNA samples so that samples are within the same range. The analysis identifying these two subsets of samples can be found in this [notebook](0_explore_input_data.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace ensembl ids with gene symbols\n",
    "template_data = process.replace_ensembl_ids(template_data,\n",
    "                                            gene_id_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>NFYA</th>\n",
       "      <th>...</th>\n",
       "      <th>MIR6830</th>\n",
       "      <th>MIR548AH</th>\n",
       "      <th>MYOCOS</th>\n",
       "      <th>MIR522</th>\n",
       "      <th>MIR6715B</th>\n",
       "      <th>MIR3116-2</th>\n",
       "      <th>MIR3202-2</th>\n",
       "      <th>HSFX3</th>\n",
       "      <th>VINAC1P</th>\n",
       "      <th>MIR4481</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493937</th>\n",
       "      <td>3.800154</td>\n",
       "      <td>0.018190</td>\n",
       "      <td>9.128459</td>\n",
       "      <td>1.585803</td>\n",
       "      <td>0.713952</td>\n",
       "      <td>84.990542</td>\n",
       "      <td>19.943002</td>\n",
       "      <td>12.065586</td>\n",
       "      <td>2.498822</td>\n",
       "      <td>7.713002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.135649</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.588513</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493938</th>\n",
       "      <td>3.798473</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.157669</td>\n",
       "      <td>1.456017</td>\n",
       "      <td>0.738967</td>\n",
       "      <td>85.788323</td>\n",
       "      <td>19.621252</td>\n",
       "      <td>11.738855</td>\n",
       "      <td>2.435308</td>\n",
       "      <td>7.125808</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005417</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.127873</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.452511</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493939</th>\n",
       "      <td>20.222134</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.522072</td>\n",
       "      <td>3.944919</td>\n",
       "      <td>2.512265</td>\n",
       "      <td>8.072948</td>\n",
       "      <td>47.246078</td>\n",
       "      <td>18.866957</td>\n",
       "      <td>2.443230</td>\n",
       "      <td>12.143670</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011112</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.446833</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493940</th>\n",
       "      <td>19.988329</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17.140304</td>\n",
       "      <td>3.977217</td>\n",
       "      <td>2.378849</td>\n",
       "      <td>7.599225</td>\n",
       "      <td>46.594458</td>\n",
       "      <td>18.921190</td>\n",
       "      <td>2.448276</td>\n",
       "      <td>12.048960</td>\n",
       "      <td>...</td>\n",
       "      <td>0.147412</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.037886</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.07756</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.462295</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.168159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493941</th>\n",
       "      <td>3.239231</td>\n",
       "      <td>0.019398</td>\n",
       "      <td>9.069316</td>\n",
       "      <td>2.181582</td>\n",
       "      <td>1.008155</td>\n",
       "      <td>17.675885</td>\n",
       "      <td>19.371137</td>\n",
       "      <td>11.759756</td>\n",
       "      <td>1.970725</td>\n",
       "      <td>6.740705</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002860</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156220</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.070955</td>\n",
       "      <td>0.739403</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37911 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TSPAN6      TNMD       DPM1     SCYL3  C1orf112        FGR  \\\n",
       "SRR493937   3.800154  0.018190   9.128459  1.585803  0.713952  84.990542   \n",
       "SRR493938   3.798473  0.000000   9.157669  1.456017  0.738967  85.788323   \n",
       "SRR493939  20.222134  0.000000  15.522072  3.944919  2.512265   8.072948   \n",
       "SRR493940  19.988329  0.000000  17.140304  3.977217  2.378849   7.599225   \n",
       "SRR493941   3.239231  0.019398   9.069316  2.181582  1.008155  17.675885   \n",
       "\n",
       "                 CFH      FUCA2      GCLC       NFYA    ...      MIR6830  \\\n",
       "SRR493937  19.943002  12.065586  2.498822   7.713002    ...     0.000000   \n",
       "SRR493938  19.621252  11.738855  2.435308   7.125808    ...     0.005417   \n",
       "SRR493939  47.246078  18.866957  2.443230  12.143670    ...     0.000000   \n",
       "SRR493940  46.594458  18.921190  2.448276  12.048960    ...     0.147412   \n",
       "SRR493941  19.371137  11.759756  1.970725   6.740705    ...     0.002860   \n",
       "\n",
       "           MIR548AH    MYOCOS  MIR522  MIR6715B  MIR3116-2  MIR3202-2  \\\n",
       "SRR493937       0.0  0.135649     0.0       0.0    0.00000   0.000000   \n",
       "SRR493938       0.0  0.127873     0.0       0.0    0.00000   0.000000   \n",
       "SRR493939       0.0  0.011112     0.0       0.0    0.00000   0.000000   \n",
       "SRR493940       0.0  0.037886     0.0       0.0    0.07756   0.000000   \n",
       "SRR493941       0.0  0.156220     0.0       0.0    0.00000   0.070955   \n",
       "\n",
       "              HSFX3  VINAC1P   MIR4481  \n",
       "SRR493937  0.588513      0.0  0.000000  \n",
       "SRR493938  0.452511      0.0  0.000000  \n",
       "SRR493939  0.446833      0.0  0.000000  \n",
       "SRR493940  0.462295      0.0  0.168159  \n",
       "SRR493941  0.739403      0.0  0.000000  \n",
       "\n",
       "[5 rows x 37911 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17785\n"
     ]
    }
   ],
   "source": [
    "# Get intersection of gene lists\n",
    "our_gene_ids_hgnc = template_data.columns\n",
    "shared_genes_hgnc = list(set(our_gene_ids_hgnc).intersection(published_generic_genes))\n",
    "print(len(shared_genes_hgnc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save shared genes\n",
    "shared_genes_file = os.path.join(\n",
    "    local_dir,\n",
    "    \"shared_gene_ids.pickle\")\n",
    "\n",
    "outfile = open(shared_genes_file,'wb')\n",
    "pickle.dump(shared_genes_hgnc,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop smRNA samples so that samples are within the same range\n",
    "smRNA_samples = [\"SRR493961\",\n",
    "                 \"SRR493962\",\n",
    "                 \"SRR493963\",\n",
    "                 \"SRR493964\",\n",
    "                 \"SRR493965\",\n",
    "                 \"SRR493966\",\n",
    "                 \"SRR493967\",\n",
    "                 \"SRR493968\",\n",
    "                 \"SRR493969\",\n",
    "                 \"SRR493970\",\n",
    "                 \"SRR493971\",\n",
    "                 \"SRR493972\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop samples\n",
    "template_data = template_data.drop(smRNA_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24, 17788)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ARHGAP45</th>\n",
       "      <th>ZNF610</th>\n",
       "      <th>PPP1R3B</th>\n",
       "      <th>RBBP7</th>\n",
       "      <th>CCDC185</th>\n",
       "      <th>GPATCH3</th>\n",
       "      <th>C1orf159</th>\n",
       "      <th>DECR2</th>\n",
       "      <th>NAA11</th>\n",
       "      <th>KRTAP4-6</th>\n",
       "      <th>...</th>\n",
       "      <th>FBXL6</th>\n",
       "      <th>UBC</th>\n",
       "      <th>CD1A</th>\n",
       "      <th>OLFML2B</th>\n",
       "      <th>CACNA2D4</th>\n",
       "      <th>EID1</th>\n",
       "      <th>CHRNA10</th>\n",
       "      <th>SCYL2</th>\n",
       "      <th>SCRT1</th>\n",
       "      <th>POLR2L</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493937</th>\n",
       "      <td>35.762946</td>\n",
       "      <td>0.271364</td>\n",
       "      <td>7.859956</td>\n",
       "      <td>7.288771</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.961903</td>\n",
       "      <td>2.380899</td>\n",
       "      <td>2.577827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013790</td>\n",
       "      <td>...</td>\n",
       "      <td>7.551714</td>\n",
       "      <td>772.176176</td>\n",
       "      <td>1.139391</td>\n",
       "      <td>7.299552</td>\n",
       "      <td>0.571510</td>\n",
       "      <td>30.028322</td>\n",
       "      <td>1.354713</td>\n",
       "      <td>5.571204</td>\n",
       "      <td>0.010249</td>\n",
       "      <td>75.147459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493938</th>\n",
       "      <td>35.645899</td>\n",
       "      <td>0.201617</td>\n",
       "      <td>7.365964</td>\n",
       "      <td>7.483894</td>\n",
       "      <td>0.012276</td>\n",
       "      <td>3.871496</td>\n",
       "      <td>2.049562</td>\n",
       "      <td>2.116853</td>\n",
       "      <td>0.010950</td>\n",
       "      <td>0.013837</td>\n",
       "      <td>...</td>\n",
       "      <td>7.729311</td>\n",
       "      <td>766.487971</td>\n",
       "      <td>1.244618</td>\n",
       "      <td>7.350872</td>\n",
       "      <td>0.576157</td>\n",
       "      <td>31.325761</td>\n",
       "      <td>1.232702</td>\n",
       "      <td>5.568605</td>\n",
       "      <td>0.024432</td>\n",
       "      <td>73.244201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493939</th>\n",
       "      <td>12.084931</td>\n",
       "      <td>1.146171</td>\n",
       "      <td>7.689973</td>\n",
       "      <td>16.502015</td>\n",
       "      <td>0.115370</td>\n",
       "      <td>4.096778</td>\n",
       "      <td>4.413026</td>\n",
       "      <td>8.868962</td>\n",
       "      <td>0.010627</td>\n",
       "      <td>0.006312</td>\n",
       "      <td>...</td>\n",
       "      <td>18.543304</td>\n",
       "      <td>517.161381</td>\n",
       "      <td>4.463290</td>\n",
       "      <td>5.384466</td>\n",
       "      <td>0.387584</td>\n",
       "      <td>39.226702</td>\n",
       "      <td>0.495944</td>\n",
       "      <td>13.582267</td>\n",
       "      <td>0.065016</td>\n",
       "      <td>100.520026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493940</th>\n",
       "      <td>12.239497</td>\n",
       "      <td>1.104245</td>\n",
       "      <td>7.734640</td>\n",
       "      <td>16.218601</td>\n",
       "      <td>0.139438</td>\n",
       "      <td>4.838954</td>\n",
       "      <td>4.288688</td>\n",
       "      <td>8.957386</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>17.779922</td>\n",
       "      <td>518.814696</td>\n",
       "      <td>4.515256</td>\n",
       "      <td>6.079031</td>\n",
       "      <td>0.448529</td>\n",
       "      <td>39.771272</td>\n",
       "      <td>0.668580</td>\n",
       "      <td>13.771750</td>\n",
       "      <td>0.068446</td>\n",
       "      <td>94.962601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493941</th>\n",
       "      <td>19.122830</td>\n",
       "      <td>0.722249</td>\n",
       "      <td>6.414456</td>\n",
       "      <td>7.214821</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.727227</td>\n",
       "      <td>2.236915</td>\n",
       "      <td>3.578029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>7.752204</td>\n",
       "      <td>356.842882</td>\n",
       "      <td>0.599057</td>\n",
       "      <td>6.977339</td>\n",
       "      <td>0.539968</td>\n",
       "      <td>35.109401</td>\n",
       "      <td>0.518152</td>\n",
       "      <td>5.237986</td>\n",
       "      <td>0.005351</td>\n",
       "      <td>104.470958</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17788 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ARHGAP45    ZNF610   PPP1R3B      RBBP7   CCDC185   GPATCH3  \\\n",
       "SRR493937  35.762946  0.271364  7.859956   7.288771  0.000000  3.961903   \n",
       "SRR493938  35.645899  0.201617  7.365964   7.483894  0.012276  3.871496   \n",
       "SRR493939  12.084931  1.146171  7.689973  16.502015  0.115370  4.096778   \n",
       "SRR493940  12.239497  1.104245  7.734640  16.218601  0.139438  4.838954   \n",
       "SRR493941  19.122830  0.722249  6.414456   7.214821  0.000000  3.727227   \n",
       "\n",
       "           C1orf159     DECR2     NAA11  KRTAP4-6     ...          FBXL6  \\\n",
       "SRR493937  2.380899  2.577827  0.000000  0.013790     ...       7.551714   \n",
       "SRR493938  2.049562  2.116853  0.010950  0.013837     ...       7.729311   \n",
       "SRR493939  4.413026  8.868962  0.010627  0.006312     ...      18.543304   \n",
       "SRR493940  4.288688  8.957386  0.000000  0.000000     ...      17.779922   \n",
       "SRR493941  2.236915  3.578029  0.000000  0.000000     ...       7.752204   \n",
       "\n",
       "                  UBC      CD1A   OLFML2B  CACNA2D4       EID1   CHRNA10  \\\n",
       "SRR493937  772.176176  1.139391  7.299552  0.571510  30.028322  1.354713   \n",
       "SRR493938  766.487971  1.244618  7.350872  0.576157  31.325761  1.232702   \n",
       "SRR493939  517.161381  4.463290  5.384466  0.387584  39.226702  0.495944   \n",
       "SRR493940  518.814696  4.515256  6.079031  0.448529  39.771272  0.668580   \n",
       "SRR493941  356.842882  0.599057  6.977339  0.539968  35.109401  0.518152   \n",
       "\n",
       "               SCYL2     SCRT1      POLR2L  \n",
       "SRR493937   5.571204  0.010249   75.147459  \n",
       "SRR493938   5.568605  0.024432   73.244201  \n",
       "SRR493939  13.582267  0.065016  100.520026  \n",
       "SRR493940  13.771750  0.068446   94.962601  \n",
       "SRR493941   5.237986  0.005351  104.470958  \n",
       "\n",
       "[5 rows x 17788 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop genes\n",
    "template_data = template_data[shared_genes_hgnc]\n",
    "\n",
    "print(template_data.shape)\n",
    "template_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    }
   ],
   "source": [
    "print(len(template_data.columns) - len(shared_genes_hgnc))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Note:* There is a difference in the number of `shared_genes_hgnc` and genes in the template experiment because 3 genes have 2 different ensembl gene ids have map to the same hgnc symbol (one forward, one reverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save \n",
    "template_data.to_csv(template_data_file, float_format='%.5f', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalize compendium "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load real gene expression data\n",
    "original_compendium_file = params['compendium_data_file']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1552, 17785)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GIN1</th>\n",
       "      <th>ATP7B</th>\n",
       "      <th>STX2</th>\n",
       "      <th>CALML4</th>\n",
       "      <th>IL33</th>\n",
       "      <th>NDUFA2</th>\n",
       "      <th>NIPA2</th>\n",
       "      <th>CC2D1B</th>\n",
       "      <th>KRT36</th>\n",
       "      <th>SLC11A2</th>\n",
       "      <th>...</th>\n",
       "      <th>FHL1</th>\n",
       "      <th>FAM217B</th>\n",
       "      <th>SLC35A3</th>\n",
       "      <th>FOXN3-AS2</th>\n",
       "      <th>ADA</th>\n",
       "      <th>MAP2K3</th>\n",
       "      <th>NALT1</th>\n",
       "      <th>TAGLN2</th>\n",
       "      <th>BNIPL</th>\n",
       "      <th>LINC00347</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR1604987</th>\n",
       "      <td>2.390</td>\n",
       "      <td>0.718</td>\n",
       "      <td>1.672</td>\n",
       "      <td>1.643</td>\n",
       "      <td>0.007</td>\n",
       "      <td>24.263</td>\n",
       "      <td>10.499</td>\n",
       "      <td>9.866</td>\n",
       "      <td>0.001</td>\n",
       "      <td>10.171</td>\n",
       "      <td>...</td>\n",
       "      <td>0.774</td>\n",
       "      <td>4.924</td>\n",
       "      <td>2.727</td>\n",
       "      <td>0.000</td>\n",
       "      <td>6.939</td>\n",
       "      <td>32.849</td>\n",
       "      <td>0.155</td>\n",
       "      <td>237.958</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604988</th>\n",
       "      <td>2.462</td>\n",
       "      <td>0.711</td>\n",
       "      <td>1.455</td>\n",
       "      <td>1.586</td>\n",
       "      <td>0.037</td>\n",
       "      <td>24.275</td>\n",
       "      <td>11.261</td>\n",
       "      <td>9.867</td>\n",
       "      <td>0.024</td>\n",
       "      <td>10.181</td>\n",
       "      <td>...</td>\n",
       "      <td>0.702</td>\n",
       "      <td>4.955</td>\n",
       "      <td>2.866</td>\n",
       "      <td>0.013</td>\n",
       "      <td>6.641</td>\n",
       "      <td>33.477</td>\n",
       "      <td>0.209</td>\n",
       "      <td>238.721</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604989</th>\n",
       "      <td>1.603</td>\n",
       "      <td>0.814</td>\n",
       "      <td>2.398</td>\n",
       "      <td>2.534</td>\n",
       "      <td>0.010</td>\n",
       "      <td>14.434</td>\n",
       "      <td>15.402</td>\n",
       "      <td>7.834</td>\n",
       "      <td>0.001</td>\n",
       "      <td>8.509</td>\n",
       "      <td>...</td>\n",
       "      <td>0.689</td>\n",
       "      <td>4.520</td>\n",
       "      <td>4.233</td>\n",
       "      <td>0.000</td>\n",
       "      <td>7.546</td>\n",
       "      <td>28.347</td>\n",
       "      <td>0.123</td>\n",
       "      <td>149.598</td>\n",
       "      <td>0.144</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604990</th>\n",
       "      <td>1.872</td>\n",
       "      <td>0.748</td>\n",
       "      <td>2.193</td>\n",
       "      <td>2.294</td>\n",
       "      <td>0.020</td>\n",
       "      <td>14.673</td>\n",
       "      <td>14.667</td>\n",
       "      <td>7.892</td>\n",
       "      <td>0.003</td>\n",
       "      <td>8.536</td>\n",
       "      <td>...</td>\n",
       "      <td>0.802</td>\n",
       "      <td>4.469</td>\n",
       "      <td>4.414</td>\n",
       "      <td>0.000</td>\n",
       "      <td>7.558</td>\n",
       "      <td>27.931</td>\n",
       "      <td>0.193</td>\n",
       "      <td>152.254</td>\n",
       "      <td>0.183</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604991</th>\n",
       "      <td>0.786</td>\n",
       "      <td>1.362</td>\n",
       "      <td>6.530</td>\n",
       "      <td>1.390</td>\n",
       "      <td>0.012</td>\n",
       "      <td>18.238</td>\n",
       "      <td>17.125</td>\n",
       "      <td>10.627</td>\n",
       "      <td>0.001</td>\n",
       "      <td>3.683</td>\n",
       "      <td>...</td>\n",
       "      <td>7.346</td>\n",
       "      <td>5.528</td>\n",
       "      <td>2.535</td>\n",
       "      <td>0.000</td>\n",
       "      <td>9.323</td>\n",
       "      <td>32.487</td>\n",
       "      <td>0.311</td>\n",
       "      <td>241.787</td>\n",
       "      <td>0.105</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17785 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             GIN1  ATP7B   STX2  CALML4   IL33  NDUFA2   NIPA2  CC2D1B  KRT36  \\\n",
       "SRR1604987  2.390  0.718  1.672   1.643  0.007  24.263  10.499   9.866  0.001   \n",
       "SRR1604988  2.462  0.711  1.455   1.586  0.037  24.275  11.261   9.867  0.024   \n",
       "SRR1604989  1.603  0.814  2.398   2.534  0.010  14.434  15.402   7.834  0.001   \n",
       "SRR1604990  1.872  0.748  2.193   2.294  0.020  14.673  14.667   7.892  0.003   \n",
       "SRR1604991  0.786  1.362  6.530   1.390  0.012  18.238  17.125  10.627  0.001   \n",
       "\n",
       "            SLC11A2    ...       FHL1  FAM217B  SLC35A3  FOXN3-AS2    ADA  \\\n",
       "SRR1604987   10.171    ...      0.774    4.924    2.727      0.000  6.939   \n",
       "SRR1604988   10.181    ...      0.702    4.955    2.866      0.013  6.641   \n",
       "SRR1604989    8.509    ...      0.689    4.520    4.233      0.000  7.546   \n",
       "SRR1604990    8.536    ...      0.802    4.469    4.414      0.000  7.558   \n",
       "SRR1604991    3.683    ...      7.346    5.528    2.535      0.000  9.323   \n",
       "\n",
       "            MAP2K3  NALT1   TAGLN2  BNIPL  LINC00347  \n",
       "SRR1604987  32.849  0.155  237.958  0.070      0.000  \n",
       "SRR1604988  33.477  0.209  238.721  0.081      0.000  \n",
       "SRR1604989  28.347  0.123  149.598  0.144      0.000  \n",
       "SRR1604990  27.931  0.193  152.254  0.183      0.000  \n",
       "SRR1604991  32.487  0.311  241.787  0.105      0.004  \n",
       "\n",
       "[5 rows x 17785 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read data\n",
    "original_compendium = pd.read_table(\n",
    "    original_compendium_file,\n",
    "    header=0,\n",
    "    sep='\\t',\n",
    "    index_col=0)\n",
    "\n",
    "print(original_compendium.shape)\n",
    "original_compendium.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace ensembl ids with gene symbols\n",
    "#original_compendium = process.replace_ensembl_ids(original_compendium,\n",
    "#                                                gene_id_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ARHGAP45</th>\n",
       "      <th>ZNF610</th>\n",
       "      <th>PPP1R3B</th>\n",
       "      <th>RBBP7</th>\n",
       "      <th>CCDC185</th>\n",
       "      <th>GPATCH3</th>\n",
       "      <th>C1orf159</th>\n",
       "      <th>DECR2</th>\n",
       "      <th>NAA11</th>\n",
       "      <th>KRTAP4-6</th>\n",
       "      <th>...</th>\n",
       "      <th>FBXL6</th>\n",
       "      <th>UBC</th>\n",
       "      <th>CD1A</th>\n",
       "      <th>OLFML2B</th>\n",
       "      <th>CACNA2D4</th>\n",
       "      <th>EID1</th>\n",
       "      <th>CHRNA10</th>\n",
       "      <th>SCYL2</th>\n",
       "      <th>SCRT1</th>\n",
       "      <th>POLR2L</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR1604987</th>\n",
       "      <td>4.686</td>\n",
       "      <td>2.914</td>\n",
       "      <td>6.349</td>\n",
       "      <td>12.617</td>\n",
       "      <td>0.034</td>\n",
       "      <td>4.167</td>\n",
       "      <td>1.487</td>\n",
       "      <td>4.824</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.107</td>\n",
       "      <td>...</td>\n",
       "      <td>6.677</td>\n",
       "      <td>418.218</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.050</td>\n",
       "      <td>40.078</td>\n",
       "      <td>0.194</td>\n",
       "      <td>9.640</td>\n",
       "      <td>0.012</td>\n",
       "      <td>52.679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604988</th>\n",
       "      <td>4.968</td>\n",
       "      <td>2.993</td>\n",
       "      <td>6.243</td>\n",
       "      <td>12.530</td>\n",
       "      <td>0.011</td>\n",
       "      <td>4.031</td>\n",
       "      <td>1.397</td>\n",
       "      <td>4.778</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.065</td>\n",
       "      <td>...</td>\n",
       "      <td>6.247</td>\n",
       "      <td>413.291</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.042</td>\n",
       "      <td>40.601</td>\n",
       "      <td>0.090</td>\n",
       "      <td>9.914</td>\n",
       "      <td>0.007</td>\n",
       "      <td>51.503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604989</th>\n",
       "      <td>5.910</td>\n",
       "      <td>2.855</td>\n",
       "      <td>13.585</td>\n",
       "      <td>16.710</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3.882</td>\n",
       "      <td>1.172</td>\n",
       "      <td>2.845</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.733</td>\n",
       "      <td>297.397</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.003</td>\n",
       "      <td>38.530</td>\n",
       "      <td>0.154</td>\n",
       "      <td>10.461</td>\n",
       "      <td>0.002</td>\n",
       "      <td>39.457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604990</th>\n",
       "      <td>5.737</td>\n",
       "      <td>2.884</td>\n",
       "      <td>14.170</td>\n",
       "      <td>16.927</td>\n",
       "      <td>0.016</td>\n",
       "      <td>3.883</td>\n",
       "      <td>1.199</td>\n",
       "      <td>2.745</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.552</td>\n",
       "      <td>298.423</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.012</td>\n",
       "      <td>37.747</td>\n",
       "      <td>0.114</td>\n",
       "      <td>10.902</td>\n",
       "      <td>0.000</td>\n",
       "      <td>38.281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604991</th>\n",
       "      <td>2.580</td>\n",
       "      <td>2.653</td>\n",
       "      <td>8.768</td>\n",
       "      <td>17.953</td>\n",
       "      <td>0.067</td>\n",
       "      <td>4.608</td>\n",
       "      <td>1.624</td>\n",
       "      <td>3.556</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.893</td>\n",
       "      <td>281.337</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0.012</td>\n",
       "      <td>64.853</td>\n",
       "      <td>0.215</td>\n",
       "      <td>8.097</td>\n",
       "      <td>0.010</td>\n",
       "      <td>83.953</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17785 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ARHGAP45  ZNF610  PPP1R3B   RBBP7  CCDC185  GPATCH3  C1orf159  \\\n",
       "SRR1604987     4.686   2.914    6.349  12.617    0.034    4.167     1.487   \n",
       "SRR1604988     4.968   2.993    6.243  12.530    0.011    4.031     1.397   \n",
       "SRR1604989     5.910   2.855   13.585  16.710    0.000    3.882     1.172   \n",
       "SRR1604990     5.737   2.884   14.170  16.927    0.016    3.883     1.199   \n",
       "SRR1604991     2.580   2.653    8.768  17.953    0.067    4.608     1.624   \n",
       "\n",
       "            DECR2  NAA11  KRTAP4-6   ...    FBXL6      UBC  CD1A  OLFML2B  \\\n",
       "SRR1604987  4.824  0.000     0.107   ...    6.677  418.218   0.0    0.070   \n",
       "SRR1604988  4.778  0.006     0.065   ...    6.247  413.291   0.0    0.040   \n",
       "SRR1604989  2.845  0.000     0.000   ...    5.733  297.397   0.0    0.019   \n",
       "SRR1604990  2.745  0.000     0.000   ...    5.552  298.423   0.0    0.038   \n",
       "SRR1604991  3.556  0.000     0.000   ...    5.893  281.337   0.0    0.986   \n",
       "\n",
       "            CACNA2D4    EID1  CHRNA10   SCYL2  SCRT1  POLR2L  \n",
       "SRR1604987     0.050  40.078    0.194   9.640  0.012  52.679  \n",
       "SRR1604988     0.042  40.601    0.090   9.914  0.007  51.503  \n",
       "SRR1604989     0.003  38.530    0.154  10.461  0.002  39.457  \n",
       "SRR1604990     0.012  37.747    0.114  10.902  0.000  38.281  \n",
       "SRR1604991     0.012  64.853    0.215   8.097  0.010  83.953  \n",
       "\n",
       "[5 rows x 17785 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop genes\n",
    "original_compendium = original_compendium[shared_genes_hgnc]\n",
    "\n",
    "original_compendium.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1552, 17785)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ARHGAP45</th>\n",
       "      <th>ZNF610</th>\n",
       "      <th>PPP1R3B</th>\n",
       "      <th>RBBP7</th>\n",
       "      <th>CCDC185</th>\n",
       "      <th>GPATCH3</th>\n",
       "      <th>C1orf159</th>\n",
       "      <th>DECR2</th>\n",
       "      <th>NAA11</th>\n",
       "      <th>KRTAP4-6</th>\n",
       "      <th>...</th>\n",
       "      <th>FBXL6</th>\n",
       "      <th>UBC</th>\n",
       "      <th>CD1A</th>\n",
       "      <th>OLFML2B</th>\n",
       "      <th>CACNA2D4</th>\n",
       "      <th>EID1</th>\n",
       "      <th>CHRNA10</th>\n",
       "      <th>SCYL2</th>\n",
       "      <th>SCRT1</th>\n",
       "      <th>POLR2L</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR1604987</th>\n",
       "      <td>0.008520</td>\n",
       "      <td>0.285798</td>\n",
       "      <td>0.049474</td>\n",
       "      <td>0.123779</td>\n",
       "      <td>0.001503</td>\n",
       "      <td>0.111375</td>\n",
       "      <td>0.045315</td>\n",
       "      <td>0.141824</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001282</td>\n",
       "      <td>...</td>\n",
       "      <td>0.089699</td>\n",
       "      <td>0.198485</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000616</td>\n",
       "      <td>0.002863</td>\n",
       "      <td>0.069023</td>\n",
       "      <td>0.011408</td>\n",
       "      <td>0.372906</td>\n",
       "      <td>0.000190</td>\n",
       "      <td>0.017902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604988</th>\n",
       "      <td>0.009032</td>\n",
       "      <td>0.293546</td>\n",
       "      <td>0.048648</td>\n",
       "      <td>0.122925</td>\n",
       "      <td>0.000486</td>\n",
       "      <td>0.107740</td>\n",
       "      <td>0.042572</td>\n",
       "      <td>0.140472</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>...</td>\n",
       "      <td>0.083922</td>\n",
       "      <td>0.196147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000352</td>\n",
       "      <td>0.002405</td>\n",
       "      <td>0.069924</td>\n",
       "      <td>0.005292</td>\n",
       "      <td>0.383505</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>0.017503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604989</th>\n",
       "      <td>0.010745</td>\n",
       "      <td>0.280012</td>\n",
       "      <td>0.105860</td>\n",
       "      <td>0.163933</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.103758</td>\n",
       "      <td>0.035715</td>\n",
       "      <td>0.083642</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.077017</td>\n",
       "      <td>0.141144</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000172</td>\n",
       "      <td>0.066357</td>\n",
       "      <td>0.009056</td>\n",
       "      <td>0.404665</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.013409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604990</th>\n",
       "      <td>0.010430</td>\n",
       "      <td>0.282856</td>\n",
       "      <td>0.110418</td>\n",
       "      <td>0.166062</td>\n",
       "      <td>0.000708</td>\n",
       "      <td>0.103785</td>\n",
       "      <td>0.036538</td>\n",
       "      <td>0.080702</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074586</td>\n",
       "      <td>0.141631</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.000687</td>\n",
       "      <td>0.065009</td>\n",
       "      <td>0.006704</td>\n",
       "      <td>0.421724</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604991</th>\n",
       "      <td>0.004691</td>\n",
       "      <td>0.260200</td>\n",
       "      <td>0.068324</td>\n",
       "      <td>0.176127</td>\n",
       "      <td>0.002963</td>\n",
       "      <td>0.123162</td>\n",
       "      <td>0.049490</td>\n",
       "      <td>0.104545</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.079167</td>\n",
       "      <td>0.133522</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008671</td>\n",
       "      <td>0.000687</td>\n",
       "      <td>0.111692</td>\n",
       "      <td>0.012643</td>\n",
       "      <td>0.313218</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.028531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17785 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ARHGAP45    ZNF610   PPP1R3B     RBBP7   CCDC185   GPATCH3  \\\n",
       "SRR1604987  0.008520  0.285798  0.049474  0.123779  0.001503  0.111375   \n",
       "SRR1604988  0.009032  0.293546  0.048648  0.122925  0.000486  0.107740   \n",
       "SRR1604989  0.010745  0.280012  0.105860  0.163933  0.000000  0.103758   \n",
       "SRR1604990  0.010430  0.282856  0.110418  0.166062  0.000708  0.103785   \n",
       "SRR1604991  0.004691  0.260200  0.068324  0.176127  0.002963  0.123162   \n",
       "\n",
       "            C1orf159     DECR2     NAA11  KRTAP4-6    ...        FBXL6  \\\n",
       "SRR1604987  0.045315  0.141824  0.000000  0.001282    ...     0.089699   \n",
       "SRR1604988  0.042572  0.140472  0.000105  0.000779    ...     0.083922   \n",
       "SRR1604989  0.035715  0.083642  0.000000  0.000000    ...     0.077017   \n",
       "SRR1604990  0.036538  0.080702  0.000000  0.000000    ...     0.074586   \n",
       "SRR1604991  0.049490  0.104545  0.000000  0.000000    ...     0.079167   \n",
       "\n",
       "                 UBC  CD1A   OLFML2B  CACNA2D4      EID1   CHRNA10     SCYL2  \\\n",
       "SRR1604987  0.198485   0.0  0.000616  0.002863  0.069023  0.011408  0.372906   \n",
       "SRR1604988  0.196147   0.0  0.000352  0.002405  0.069924  0.005292  0.383505   \n",
       "SRR1604989  0.141144   0.0  0.000167  0.000172  0.066357  0.009056  0.404665   \n",
       "SRR1604990  0.141631   0.0  0.000334  0.000687  0.065009  0.006704  0.421724   \n",
       "SRR1604991  0.133522   0.0  0.008671  0.000687  0.111692  0.012643  0.313218   \n",
       "\n",
       "               SCRT1    POLR2L  \n",
       "SRR1604987  0.000190  0.017902  \n",
       "SRR1604988  0.000111  0.017503  \n",
       "SRR1604989  0.000032  0.013409  \n",
       "SRR1604990  0.000000  0.013009  \n",
       "SRR1604991  0.000158  0.028531  \n",
       "\n",
       "[5 rows x 17785 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 0-1 normalize per gene\n",
    "scaler = preprocessing.MinMaxScaler()\n",
    "original_data_scaled = scaler.fit_transform(original_compendium)\n",
    "original_data_scaled_df = pd.DataFrame(original_data_scaled,\n",
    "                                columns=original_compendium.columns,\n",
    "                                index=original_compendium.index)\n",
    "\n",
    "print(original_data_scaled_df.shape)\n",
    "original_data_scaled_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "normalized_data_file = params['normalized_compendium_data_file']\n",
    "\n",
    "original_data_scaled_df.to_csv(\n",
    "    normalized_data_file, float_format='%.3f', sep='\\t')\n",
    "\n",
    "original_compendium.to_csv(\n",
    "    original_compendium_file, float_format='%.3f', sep='\\t')\n",
    "\n",
    "# Save scaler transform\n",
    "scaler_file = params['scaler_transform_file']\n",
    "\n",
    "outfile = open(scaler_file,'wb')\n",
    "pickle.dump(scaler,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train VAE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup directories\n",
    "# Create VAE directories\n",
    "output_dirs = [os.path.join(base_dir, dataset_name, \"models\"),\n",
    "               os.path.join(base_dir, dataset_name, \"logs\")]\n",
    "\n",
    "# Check if analysis output directory exist otherwise create\n",
    "for each_dir in output_dirs:\n",
    "    if os.path.exists(each_dir) == False:\n",
    "        print('creating new directory: {}'.format(each_dir))\n",
    "        os.makedirs(each_dir, exist_ok=True)\n",
    "\n",
    "# Check if NN architecture directory exist otherwise create\n",
    "for each_dir in output_dirs:\n",
    "    new_dir = os.path.join(each_dir, NN_architecture)\n",
    "    if os.path.exists(new_dir) == False:\n",
    "        print('creating new directory: {}'.format(new_dir))\n",
    "        os.makedirs(new_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Train VAE on new compendium data\n",
    "#pipeline.train_vae(config_file,\n",
    "#                   normalized_data_file)"
   ]
  }
 ],
 "metadata": {
  "finalized": {
   "timestamp": 1592246126078,
   "trusted": true
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
