{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process recount2 data\n",
    "This notebook does the following:\n",
    "\n",
    "1. Selects template experiment\n",
    "2. Downloads subset of recount2 data, including the template experiment (50 random experiments + 1 template experiment)\n",
    "3. Train VAE on subset of recount2 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import rpy2\n",
    "import seaborn as sns\n",
    "from sklearn import preprocessing\n",
    "import pickle\n",
    "\n",
    "from ponyo import generate_template_data, utils, pipeline\n",
    "from generic_expression_patterns_modules import process, calc\n",
    "\n",
    "from numpy.random import seed\n",
    "random_state = 123\n",
    "seed(random_state)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(),\"../\"))\n",
    "\n",
    "config_file = os.path.abspath(os.path.join(base_dir,\n",
    "                                           \"config_human.tsv\"))\n",
    "params = utils.read_config(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select template experiment\n",
    "\n",
    "We manually selected bioproject [SRP012656](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37764), which contains primary non-small cell lung adenocarcinoma tumors and adjacent normal tissues of 6 never-smoker Korean female patients with 2 replicates each."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load params\n",
    "local_dir = params[\"local_dir\"]\n",
    "dataset_name = params['dataset_name']\n",
    "NN_architecture = params['NN_architecture']\n",
    "project_id = params['project_id']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download subset of recount2 to use as a compendium\n",
    "The compendium will be composed of random experiments + the selected template experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "# Select 59\n",
    "# Select a\n",
    "# Run one time\n",
    "#if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "#    install.packages(\"BiocManager\")\n",
    "#BiocManager::install(\"derfinder\")\n",
    "#BiocManager::install(\"recount\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: SummarizedExperiment\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomicRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: stats4\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocGenerics\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: parallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: S4Vectors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: IRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomeInfoDb\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Biobase\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: DelayedArray\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: matrixStats\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:genefilter’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Setting options('download.file.method.GEOquery'='auto')\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Setting options('GEOquery.inmemory.gpl'=FALSE)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library('recount')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#%%R -i project_id -i local_dir\n",
    "\n",
    "#source('../generic_expression_patterns_modules/download_recount2_data.R')\n",
    "\n",
    "#get_recount2_compendium(project_id, local_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download expression data for selected project id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: 2020-06-22 15:22:12 downloading file rse_gene.Rdata to SRP012656\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: trying URL 'http://duffel.rail.bio/recount/v2/SRP012656/rse_gene.Rdata'\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Content type 'application/octet-stream'\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning:  length 5841450 bytes (5.6 MB)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: =\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: downloaded 5.6 MB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Loading objects:\n",
       "  rse_gene\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i project_id -i local_dir\n",
    "\n",
    "source('../generic_expression_patterns_modules/download_recount2_data.R')\n",
    "\n",
    "get_recount2_template_experiment(project_id, local_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset genes\n",
    "For our downstream analysis we will be comparing our set of differentially expression genes against the set found in [Crow et. al. publication](https://www.pnas.org/content/pnas/116/13/6491.full.pdf), therefore we will limit our genes to include only those genes shared between our starting set of genes and those in publication. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Order</th>\n",
       "      <th>Gene_EntrezID</th>\n",
       "      <th>N_HitLists</th>\n",
       "      <th>DE_Prior_Rank</th>\n",
       "      <th>Gene_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>7503</td>\n",
       "      <td>79</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>8653</td>\n",
       "      <td>64</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>DDX3Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>9086</td>\n",
       "      <td>62</td>\n",
       "      <td>0.999870</td>\n",
       "      <td>EIF1AY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>8284</td>\n",
       "      <td>52</td>\n",
       "      <td>0.999870</td>\n",
       "      <td>KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>8287</td>\n",
       "      <td>46</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>USP9Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene_Order  Gene_EntrezID  N_HitLists  DE_Prior_Rank Gene_Name\n",
       "0           1           7503          79       1.000000      XIST\n",
       "1           2           8653          64       0.999948     DDX3Y\n",
       "2           3           9086          62       0.999870    EIF1AY\n",
       "3           4           8284          52       0.999870     KDM5D\n",
       "4           5           8287          46       0.999791     USP9Y"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get generic genes identified by Crow et. al.\n",
    "DE_prior_file = \"https://raw.githubusercontent.com/maggiecrow/DEprior/master/DE_Prior.txt\"\n",
    "\n",
    "DE_prior = pd.read_csv(DE_prior_file,\n",
    "                       header=0,\n",
    "                       sep=\"\\t\")\n",
    "\n",
    "DE_prior.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of published generic genes\n",
    "published_generic_genes = list(DE_prior['Gene_Name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of our genes\n",
    "# Load real template experiment\n",
    "template_data_file = params['template_data_file']\n",
    "\n",
    "# Read template data\n",
    "template_data = pd.read_csv(\n",
    "    template_data_file,\n",
    "    header=0,\n",
    "    sep='\\t',\n",
    "    index_col=0)\n",
    "\n",
    "our_gene_ids = list(template_data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# File mapping ensembl ids to hgnc symbols\n",
    "gene_id_file = os.path.join(\n",
    "    local_dir,\n",
    "    \"ensembl_hgnc_mapping.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "suppressWarnings(library(\"biomaRt\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R -i template_data_file -i gene_id_file\n",
    "\n",
    "# Get mapping between ensembl gene ids (ours) to HGNC gene symbols (published)\n",
    "\n",
    "source('../generic_expression_patterns_modules/process_names.R')\n",
    "\n",
    "if (file.exists(gene_id_file) == FALSE){\n",
    "    gene_id_mapping <- get_ensembl_symbol_mapping(template_data_file, gene_id_file)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(57210, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000003137</td>\n",
       "      <td>CYP26B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000003249</td>\n",
       "      <td>DBNDD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000004799</td>\n",
       "      <td>PDK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000006062</td>\n",
       "      <td>MAP3K14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id hgnc_symbol\n",
       "1  ENSG00000002330         BAD\n",
       "2  ENSG00000003137     CYP26B1\n",
       "3  ENSG00000003249      DBNDD1\n",
       "4  ENSG00000004799        PDK4\n",
       "5  ENSG00000006062     MAP3K14"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read gene id mapping\n",
    "gene_id_mapping = pd.read_csv(\n",
    "        gene_id_file,\n",
    "        header=0,\n",
    "        sep='\\t',\n",
    "        index_col=0)\n",
    "\n",
    "print(gene_id_mapping.shape)\n",
    "gene_id_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58037, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_version</th>\n",
       "      <th>ensembl_parsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ensembl_version   ensembl_parsed\n",
       "0  ENSG00000000003.14  ENSG00000000003\n",
       "1   ENSG00000000005.5  ENSG00000000005\n",
       "2  ENSG00000000419.12  ENSG00000000419\n",
       "3  ENSG00000000457.13  ENSG00000000457\n",
       "4  ENSG00000000460.16  ENSG00000000460"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get mapping between ensembl ids with and without version numbers\n",
    "# Expressiond data uses ensembl gene ids with version number \n",
    "ensembl_gene_ids = pd.DataFrame(data={'ensembl_version': our_gene_ids,\n",
    "                                      'ensembl_parsed': [gene_id.split('.')[0] for gene_id in our_gene_ids]})\n",
    "\n",
    "print(ensembl_gene_ids.shape)\n",
    "ensembl_gene_ids.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58129, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>ensembl_parsed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_version</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000002330.13</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003137.8</th>\n",
       "      <td>ENSG00000003137</td>\n",
       "      <td>CYP26B1</td>\n",
       "      <td>ENSG00000003137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003249.13</th>\n",
       "      <td>ENSG00000003249</td>\n",
       "      <td>DBNDD1</td>\n",
       "      <td>ENSG00000003249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000004799.7</th>\n",
       "      <td>ENSG00000004799</td>\n",
       "      <td>PDK4</td>\n",
       "      <td>ENSG00000004799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000006062.13</th>\n",
       "      <td>ENSG00000006062</td>\n",
       "      <td>MAP3K14</td>\n",
       "      <td>ENSG00000006062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ensembl_gene_id hgnc_symbol   ensembl_parsed\n",
       "ensembl_version                                                 \n",
       "ENSG00000002330.13  ENSG00000002330         BAD  ENSG00000002330\n",
       "ENSG00000003137.8   ENSG00000003137     CYP26B1  ENSG00000003137\n",
       "ENSG00000003249.13  ENSG00000003249      DBNDD1  ENSG00000003249\n",
       "ENSG00000004799.7   ENSG00000004799        PDK4  ENSG00000004799\n",
       "ENSG00000006062.13  ENSG00000006062     MAP3K14  ENSG00000006062"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map ensembl ids with version number to gene_id_mapping_df\n",
    "gene_id_mapping = pd.merge(gene_id_mapping, \n",
    "                           ensembl_gene_ids, \n",
    "                           left_on='ensembl_gene_id',\n",
    "                           right_on='ensembl_parsed', \n",
    "                           how='outer')\n",
    "\n",
    "print(gene_id_mapping.shape)\n",
    "gene_id_mapping.set_index('ensembl_version', inplace=True)\n",
    "gene_id_mapping.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since this experiment contains both RNA-seq and smRNA-seq samples which are in different ranges so we will drop smRNA samples so that samples are within the same range. The analysis identifying these two subsets of samples can be found in this [notebook](0_explore_input_data.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/frame.py:3781: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  return super(DataFrame, self).rename(**kwargs)\n"
     ]
    }
   ],
   "source": [
    "# Replace ensembl ids with gene symbols\n",
    "template_data = process.replace_ensembl_ids(template_data,\n",
    "                                            gene_id_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>NFYA</th>\n",
       "      <th>...</th>\n",
       "      <th>MIR6830</th>\n",
       "      <th>MIR548AH</th>\n",
       "      <th>MYOCOS</th>\n",
       "      <th>MIR522</th>\n",
       "      <th>MIR6715B</th>\n",
       "      <th>MIR3116-2</th>\n",
       "      <th>MIR3202-2</th>\n",
       "      <th>HSFX3</th>\n",
       "      <th>VINAC1P</th>\n",
       "      <th>MIR4481</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493937</th>\n",
       "      <td>3.800154</td>\n",
       "      <td>0.018190</td>\n",
       "      <td>9.128459</td>\n",
       "      <td>1.585803</td>\n",
       "      <td>0.713952</td>\n",
       "      <td>84.990542</td>\n",
       "      <td>19.943002</td>\n",
       "      <td>12.065586</td>\n",
       "      <td>2.498822</td>\n",
       "      <td>7.713002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.135649</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.588513</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493938</th>\n",
       "      <td>3.798473</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.157669</td>\n",
       "      <td>1.456017</td>\n",
       "      <td>0.738967</td>\n",
       "      <td>85.788323</td>\n",
       "      <td>19.621252</td>\n",
       "      <td>11.738855</td>\n",
       "      <td>2.435308</td>\n",
       "      <td>7.125808</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005417</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.127873</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.452511</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493939</th>\n",
       "      <td>20.222134</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.522072</td>\n",
       "      <td>3.944919</td>\n",
       "      <td>2.512265</td>\n",
       "      <td>8.072948</td>\n",
       "      <td>47.246078</td>\n",
       "      <td>18.866957</td>\n",
       "      <td>2.443230</td>\n",
       "      <td>12.143670</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011112</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.446833</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493940</th>\n",
       "      <td>19.988329</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17.140304</td>\n",
       "      <td>3.977217</td>\n",
       "      <td>2.378849</td>\n",
       "      <td>7.599225</td>\n",
       "      <td>46.594458</td>\n",
       "      <td>18.921190</td>\n",
       "      <td>2.448276</td>\n",
       "      <td>12.048960</td>\n",
       "      <td>...</td>\n",
       "      <td>0.147412</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.037886</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.07756</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.462295</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.168159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493941</th>\n",
       "      <td>3.239231</td>\n",
       "      <td>0.019398</td>\n",
       "      <td>9.069316</td>\n",
       "      <td>2.181582</td>\n",
       "      <td>1.008155</td>\n",
       "      <td>17.675885</td>\n",
       "      <td>19.371137</td>\n",
       "      <td>11.759756</td>\n",
       "      <td>1.970725</td>\n",
       "      <td>6.740705</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002860</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156220</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.070955</td>\n",
       "      <td>0.739403</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37911 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TSPAN6      TNMD       DPM1     SCYL3  C1orf112        FGR  \\\n",
       "SRR493937   3.800154  0.018190   9.128459  1.585803  0.713952  84.990542   \n",
       "SRR493938   3.798473  0.000000   9.157669  1.456017  0.738967  85.788323   \n",
       "SRR493939  20.222134  0.000000  15.522072  3.944919  2.512265   8.072948   \n",
       "SRR493940  19.988329  0.000000  17.140304  3.977217  2.378849   7.599225   \n",
       "SRR493941   3.239231  0.019398   9.069316  2.181582  1.008155  17.675885   \n",
       "\n",
       "                 CFH      FUCA2      GCLC       NFYA    ...      MIR6830  \\\n",
       "SRR493937  19.943002  12.065586  2.498822   7.713002    ...     0.000000   \n",
       "SRR493938  19.621252  11.738855  2.435308   7.125808    ...     0.005417   \n",
       "SRR493939  47.246078  18.866957  2.443230  12.143670    ...     0.000000   \n",
       "SRR493940  46.594458  18.921190  2.448276  12.048960    ...     0.147412   \n",
       "SRR493941  19.371137  11.759756  1.970725   6.740705    ...     0.002860   \n",
       "\n",
       "           MIR548AH    MYOCOS  MIR522  MIR6715B  MIR3116-2  MIR3202-2  \\\n",
       "SRR493937       0.0  0.135649     0.0       0.0    0.00000   0.000000   \n",
       "SRR493938       0.0  0.127873     0.0       0.0    0.00000   0.000000   \n",
       "SRR493939       0.0  0.011112     0.0       0.0    0.00000   0.000000   \n",
       "SRR493940       0.0  0.037886     0.0       0.0    0.07756   0.000000   \n",
       "SRR493941       0.0  0.156220     0.0       0.0    0.00000   0.070955   \n",
       "\n",
       "              HSFX3  VINAC1P   MIR4481  \n",
       "SRR493937  0.588513      0.0  0.000000  \n",
       "SRR493938  0.452511      0.0  0.000000  \n",
       "SRR493939  0.446833      0.0  0.000000  \n",
       "SRR493940  0.462295      0.0  0.168159  \n",
       "SRR493941  0.739403      0.0  0.000000  \n",
       "\n",
       "[5 rows x 37911 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17785\n"
     ]
    }
   ],
   "source": [
    "# Get intersection of gene lists\n",
    "our_gene_ids_hgnc = template_data.columns\n",
    "shared_genes_hgnc = list(set(our_gene_ids_hgnc).intersection(published_generic_genes))\n",
    "print(len(shared_genes_hgnc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save shared genes\n",
    "shared_genes_file = os.path.join(\n",
    "    local_dir,\n",
    "    \"shared_gene_ids.pickle\")\n",
    "\n",
    "outfile = open(shared_genes_file,'wb')\n",
    "pickle.dump(shared_genes_hgnc,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop smRNA samples so that samples are within the same range\n",
    "smRNA_samples = [\"SRR493961\",\n",
    "                 \"SRR493962\",\n",
    "                 \"SRR493963\",\n",
    "                 \"SRR493964\",\n",
    "                 \"SRR493965\",\n",
    "                 \"SRR493966\",\n",
    "                 \"SRR493967\",\n",
    "                 \"SRR493968\",\n",
    "                 \"SRR493969\",\n",
    "                 \"SRR493970\",\n",
    "                 \"SRR493971\",\n",
    "                 \"SRR493972\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop samples\n",
    "template_data = template_data.drop(smRNA_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24, 17788)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POLR3G</th>\n",
       "      <th>SSTR4</th>\n",
       "      <th>CORIN</th>\n",
       "      <th>BBOX1-AS1</th>\n",
       "      <th>PROSER2-AS1</th>\n",
       "      <th>LRRC37A3</th>\n",
       "      <th>SERPINA7</th>\n",
       "      <th>GABRA3</th>\n",
       "      <th>SNX22</th>\n",
       "      <th>MATN1</th>\n",
       "      <th>...</th>\n",
       "      <th>LRP5</th>\n",
       "      <th>SRD5A3-AS1</th>\n",
       "      <th>CFP</th>\n",
       "      <th>RHBDF1</th>\n",
       "      <th>SUPT5H</th>\n",
       "      <th>TEFM</th>\n",
       "      <th>DMRTC2</th>\n",
       "      <th>SLC35E3</th>\n",
       "      <th>SOD2</th>\n",
       "      <th>TMEM70</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493937</th>\n",
       "      <td>0.247983</td>\n",
       "      <td>0.092681</td>\n",
       "      <td>0.155784</td>\n",
       "      <td>0.022465</td>\n",
       "      <td>0.156563</td>\n",
       "      <td>1.039957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022215</td>\n",
       "      <td>20.697529</td>\n",
       "      <td>0.358349</td>\n",
       "      <td>...</td>\n",
       "      <td>17.995453</td>\n",
       "      <td>0.659922</td>\n",
       "      <td>3.781465</td>\n",
       "      <td>7.228667</td>\n",
       "      <td>19.568409</td>\n",
       "      <td>0.869790</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.667853</td>\n",
       "      <td>61.457390</td>\n",
       "      <td>9.299350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493938</th>\n",
       "      <td>0.399930</td>\n",
       "      <td>0.041186</td>\n",
       "      <td>0.159909</td>\n",
       "      <td>0.032658</td>\n",
       "      <td>0.123457</td>\n",
       "      <td>1.010312</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009993</td>\n",
       "      <td>20.712299</td>\n",
       "      <td>0.305661</td>\n",
       "      <td>...</td>\n",
       "      <td>18.666885</td>\n",
       "      <td>0.571155</td>\n",
       "      <td>4.032993</td>\n",
       "      <td>7.338301</td>\n",
       "      <td>19.840440</td>\n",
       "      <td>0.839970</td>\n",
       "      <td>0.008727</td>\n",
       "      <td>0.708261</td>\n",
       "      <td>61.391806</td>\n",
       "      <td>9.188426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493939</th>\n",
       "      <td>1.115604</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.288992</td>\n",
       "      <td>1.351308</td>\n",
       "      <td>0.316169</td>\n",
       "      <td>1.792745</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.501358</td>\n",
       "      <td>37.996993</td>\n",
       "      <td>0.433920</td>\n",
       "      <td>...</td>\n",
       "      <td>32.938705</td>\n",
       "      <td>2.220845</td>\n",
       "      <td>0.419704</td>\n",
       "      <td>17.261978</td>\n",
       "      <td>30.272669</td>\n",
       "      <td>2.069097</td>\n",
       "      <td>0.152653</td>\n",
       "      <td>0.943442</td>\n",
       "      <td>10.636130</td>\n",
       "      <td>7.736250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493940</th>\n",
       "      <td>1.150075</td>\n",
       "      <td>0.108467</td>\n",
       "      <td>0.234150</td>\n",
       "      <td>1.832712</td>\n",
       "      <td>0.174537</td>\n",
       "      <td>1.951469</td>\n",
       "      <td>0.026073</td>\n",
       "      <td>0.386864</td>\n",
       "      <td>37.201605</td>\n",
       "      <td>0.569087</td>\n",
       "      <td>...</td>\n",
       "      <td>33.044193</td>\n",
       "      <td>2.016507</td>\n",
       "      <td>0.578981</td>\n",
       "      <td>17.409587</td>\n",
       "      <td>30.783963</td>\n",
       "      <td>2.084962</td>\n",
       "      <td>0.078150</td>\n",
       "      <td>0.970944</td>\n",
       "      <td>10.656667</td>\n",
       "      <td>8.143106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493941</th>\n",
       "      <td>0.506490</td>\n",
       "      <td>0.054713</td>\n",
       "      <td>0.341643</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.145713</td>\n",
       "      <td>1.273116</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007448</td>\n",
       "      <td>16.055489</td>\n",
       "      <td>0.153819</td>\n",
       "      <td>...</td>\n",
       "      <td>36.473139</td>\n",
       "      <td>0.584874</td>\n",
       "      <td>3.538937</td>\n",
       "      <td>13.199591</td>\n",
       "      <td>15.164681</td>\n",
       "      <td>1.165465</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.717657</td>\n",
       "      <td>10.316851</td>\n",
       "      <td>4.318829</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17788 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             POLR3G     SSTR4     CORIN  BBOX1-AS1  PROSER2-AS1  LRRC37A3  \\\n",
       "SRR493937  0.247983  0.092681  0.155784   0.022465     0.156563  1.039957   \n",
       "SRR493938  0.399930  0.041186  0.159909   0.032658     0.123457  1.010312   \n",
       "SRR493939  1.115604  0.000000  0.288992   1.351308     0.316169  1.792745   \n",
       "SRR493940  1.150075  0.108467  0.234150   1.832712     0.174537  1.951469   \n",
       "SRR493941  0.506490  0.054713  0.341643   0.000000     0.145713  1.273116   \n",
       "\n",
       "           SERPINA7    GABRA3      SNX22     MATN1    ...          LRP5  \\\n",
       "SRR493937  0.000000  0.022215  20.697529  0.358349    ...     17.995453   \n",
       "SRR493938  0.000000  0.009993  20.712299  0.305661    ...     18.666885   \n",
       "SRR493939  0.000000  0.501358  37.996993  0.433920    ...     32.938705   \n",
       "SRR493940  0.026073  0.386864  37.201605  0.569087    ...     33.044193   \n",
       "SRR493941  0.000000  0.007448  16.055489  0.153819    ...     36.473139   \n",
       "\n",
       "           SRD5A3-AS1       CFP     RHBDF1     SUPT5H      TEFM    DMRTC2  \\\n",
       "SRR493937    0.659922  3.781465   7.228667  19.568409  0.869790  0.000000   \n",
       "SRR493938    0.571155  4.032993   7.338301  19.840440  0.839970  0.008727   \n",
       "SRR493939    2.220845  0.419704  17.261978  30.272669  2.069097  0.152653   \n",
       "SRR493940    2.016507  0.578981  17.409587  30.783963  2.084962  0.078150   \n",
       "SRR493941    0.584874  3.538937  13.199591  15.164681  1.165465  0.000000   \n",
       "\n",
       "            SLC35E3       SOD2    TMEM70  \n",
       "SRR493937  0.667853  61.457390  9.299350  \n",
       "SRR493938  0.708261  61.391806  9.188426  \n",
       "SRR493939  0.943442  10.636130  7.736250  \n",
       "SRR493940  0.970944  10.656667  8.143106  \n",
       "SRR493941  0.717657  10.316851  4.318829  \n",
       "\n",
       "[5 rows x 17788 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop genes\n",
    "template_data = template_data[shared_genes_hgnc]\n",
    "\n",
    "print(template_data.shape)\n",
    "template_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    }
   ],
   "source": [
    "print(len(template_data.columns) - len(shared_genes_hgnc))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Note:* There is a difference in the number of `shared_genes_hgnc` and genes in the template experiment because 3 genes have 2 different ensembl gene ids have map to the same hgnc symbol (one forward, one reverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save \n",
    "template_data.to_csv(template_data_file, float_format='%.5f', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalize compendium "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load real gene expression data\n",
    "original_compendium_file = params['compendium_data_file']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4119, 17788)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ANKRD36B</th>\n",
       "      <th>DMBT1</th>\n",
       "      <th>CFAP45</th>\n",
       "      <th>TP53I11</th>\n",
       "      <th>AARS2</th>\n",
       "      <th>KLRA1P</th>\n",
       "      <th>SOCS6</th>\n",
       "      <th>USP9X</th>\n",
       "      <th>ABI3BP</th>\n",
       "      <th>LCMT1-AS2</th>\n",
       "      <th>...</th>\n",
       "      <th>IL7</th>\n",
       "      <th>KCNJ3</th>\n",
       "      <th>SLC30A3</th>\n",
       "      <th>RTTN</th>\n",
       "      <th>FAM32A</th>\n",
       "      <th>ERBB3</th>\n",
       "      <th>SMYD5</th>\n",
       "      <th>BCAS1</th>\n",
       "      <th>TYRO3</th>\n",
       "      <th>ZNF283</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR223496</th>\n",
       "      <td>0.967</td>\n",
       "      <td>0.154</td>\n",
       "      <td>0.918</td>\n",
       "      <td>12.729</td>\n",
       "      <td>7.732</td>\n",
       "      <td>0.530</td>\n",
       "      <td>3.206</td>\n",
       "      <td>23.838</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.085</td>\n",
       "      <td>...</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.170</td>\n",
       "      <td>0.861</td>\n",
       "      <td>23.540</td>\n",
       "      <td>7.742</td>\n",
       "      <td>11.772</td>\n",
       "      <td>0.206</td>\n",
       "      <td>2.488</td>\n",
       "      <td>0.935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223497</th>\n",
       "      <td>1.458</td>\n",
       "      <td>55.136</td>\n",
       "      <td>0.177</td>\n",
       "      <td>0.616</td>\n",
       "      <td>8.166</td>\n",
       "      <td>0.608</td>\n",
       "      <td>4.714</td>\n",
       "      <td>7.537</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.218</td>\n",
       "      <td>...</td>\n",
       "      <td>0.513</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.118</td>\n",
       "      <td>2.169</td>\n",
       "      <td>26.574</td>\n",
       "      <td>4.492</td>\n",
       "      <td>12.299</td>\n",
       "      <td>0.007</td>\n",
       "      <td>2.343</td>\n",
       "      <td>0.474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223506</th>\n",
       "      <td>0.530</td>\n",
       "      <td>0.006</td>\n",
       "      <td>3.627</td>\n",
       "      <td>9.054</td>\n",
       "      <td>11.241</td>\n",
       "      <td>0.119</td>\n",
       "      <td>2.736</td>\n",
       "      <td>6.762</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.039</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.397</td>\n",
       "      <td>1.904</td>\n",
       "      <td>47.685</td>\n",
       "      <td>3.783</td>\n",
       "      <td>8.768</td>\n",
       "      <td>0.085</td>\n",
       "      <td>4.555</td>\n",
       "      <td>2.435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223524</th>\n",
       "      <td>0.355</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.244</td>\n",
       "      <td>0.687</td>\n",
       "      <td>5.419</td>\n",
       "      <td>0.154</td>\n",
       "      <td>4.513</td>\n",
       "      <td>16.814</td>\n",
       "      <td>0.294</td>\n",
       "      <td>0.022</td>\n",
       "      <td>...</td>\n",
       "      <td>0.161</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.181</td>\n",
       "      <td>1.229</td>\n",
       "      <td>59.911</td>\n",
       "      <td>0.057</td>\n",
       "      <td>7.475</td>\n",
       "      <td>0.012</td>\n",
       "      <td>1.040</td>\n",
       "      <td>0.686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223525</th>\n",
       "      <td>1.346</td>\n",
       "      <td>0.910</td>\n",
       "      <td>0.448</td>\n",
       "      <td>6.607</td>\n",
       "      <td>4.814</td>\n",
       "      <td>0.297</td>\n",
       "      <td>2.515</td>\n",
       "      <td>5.427</td>\n",
       "      <td>0.115</td>\n",
       "      <td>0.064</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.317</td>\n",
       "      <td>1.214</td>\n",
       "      <td>24.410</td>\n",
       "      <td>7.884</td>\n",
       "      <td>10.431</td>\n",
       "      <td>0.069</td>\n",
       "      <td>0.617</td>\n",
       "      <td>0.603</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17788 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           ANKRD36B   DMBT1  CFAP45  TP53I11   AARS2  KLRA1P  SOCS6   USP9X  \\\n",
       "SRR223496     0.967   0.154   0.918   12.729   7.732   0.530  3.206  23.838   \n",
       "SRR223497     1.458  55.136   0.177    0.616   8.166   0.608  4.714   7.537   \n",
       "SRR223506     0.530   0.006   3.627    9.054  11.241   0.119  2.736   6.762   \n",
       "SRR223524     0.355   0.013   0.244    0.687   5.419   0.154  4.513  16.814   \n",
       "SRR223525     1.346   0.910   0.448    6.607   4.814   0.297  2.515   5.427   \n",
       "\n",
       "           ABI3BP  LCMT1-AS2   ...      IL7  KCNJ3  SLC30A3   RTTN  FAM32A  \\\n",
       "SRR223496   0.033      0.085   ...    0.240  0.000    0.170  0.861  23.540   \n",
       "SRR223497   0.006      0.218   ...    0.513  0.014    0.118  2.169  26.574   \n",
       "SRR223506   0.033      0.039   ...    0.001  0.001    0.397  1.904  47.685   \n",
       "SRR223524   0.294      0.022   ...    0.161  0.000    0.181  1.229  59.911   \n",
       "SRR223525   0.115      0.064   ...    0.033  0.000    0.317  1.214  24.410   \n",
       "\n",
       "           ERBB3   SMYD5  BCAS1  TYRO3  ZNF283  \n",
       "SRR223496  7.742  11.772  0.206  2.488   0.935  \n",
       "SRR223497  4.492  12.299  0.007  2.343   0.474  \n",
       "SRR223506  3.783   8.768  0.085  4.555   2.435  \n",
       "SRR223524  0.057   7.475  0.012  1.040   0.686  \n",
       "SRR223525  7.884  10.431  0.069  0.617   0.603  \n",
       "\n",
       "[5 rows x 17788 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read data\n",
    "original_compendium = pd.read_table(\n",
    "    original_compendium_file,\n",
    "    header=0,\n",
    "    sep='\\t',\n",
    "    index_col=0)\n",
    "\n",
    "print(original_compendium.shape)\n",
    "original_compendium.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace ensembl ids with gene symbols\n",
    "#original_compendium = process.replace_ensembl_ids(original_compendium,\n",
    "#                                                gene_id_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POLR3G</th>\n",
       "      <th>SSTR4</th>\n",
       "      <th>CORIN</th>\n",
       "      <th>BBOX1-AS1</th>\n",
       "      <th>PROSER2-AS1</th>\n",
       "      <th>LRRC37A3</th>\n",
       "      <th>SERPINA7</th>\n",
       "      <th>GABRA3</th>\n",
       "      <th>SNX22</th>\n",
       "      <th>MATN1</th>\n",
       "      <th>...</th>\n",
       "      <th>LRP5</th>\n",
       "      <th>SRD5A3-AS1</th>\n",
       "      <th>CFP</th>\n",
       "      <th>RHBDF1</th>\n",
       "      <th>SUPT5H</th>\n",
       "      <th>TEFM</th>\n",
       "      <th>DMRTC2</th>\n",
       "      <th>SLC35E3</th>\n",
       "      <th>SOD2</th>\n",
       "      <th>TMEM70</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR223496</th>\n",
       "      <td>0.655</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.736</td>\n",
       "      <td>0.846</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.134</td>\n",
       "      <td>23.167</td>\n",
       "      <td>0.231</td>\n",
       "      <td>...</td>\n",
       "      <td>28.805</td>\n",
       "      <td>0.709</td>\n",
       "      <td>0.098</td>\n",
       "      <td>5.882</td>\n",
       "      <td>31.860</td>\n",
       "      <td>3.619</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.424</td>\n",
       "      <td>6.531</td>\n",
       "      <td>7.811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223497</th>\n",
       "      <td>3.068</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.620</td>\n",
       "      <td>0.925</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.039</td>\n",
       "      <td>23.608</td>\n",
       "      <td>0.566</td>\n",
       "      <td>...</td>\n",
       "      <td>38.817</td>\n",
       "      <td>0.379</td>\n",
       "      <td>0.066</td>\n",
       "      <td>11.390</td>\n",
       "      <td>15.556</td>\n",
       "      <td>1.605</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.539</td>\n",
       "      <td>21.540</td>\n",
       "      <td>6.549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223506</th>\n",
       "      <td>1.871</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.035</td>\n",
       "      <td>0.034</td>\n",
       "      <td>1.003</td>\n",
       "      <td>0.898</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.425</td>\n",
       "      <td>33.519</td>\n",
       "      <td>0.227</td>\n",
       "      <td>...</td>\n",
       "      <td>19.708</td>\n",
       "      <td>0.431</td>\n",
       "      <td>0.265</td>\n",
       "      <td>2.327</td>\n",
       "      <td>49.270</td>\n",
       "      <td>3.775</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.269</td>\n",
       "      <td>5.545</td>\n",
       "      <td>4.361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223524</th>\n",
       "      <td>11.650</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.059</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.013</td>\n",
       "      <td>2.255</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.248</td>\n",
       "      <td>43.632</td>\n",
       "      <td>0.328</td>\n",
       "      <td>...</td>\n",
       "      <td>3.314</td>\n",
       "      <td>2.096</td>\n",
       "      <td>0.024</td>\n",
       "      <td>1.574</td>\n",
       "      <td>20.061</td>\n",
       "      <td>2.512</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.493</td>\n",
       "      <td>8.279</td>\n",
       "      <td>9.685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223525</th>\n",
       "      <td>0.980</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.056</td>\n",
       "      <td>1.472</td>\n",
       "      <td>0.153</td>\n",
       "      <td>1.256</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.123</td>\n",
       "      <td>144.121</td>\n",
       "      <td>0.087</td>\n",
       "      <td>...</td>\n",
       "      <td>14.225</td>\n",
       "      <td>2.375</td>\n",
       "      <td>0.367</td>\n",
       "      <td>22.861</td>\n",
       "      <td>19.630</td>\n",
       "      <td>1.031</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.584</td>\n",
       "      <td>4.113</td>\n",
       "      <td>3.822</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17785 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           POLR3G  SSTR4  CORIN  BBOX1-AS1  PROSER2-AS1  LRRC37A3  SERPINA7  \\\n",
       "SRR223496   0.655  0.000  0.010      0.021        0.736     0.846       0.0   \n",
       "SRR223497   3.068  0.000  0.000      0.000        1.620     0.925       0.0   \n",
       "SRR223506   1.871  0.000  0.035      0.034        1.003     0.898       0.0   \n",
       "SRR223524  11.650  0.006  0.059      0.039        0.013     2.255       0.0   \n",
       "SRR223525   0.980  0.019  0.056      1.472        0.153     1.256       0.0   \n",
       "\n",
       "           GABRA3    SNX22  MATN1   ...      LRP5  SRD5A3-AS1    CFP  RHBDF1  \\\n",
       "SRR223496   0.134   23.167  0.231   ...    28.805       0.709  0.098   5.882   \n",
       "SRR223497   0.039   23.608  0.566   ...    38.817       0.379  0.066  11.390   \n",
       "SRR223506   2.425   33.519  0.227   ...    19.708       0.431  0.265   2.327   \n",
       "SRR223524   4.248   43.632  0.328   ...     3.314       2.096  0.024   1.574   \n",
       "SRR223525   2.123  144.121  0.087   ...    14.225       2.375  0.367  22.861   \n",
       "\n",
       "           SUPT5H   TEFM  DMRTC2  SLC35E3    SOD2  TMEM70  \n",
       "SRR223496  31.860  3.619   0.000    0.424   6.531   7.811  \n",
       "SRR223497  15.556  1.605   0.000    0.539  21.540   6.549  \n",
       "SRR223506  49.270  3.775   0.000    0.269   5.545   4.361  \n",
       "SRR223524  20.061  2.512   0.001    0.493   8.279   9.685  \n",
       "SRR223525  19.630  1.031   0.039    0.584   4.113   3.822  \n",
       "\n",
       "[5 rows x 17785 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop genes\n",
    "original_compendium = original_compendium[shared_genes_hgnc]\n",
    "\n",
    "original_compendium.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4119, 17785)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POLR3G</th>\n",
       "      <th>SSTR4</th>\n",
       "      <th>CORIN</th>\n",
       "      <th>BBOX1-AS1</th>\n",
       "      <th>PROSER2-AS1</th>\n",
       "      <th>LRRC37A3</th>\n",
       "      <th>SERPINA7</th>\n",
       "      <th>GABRA3</th>\n",
       "      <th>SNX22</th>\n",
       "      <th>MATN1</th>\n",
       "      <th>...</th>\n",
       "      <th>LRP5</th>\n",
       "      <th>SRD5A3-AS1</th>\n",
       "      <th>CFP</th>\n",
       "      <th>RHBDF1</th>\n",
       "      <th>SUPT5H</th>\n",
       "      <th>TEFM</th>\n",
       "      <th>DMRTC2</th>\n",
       "      <th>SLC35E3</th>\n",
       "      <th>SOD2</th>\n",
       "      <th>TMEM70</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR223496</th>\n",
       "      <td>0.006218</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000388</td>\n",
       "      <td>0.000331</td>\n",
       "      <td>0.044985</td>\n",
       "      <td>0.019284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0.082804</td>\n",
       "      <td>0.019822</td>\n",
       "      <td>...</td>\n",
       "      <td>0.189333</td>\n",
       "      <td>0.046446</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>0.077203</td>\n",
       "      <td>0.091867</td>\n",
       "      <td>0.109810</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000563</td>\n",
       "      <td>0.007360</td>\n",
       "      <td>0.015749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223497</th>\n",
       "      <td>0.029124</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.099016</td>\n",
       "      <td>0.021085</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.084380</td>\n",
       "      <td>0.048567</td>\n",
       "      <td>...</td>\n",
       "      <td>0.255142</td>\n",
       "      <td>0.024828</td>\n",
       "      <td>0.000316</td>\n",
       "      <td>0.149497</td>\n",
       "      <td>0.044855</td>\n",
       "      <td>0.048700</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000716</td>\n",
       "      <td>0.024276</td>\n",
       "      <td>0.013204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223506</th>\n",
       "      <td>0.017761</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001359</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.061304</td>\n",
       "      <td>0.020470</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002489</td>\n",
       "      <td>0.119804</td>\n",
       "      <td>0.019478</td>\n",
       "      <td>...</td>\n",
       "      <td>0.129539</td>\n",
       "      <td>0.028235</td>\n",
       "      <td>0.001268</td>\n",
       "      <td>0.030542</td>\n",
       "      <td>0.142069</td>\n",
       "      <td>0.114543</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000357</td>\n",
       "      <td>0.006249</td>\n",
       "      <td>0.008793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223524</th>\n",
       "      <td>0.110591</td>\n",
       "      <td>0.000498</td>\n",
       "      <td>0.002292</td>\n",
       "      <td>0.000615</td>\n",
       "      <td>0.000795</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004360</td>\n",
       "      <td>0.155950</td>\n",
       "      <td>0.028145</td>\n",
       "      <td>...</td>\n",
       "      <td>0.021783</td>\n",
       "      <td>0.137308</td>\n",
       "      <td>0.000115</td>\n",
       "      <td>0.020659</td>\n",
       "      <td>0.057845</td>\n",
       "      <td>0.076221</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000655</td>\n",
       "      <td>0.009331</td>\n",
       "      <td>0.019527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR223525</th>\n",
       "      <td>0.009303</td>\n",
       "      <td>0.001577</td>\n",
       "      <td>0.002175</td>\n",
       "      <td>0.023217</td>\n",
       "      <td>0.009352</td>\n",
       "      <td>0.028630</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002179</td>\n",
       "      <td>0.515119</td>\n",
       "      <td>0.007465</td>\n",
       "      <td>...</td>\n",
       "      <td>0.093500</td>\n",
       "      <td>0.155585</td>\n",
       "      <td>0.001756</td>\n",
       "      <td>0.300056</td>\n",
       "      <td>0.056603</td>\n",
       "      <td>0.031283</td>\n",
       "      <td>0.000589</td>\n",
       "      <td>0.000775</td>\n",
       "      <td>0.004635</td>\n",
       "      <td>0.007706</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17785 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             POLR3G     SSTR4     CORIN  BBOX1-AS1  PROSER2-AS1  LRRC37A3  \\\n",
       "SRR223496  0.006218  0.000000  0.000388   0.000331     0.044985  0.019284   \n",
       "SRR223497  0.029124  0.000000  0.000000   0.000000     0.099016  0.021085   \n",
       "SRR223506  0.017761  0.000000  0.001359   0.000536     0.061304  0.020470   \n",
       "SRR223524  0.110591  0.000498  0.002292   0.000615     0.000795  0.051402   \n",
       "SRR223525  0.009303  0.001577  0.002175   0.023217     0.009352  0.028630   \n",
       "\n",
       "           SERPINA7    GABRA3     SNX22     MATN1    ...         LRP5  \\\n",
       "SRR223496       0.0  0.000138  0.082804  0.019822    ...     0.189333   \n",
       "SRR223497       0.0  0.000040  0.084380  0.048567    ...     0.255142   \n",
       "SRR223506       0.0  0.002489  0.119804  0.019478    ...     0.129539   \n",
       "SRR223524       0.0  0.004360  0.155950  0.028145    ...     0.021783   \n",
       "SRR223525       0.0  0.002179  0.515119  0.007465    ...     0.093500   \n",
       "\n",
       "           SRD5A3-AS1       CFP    RHBDF1    SUPT5H      TEFM    DMRTC2  \\\n",
       "SRR223496    0.046446  0.000469  0.077203  0.091867  0.109810  0.000000   \n",
       "SRR223497    0.024828  0.000316  0.149497  0.044855  0.048700  0.000000   \n",
       "SRR223506    0.028235  0.001268  0.030542  0.142069  0.114543  0.000000   \n",
       "SRR223524    0.137308  0.000115  0.020659  0.057845  0.076221  0.000015   \n",
       "SRR223525    0.155585  0.001756  0.300056  0.056603  0.031283  0.000589   \n",
       "\n",
       "            SLC35E3      SOD2    TMEM70  \n",
       "SRR223496  0.000563  0.007360  0.015749  \n",
       "SRR223497  0.000716  0.024276  0.013204  \n",
       "SRR223506  0.000357  0.006249  0.008793  \n",
       "SRR223524  0.000655  0.009331  0.019527  \n",
       "SRR223525  0.000775  0.004635  0.007706  \n",
       "\n",
       "[5 rows x 17785 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 0-1 normalize per gene\n",
    "scaler = preprocessing.MinMaxScaler()\n",
    "original_data_scaled = scaler.fit_transform(original_compendium)\n",
    "original_data_scaled_df = pd.DataFrame(original_data_scaled,\n",
    "                                columns=original_compendium.columns,\n",
    "                                index=original_compendium.index)\n",
    "\n",
    "print(original_data_scaled_df.shape)\n",
    "original_data_scaled_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "normalized_data_file = params['normalized_compendium_data_file']\n",
    "\n",
    "original_data_scaled_df.to_csv(\n",
    "    normalized_data_file, float_format='%.3f', sep='\\t')\n",
    "\n",
    "original_compendium.to_csv(\n",
    "    original_compendium_file, float_format='%.3f', sep='\\t')\n",
    "\n",
    "# Save scaler transform\n",
    "scaler_file = params['scaler_transform_file']\n",
    "\n",
    "outfile = open(scaler_file,'wb')\n",
    "pickle.dump(scaler,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train VAE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup directories\n",
    "# Create VAE directories\n",
    "output_dirs = [os.path.join(base_dir, dataset_name, \"models\"),\n",
    "               os.path.join(base_dir, dataset_name, \"logs\")]\n",
    "\n",
    "# Check if analysis output directory exist otherwise create\n",
    "for each_dir in output_dirs:\n",
    "    if os.path.exists(each_dir) == False:\n",
    "        print('creating new directory: {}'.format(each_dir))\n",
    "        os.makedirs(each_dir, exist_ok=True)\n",
    "\n",
    "# Check if NN architecture directory exist otherwise create\n",
    "for each_dir in output_dirs:\n",
    "    new_dir = os.path.join(each_dir, NN_architecture)\n",
    "    if os.path.exists(new_dir) == False:\n",
    "        print('creating new directory: {}'.format(new_dir))\n",
    "        os.makedirs(new_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Train VAE on new compendium data\n",
    "#pipeline.train_vae(config_file,\n",
    "#                   normalized_data_file)"
   ]
  }
 ],
 "metadata": {
  "finalized": {
   "timestamp": 1592246126078,
   "trusted": true
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
