{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process recount2 data\n",
    "This notebook does the following:\n",
    "\n",
    "1. Selects template experiment\n",
    "2. Downloads subset of recount2 data, including the template experiment (50 random experiments + 1 template experiment)\n",
    "3. Train VAE on subset of recount2 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import rpy2\n",
    "import seaborn as sns\n",
    "from sklearn import preprocessing\n",
    "import pickle\n",
    "\n",
    "from ponyo import generate_template_data, utils, pipeline\n",
    "from generic_modules import process, calc\n",
    "\n",
    "from numpy.random import seed\n",
    "random_state = 123\n",
    "seed(random_state)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(),\"../\"))\n",
    "\n",
    "config_file = os.path.abspath(os.path.join(base_dir,\n",
    "                                           \"config_human.tsv\"))\n",
    "params = utils.read_config(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select template experiment\n",
    "\n",
    "We manually selected bioproject [SRP012656](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37764), which contains primary non-small cell lung adenocarcinoma tumors and adjacent normal tissues of 6 never-smoker Korean female patients with 2 replicates each."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load params\n",
    "local_dir = params[\"local_dir\"]\n",
    "dataset_name = params['dataset_name']\n",
    "NN_architecture = params['NN_architecture']\n",
    "project_id = params['project_id']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download subset of recount2 to use as a compendium\n",
    "The compendium will be composed of random experiments + the selected template experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NULL\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "# Select 59\n",
    "# Select a\n",
    "# Run one time\n",
    "#if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "#    install.packages(\"BiocManager\")\n",
    "#BiocManager::install(\"derfinder\")\n",
    "#BiocManager::install(\"recount\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library('recount')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#%%R -i project_id -i local_dir\n",
    "\n",
    "#source('../generic_modules/download_recount2_data.R')\n",
    "\n",
    "#get_recount2_compendium(project_id, local_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download expression data for selected project id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Loading objects:\n",
       "  rse_gene\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i project_id -i local_dir\n",
    "\n",
    "source('../generic_modules/download_recount2_data.R')\n",
    "\n",
    "get_recount2_template_experiment(project_id, local_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset genes\n",
    "For our downstream analysis we will be comparing our set of differentially expression genes against the set found in [Crow et. al. publication](https://www.pnas.org/content/pnas/116/13/6491.full.pdf), therefore we will limit our genes to include only those genes shared between our starting set of genes and those in publication. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Order</th>\n",
       "      <th>Gene_EntrezID</th>\n",
       "      <th>N_HitLists</th>\n",
       "      <th>DE_Prior_Rank</th>\n",
       "      <th>Gene_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>7503</td>\n",
       "      <td>79</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>8653</td>\n",
       "      <td>64</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>DDX3Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>9086</td>\n",
       "      <td>62</td>\n",
       "      <td>0.999870</td>\n",
       "      <td>EIF1AY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>8284</td>\n",
       "      <td>52</td>\n",
       "      <td>0.999870</td>\n",
       "      <td>KDM5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>8287</td>\n",
       "      <td>46</td>\n",
       "      <td>0.999791</td>\n",
       "      <td>USP9Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene_Order  Gene_EntrezID  N_HitLists  DE_Prior_Rank Gene_Name\n",
       "0           1           7503          79       1.000000      XIST\n",
       "1           2           8653          64       0.999948     DDX3Y\n",
       "2           3           9086          62       0.999870    EIF1AY\n",
       "3           4           8284          52       0.999870     KDM5D\n",
       "4           5           8287          46       0.999791     USP9Y"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get generic genes identified by Crow et. al.\n",
    "DE_prior_file = \"https://raw.githubusercontent.com/maggiecrow/DEprior/master/DE_Prior.txt\"\n",
    "\n",
    "DE_prior = pd.read_csv(DE_prior_file,\n",
    "                       header=0,\n",
    "                       sep=\"\\t\")\n",
    "\n",
    "DE_prior.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of published generic genes\n",
    "published_generic_genes = list(DE_prior['Gene_Name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of our genes\n",
    "# Load real template experiment\n",
    "template_data_file = params['template_data_file']\n",
    "\n",
    "# Read template data\n",
    "template_data = pd.read_csv(\n",
    "    template_data_file,\n",
    "    header=0,\n",
    "    sep='\\t',\n",
    "    index_col=0)\n",
    "\n",
    "our_gene_ids = list(template_data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# File mapping ensembl ids to hgnc symbols\n",
    "gene_id_file = os.path.join(\n",
    "    local_dir,\n",
    "    \"ensembl_hgnc_mapping.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "suppressWarnings(library(\"biomaRt\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R -i template_data_file -i gene_id_file\n",
    "\n",
    "# Get mapping between ensembl gene ids (ours) to HGNC gene symbols (published)\n",
    "\n",
    "source('../generic_modules/process_names.R')\n",
    "\n",
    "if (file.exists(gene_id_file) == FALSE){\n",
    "    gene_id_mapping <- get_ensembl_symbol_mapping(template_data_file, gene_id_file)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(57210, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000003137</td>\n",
       "      <td>CYP26B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000003249</td>\n",
       "      <td>DBNDD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000004799</td>\n",
       "      <td>PDK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000006062</td>\n",
       "      <td>MAP3K14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_gene_id hgnc_symbol\n",
       "1  ENSG00000002330         BAD\n",
       "2  ENSG00000003137     CYP26B1\n",
       "3  ENSG00000003249      DBNDD1\n",
       "4  ENSG00000004799        PDK4\n",
       "5  ENSG00000006062     MAP3K14"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read gene id mapping\n",
    "gene_id_mapping = pd.read_csv(\n",
    "        gene_id_file,\n",
    "        header=0,\n",
    "        sep='\\t',\n",
    "        index_col=0)\n",
    "\n",
    "print(gene_id_mapping.shape)\n",
    "gene_id_mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58037, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_version</th>\n",
       "      <th>ensembl_parsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ensembl_version   ensembl_parsed\n",
       "0  ENSG00000000003.14  ENSG00000000003\n",
       "1   ENSG00000000005.5  ENSG00000000005\n",
       "2  ENSG00000000419.12  ENSG00000000419\n",
       "3  ENSG00000000457.13  ENSG00000000457\n",
       "4  ENSG00000000460.16  ENSG00000000460"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get mapping between ensembl ids with and without version numbers\n",
    "# Expressiond data uses ensembl gene ids with version number \n",
    "ensembl_gene_ids = pd.DataFrame(data={'ensembl_version': our_gene_ids,\n",
    "                                      'ensembl_parsed': [gene_id.split('.')[0] for gene_id in our_gene_ids]})\n",
    "\n",
    "print(ensembl_gene_ids.shape)\n",
    "ensembl_gene_ids.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58129, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>ensembl_parsed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_version</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000002330.13</th>\n",
       "      <td>ENSG00000002330</td>\n",
       "      <td>BAD</td>\n",
       "      <td>ENSG00000002330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003137.8</th>\n",
       "      <td>ENSG00000003137</td>\n",
       "      <td>CYP26B1</td>\n",
       "      <td>ENSG00000003137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000003249.13</th>\n",
       "      <td>ENSG00000003249</td>\n",
       "      <td>DBNDD1</td>\n",
       "      <td>ENSG00000003249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000004799.7</th>\n",
       "      <td>ENSG00000004799</td>\n",
       "      <td>PDK4</td>\n",
       "      <td>ENSG00000004799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000006062.13</th>\n",
       "      <td>ENSG00000006062</td>\n",
       "      <td>MAP3K14</td>\n",
       "      <td>ENSG00000006062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ensembl_gene_id hgnc_symbol   ensembl_parsed\n",
       "ensembl_version                                                 \n",
       "ENSG00000002330.13  ENSG00000002330         BAD  ENSG00000002330\n",
       "ENSG00000003137.8   ENSG00000003137     CYP26B1  ENSG00000003137\n",
       "ENSG00000003249.13  ENSG00000003249      DBNDD1  ENSG00000003249\n",
       "ENSG00000004799.7   ENSG00000004799        PDK4  ENSG00000004799\n",
       "ENSG00000006062.13  ENSG00000006062     MAP3K14  ENSG00000006062"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map ensembl ids with version number to gene_id_mapping_df\n",
    "gene_id_mapping = pd.merge(gene_id_mapping, \n",
    "                           ensembl_gene_ids, \n",
    "                           left_on='ensembl_gene_id',\n",
    "                           right_on='ensembl_parsed', \n",
    "                           how='outer')\n",
    "\n",
    "print(gene_id_mapping.shape)\n",
    "gene_id_mapping.set_index('ensembl_version', inplace=True)\n",
    "gene_id_mapping.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since this experiment contains both RNA-seq and smRNA-seq samples which are in different ranges so we will drop smRNA samples so that samples are within the same range. The analysis identifying these two subsets of samples can be found in this [notebook](0_explore_input_data.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace ensembl ids with gene symbols\n",
    "template_data = process.replace_ensembl_ids(template_data,\n",
    "                                            gene_id_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>NFYA</th>\n",
       "      <th>...</th>\n",
       "      <th>MIR6830</th>\n",
       "      <th>MIR548AH</th>\n",
       "      <th>MYOCOS</th>\n",
       "      <th>MIR522</th>\n",
       "      <th>MIR6715B</th>\n",
       "      <th>MIR3116-2</th>\n",
       "      <th>MIR3202-2</th>\n",
       "      <th>HSFX3</th>\n",
       "      <th>VINAC1P</th>\n",
       "      <th>MIR4481</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493937</th>\n",
       "      <td>3.800154</td>\n",
       "      <td>0.018190</td>\n",
       "      <td>9.128459</td>\n",
       "      <td>1.585803</td>\n",
       "      <td>0.713952</td>\n",
       "      <td>84.990542</td>\n",
       "      <td>19.943002</td>\n",
       "      <td>12.065586</td>\n",
       "      <td>2.498822</td>\n",
       "      <td>7.713002</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.135649</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.588513</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493938</th>\n",
       "      <td>3.798473</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.157669</td>\n",
       "      <td>1.456017</td>\n",
       "      <td>0.738967</td>\n",
       "      <td>85.788323</td>\n",
       "      <td>19.621252</td>\n",
       "      <td>11.738855</td>\n",
       "      <td>2.435308</td>\n",
       "      <td>7.125808</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005417</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.127873</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.452511</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493939</th>\n",
       "      <td>20.222134</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.522072</td>\n",
       "      <td>3.944919</td>\n",
       "      <td>2.512265</td>\n",
       "      <td>8.072948</td>\n",
       "      <td>47.246078</td>\n",
       "      <td>18.866957</td>\n",
       "      <td>2.443230</td>\n",
       "      <td>12.143670</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011112</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.446833</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493940</th>\n",
       "      <td>19.988329</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17.140304</td>\n",
       "      <td>3.977217</td>\n",
       "      <td>2.378849</td>\n",
       "      <td>7.599225</td>\n",
       "      <td>46.594458</td>\n",
       "      <td>18.921190</td>\n",
       "      <td>2.448276</td>\n",
       "      <td>12.048960</td>\n",
       "      <td>...</td>\n",
       "      <td>0.147412</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.037886</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.07756</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.462295</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.168159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493941</th>\n",
       "      <td>3.239231</td>\n",
       "      <td>0.019398</td>\n",
       "      <td>9.069316</td>\n",
       "      <td>2.181582</td>\n",
       "      <td>1.008155</td>\n",
       "      <td>17.675885</td>\n",
       "      <td>19.371137</td>\n",
       "      <td>11.759756</td>\n",
       "      <td>1.970725</td>\n",
       "      <td>6.740705</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002860</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156220</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.070955</td>\n",
       "      <td>0.739403</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37911 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TSPAN6      TNMD       DPM1     SCYL3  C1orf112        FGR  \\\n",
       "SRR493937   3.800154  0.018190   9.128459  1.585803  0.713952  84.990542   \n",
       "SRR493938   3.798473  0.000000   9.157669  1.456017  0.738967  85.788323   \n",
       "SRR493939  20.222134  0.000000  15.522072  3.944919  2.512265   8.072948   \n",
       "SRR493940  19.988329  0.000000  17.140304  3.977217  2.378849   7.599225   \n",
       "SRR493941   3.239231  0.019398   9.069316  2.181582  1.008155  17.675885   \n",
       "\n",
       "                 CFH      FUCA2      GCLC       NFYA    ...      MIR6830  \\\n",
       "SRR493937  19.943002  12.065586  2.498822   7.713002    ...     0.000000   \n",
       "SRR493938  19.621252  11.738855  2.435308   7.125808    ...     0.005417   \n",
       "SRR493939  47.246078  18.866957  2.443230  12.143670    ...     0.000000   \n",
       "SRR493940  46.594458  18.921190  2.448276  12.048960    ...     0.147412   \n",
       "SRR493941  19.371137  11.759756  1.970725   6.740705    ...     0.002860   \n",
       "\n",
       "           MIR548AH    MYOCOS  MIR522  MIR6715B  MIR3116-2  MIR3202-2  \\\n",
       "SRR493937       0.0  0.135649     0.0       0.0    0.00000   0.000000   \n",
       "SRR493938       0.0  0.127873     0.0       0.0    0.00000   0.000000   \n",
       "SRR493939       0.0  0.011112     0.0       0.0    0.00000   0.000000   \n",
       "SRR493940       0.0  0.037886     0.0       0.0    0.07756   0.000000   \n",
       "SRR493941       0.0  0.156220     0.0       0.0    0.00000   0.070955   \n",
       "\n",
       "              HSFX3  VINAC1P   MIR4481  \n",
       "SRR493937  0.588513      0.0  0.000000  \n",
       "SRR493938  0.452511      0.0  0.000000  \n",
       "SRR493939  0.446833      0.0  0.000000  \n",
       "SRR493940  0.462295      0.0  0.168159  \n",
       "SRR493941  0.739403      0.0  0.000000  \n",
       "\n",
       "[5 rows x 37911 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17785\n"
     ]
    }
   ],
   "source": [
    "# Get intersection of gene lists\n",
    "our_gene_ids_hgnc = template_data.columns\n",
    "shared_genes_hgnc = list(set(our_gene_ids_hgnc).intersection(published_generic_genes))\n",
    "print(len(shared_genes_hgnc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save shared genes\n",
    "shared_genes_file = os.path.join(\n",
    "    local_dir,\n",
    "    \"shared_gene_ids.pickle\")\n",
    "\n",
    "outfile = open(shared_genes_file,'wb')\n",
    "pickle.dump(shared_genes_hgnc,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop smRNA samples so that samples are within the same range\n",
    "smRNA_samples = [\"SRR493961\",\n",
    "                 \"SRR493962\",\n",
    "                 \"SRR493963\",\n",
    "                 \"SRR493964\",\n",
    "                 \"SRR493965\",\n",
    "                 \"SRR493966\",\n",
    "                 \"SRR493967\",\n",
    "                 \"SRR493968\",\n",
    "                 \"SRR493969\",\n",
    "                 \"SRR493970\",\n",
    "                 \"SRR493971\",\n",
    "                 \"SRR493972\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop samples\n",
    "template_data = template_data.drop(smRNA_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24, 17788)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RPL23A</th>\n",
       "      <th>HOPX</th>\n",
       "      <th>CNTLN</th>\n",
       "      <th>YEATS2</th>\n",
       "      <th>ZEB2</th>\n",
       "      <th>DAB2</th>\n",
       "      <th>LLGL2</th>\n",
       "      <th>TEKT3</th>\n",
       "      <th>CCNG1</th>\n",
       "      <th>HNRNPDL</th>\n",
       "      <th>...</th>\n",
       "      <th>TMEM52</th>\n",
       "      <th>SPINT3</th>\n",
       "      <th>ERICH3</th>\n",
       "      <th>GOLGA2P5</th>\n",
       "      <th>PIH1D2</th>\n",
       "      <th>GABARAPL2</th>\n",
       "      <th>PHKA2</th>\n",
       "      <th>AHCYL2</th>\n",
       "      <th>FAM170A</th>\n",
       "      <th>CAMSAP3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR493937</th>\n",
       "      <td>26.380195</td>\n",
       "      <td>7.594717</td>\n",
       "      <td>0.486687</td>\n",
       "      <td>3.244362</td>\n",
       "      <td>4.594237</td>\n",
       "      <td>10.716403</td>\n",
       "      <td>12.610643</td>\n",
       "      <td>1.133912</td>\n",
       "      <td>9.398151</td>\n",
       "      <td>31.639392</td>\n",
       "      <td>...</td>\n",
       "      <td>0.841445</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.098554</td>\n",
       "      <td>1.865488</td>\n",
       "      <td>1.264137</td>\n",
       "      <td>11.905335</td>\n",
       "      <td>3.857274</td>\n",
       "      <td>18.718648</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.894070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493938</th>\n",
       "      <td>26.043990</td>\n",
       "      <td>7.449357</td>\n",
       "      <td>0.465863</td>\n",
       "      <td>3.453337</td>\n",
       "      <td>4.655460</td>\n",
       "      <td>10.348601</td>\n",
       "      <td>12.898778</td>\n",
       "      <td>1.178870</td>\n",
       "      <td>8.950152</td>\n",
       "      <td>31.577780</td>\n",
       "      <td>...</td>\n",
       "      <td>0.762630</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.248183</td>\n",
       "      <td>1.838183</td>\n",
       "      <td>1.335846</td>\n",
       "      <td>11.891560</td>\n",
       "      <td>3.798020</td>\n",
       "      <td>18.645412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.979511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493939</th>\n",
       "      <td>23.938892</td>\n",
       "      <td>17.296808</td>\n",
       "      <td>0.309277</td>\n",
       "      <td>2.774270</td>\n",
       "      <td>1.626023</td>\n",
       "      <td>4.724359</td>\n",
       "      <td>21.224896</td>\n",
       "      <td>0.599888</td>\n",
       "      <td>16.530086</td>\n",
       "      <td>34.013420</td>\n",
       "      <td>...</td>\n",
       "      <td>1.633144</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.602161</td>\n",
       "      <td>3.885646</td>\n",
       "      <td>1.750385</td>\n",
       "      <td>11.529645</td>\n",
       "      <td>5.524931</td>\n",
       "      <td>13.243615</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.817652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493940</th>\n",
       "      <td>23.536737</td>\n",
       "      <td>17.226186</td>\n",
       "      <td>0.303837</td>\n",
       "      <td>2.888214</td>\n",
       "      <td>1.664242</td>\n",
       "      <td>4.715913</td>\n",
       "      <td>21.067640</td>\n",
       "      <td>0.587902</td>\n",
       "      <td>17.026152</td>\n",
       "      <td>35.165989</td>\n",
       "      <td>...</td>\n",
       "      <td>1.475681</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.640175</td>\n",
       "      <td>3.943578</td>\n",
       "      <td>1.805356</td>\n",
       "      <td>11.934644</td>\n",
       "      <td>5.547112</td>\n",
       "      <td>12.788488</td>\n",
       "      <td>0.025959</td>\n",
       "      <td>8.600757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR493941</th>\n",
       "      <td>19.063970</td>\n",
       "      <td>20.052478</td>\n",
       "      <td>0.572087</td>\n",
       "      <td>4.763826</td>\n",
       "      <td>5.982738</td>\n",
       "      <td>17.533812</td>\n",
       "      <td>18.588802</td>\n",
       "      <td>0.296801</td>\n",
       "      <td>12.384805</td>\n",
       "      <td>24.522430</td>\n",
       "      <td>...</td>\n",
       "      <td>0.605483</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.583209</td>\n",
       "      <td>1.573460</td>\n",
       "      <td>0.705457</td>\n",
       "      <td>15.842725</td>\n",
       "      <td>4.316776</td>\n",
       "      <td>18.247787</td>\n",
       "      <td>0.011332</td>\n",
       "      <td>7.776162</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17788 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              RPL23A       HOPX     CNTLN    YEATS2      ZEB2       DAB2  \\\n",
       "SRR493937  26.380195   7.594717  0.486687  3.244362  4.594237  10.716403   \n",
       "SRR493938  26.043990   7.449357  0.465863  3.453337  4.655460  10.348601   \n",
       "SRR493939  23.938892  17.296808  0.309277  2.774270  1.626023   4.724359   \n",
       "SRR493940  23.536737  17.226186  0.303837  2.888214  1.664242   4.715913   \n",
       "SRR493941  19.063970  20.052478  0.572087  4.763826  5.982738  17.533812   \n",
       "\n",
       "               LLGL2     TEKT3      CCNG1    HNRNPDL    ...       TMEM52  \\\n",
       "SRR493937  12.610643  1.133912   9.398151  31.639392    ...     0.841445   \n",
       "SRR493938  12.898778  1.178870   8.950152  31.577780    ...     0.762630   \n",
       "SRR493939  21.224896  0.599888  16.530086  34.013420    ...     1.633144   \n",
       "SRR493940  21.067640  0.587902  17.026152  35.165989    ...     1.475681   \n",
       "SRR493941  18.588802  0.296801  12.384805  24.522430    ...     0.605483   \n",
       "\n",
       "           SPINT3    ERICH3  GOLGA2P5    PIH1D2  GABARAPL2     PHKA2  \\\n",
       "SRR493937     0.0  3.098554  1.865488  1.264137  11.905335  3.857274   \n",
       "SRR493938     0.0  3.248183  1.838183  1.335846  11.891560  3.798020   \n",
       "SRR493939     0.0  0.602161  3.885646  1.750385  11.529645  5.524931   \n",
       "SRR493940     0.0  0.640175  3.943578  1.805356  11.934644  5.547112   \n",
       "SRR493941     0.0  0.583209  1.573460  0.705457  15.842725  4.316776   \n",
       "\n",
       "              AHCYL2   FAM170A   CAMSAP3  \n",
       "SRR493937  18.718648  0.000000  2.894070  \n",
       "SRR493938  18.645412  0.000000  2.979511  \n",
       "SRR493939  13.243615  0.000000  8.817652  \n",
       "SRR493940  12.788488  0.025959  8.600757  \n",
       "SRR493941  18.247787  0.011332  7.776162  \n",
       "\n",
       "[5 rows x 17788 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop genes\n",
    "template_data = template_data[shared_genes_hgnc]\n",
    "\n",
    "print(template_data.shape)\n",
    "template_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    }
   ],
   "source": [
    "print(len(template_data.columns) - len(shared_genes_hgnc))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Note:* There is a difference in the number of `shared_genes_hgnc` and genes in the template experiment because 3 genes have 2 different ensembl gene ids have map to the same hgnc symbol (one forward, one reverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save \n",
    "template_data.to_csv(template_data_file, float_format='%.5f', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalize compendium "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load real gene expression data\n",
    "original_compendium_file = params['compendium_data_file']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1552, 17785)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ARHGAP45</th>\n",
       "      <th>ZNF610</th>\n",
       "      <th>PPP1R3B</th>\n",
       "      <th>RBBP7</th>\n",
       "      <th>CCDC185</th>\n",
       "      <th>GPATCH3</th>\n",
       "      <th>C1orf159</th>\n",
       "      <th>DECR2</th>\n",
       "      <th>NAA11</th>\n",
       "      <th>KRTAP4-6</th>\n",
       "      <th>...</th>\n",
       "      <th>FBXL6</th>\n",
       "      <th>UBC</th>\n",
       "      <th>CD1A</th>\n",
       "      <th>OLFML2B</th>\n",
       "      <th>CACNA2D4</th>\n",
       "      <th>EID1</th>\n",
       "      <th>CHRNA10</th>\n",
       "      <th>SCYL2</th>\n",
       "      <th>SCRT1</th>\n",
       "      <th>POLR2L</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR1604987</th>\n",
       "      <td>4.686</td>\n",
       "      <td>2.914</td>\n",
       "      <td>6.349</td>\n",
       "      <td>12.617</td>\n",
       "      <td>0.034</td>\n",
       "      <td>4.167</td>\n",
       "      <td>1.487</td>\n",
       "      <td>4.824</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.107</td>\n",
       "      <td>...</td>\n",
       "      <td>6.677</td>\n",
       "      <td>418.218</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.050</td>\n",
       "      <td>40.078</td>\n",
       "      <td>0.194</td>\n",
       "      <td>9.640</td>\n",
       "      <td>0.012</td>\n",
       "      <td>52.679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604988</th>\n",
       "      <td>4.968</td>\n",
       "      <td>2.993</td>\n",
       "      <td>6.243</td>\n",
       "      <td>12.530</td>\n",
       "      <td>0.011</td>\n",
       "      <td>4.031</td>\n",
       "      <td>1.397</td>\n",
       "      <td>4.778</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.065</td>\n",
       "      <td>...</td>\n",
       "      <td>6.247</td>\n",
       "      <td>413.291</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.042</td>\n",
       "      <td>40.601</td>\n",
       "      <td>0.090</td>\n",
       "      <td>9.914</td>\n",
       "      <td>0.007</td>\n",
       "      <td>51.503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604989</th>\n",
       "      <td>5.910</td>\n",
       "      <td>2.855</td>\n",
       "      <td>13.585</td>\n",
       "      <td>16.710</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3.882</td>\n",
       "      <td>1.172</td>\n",
       "      <td>2.845</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.733</td>\n",
       "      <td>297.397</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.003</td>\n",
       "      <td>38.530</td>\n",
       "      <td>0.154</td>\n",
       "      <td>10.461</td>\n",
       "      <td>0.002</td>\n",
       "      <td>39.457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604990</th>\n",
       "      <td>5.737</td>\n",
       "      <td>2.884</td>\n",
       "      <td>14.170</td>\n",
       "      <td>16.927</td>\n",
       "      <td>0.016</td>\n",
       "      <td>3.883</td>\n",
       "      <td>1.199</td>\n",
       "      <td>2.745</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.552</td>\n",
       "      <td>298.423</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.012</td>\n",
       "      <td>37.747</td>\n",
       "      <td>0.114</td>\n",
       "      <td>10.902</td>\n",
       "      <td>0.000</td>\n",
       "      <td>38.281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604991</th>\n",
       "      <td>2.580</td>\n",
       "      <td>2.653</td>\n",
       "      <td>8.768</td>\n",
       "      <td>17.953</td>\n",
       "      <td>0.067</td>\n",
       "      <td>4.608</td>\n",
       "      <td>1.624</td>\n",
       "      <td>3.556</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.893</td>\n",
       "      <td>281.337</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0.012</td>\n",
       "      <td>64.853</td>\n",
       "      <td>0.215</td>\n",
       "      <td>8.097</td>\n",
       "      <td>0.010</td>\n",
       "      <td>83.953</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17785 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ARHGAP45  ZNF610  PPP1R3B   RBBP7  CCDC185  GPATCH3  C1orf159  \\\n",
       "SRR1604987     4.686   2.914    6.349  12.617    0.034    4.167     1.487   \n",
       "SRR1604988     4.968   2.993    6.243  12.530    0.011    4.031     1.397   \n",
       "SRR1604989     5.910   2.855   13.585  16.710    0.000    3.882     1.172   \n",
       "SRR1604990     5.737   2.884   14.170  16.927    0.016    3.883     1.199   \n",
       "SRR1604991     2.580   2.653    8.768  17.953    0.067    4.608     1.624   \n",
       "\n",
       "            DECR2  NAA11  KRTAP4-6   ...    FBXL6      UBC  CD1A  OLFML2B  \\\n",
       "SRR1604987  4.824  0.000     0.107   ...    6.677  418.218   0.0    0.070   \n",
       "SRR1604988  4.778  0.006     0.065   ...    6.247  413.291   0.0    0.040   \n",
       "SRR1604989  2.845  0.000     0.000   ...    5.733  297.397   0.0    0.019   \n",
       "SRR1604990  2.745  0.000     0.000   ...    5.552  298.423   0.0    0.038   \n",
       "SRR1604991  3.556  0.000     0.000   ...    5.893  281.337   0.0    0.986   \n",
       "\n",
       "            CACNA2D4    EID1  CHRNA10   SCYL2  SCRT1  POLR2L  \n",
       "SRR1604987     0.050  40.078    0.194   9.640  0.012  52.679  \n",
       "SRR1604988     0.042  40.601    0.090   9.914  0.007  51.503  \n",
       "SRR1604989     0.003  38.530    0.154  10.461  0.002  39.457  \n",
       "SRR1604990     0.012  37.747    0.114  10.902  0.000  38.281  \n",
       "SRR1604991     0.012  64.853    0.215   8.097  0.010  83.953  \n",
       "\n",
       "[5 rows x 17785 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read data\n",
    "original_compendium = pd.read_table(\n",
    "    original_compendium_file,\n",
    "    header=0,\n",
    "    sep='\\t',\n",
    "    index_col=0)\n",
    "\n",
    "print(original_compendium.shape)\n",
    "original_compendium.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace ensembl ids with gene symbols\n",
    "#original_compendium = process.replace_ensembl_ids(original_compendium,\n",
    "#                                                gene_id_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RPL23A</th>\n",
       "      <th>HOPX</th>\n",
       "      <th>CNTLN</th>\n",
       "      <th>YEATS2</th>\n",
       "      <th>ZEB2</th>\n",
       "      <th>DAB2</th>\n",
       "      <th>LLGL2</th>\n",
       "      <th>TEKT3</th>\n",
       "      <th>CCNG1</th>\n",
       "      <th>HNRNPDL</th>\n",
       "      <th>...</th>\n",
       "      <th>TMEM52</th>\n",
       "      <th>SPINT3</th>\n",
       "      <th>ERICH3</th>\n",
       "      <th>GOLGA2P5</th>\n",
       "      <th>PIH1D2</th>\n",
       "      <th>GABARAPL2</th>\n",
       "      <th>PHKA2</th>\n",
       "      <th>AHCYL2</th>\n",
       "      <th>FAM170A</th>\n",
       "      <th>CAMSAP3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR1604987</th>\n",
       "      <td>75.874</td>\n",
       "      <td>0.469</td>\n",
       "      <td>1.026</td>\n",
       "      <td>6.637</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.108</td>\n",
       "      <td>8.724</td>\n",
       "      <td>0.148</td>\n",
       "      <td>21.094</td>\n",
       "      <td>32.100</td>\n",
       "      <td>...</td>\n",
       "      <td>0.507</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.768</td>\n",
       "      <td>2.511</td>\n",
       "      <td>25.181</td>\n",
       "      <td>1.838</td>\n",
       "      <td>10.666</td>\n",
       "      <td>0.017</td>\n",
       "      <td>8.872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604988</th>\n",
       "      <td>76.305</td>\n",
       "      <td>0.462</td>\n",
       "      <td>0.934</td>\n",
       "      <td>6.621</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.096</td>\n",
       "      <td>8.777</td>\n",
       "      <td>0.160</td>\n",
       "      <td>21.491</td>\n",
       "      <td>32.700</td>\n",
       "      <td>...</td>\n",
       "      <td>0.511</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.781</td>\n",
       "      <td>2.293</td>\n",
       "      <td>24.445</td>\n",
       "      <td>1.834</td>\n",
       "      <td>10.647</td>\n",
       "      <td>0.016</td>\n",
       "      <td>9.349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604989</th>\n",
       "      <td>87.869</td>\n",
       "      <td>0.069</td>\n",
       "      <td>1.145</td>\n",
       "      <td>14.261</td>\n",
       "      <td>0.044</td>\n",
       "      <td>0.170</td>\n",
       "      <td>8.784</td>\n",
       "      <td>0.157</td>\n",
       "      <td>29.383</td>\n",
       "      <td>30.950</td>\n",
       "      <td>...</td>\n",
       "      <td>0.382</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.125</td>\n",
       "      <td>1.918</td>\n",
       "      <td>21.214</td>\n",
       "      <td>2.574</td>\n",
       "      <td>4.115</td>\n",
       "      <td>0.000</td>\n",
       "      <td>10.486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604990</th>\n",
       "      <td>85.461</td>\n",
       "      <td>0.060</td>\n",
       "      <td>1.221</td>\n",
       "      <td>14.436</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.156</td>\n",
       "      <td>8.665</td>\n",
       "      <td>0.070</td>\n",
       "      <td>30.541</td>\n",
       "      <td>31.362</td>\n",
       "      <td>...</td>\n",
       "      <td>0.450</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.074</td>\n",
       "      <td>2.168</td>\n",
       "      <td>20.698</td>\n",
       "      <td>2.530</td>\n",
       "      <td>4.177</td>\n",
       "      <td>0.000</td>\n",
       "      <td>10.430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604991</th>\n",
       "      <td>67.554</td>\n",
       "      <td>0.088</td>\n",
       "      <td>1.208</td>\n",
       "      <td>9.838</td>\n",
       "      <td>1.624</td>\n",
       "      <td>6.820</td>\n",
       "      <td>5.938</td>\n",
       "      <td>0.179</td>\n",
       "      <td>33.542</td>\n",
       "      <td>37.241</td>\n",
       "      <td>...</td>\n",
       "      <td>0.269</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.360</td>\n",
       "      <td>1.616</td>\n",
       "      <td>23.423</td>\n",
       "      <td>1.885</td>\n",
       "      <td>5.375</td>\n",
       "      <td>0.000</td>\n",
       "      <td>4.421</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17785 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            RPL23A   HOPX  CNTLN  YEATS2   ZEB2   DAB2  LLGL2  TEKT3   CCNG1  \\\n",
       "SRR1604987  75.874  0.469  1.026   6.637  0.020  0.108  8.724  0.148  21.094   \n",
       "SRR1604988  76.305  0.462  0.934   6.621  0.019  0.096  8.777  0.160  21.491   \n",
       "SRR1604989  87.869  0.069  1.145  14.261  0.044  0.170  8.784  0.157  29.383   \n",
       "SRR1604990  85.461  0.060  1.221  14.436  0.031  0.156  8.665  0.070  30.541   \n",
       "SRR1604991  67.554  0.088  1.208   9.838  1.624  6.820  5.938  0.179  33.542   \n",
       "\n",
       "            HNRNPDL   ...     TMEM52  SPINT3  ERICH3  GOLGA2P5  PIH1D2  \\\n",
       "SRR1604987   32.100   ...      0.507     0.0   0.000     0.768   2.511   \n",
       "SRR1604988   32.700   ...      0.511     0.0   0.005     0.781   2.293   \n",
       "SRR1604989   30.950   ...      0.382     0.0   0.000     1.125   1.918   \n",
       "SRR1604990   31.362   ...      0.450     0.0   0.000     1.074   2.168   \n",
       "SRR1604991   37.241   ...      0.269     0.0   0.000     0.360   1.616   \n",
       "\n",
       "            GABARAPL2  PHKA2  AHCYL2  FAM170A  CAMSAP3  \n",
       "SRR1604987     25.181  1.838  10.666    0.017    8.872  \n",
       "SRR1604988     24.445  1.834  10.647    0.016    9.349  \n",
       "SRR1604989     21.214  2.574   4.115    0.000   10.486  \n",
       "SRR1604990     20.698  2.530   4.177    0.000   10.430  \n",
       "SRR1604991     23.423  1.885   5.375    0.000    4.421  \n",
       "\n",
       "[5 rows x 17785 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop genes\n",
    "original_compendium = original_compendium[shared_genes_hgnc]\n",
    "\n",
    "original_compendium.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1552, 17785)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RPL23A</th>\n",
       "      <th>HOPX</th>\n",
       "      <th>CNTLN</th>\n",
       "      <th>YEATS2</th>\n",
       "      <th>ZEB2</th>\n",
       "      <th>DAB2</th>\n",
       "      <th>LLGL2</th>\n",
       "      <th>TEKT3</th>\n",
       "      <th>CCNG1</th>\n",
       "      <th>HNRNPDL</th>\n",
       "      <th>...</th>\n",
       "      <th>TMEM52</th>\n",
       "      <th>SPINT3</th>\n",
       "      <th>ERICH3</th>\n",
       "      <th>GOLGA2P5</th>\n",
       "      <th>PIH1D2</th>\n",
       "      <th>GABARAPL2</th>\n",
       "      <th>PHKA2</th>\n",
       "      <th>AHCYL2</th>\n",
       "      <th>FAM170A</th>\n",
       "      <th>CAMSAP3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR1604987</th>\n",
       "      <td>0.026935</td>\n",
       "      <td>0.002304</td>\n",
       "      <td>0.056414</td>\n",
       "      <td>0.286263</td>\n",
       "      <td>0.000169</td>\n",
       "      <td>0.000591</td>\n",
       "      <td>0.099707</td>\n",
       "      <td>0.007618</td>\n",
       "      <td>0.058015</td>\n",
       "      <td>0.190397</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006762</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.075882</td>\n",
       "      <td>0.104425</td>\n",
       "      <td>0.076349</td>\n",
       "      <td>0.019088</td>\n",
       "      <td>0.385667</td>\n",
       "      <td>0.001143</td>\n",
       "      <td>0.196963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604988</th>\n",
       "      <td>0.027088</td>\n",
       "      <td>0.002269</td>\n",
       "      <td>0.051355</td>\n",
       "      <td>0.285573</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0.000526</td>\n",
       "      <td>0.100313</td>\n",
       "      <td>0.008236</td>\n",
       "      <td>0.059107</td>\n",
       "      <td>0.193956</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006815</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000387</td>\n",
       "      <td>0.077166</td>\n",
       "      <td>0.095359</td>\n",
       "      <td>0.074117</td>\n",
       "      <td>0.019047</td>\n",
       "      <td>0.384980</td>\n",
       "      <td>0.001075</td>\n",
       "      <td>0.207553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604989</th>\n",
       "      <td>0.031193</td>\n",
       "      <td>0.000339</td>\n",
       "      <td>0.062957</td>\n",
       "      <td>0.615096</td>\n",
       "      <td>0.000373</td>\n",
       "      <td>0.000931</td>\n",
       "      <td>0.100393</td>\n",
       "      <td>0.008082</td>\n",
       "      <td>0.080812</td>\n",
       "      <td>0.183576</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005095</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.111155</td>\n",
       "      <td>0.079764</td>\n",
       "      <td>0.064321</td>\n",
       "      <td>0.026732</td>\n",
       "      <td>0.148792</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.232795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604990</th>\n",
       "      <td>0.030338</td>\n",
       "      <td>0.000295</td>\n",
       "      <td>0.067136</td>\n",
       "      <td>0.622644</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>0.000854</td>\n",
       "      <td>0.099033</td>\n",
       "      <td>0.003603</td>\n",
       "      <td>0.083997</td>\n",
       "      <td>0.186020</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.106116</td>\n",
       "      <td>0.090161</td>\n",
       "      <td>0.062756</td>\n",
       "      <td>0.026275</td>\n",
       "      <td>0.151034</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.231551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1604991</th>\n",
       "      <td>0.023981</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.066421</td>\n",
       "      <td>0.424326</td>\n",
       "      <td>0.013754</td>\n",
       "      <td>0.037351</td>\n",
       "      <td>0.067866</td>\n",
       "      <td>0.009214</td>\n",
       "      <td>0.092250</td>\n",
       "      <td>0.220890</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003588</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035570</td>\n",
       "      <td>0.067205</td>\n",
       "      <td>0.071018</td>\n",
       "      <td>0.019576</td>\n",
       "      <td>0.194352</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.098148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17785 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              RPL23A      HOPX     CNTLN    YEATS2      ZEB2      DAB2  \\\n",
       "SRR1604987  0.026935  0.002304  0.056414  0.286263  0.000169  0.000591   \n",
       "SRR1604988  0.027088  0.002269  0.051355  0.285573  0.000161  0.000526   \n",
       "SRR1604989  0.031193  0.000339  0.062957  0.615096  0.000373  0.000931   \n",
       "SRR1604990  0.030338  0.000295  0.067136  0.622644  0.000263  0.000854   \n",
       "SRR1604991  0.023981  0.000432  0.066421  0.424326  0.013754  0.037351   \n",
       "\n",
       "               LLGL2     TEKT3     CCNG1   HNRNPDL    ...       TMEM52  \\\n",
       "SRR1604987  0.099707  0.007618  0.058015  0.190397    ...     0.006762   \n",
       "SRR1604988  0.100313  0.008236  0.059107  0.193956    ...     0.006815   \n",
       "SRR1604989  0.100393  0.008082  0.080812  0.183576    ...     0.005095   \n",
       "SRR1604990  0.099033  0.003603  0.083997  0.186020    ...     0.006001   \n",
       "SRR1604991  0.067866  0.009214  0.092250  0.220890    ...     0.003588   \n",
       "\n",
       "            SPINT3    ERICH3  GOLGA2P5    PIH1D2  GABARAPL2     PHKA2  \\\n",
       "SRR1604987     0.0  0.000000  0.075882  0.104425   0.076349  0.019088   \n",
       "SRR1604988     0.0  0.000387  0.077166  0.095359   0.074117  0.019047   \n",
       "SRR1604989     0.0  0.000000  0.111155  0.079764   0.064321  0.026732   \n",
       "SRR1604990     0.0  0.000000  0.106116  0.090161   0.062756  0.026275   \n",
       "SRR1604991     0.0  0.000000  0.035570  0.067205   0.071018  0.019576   \n",
       "\n",
       "              AHCYL2   FAM170A   CAMSAP3  \n",
       "SRR1604987  0.385667  0.001143  0.196963  \n",
       "SRR1604988  0.384980  0.001075  0.207553  \n",
       "SRR1604989  0.148792  0.000000  0.232795  \n",
       "SRR1604990  0.151034  0.000000  0.231551  \n",
       "SRR1604991  0.194352  0.000000  0.098148  \n",
       "\n",
       "[5 rows x 17785 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 0-1 normalize per gene\n",
    "scaler = preprocessing.MinMaxScaler()\n",
    "original_data_scaled = scaler.fit_transform(original_compendium)\n",
    "original_data_scaled_df = pd.DataFrame(original_data_scaled,\n",
    "                                columns=original_compendium.columns,\n",
    "                                index=original_compendium.index)\n",
    "\n",
    "print(original_data_scaled_df.shape)\n",
    "original_data_scaled_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "normalized_data_file = params['normalized_compendium_data_file']\n",
    "\n",
    "original_data_scaled_df.to_csv(\n",
    "    normalized_data_file, float_format='%.3f', sep='\\t')\n",
    "\n",
    "original_compendium.to_csv(\n",
    "    original_compendium_file, float_format='%.3f', sep='\\t')\n",
    "\n",
    "# Save scaler transform\n",
    "scaler_file = params['scaler_transform_file']\n",
    "\n",
    "outfile = open(scaler_file,'wb')\n",
    "pickle.dump(scaler,outfile)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train VAE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup directories\n",
    "# Create VAE directories\n",
    "output_dirs = [os.path.join(base_dir, dataset_name, \"models\"),\n",
    "               os.path.join(base_dir, dataset_name, \"logs\")]\n",
    "\n",
    "# Check if analysis output directory exist otherwise create\n",
    "for each_dir in output_dirs:\n",
    "    if os.path.exists(each_dir) == False:\n",
    "        print('creating new directory: {}'.format(each_dir))\n",
    "        os.makedirs(each_dir, exist_ok=True)\n",
    "\n",
    "# Check if NN architecture directory exist otherwise create\n",
    "for each_dir in output_dirs:\n",
    "    new_dir = os.path.join(each_dir, NN_architecture)\n",
    "    if os.path.exists(new_dir) == False:\n",
    "        print('creating new directory: {}'.format(new_dir))\n",
    "        os.makedirs(new_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Train VAE on new compendium data\n",
    "#pipeline.train_vae(config_file,\n",
    "#                   normalized_data_file)"
   ]
  }
 ],
 "metadata": {
  "finalized": {
   "timestamp": 1592246126078,
   "trusted": true
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
