# Explore RNA-seq generic only genes

When we compared the correlation between gene percentiles generated by SOPHIE versus the manually curated dataset [here](https://github.com/greenelab/generic-expression-patterns/blob/master/human_general_analysis/gene_ranking_log2FoldChange.svg), we noticed that there was a group of genes that SOPHIE identified as generic but were not found to be generic using the manually curated dataset.
In this case, SOPHIE was trained on recount2 (RNA-seq) dataset while the manually curated dataset was using array platform.

The goal of these notebooks is to examine why these genes are uncorrelated (i.e. generic in RNA-seq and not array).
Specifically, we believe that these RNA-seq generic genes are due to differences in platform so we will test two different hypotheses that get are this.

## Processing differences

**Hypothesis 1:** These RNA-seq generic genes are due to difference in platforms.
Specifically, we believe that these genes are those that are differentially expressed using ribo-depleted vs polyA-selection.
There may exist some genes that are frequently found to be differentially expressed using polyA-selection and some genes that are frequently found to be differentially expressed using the ribo-depletion strategy.
Since our training compendium is recount2, which includes many different types of experiments therefore different selection strategies, its possible that these gene biases that result from differences in processing are found by SOPHIE but not found by Crow et. al. which uses array technology.

**Dataset:**
We were able to obtain matched sample data from [Pediatric Brain Tumor Atlas repository](https://github.com/AlexsLemonade/OpenPBTA-analysis).

To get the data for this analysis we performed the following steps:

Github issue: https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/955

1. Clone repository
```
clone: https://github.com/AlexsLemonade/OpenPBTA-analysis.git
```
2. Install docker following the instructions provided [here](https://docs.docker.com/buildx/working-with-buildx/)
3. Pull latest docker image
```
sudo docker pull ccdlopenpbta/open-pbta:latest
```
4. Navigate to root of repository
```
cd /home/alexandra/Documents/Repos/OpenPBTA-analysis
```
5. Run pre-built docker image:
```
sudo docker run --mount type=bind,target=/home/rstudio/kitematic,source=$PWD -e PASSWORD=<password> -p 8787:8787 ccdlopenpbta/open-pbta:latest
```
6. Open localhost:8787 in browser and login to RStudio using username=rstudio and the password you set above

7. In the R terminal run:
```
bash download-data.sh
```
Data files in data/

8. Get combined count data with HGNC symbols by running the following in the RStudio terminal:
```
Rscript analyses/collapse-rnaseq/01-summarize_matrices.R -i data/pbta-gene-counts-rsem-expected_count.polya.rds -g data/gencode.v27.primary_assembly.annotation.gtf.gz -m analyses/collapse-rnaseq/pbta-gene-counts-rsem-expected_counts-collapsed.polya.rds -t analyses/collapse-rnaseq/pbta-gene-counts-rsem-expected_count-collapsed_table.polya.rds
```
Same for stranded file

9. Save the output files locally

**Results:**
![hypothesis_1](https://github.com/greenelab/generic-expression-patterns/blob/master/explore_RNAseq_only_generic_genes/gene_ranking_log2FoldChange_highlight_polyA_vs_ribo.svg)

Looks like the genes that differentiate between the two protocals are not localized to the bottom right.

## Poorly captured genes

**Hypothesis 2:** These RNA-seq generic genes are those that are not well captured on microarray technology.
[Zhao et. al.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3894192/) found that one possible reason for differences in differentially expressed genes detected on different platforms is due to how genes are captured.
RNA-Seq is more likely to detect the changes at two different conditions for genes with very low expression or very high expression compared to arrays.

**Dataset:**
Expression dataset from Crow et. al. downloaded from https://github.com/PavlidisLab/gemmaAPI.R using script `download_Crow_data.R`

**Results:**
![hypothesis_2](https://github.com/greenelab/generic-expression-patterns/blob/master/explore_RNAseq_only_generic_genes/array_expression_dist_gene_groups_highlight.svg)

Looks like the RNA-seq generic genes (light blue) are lowly expressed, but not more than other RNA-seq and array generic genes (dark blue) captured on the array.

## VAE compression

Based on experiment for hypothesis 2, we noticed that the RNA-seq generic genes are low compared to RNA-seq/array generic genes in RNA-seq data.

**Hypothesis 3:** The VAE isnâ€™t adequately accounting for low read counts.
Say we have an experiment with low read depth where a set of genes have 0 expression (i.e. not detectable).
After going through the VAE shifting process, the gene counts might get compressed (i.e. very low gene counts are increased and very high gene counts are decreased) so that the new simulated counts for these originally 0-expressed genes, which should have a high error rate and therefore not found to be DE by DESeq, can be found as DE by DESeq with the artificial shift in values.
If this is the case, we will need a way to re-scale values to account for the read depth differences.
