{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Application: new experiment\n",
    "\n",
    "This notebook allows users to find specific genes in their experiment of interest using an existing VAE model\n",
    "\n",
    "This notebook will generate a `generic_gene_summary_<experiment id>.tsv` file that contains a z-score per gene that indicates how specific a gene is the experiment in question."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from ponyo import utils\n",
    "from generic_expression_patterns_modules import process, new_experiment_process, stats, ranking"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User input\n",
    "\n",
    "User needs to define the following in the [config file](../configs/config_new_experiment.tsv):\n",
    "\n",
    "1. Template experiment. This is the experiment you are interested in studying\n",
    "2. Training compendium used to train VAE, including unnormalized gene mapped version and normalized version\n",
    "3. Scaler transform used to normalize the training compendium\n",
    "4. Directory containing trained VAE model\n",
    "5. Experiment id to label newly create simulated experiments\n",
    "\n",
    "The user also needs to provide metadata files:\n",
    "1. `<experiment id>_process_samples.tsv` contains 2 columns (sample ids, label that indicates if the sample is kept or removed). See [example](data/metadata/cis-gem-par-KU1919_process_samples.tsv). **Note: This file is not required if the user wishes to use all the samples in the template experiment file.**\n",
    "2. `<experiment id>_groups.tsv` contains 2 columns: sample ids, group label to perform DE analysis. See [example](data/metadata/cis-gem-par-KU1919_groups.tsv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = os.path.abspath(\n",
    "    os.path.join(base_dir, \"configs\", \"config_new_experiment.tsv\")\n",
    ")\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load config params\n",
    "\n",
    "# Local directory to store intermediate files\n",
    "local_dir = params['local_dir']\n",
    "\n",
    "# Number of simulated experiments to generate\n",
    "num_runs = params['num_simulated']\n",
    "\n",
    "# Directory containing trained VAE model\n",
    "vae_model_dir = params['vae_model_dir']\n",
    "\n",
    "# Dimension of latent space used in VAE model\n",
    "latent_dim = params['latent_dim']\n",
    "\n",
    "# ID for template experiment\n",
    "# This ID will be used to label new simulated experiments\n",
    "project_id = params['project_id']\n",
    "\n",
    "# Template experiment filename\n",
    "template_filename = params['raw_template_filename']\n",
    "mapped_template_filename = params['mapped_template_filename']\n",
    "normalized_template_filename = params['normalized_template_filename']\n",
    "processed_template_filename = params['processed_template_filename']\n",
    "\n",
    "# Training dataset used for existing VAE model\n",
    "mapped_compendium_filename = params['mapped_compendium_filename']\n",
    "\n",
    "# Normalized compendium filename\n",
    "normalized_compendium_filename = params['normalized_compendium_filename']\n",
    "\n",
    "# Scaler transform used to scale compendium data into 0-1 range for training\n",
    "scaler_filename = params['scaler_filename']\n",
    "\n",
    "# Test statistic used to rank genes by\n",
    "col_to_rank_genes = params['rank_genes_by']\n",
    "\n",
    "# Minimum mean count per gene\n",
    "count_threshold = params['count_threshold']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata files\n",
    "\n",
    "# Load metadata file with processing information\n",
    "sample_id_metadata_filename = os.path.join(\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    f\"{project_id}_process_samples.tsv\"\n",
    ")\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "metadata_filename = os.path.join(\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    f\"{project_id}_groups.tsv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output filename\n",
    "gene_summary_filename = f\"generic_gene_summary_{project_id}.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map template experiment to same feature space as training compendium\n",
    "\n",
    "In order to simulate a new gene expression experiment, we will need to encode this experiment into the learned latent space. This requires that the feature space (i.e. genes) in the template experiment match the features in the compendium used to train the VAE model. These cells process the template experiment to be of the expected input format:\n",
    "* Template data is expected to be a matrix that is sample x gene\n",
    "* Template experiment is expected to have the same genes as the compendium experiment. Genes that are in the template experiment but not in the compendium are removed. Genes that are in the compendium but missing in the template experiment are added and the gene expression value is set to the median gene expression value of that gene across the samples in the compendium."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Template experiment needs to be of the form sample x gene\n",
    "template_filename_only = template_filename.split(\"/\")[-1].split(\".\")[0]\n",
    "transposed_template_filename = os.path.join(local_dir, template_filename_only+\"_transposed.txt\")\n",
    "\n",
    "new_experiment_process.transpose_save(template_filename, transposed_template_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(72, 58528)\n",
      "(49651, 17755)\n"
     ]
    }
   ],
   "source": [
    "new_experiment_process.process_template_experiment(\n",
    "    transposed_template_filename,\n",
    "    mapped_compendium_filename,\n",
    "    scaler_filename,\n",
    "    mapped_template_filename,\n",
    "    normalized_template_filename,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate experiments based on template experiment\n",
    "\n",
    "Embed template experiment into learned latent space and linearly shift template experiment to different locations of the latent space to create new experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate experiments based on template experiment\n",
    "normalized_compendium_data = pd.read_csv(normalized_compendium_filename, sep=\"\\t\", index_col=0, header=0)\n",
    "normalized_template_data = pd.read_csv(normalized_template_filename, sep=\"\\t\", index_col=0, header=0)\n",
    "\n",
    "for run_id in range(num_runs):\n",
    "    new_experiment_process.embed_shift_template_experiment(\n",
    "        normalized_compendium_data,\n",
    "        normalized_template_data,\n",
    "        vae_model_dir,\n",
    "        project_id,\n",
    "        scaler_filename,\n",
    "        local_dir,\n",
    "        latent_dim,\n",
    "        run_id\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process template and simulated experiments\n",
    "\n",
    "* Remove samples not required for comparison\n",
    "* Make sure ordering of samples matches metadata for proper comparison\n",
    "* Make sure values are cast as integers if using DESeq\n",
    "* Filter lowly expressed genes if using DESeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"human_general_analysis\" in vae_model_dir:\n",
    "    method = \"deseq\"\n",
    "else:\n",
    "    method = \"limma\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(sample_id_metadata_filename):\n",
    "    sample_id_metadata_filename = None\n",
    "    \n",
    "if method == \"deseq\":\n",
    "    stats.process_samples_for_DESeq(\n",
    "        mapped_template_filename,\n",
    "        metadata_filename,\n",
    "        processed_template_filename,\n",
    "        count_threshold,\n",
    "        sample_id_metadata_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\"\n",
    "        )\n",
    "        out_simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}_processed.txt\"\n",
    "        )\n",
    "        stats.process_samples_for_DESeq(\n",
    "            simulated_filename,\n",
    "            metadata_filename,\n",
    "            out_simulated_filename,\n",
    "            count_threshold,\n",
    "            sample_id_metadata_filename,\n",
    "    )\n",
    "else:\n",
    "    stats.process_samples_for_limma(\n",
    "        mapped_template_filename,\n",
    "        metadata_filename,\n",
    "        processed_template_filename,\n",
    "        sample_id_metadata_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\"\n",
    "        )\n",
    "        stats.process_samples_for_limma(\n",
    "            simulated_filename,\n",
    "            metadata_filename,\n",
    "            None,\n",
    "            sample_id_metadata_filename,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential expression analysis\n",
    "\n",
    "* If data is RNA-seq then use DESeq2 (using human_general_analysis model)\n",
    "* If data is microarray then use Limma (using human_cancer_analysis, pseudomonas_analysis models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: S4Vectors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: stats4\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocGenerics\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: parallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: IRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomicRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomeInfoDb\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: SummarizedExperiment\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Biobase\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: DelayedArray\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: matrixStats\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocParallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: estimating size factors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: estimating dispersions\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: gene-wise dispersion estimates\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: mean-dispersion relationship\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: final dispersion estimates\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: fitting model and testing\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 112 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        metadata_filename,\n",
    "        project_id, \n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        metadata_filename,\n",
    "        project_id, \n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    ) \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 235 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 438 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 104 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 143 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 105 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 102 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 84 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 48 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 129 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 29 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 609 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 892 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 889 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 409 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 24 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 596 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 601 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 79 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 260 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 214 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 901 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 523 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 205 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 117 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i base_dir -i local_dir -i num_runs -i method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# Files created: \"<local_dir>/DE_stats/DE_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir, \n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\", \n",
    "        i,\n",
    "        \"_processed.txt\",\n",
    "        sep = \"\"\n",
    "    )\n",
    "    if (method == \"deseq\"){\n",
    "        get_DE_stats_DESeq(\n",
    "            metadata_filename,\n",
    "            project_id, \n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "    }\n",
    "    else {\n",
    "        get_DE_stats_limma(\n",
    "            metadata_filename,\n",
    "            project_id, \n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rank genes\n",
    "\n",
    "Genes are ranked by their \"generic-ness\" - how frequently these genes are changed across the simulated experiments using user-specific test statistic (i.e. log2 fold change)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir,\n",
    "    \"DE_stats\",\n",
    "    f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary table\n",
    "\n",
    "* Gene ID: Gene identifier (hgnc symbols for human data or PA number for *P. aeruginosa* data)\n",
    "* (Real): Statistics for template experiment\n",
    "* (Simulated): Statistics across simulated experiments\n",
    "* Number of experiments: Number of simulated experiments\n",
    "* Z-score: High z-score indicates that gene is more changed in template compared to the null set of simulated experiments (high z-score = highly specific to template experiment)\n",
    "\n",
    "\n",
    "Note: If using DESeq, genes with NaN in `Adj P-value (Real)` column are those genes flagged because of the `cooksCutoff` parameter. The cook's distance as a diagnostic to tell if a single sample has a count which has a disproportionate impact on the log fold change and p-values. These genes are flagged with an NA in the pvalue and padj columns of the result table. For more information you can read [DESeq FAQs](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (1 levels on the left, 2 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>APBB1IP</th>\n",
       "      <td>APBB1IP</td>\n",
       "      <td>0.254791</td>\n",
       "      <td>7750.0</td>\n",
       "      <td>3.842305</td>\n",
       "      <td>-3.842305</td>\n",
       "      <td>0.983803</td>\n",
       "      <td>239.0</td>\n",
       "      <td>0.040550</td>\n",
       "      <td>0.014524</td>\n",
       "      <td>4</td>\n",
       "      <td>261.763110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZBTB17</th>\n",
       "      <td>ZBTB17</td>\n",
       "      <td>0.690399</td>\n",
       "      <td>5047.0</td>\n",
       "      <td>0.230101</td>\n",
       "      <td>-0.230101</td>\n",
       "      <td>0.966235</td>\n",
       "      <td>1628.0</td>\n",
       "      <td>0.094657</td>\n",
       "      <td>0.000574</td>\n",
       "      <td>2</td>\n",
       "      <td>236.051547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DIO2</th>\n",
       "      <td>DIO2</td>\n",
       "      <td>0.234612</td>\n",
       "      <td>7672.0</td>\n",
       "      <td>1.361258</td>\n",
       "      <td>-1.361258</td>\n",
       "      <td>0.985542</td>\n",
       "      <td>91.0</td>\n",
       "      <td>0.026934</td>\n",
       "      <td>0.005870</td>\n",
       "      <td>2</td>\n",
       "      <td>227.301178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIPG</th>\n",
       "      <td>LIPG</td>\n",
       "      <td>0.362365</td>\n",
       "      <td>7618.0</td>\n",
       "      <td>1.065054</td>\n",
       "      <td>-1.065054</td>\n",
       "      <td>0.947206</td>\n",
       "      <td>922.0</td>\n",
       "      <td>0.076981</td>\n",
       "      <td>0.007885</td>\n",
       "      <td>2</td>\n",
       "      <td>125.306419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLHL8</th>\n",
       "      <td>KLHL8</td>\n",
       "      <td>0.355956</td>\n",
       "      <td>5790.0</td>\n",
       "      <td>0.290549</td>\n",
       "      <td>-0.290549</td>\n",
       "      <td>0.860544</td>\n",
       "      <td>3208.0</td>\n",
       "      <td>0.126050</td>\n",
       "      <td>0.001465</td>\n",
       "      <td>2</td>\n",
       "      <td>112.287286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DCAKD</th>\n",
       "      <td>DCAKD</td>\n",
       "      <td>0.076283</td>\n",
       "      <td>7290.0</td>\n",
       "      <td>0.590127</td>\n",
       "      <td>0.590127</td>\n",
       "      <td>0.995908</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.011037</td>\n",
       "      <td>0.005176</td>\n",
       "      <td>2</td>\n",
       "      <td>111.889617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN25</th>\n",
       "      <td>ZSCAN25</td>\n",
       "      <td>0.346296</td>\n",
       "      <td>5786.0</td>\n",
       "      <td>0.289934</td>\n",
       "      <td>0.289934</td>\n",
       "      <td>0.973542</td>\n",
       "      <td>179.0</td>\n",
       "      <td>0.040688</td>\n",
       "      <td>0.003167</td>\n",
       "      <td>2</td>\n",
       "      <td>78.694865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACKR3</th>\n",
       "      <td>ACKR3</td>\n",
       "      <td>0.522792</td>\n",
       "      <td>7724.0</td>\n",
       "      <td>2.181050</td>\n",
       "      <td>-2.181050</td>\n",
       "      <td>0.973542</td>\n",
       "      <td>227.0</td>\n",
       "      <td>0.045143</td>\n",
       "      <td>0.028716</td>\n",
       "      <td>2</td>\n",
       "      <td>74.380158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3H2</th>\n",
       "      <td>P3H2</td>\n",
       "      <td>0.836610</td>\n",
       "      <td>6989.0</td>\n",
       "      <td>0.469682</td>\n",
       "      <td>-0.469682</td>\n",
       "      <td>0.980255</td>\n",
       "      <td>76.0</td>\n",
       "      <td>0.023071</td>\n",
       "      <td>0.006214</td>\n",
       "      <td>2</td>\n",
       "      <td>71.875236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIPK4</th>\n",
       "      <td>RIPK4</td>\n",
       "      <td>0.835102</td>\n",
       "      <td>5679.0</td>\n",
       "      <td>0.279611</td>\n",
       "      <td>-0.279611</td>\n",
       "      <td>0.950200</td>\n",
       "      <td>4237.0</td>\n",
       "      <td>0.145883</td>\n",
       "      <td>0.002468</td>\n",
       "      <td>3</td>\n",
       "      <td>54.183316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "APBB1IP  APBB1IP            0.254791       7750.0                    3.842305   \n",
       "ZBTB17    ZBTB17            0.690399       5047.0                    0.230101   \n",
       "DIO2        DIO2            0.234612       7672.0                    1.361258   \n",
       "LIPG        LIPG            0.362365       7618.0                    1.065054   \n",
       "KLHL8      KLHL8            0.355956       5790.0                    0.290549   \n",
       "DCAKD      DCAKD            0.076283       7290.0                    0.590127   \n",
       "ZSCAN25  ZSCAN25            0.346296       5786.0                    0.289934   \n",
       "ACKR3      ACKR3            0.522792       7724.0                    2.181050   \n",
       "P3H2        P3H2            0.836610       6989.0                    0.469682   \n",
       "RIPK4      RIPK4            0.835102       5679.0                    0.279611   \n",
       "\n",
       "         log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "APBB1IP              -3.842305                        0.983803   \n",
       "ZBTB17               -0.230101                        0.966235   \n",
       "DIO2                 -1.361258                        0.985542   \n",
       "LIPG                 -1.065054                        0.947206   \n",
       "KLHL8                -0.290549                        0.860544   \n",
       "DCAKD                 0.590127                        0.995908   \n",
       "ZSCAN25               0.289934                        0.973542   \n",
       "ACKR3                -2.181050                        0.973542   \n",
       "P3H2                 -0.469682                        0.980255   \n",
       "RIPK4                -0.279611                        0.950200   \n",
       "\n",
       "         Rank (simulated)  Mean abs(log2FoldChange) (simulated)  \\\n",
       "APBB1IP             239.0                              0.040550   \n",
       "ZBTB17             1628.0                              0.094657   \n",
       "DIO2                 91.0                              0.026934   \n",
       "LIPG                922.0                              0.076981   \n",
       "KLHL8              3208.0                              0.126050   \n",
       "DCAKD                34.0                              0.011037   \n",
       "ZSCAN25             179.0                              0.040688   \n",
       "ACKR3               227.0                              0.045143   \n",
       "P3H2                 76.0                              0.023071   \n",
       "RIPK4              4237.0                              0.145883   \n",
       "\n",
       "         Std deviation (simulated)  Number of experiments (simulated)  \\\n",
       "APBB1IP                   0.014524                                  4   \n",
       "ZBTB17                    0.000574                                  2   \n",
       "DIO2                      0.005870                                  2   \n",
       "LIPG                      0.007885                                  2   \n",
       "KLHL8                     0.001465                                  2   \n",
       "DCAKD                     0.005176                                  2   \n",
       "ZSCAN25                   0.003167                                  2   \n",
       "ACKR3                     0.028716                                  2   \n",
       "P3H2                      0.006214                                  2   \n",
       "RIPK4                     0.002468                                  3   \n",
       "\n",
       "            Z score  \n",
       "APBB1IP  261.763110  \n",
       "ZBTB17   236.051547  \n",
       "DIO2     227.301178  \n",
       "LIPG     125.306419  \n",
       "KLHL8    112.287286  \n",
       "DCAKD    111.889617  \n",
       "ZSCAN25   78.694865  \n",
       "ACKR3     74.380158  \n",
       "P3H2      71.875236  \n",
       "RIPK4     54.183316  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get summary table\n",
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    'gene',\n",
    "    params\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene ID                                 False\n",
       "Adj P-value (Real)                      False\n",
       "Rank (Real)                             False\n",
       "abs(log2FoldChange) (Real)              False\n",
       "log2FoldChange (Real)                   False\n",
       "Median adj p-value (simulated)          False\n",
       "Rank (simulated)                        False\n",
       "Mean abs(log2FoldChange) (simulated)    False\n",
       "Std deviation (simulated)                True\n",
       "Number of experiments (simulated)       False\n",
       "Z score                                  True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks.isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PLCH2</th>\n",
       "      <td>PLCH2</td>\n",
       "      <td>0.004045</td>\n",
       "      <td>7747.0</td>\n",
       "      <td>3.468302</td>\n",
       "      <td>-3.468302</td>\n",
       "      <td>7.407263e-01</td>\n",
       "      <td>9706.0</td>\n",
       "      <td>0.654560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNCG</th>\n",
       "      <td>SNCG</td>\n",
       "      <td>0.363589</td>\n",
       "      <td>7692.0</td>\n",
       "      <td>1.530144</td>\n",
       "      <td>-1.530144</td>\n",
       "      <td>9.476316e-01</td>\n",
       "      <td>4688.0</td>\n",
       "      <td>0.157218</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAT8L</th>\n",
       "      <td>NAT8L</td>\n",
       "      <td>0.063518</td>\n",
       "      <td>7679.0</td>\n",
       "      <td>1.423398</td>\n",
       "      <td>-1.423398</td>\n",
       "      <td>8.987325e-01</td>\n",
       "      <td>9233.0</td>\n",
       "      <td>0.432301</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFB2</th>\n",
       "      <td>TGFB2</td>\n",
       "      <td>0.538644</td>\n",
       "      <td>7676.0</td>\n",
       "      <td>1.401821</td>\n",
       "      <td>-1.401821</td>\n",
       "      <td>9.133891e-01</td>\n",
       "      <td>9637.0</td>\n",
       "      <td>0.603258</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MROH6</th>\n",
       "      <td>MROH6</td>\n",
       "      <td>0.262993</td>\n",
       "      <td>7669.0</td>\n",
       "      <td>1.336284</td>\n",
       "      <td>1.336284</td>\n",
       "      <td>9.978440e-01</td>\n",
       "      <td>8841.0</td>\n",
       "      <td>0.365070</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCBE1</th>\n",
       "      <td>CCBE1</td>\n",
       "      <td>0.290052</td>\n",
       "      <td>7661.0</td>\n",
       "      <td>1.310070</td>\n",
       "      <td>1.310070</td>\n",
       "      <td>9.133891e-01</td>\n",
       "      <td>9876.0</td>\n",
       "      <td>0.863658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL11</th>\n",
       "      <td>IL11</td>\n",
       "      <td>0.583795</td>\n",
       "      <td>7648.0</td>\n",
       "      <td>1.177909</td>\n",
       "      <td>-1.177909</td>\n",
       "      <td>9.998920e-01</td>\n",
       "      <td>9556.0</td>\n",
       "      <td>0.550287</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR</th>\n",
       "      <td>HR</td>\n",
       "      <td>0.565028</td>\n",
       "      <td>7647.0</td>\n",
       "      <td>1.176085</td>\n",
       "      <td>1.176085</td>\n",
       "      <td>8.618461e-01</td>\n",
       "      <td>7481.0</td>\n",
       "      <td>0.251641</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMCC3</th>\n",
       "      <td>TMCC3</td>\n",
       "      <td>0.531672</td>\n",
       "      <td>7640.0</td>\n",
       "      <td>1.140188</td>\n",
       "      <td>-1.140188</td>\n",
       "      <td>9.697681e-01</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0.018445</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLCF1</th>\n",
       "      <td>CLCF1</td>\n",
       "      <td>0.110332</td>\n",
       "      <td>7595.0</td>\n",
       "      <td>1.004859</td>\n",
       "      <td>-1.004859</td>\n",
       "      <td>9.978440e-01</td>\n",
       "      <td>7824.0</td>\n",
       "      <td>0.272146</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCDH10</th>\n",
       "      <td>PCDH10</td>\n",
       "      <td>0.648327</td>\n",
       "      <td>7557.0</td>\n",
       "      <td>0.885865</td>\n",
       "      <td>-0.885865</td>\n",
       "      <td>8.703668e-01</td>\n",
       "      <td>9093.0</td>\n",
       "      <td>0.406424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TENM4</th>\n",
       "      <td>TENM4</td>\n",
       "      <td>0.836610</td>\n",
       "      <td>7553.0</td>\n",
       "      <td>0.881028</td>\n",
       "      <td>0.881028</td>\n",
       "      <td>9.998920e-01</td>\n",
       "      <td>470.0</td>\n",
       "      <td>0.059349</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GALC</th>\n",
       "      <td>GALC</td>\n",
       "      <td>0.866870</td>\n",
       "      <td>7540.0</td>\n",
       "      <td>0.853732</td>\n",
       "      <td>-0.853732</td>\n",
       "      <td>9.994518e-01</td>\n",
       "      <td>7925.0</td>\n",
       "      <td>0.279213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADGRB2</th>\n",
       "      <td>ADGRB2</td>\n",
       "      <td>0.463667</td>\n",
       "      <td>7524.0</td>\n",
       "      <td>0.834131</td>\n",
       "      <td>-0.834131</td>\n",
       "      <td>9.954900e-01</td>\n",
       "      <td>315.0</td>\n",
       "      <td>0.051834</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHLDA2</th>\n",
       "      <td>PHLDA2</td>\n",
       "      <td>0.097089</td>\n",
       "      <td>7521.0</td>\n",
       "      <td>0.832112</td>\n",
       "      <td>-0.832112</td>\n",
       "      <td>9.978440e-01</td>\n",
       "      <td>9575.0</td>\n",
       "      <td>0.565791</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDLIM4</th>\n",
       "      <td>PDLIM4</td>\n",
       "      <td>0.888856</td>\n",
       "      <td>7514.0</td>\n",
       "      <td>0.813720</td>\n",
       "      <td>-0.813720</td>\n",
       "      <td>9.133891e-01</td>\n",
       "      <td>8683.0</td>\n",
       "      <td>0.345311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF503</th>\n",
       "      <td>ZNF503</td>\n",
       "      <td>0.417130</td>\n",
       "      <td>7508.0</td>\n",
       "      <td>0.808946</td>\n",
       "      <td>-0.808946</td>\n",
       "      <td>9.224550e-01</td>\n",
       "      <td>6671.0</td>\n",
       "      <td>0.215211</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAP3K14</th>\n",
       "      <td>MAP3K14</td>\n",
       "      <td>0.112500</td>\n",
       "      <td>7506.0</td>\n",
       "      <td>0.807813</td>\n",
       "      <td>0.807813</td>\n",
       "      <td>8.885463e-01</td>\n",
       "      <td>338.0</td>\n",
       "      <td>0.053144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HKDC1</th>\n",
       "      <td>HKDC1</td>\n",
       "      <td>0.561052</td>\n",
       "      <td>7492.0</td>\n",
       "      <td>0.792325</td>\n",
       "      <td>-0.792325</td>\n",
       "      <td>9.978440e-01</td>\n",
       "      <td>9656.0</td>\n",
       "      <td>0.614613</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ITPKC</th>\n",
       "      <td>ITPKC</td>\n",
       "      <td>0.076283</td>\n",
       "      <td>7482.0</td>\n",
       "      <td>0.773988</td>\n",
       "      <td>-0.773988</td>\n",
       "      <td>9.978440e-01</td>\n",
       "      <td>7298.0</td>\n",
       "      <td>0.242714</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VEGFC</th>\n",
       "      <td>VEGFC</td>\n",
       "      <td>0.499695</td>\n",
       "      <td>7478.0</td>\n",
       "      <td>0.766217</td>\n",
       "      <td>-0.766217</td>\n",
       "      <td>9.133891e-01</td>\n",
       "      <td>8907.0</td>\n",
       "      <td>0.377006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OLR1</th>\n",
       "      <td>OLR1</td>\n",
       "      <td>0.827710</td>\n",
       "      <td>7473.0</td>\n",
       "      <td>0.759269</td>\n",
       "      <td>-0.759269</td>\n",
       "      <td>9.994518e-01</td>\n",
       "      <td>5648.0</td>\n",
       "      <td>0.182386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF128</th>\n",
       "      <td>RNF128</td>\n",
       "      <td>0.928709</td>\n",
       "      <td>7462.0</td>\n",
       "      <td>0.742530</td>\n",
       "      <td>-0.742530</td>\n",
       "      <td>9.994518e-01</td>\n",
       "      <td>4211.0</td>\n",
       "      <td>0.146576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHMT2</th>\n",
       "      <td>BHMT2</td>\n",
       "      <td>0.322185</td>\n",
       "      <td>7449.0</td>\n",
       "      <td>0.723293</td>\n",
       "      <td>-0.723293</td>\n",
       "      <td>9.476316e-01</td>\n",
       "      <td>9367.0</td>\n",
       "      <td>0.467303</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM89B</th>\n",
       "      <td>FAM89B</td>\n",
       "      <td>0.156295</td>\n",
       "      <td>7431.0</td>\n",
       "      <td>0.704381</td>\n",
       "      <td>-0.704381</td>\n",
       "      <td>9.730709e-01</td>\n",
       "      <td>210.0</td>\n",
       "      <td>0.044318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LDOC1</th>\n",
       "      <td>LDOC1</td>\n",
       "      <td>0.326268</td>\n",
       "      <td>7411.0</td>\n",
       "      <td>0.673671</td>\n",
       "      <td>0.673671</td>\n",
       "      <td>9.965330e-01</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.008601</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VANGL2</th>\n",
       "      <td>VANGL2</td>\n",
       "      <td>0.922745</td>\n",
       "      <td>7398.0</td>\n",
       "      <td>0.665270</td>\n",
       "      <td>0.665270</td>\n",
       "      <td>9.476316e-01</td>\n",
       "      <td>6156.0</td>\n",
       "      <td>0.198164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MCC</th>\n",
       "      <td>MCC</td>\n",
       "      <td>0.672879</td>\n",
       "      <td>7386.0</td>\n",
       "      <td>0.652127</td>\n",
       "      <td>-0.652127</td>\n",
       "      <td>9.940326e-01</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.004343</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PALM</th>\n",
       "      <td>PALM</td>\n",
       "      <td>0.885249</td>\n",
       "      <td>7383.0</td>\n",
       "      <td>0.649962</td>\n",
       "      <td>-0.649962</td>\n",
       "      <td>9.757951e-01</td>\n",
       "      <td>4056.0</td>\n",
       "      <td>0.143254</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSPYL5</th>\n",
       "      <td>TSPYL5</td>\n",
       "      <td>0.953404</td>\n",
       "      <td>7374.0</td>\n",
       "      <td>0.643369</td>\n",
       "      <td>0.643369</td>\n",
       "      <td>7.129761e-01</td>\n",
       "      <td>9256.0</td>\n",
       "      <td>0.436972</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CENPW</th>\n",
       "      <td>CENPW</td>\n",
       "      <td>0.979196</td>\n",
       "      <td>911.0</td>\n",
       "      <td>0.034627</td>\n",
       "      <td>0.034627</td>\n",
       "      <td>7.130780e-01</td>\n",
       "      <td>9738.0</td>\n",
       "      <td>0.672497</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCL1</th>\n",
       "      <td>RCL1</td>\n",
       "      <td>0.968518</td>\n",
       "      <td>880.0</td>\n",
       "      <td>0.033706</td>\n",
       "      <td>0.033706</td>\n",
       "      <td>7.245875e-01</td>\n",
       "      <td>9707.0</td>\n",
       "      <td>0.654708</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTC13</th>\n",
       "      <td>TTC13</td>\n",
       "      <td>0.979776</td>\n",
       "      <td>846.0</td>\n",
       "      <td>0.032673</td>\n",
       "      <td>-0.032673</td>\n",
       "      <td>5.825290e-01</td>\n",
       "      <td>8734.0</td>\n",
       "      <td>0.351125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MED30</th>\n",
       "      <td>MED30</td>\n",
       "      <td>0.982446</td>\n",
       "      <td>782.0</td>\n",
       "      <td>0.030191</td>\n",
       "      <td>0.030191</td>\n",
       "      <td>4.797607e-08</td>\n",
       "      <td>10156.0</td>\n",
       "      <td>4.490571</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MED20</th>\n",
       "      <td>MED20</td>\n",
       "      <td>0.979418</td>\n",
       "      <td>779.0</td>\n",
       "      <td>0.030100</td>\n",
       "      <td>-0.030100</td>\n",
       "      <td>9.346254e-01</td>\n",
       "      <td>712.0</td>\n",
       "      <td>0.069696</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PICK1</th>\n",
       "      <td>PICK1</td>\n",
       "      <td>0.965384</td>\n",
       "      <td>763.0</td>\n",
       "      <td>0.029556</td>\n",
       "      <td>-0.029556</td>\n",
       "      <td>8.389070e-01</td>\n",
       "      <td>6058.0</td>\n",
       "      <td>0.195148</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFAP4</th>\n",
       "      <td>TFAP4</td>\n",
       "      <td>0.978828</td>\n",
       "      <td>752.0</td>\n",
       "      <td>0.029230</td>\n",
       "      <td>-0.029230</td>\n",
       "      <td>1.406457e-04</td>\n",
       "      <td>10074.0</td>\n",
       "      <td>2.259566</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETAA1</th>\n",
       "      <td>ETAA1</td>\n",
       "      <td>0.969299</td>\n",
       "      <td>749.0</td>\n",
       "      <td>0.029165</td>\n",
       "      <td>-0.029165</td>\n",
       "      <td>9.927579e-01</td>\n",
       "      <td>6068.0</td>\n",
       "      <td>0.195418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCORL1</th>\n",
       "      <td>BCORL1</td>\n",
       "      <td>0.981293</td>\n",
       "      <td>689.0</td>\n",
       "      <td>0.026318</td>\n",
       "      <td>-0.026318</td>\n",
       "      <td>9.965330e-01</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.010269</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIAA0556</th>\n",
       "      <td>KIAA0556</td>\n",
       "      <td>0.979196</td>\n",
       "      <td>648.0</td>\n",
       "      <td>0.024663</td>\n",
       "      <td>-0.024663</td>\n",
       "      <td>7.943372e-01</td>\n",
       "      <td>6279.0</td>\n",
       "      <td>0.202275</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRBD1</th>\n",
       "      <td>SRBD1</td>\n",
       "      <td>0.979622</td>\n",
       "      <td>624.0</td>\n",
       "      <td>0.023764</td>\n",
       "      <td>-0.023764</td>\n",
       "      <td>9.927579e-01</td>\n",
       "      <td>560.0</td>\n",
       "      <td>0.062803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NKRF</th>\n",
       "      <td>NKRF</td>\n",
       "      <td>0.976875</td>\n",
       "      <td>617.0</td>\n",
       "      <td>0.023534</td>\n",
       "      <td>0.023534</td>\n",
       "      <td>7.891151e-01</td>\n",
       "      <td>5693.0</td>\n",
       "      <td>0.183607</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GPR180</th>\n",
       "      <td>GPR180</td>\n",
       "      <td>0.980273</td>\n",
       "      <td>576.0</td>\n",
       "      <td>0.022304</td>\n",
       "      <td>-0.022304</td>\n",
       "      <td>9.133891e-01</td>\n",
       "      <td>8316.0</td>\n",
       "      <td>0.309923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAPK8IP1</th>\n",
       "      <td>MAPK8IP1</td>\n",
       "      <td>0.993625</td>\n",
       "      <td>552.0</td>\n",
       "      <td>0.021502</td>\n",
       "      <td>-0.021502</td>\n",
       "      <td>9.299498e-01</td>\n",
       "      <td>8809.0</td>\n",
       "      <td>0.361785</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM217B</th>\n",
       "      <td>FAM217B</td>\n",
       "      <td>0.979196</td>\n",
       "      <td>546.0</td>\n",
       "      <td>0.021232</td>\n",
       "      <td>-0.021232</td>\n",
       "      <td>7.403247e-02</td>\n",
       "      <td>7706.0</td>\n",
       "      <td>0.264126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAAP1</th>\n",
       "      <td>CAAP1</td>\n",
       "      <td>0.981198</td>\n",
       "      <td>530.0</td>\n",
       "      <td>0.020565</td>\n",
       "      <td>0.020565</td>\n",
       "      <td>5.585322e-01</td>\n",
       "      <td>2579.0</td>\n",
       "      <td>0.114026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALDH4A1</th>\n",
       "      <td>ALDH4A1</td>\n",
       "      <td>0.990391</td>\n",
       "      <td>468.0</td>\n",
       "      <td>0.017972</td>\n",
       "      <td>0.017972</td>\n",
       "      <td>9.698867e-01</td>\n",
       "      <td>151.0</td>\n",
       "      <td>0.037700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOXJ2</th>\n",
       "      <td>FOXJ2</td>\n",
       "      <td>0.982868</td>\n",
       "      <td>453.0</td>\n",
       "      <td>0.017408</td>\n",
       "      <td>0.017408</td>\n",
       "      <td>9.806547e-01</td>\n",
       "      <td>101.0</td>\n",
       "      <td>0.029115</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF362</th>\n",
       "      <td>ZNF362</td>\n",
       "      <td>0.988223</td>\n",
       "      <td>443.0</td>\n",
       "      <td>0.017153</td>\n",
       "      <td>-0.017153</td>\n",
       "      <td>9.978440e-01</td>\n",
       "      <td>3336.0</td>\n",
       "      <td>0.128810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FDX1</th>\n",
       "      <td>FDX1</td>\n",
       "      <td>0.990391</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.014154</td>\n",
       "      <td>-0.014154</td>\n",
       "      <td>9.994518e-01</td>\n",
       "      <td>1523.0</td>\n",
       "      <td>0.092129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB32</th>\n",
       "      <td>RAB32</td>\n",
       "      <td>0.992639</td>\n",
       "      <td>334.0</td>\n",
       "      <td>0.012810</td>\n",
       "      <td>0.012810</td>\n",
       "      <td>9.211473e-01</td>\n",
       "      <td>2413.0</td>\n",
       "      <td>0.110776</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF473</th>\n",
       "      <td>ZNF473</td>\n",
       "      <td>0.990258</td>\n",
       "      <td>317.0</td>\n",
       "      <td>0.012171</td>\n",
       "      <td>-0.012171</td>\n",
       "      <td>9.994518e-01</td>\n",
       "      <td>1932.0</td>\n",
       "      <td>0.100717</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EOGT</th>\n",
       "      <td>EOGT</td>\n",
       "      <td>0.993258</td>\n",
       "      <td>314.0</td>\n",
       "      <td>0.012006</td>\n",
       "      <td>0.012006</td>\n",
       "      <td>9.133891e-01</td>\n",
       "      <td>6407.0</td>\n",
       "      <td>0.205845</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCCA</th>\n",
       "      <td>PCCA</td>\n",
       "      <td>0.993258</td>\n",
       "      <td>297.0</td>\n",
       "      <td>0.011520</td>\n",
       "      <td>-0.011520</td>\n",
       "      <td>9.994518e-01</td>\n",
       "      <td>5865.0</td>\n",
       "      <td>0.189106</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIN54</th>\n",
       "      <td>LIN54</td>\n",
       "      <td>0.986786</td>\n",
       "      <td>278.0</td>\n",
       "      <td>0.010914</td>\n",
       "      <td>0.010914</td>\n",
       "      <td>9.968601e-01</td>\n",
       "      <td>8491.0</td>\n",
       "      <td>0.326171</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM219A</th>\n",
       "      <td>FAM219A</td>\n",
       "      <td>0.992717</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.006547</td>\n",
       "      <td>0.006547</td>\n",
       "      <td>7.931288e-01</td>\n",
       "      <td>8748.0</td>\n",
       "      <td>0.352163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VPS37C</th>\n",
       "      <td>VPS37C</td>\n",
       "      <td>0.996219</td>\n",
       "      <td>70.0</td>\n",
       "      <td>0.002914</td>\n",
       "      <td>-0.002914</td>\n",
       "      <td>5.092639e-01</td>\n",
       "      <td>7621.0</td>\n",
       "      <td>0.259225</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SURF1</th>\n",
       "      <td>SURF1</td>\n",
       "      <td>0.999278</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.001245</td>\n",
       "      <td>0.001245</td>\n",
       "      <td>9.781185e-01</td>\n",
       "      <td>148.0</td>\n",
       "      <td>0.036865</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LPCAT2</th>\n",
       "      <td>LPCAT2</td>\n",
       "      <td>0.999857</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.001166</td>\n",
       "      <td>-0.001166</td>\n",
       "      <td>9.992124e-01</td>\n",
       "      <td>6757.0</td>\n",
       "      <td>0.218553</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRFN4</th>\n",
       "      <td>LRFN4</td>\n",
       "      <td>0.999857</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.001072</td>\n",
       "      <td>-0.001072</td>\n",
       "      <td>6.069665e-01</td>\n",
       "      <td>9414.0</td>\n",
       "      <td>0.485961</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>240 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "PLCH2        PLCH2            0.004045       7747.0   \n",
       "SNCG          SNCG            0.363589       7692.0   \n",
       "NAT8L        NAT8L            0.063518       7679.0   \n",
       "TGFB2        TGFB2            0.538644       7676.0   \n",
       "MROH6        MROH6            0.262993       7669.0   \n",
       "CCBE1        CCBE1            0.290052       7661.0   \n",
       "IL11          IL11            0.583795       7648.0   \n",
       "HR              HR            0.565028       7647.0   \n",
       "TMCC3        TMCC3            0.531672       7640.0   \n",
       "CLCF1        CLCF1            0.110332       7595.0   \n",
       "PCDH10      PCDH10            0.648327       7557.0   \n",
       "TENM4        TENM4            0.836610       7553.0   \n",
       "GALC          GALC            0.866870       7540.0   \n",
       "ADGRB2      ADGRB2            0.463667       7524.0   \n",
       "PHLDA2      PHLDA2            0.097089       7521.0   \n",
       "PDLIM4      PDLIM4            0.888856       7514.0   \n",
       "ZNF503      ZNF503            0.417130       7508.0   \n",
       "MAP3K14    MAP3K14            0.112500       7506.0   \n",
       "HKDC1        HKDC1            0.561052       7492.0   \n",
       "ITPKC        ITPKC            0.076283       7482.0   \n",
       "VEGFC        VEGFC            0.499695       7478.0   \n",
       "OLR1          OLR1            0.827710       7473.0   \n",
       "RNF128      RNF128            0.928709       7462.0   \n",
       "BHMT2        BHMT2            0.322185       7449.0   \n",
       "FAM89B      FAM89B            0.156295       7431.0   \n",
       "LDOC1        LDOC1            0.326268       7411.0   \n",
       "VANGL2      VANGL2            0.922745       7398.0   \n",
       "MCC            MCC            0.672879       7386.0   \n",
       "PALM          PALM            0.885249       7383.0   \n",
       "TSPYL5      TSPYL5            0.953404       7374.0   \n",
       "...            ...                 ...          ...   \n",
       "CENPW        CENPW            0.979196        911.0   \n",
       "RCL1          RCL1            0.968518        880.0   \n",
       "TTC13        TTC13            0.979776        846.0   \n",
       "MED30        MED30            0.982446        782.0   \n",
       "MED20        MED20            0.979418        779.0   \n",
       "PICK1        PICK1            0.965384        763.0   \n",
       "TFAP4        TFAP4            0.978828        752.0   \n",
       "ETAA1        ETAA1            0.969299        749.0   \n",
       "BCORL1      BCORL1            0.981293        689.0   \n",
       "KIAA0556  KIAA0556            0.979196        648.0   \n",
       "SRBD1        SRBD1            0.979622        624.0   \n",
       "NKRF          NKRF            0.976875        617.0   \n",
       "GPR180      GPR180            0.980273        576.0   \n",
       "MAPK8IP1  MAPK8IP1            0.993625        552.0   \n",
       "FAM217B    FAM217B            0.979196        546.0   \n",
       "CAAP1        CAAP1            0.981198        530.0   \n",
       "ALDH4A1    ALDH4A1            0.990391        468.0   \n",
       "FOXJ2        FOXJ2            0.982868        453.0   \n",
       "ZNF362      ZNF362            0.988223        443.0   \n",
       "FDX1          FDX1            0.990391        362.0   \n",
       "RAB32        RAB32            0.992639        334.0   \n",
       "ZNF473      ZNF473            0.990258        317.0   \n",
       "EOGT          EOGT            0.993258        314.0   \n",
       "PCCA          PCCA            0.993258        297.0   \n",
       "LIN54        LIN54            0.986786        278.0   \n",
       "FAM219A    FAM219A            0.992717        159.0   \n",
       "VPS37C      VPS37C            0.996219         70.0   \n",
       "SURF1        SURF1            0.999278         33.0   \n",
       "LPCAT2      LPCAT2            0.999857         29.0   \n",
       "LRFN4        LRFN4            0.999857         26.0   \n",
       "\n",
       "          abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "PLCH2                       3.468302              -3.468302   \n",
       "SNCG                        1.530144              -1.530144   \n",
       "NAT8L                       1.423398              -1.423398   \n",
       "TGFB2                       1.401821              -1.401821   \n",
       "MROH6                       1.336284               1.336284   \n",
       "CCBE1                       1.310070               1.310070   \n",
       "IL11                        1.177909              -1.177909   \n",
       "HR                          1.176085               1.176085   \n",
       "TMCC3                       1.140188              -1.140188   \n",
       "CLCF1                       1.004859              -1.004859   \n",
       "PCDH10                      0.885865              -0.885865   \n",
       "TENM4                       0.881028               0.881028   \n",
       "GALC                        0.853732              -0.853732   \n",
       "ADGRB2                      0.834131              -0.834131   \n",
       "PHLDA2                      0.832112              -0.832112   \n",
       "PDLIM4                      0.813720              -0.813720   \n",
       "ZNF503                      0.808946              -0.808946   \n",
       "MAP3K14                     0.807813               0.807813   \n",
       "HKDC1                       0.792325              -0.792325   \n",
       "ITPKC                       0.773988              -0.773988   \n",
       "VEGFC                       0.766217              -0.766217   \n",
       "OLR1                        0.759269              -0.759269   \n",
       "RNF128                      0.742530              -0.742530   \n",
       "BHMT2                       0.723293              -0.723293   \n",
       "FAM89B                      0.704381              -0.704381   \n",
       "LDOC1                       0.673671               0.673671   \n",
       "VANGL2                      0.665270               0.665270   \n",
       "MCC                         0.652127              -0.652127   \n",
       "PALM                        0.649962              -0.649962   \n",
       "TSPYL5                      0.643369               0.643369   \n",
       "...                              ...                    ...   \n",
       "CENPW                       0.034627               0.034627   \n",
       "RCL1                        0.033706               0.033706   \n",
       "TTC13                       0.032673              -0.032673   \n",
       "MED30                       0.030191               0.030191   \n",
       "MED20                       0.030100              -0.030100   \n",
       "PICK1                       0.029556              -0.029556   \n",
       "TFAP4                       0.029230              -0.029230   \n",
       "ETAA1                       0.029165              -0.029165   \n",
       "BCORL1                      0.026318              -0.026318   \n",
       "KIAA0556                    0.024663              -0.024663   \n",
       "SRBD1                       0.023764              -0.023764   \n",
       "NKRF                        0.023534               0.023534   \n",
       "GPR180                      0.022304              -0.022304   \n",
       "MAPK8IP1                    0.021502              -0.021502   \n",
       "FAM217B                     0.021232              -0.021232   \n",
       "CAAP1                       0.020565               0.020565   \n",
       "ALDH4A1                     0.017972               0.017972   \n",
       "FOXJ2                       0.017408               0.017408   \n",
       "ZNF362                      0.017153              -0.017153   \n",
       "FDX1                        0.014154              -0.014154   \n",
       "RAB32                       0.012810               0.012810   \n",
       "ZNF473                      0.012171              -0.012171   \n",
       "EOGT                        0.012006               0.012006   \n",
       "PCCA                        0.011520              -0.011520   \n",
       "LIN54                       0.010914               0.010914   \n",
       "FAM219A                     0.006547               0.006547   \n",
       "VPS37C                      0.002914              -0.002914   \n",
       "SURF1                       0.001245               0.001245   \n",
       "LPCAT2                      0.001166              -0.001166   \n",
       "LRFN4                       0.001072              -0.001072   \n",
       "\n",
       "          Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "PLCH2                       7.407263e-01            9706.0   \n",
       "SNCG                        9.476316e-01            4688.0   \n",
       "NAT8L                       8.987325e-01            9233.0   \n",
       "TGFB2                       9.133891e-01            9637.0   \n",
       "MROH6                       9.978440e-01            8841.0   \n",
       "CCBE1                       9.133891e-01            9876.0   \n",
       "IL11                        9.998920e-01            9556.0   \n",
       "HR                          8.618461e-01            7481.0   \n",
       "TMCC3                       9.697681e-01              57.0   \n",
       "CLCF1                       9.978440e-01            7824.0   \n",
       "PCDH10                      8.703668e-01            9093.0   \n",
       "TENM4                       9.998920e-01             470.0   \n",
       "GALC                        9.994518e-01            7925.0   \n",
       "ADGRB2                      9.954900e-01             315.0   \n",
       "PHLDA2                      9.978440e-01            9575.0   \n",
       "PDLIM4                      9.133891e-01            8683.0   \n",
       "ZNF503                      9.224550e-01            6671.0   \n",
       "MAP3K14                     8.885463e-01             338.0   \n",
       "HKDC1                       9.978440e-01            9656.0   \n",
       "ITPKC                       9.978440e-01            7298.0   \n",
       "VEGFC                       9.133891e-01            8907.0   \n",
       "OLR1                        9.994518e-01            5648.0   \n",
       "RNF128                      9.994518e-01            4211.0   \n",
       "BHMT2                       9.476316e-01            9367.0   \n",
       "FAM89B                      9.730709e-01             210.0   \n",
       "LDOC1                       9.965330e-01              25.0   \n",
       "VANGL2                      9.476316e-01            6156.0   \n",
       "MCC                         9.940326e-01              10.0   \n",
       "PALM                        9.757951e-01            4056.0   \n",
       "TSPYL5                      7.129761e-01            9256.0   \n",
       "...                                  ...               ...   \n",
       "CENPW                       7.130780e-01            9738.0   \n",
       "RCL1                        7.245875e-01            9707.0   \n",
       "TTC13                       5.825290e-01            8734.0   \n",
       "MED30                       4.797607e-08           10156.0   \n",
       "MED20                       9.346254e-01             712.0   \n",
       "PICK1                       8.389070e-01            6058.0   \n",
       "TFAP4                       1.406457e-04           10074.0   \n",
       "ETAA1                       9.927579e-01            6068.0   \n",
       "BCORL1                      9.965330e-01              28.0   \n",
       "KIAA0556                    7.943372e-01            6279.0   \n",
       "SRBD1                       9.927579e-01             560.0   \n",
       "NKRF                        7.891151e-01            5693.0   \n",
       "GPR180                      9.133891e-01            8316.0   \n",
       "MAPK8IP1                    9.299498e-01            8809.0   \n",
       "FAM217B                     7.403247e-02            7706.0   \n",
       "CAAP1                       5.585322e-01            2579.0   \n",
       "ALDH4A1                     9.698867e-01             151.0   \n",
       "FOXJ2                       9.806547e-01             101.0   \n",
       "ZNF362                      9.978440e-01            3336.0   \n",
       "FDX1                        9.994518e-01            1523.0   \n",
       "RAB32                       9.211473e-01            2413.0   \n",
       "ZNF473                      9.994518e-01            1932.0   \n",
       "EOGT                        9.133891e-01            6407.0   \n",
       "PCCA                        9.994518e-01            5865.0   \n",
       "LIN54                       9.968601e-01            8491.0   \n",
       "FAM219A                     7.931288e-01            8748.0   \n",
       "VPS37C                      5.092639e-01            7621.0   \n",
       "SURF1                       9.781185e-01             148.0   \n",
       "LPCAT2                      9.992124e-01            6757.0   \n",
       "LRFN4                       6.069665e-01            9414.0   \n",
       "\n",
       "          Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "PLCH2                                 0.654560                        NaN   \n",
       "SNCG                                  0.157218                        NaN   \n",
       "NAT8L                                 0.432301                        NaN   \n",
       "TGFB2                                 0.603258                        NaN   \n",
       "MROH6                                 0.365070                        NaN   \n",
       "CCBE1                                 0.863658                        NaN   \n",
       "IL11                                  0.550287                        NaN   \n",
       "HR                                    0.251641                        NaN   \n",
       "TMCC3                                 0.018445                        NaN   \n",
       "CLCF1                                 0.272146                        NaN   \n",
       "PCDH10                                0.406424                        NaN   \n",
       "TENM4                                 0.059349                        NaN   \n",
       "GALC                                  0.279213                        NaN   \n",
       "ADGRB2                                0.051834                        NaN   \n",
       "PHLDA2                                0.565791                        NaN   \n",
       "PDLIM4                                0.345311                        NaN   \n",
       "ZNF503                                0.215211                        NaN   \n",
       "MAP3K14                               0.053144                        NaN   \n",
       "HKDC1                                 0.614613                        NaN   \n",
       "ITPKC                                 0.242714                        NaN   \n",
       "VEGFC                                 0.377006                        NaN   \n",
       "OLR1                                  0.182386                        NaN   \n",
       "RNF128                                0.146576                        NaN   \n",
       "BHMT2                                 0.467303                        NaN   \n",
       "FAM89B                                0.044318                        NaN   \n",
       "LDOC1                                 0.008601                        NaN   \n",
       "VANGL2                                0.198164                        NaN   \n",
       "MCC                                   0.004343                        NaN   \n",
       "PALM                                  0.143254                        NaN   \n",
       "TSPYL5                                0.436972                        NaN   \n",
       "...                                        ...                        ...   \n",
       "CENPW                                 0.672497                        NaN   \n",
       "RCL1                                  0.654708                        NaN   \n",
       "TTC13                                 0.351125                        NaN   \n",
       "MED30                                 4.490571                        NaN   \n",
       "MED20                                 0.069696                        NaN   \n",
       "PICK1                                 0.195148                        NaN   \n",
       "TFAP4                                 2.259566                        NaN   \n",
       "ETAA1                                 0.195418                        NaN   \n",
       "BCORL1                                0.010269                        NaN   \n",
       "KIAA0556                              0.202275                        NaN   \n",
       "SRBD1                                 0.062803                        NaN   \n",
       "NKRF                                  0.183607                        NaN   \n",
       "GPR180                                0.309923                        NaN   \n",
       "MAPK8IP1                              0.361785                        NaN   \n",
       "FAM217B                               0.264126                        NaN   \n",
       "CAAP1                                 0.114026                        NaN   \n",
       "ALDH4A1                               0.037700                        NaN   \n",
       "FOXJ2                                 0.029115                        NaN   \n",
       "ZNF362                                0.128810                        NaN   \n",
       "FDX1                                  0.092129                        NaN   \n",
       "RAB32                                 0.110776                        NaN   \n",
       "ZNF473                                0.100717                        NaN   \n",
       "EOGT                                  0.205845                        NaN   \n",
       "PCCA                                  0.189106                        NaN   \n",
       "LIN54                                 0.326171                        NaN   \n",
       "FAM219A                               0.352163                        NaN   \n",
       "VPS37C                                0.259225                        NaN   \n",
       "SURF1                                 0.036865                        NaN   \n",
       "LPCAT2                                0.218553                        NaN   \n",
       "LRFN4                                 0.485961                        NaN   \n",
       "\n",
       "          Number of experiments (simulated)  Z score  \n",
       "PLCH2                                     1      NaN  \n",
       "SNCG                                      1      NaN  \n",
       "NAT8L                                     1      NaN  \n",
       "TGFB2                                     1      NaN  \n",
       "MROH6                                     1      NaN  \n",
       "CCBE1                                     1      NaN  \n",
       "IL11                                      1      NaN  \n",
       "HR                                        1      NaN  \n",
       "TMCC3                                     1      NaN  \n",
       "CLCF1                                     1      NaN  \n",
       "PCDH10                                    1      NaN  \n",
       "TENM4                                     1      NaN  \n",
       "GALC                                      1      NaN  \n",
       "ADGRB2                                    1      NaN  \n",
       "PHLDA2                                    1      NaN  \n",
       "PDLIM4                                    1      NaN  \n",
       "ZNF503                                    1      NaN  \n",
       "MAP3K14                                   1      NaN  \n",
       "HKDC1                                     1      NaN  \n",
       "ITPKC                                     1      NaN  \n",
       "VEGFC                                     1      NaN  \n",
       "OLR1                                      1      NaN  \n",
       "RNF128                                    1      NaN  \n",
       "BHMT2                                     1      NaN  \n",
       "FAM89B                                    1      NaN  \n",
       "LDOC1                                     1      NaN  \n",
       "VANGL2                                    1      NaN  \n",
       "MCC                                       1      NaN  \n",
       "PALM                                      1      NaN  \n",
       "TSPYL5                                    1      NaN  \n",
       "...                                     ...      ...  \n",
       "CENPW                                     1      NaN  \n",
       "RCL1                                      1      NaN  \n",
       "TTC13                                     1      NaN  \n",
       "MED30                                     1      NaN  \n",
       "MED20                                     1      NaN  \n",
       "PICK1                                     1      NaN  \n",
       "TFAP4                                     1      NaN  \n",
       "ETAA1                                     1      NaN  \n",
       "BCORL1                                    1      NaN  \n",
       "KIAA0556                                  1      NaN  \n",
       "SRBD1                                     1      NaN  \n",
       "NKRF                                      1      NaN  \n",
       "GPR180                                    1      NaN  \n",
       "MAPK8IP1                                  1      NaN  \n",
       "FAM217B                                   1      NaN  \n",
       "CAAP1                                     1      NaN  \n",
       "ALDH4A1                                   1      NaN  \n",
       "FOXJ2                                     1      NaN  \n",
       "ZNF362                                    1      NaN  \n",
       "FDX1                                      1      NaN  \n",
       "RAB32                                     1      NaN  \n",
       "ZNF473                                    1      NaN  \n",
       "EOGT                                      1      NaN  \n",
       "PCCA                                      1      NaN  \n",
       "LIN54                                     1      NaN  \n",
       "FAM219A                                   1      NaN  \n",
       "VPS37C                                    1      NaN  \n",
       "SURF1                                     1      NaN  \n",
       "LPCAT2                                    1      NaN  \n",
       "LRFN4                                     1      NaN  \n",
       "\n",
       "[240 rows x 11 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks[summary_gene_ranks.isna().any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "summary_gene_ranks.to_csv(gene_summary_filename, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
