{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Application: new experiment\n",
    "\n",
    "This notebook allows users to find specific genes in their experiment of interest using an existing VAE model\n",
    "\n",
    "This notebook will generate a `generic_gene_summary_<experiment id>.tsv` file that contains a z-score per gene that indicates how specific a gene is the experiment in question."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression_linux/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from ponyo import utils\n",
    "from generic_expression_patterns_modules import (\n",
    "    process,\n",
    "    new_experiment_process,\n",
    "    stats,\n",
    "    ranking,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User input\n",
    "\n",
    "User needs to define the following in the [config file](../configs/config_new_experiment.tsv):\n",
    "\n",
    "1. Directory on your local machine to store intermediate and output data files generated (`local_dir`). Make sure to end with \"\\\".\n",
    "2. Template experiment (`raw_template_filename`). This is the experiment you are interested in studying.\n",
    "3. Training compendium used to train VAE, including unnormalized gene mapped version (`mapped_compendium_filename`) and normalized version (`normalized_compendium_filename`). Links to these datasets can be found in the README.\n",
    "4. Scaler transform (`scaler_filename`) used to normalize the training compendium. This can be found in the `data/` directory within the analysis folder.\n",
    "5. Directory (`vae_model_dir`) containing trained VAE model.\n",
    "6. Experiment id (`project_id`) to label newly create simulated experiments.\n",
    "7. The number of experiments to simulate (`num_simulated`)\n",
    "8. Minimum average read count to filter data by (`count_threshold`)\n",
    "9. Size of existing models latent dimension (`latent_dim`). Should be 30.\n",
    "10. Name of column header (`rank_genes_by`) from DE association statistic results. This column will be use to rank genes. Select \"logFC\", \"P.Value\", \"adj.P.Val\", \"t\" if using Limma. Select \"log2FoldChange\", \"pvalue\", \"padj\" if using DESeq.\n",
    "11. `DE_logFC_name` is either \"logFC\" or \"log2FoldChange\". This is used for plotting volcano plots.\n",
    "12. `DE_pvalue_name` is either \"adj.P.Val\" or \"padj\". This is used for plotting volcano plots.\n",
    "\n",
    "The remaining parameters specify filenames that intermediate data files will be written to.\n",
    "\n",
    "The user also needs to provide metadata files:\n",
    "\n",
    "These files should be located in `data/metadata/` directory.\n",
    "1. `<experiment id>_process_samples.tsv` contains 2 columns (sample ids, label that indicates if the sample is kept or removed). See [example](data/metadata/cis-gem-par-KU1919_process_samples.tsv). **Note: This file is not required if the user wishes to use all the samples in the template experiment file.**\n",
    "2. `<experiment id>_groups.tsv` contains 2 columns: sample ids, group label to perform DE analysis. See [example](data/metadata/cis-gem-par-KU1919_groups.tsv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = os.path.abspath(\n",
    "    os.path.join(base_dir, \"configs\", \"config_new_experiment.tsv\")\n",
    ")\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load config params\n",
    "\n",
    "# Local directory to store intermediate files\n",
    "local_dir = params[\"local_dir\"]\n",
    "\n",
    "# Number of simulated experiments to generate\n",
    "num_runs = params[\"num_simulated\"]\n",
    "\n",
    "# Directory containing trained VAE model\n",
    "vae_model_dir = params[\"vae_model_dir\"]\n",
    "\n",
    "# Dimension of latent space used in VAE model\n",
    "latent_dim = params[\"latent_dim\"]\n",
    "\n",
    "# ID for template experiment\n",
    "# This ID will be used to label new simulated experiments\n",
    "project_id = params[\"project_id\"]\n",
    "\n",
    "# Template experiment filename\n",
    "template_filename = params[\"raw_template_filename\"]\n",
    "mapped_template_filename = params[\"mapped_template_filename\"]\n",
    "normalized_template_filename = params[\"normalized_template_filename\"]\n",
    "processed_template_filename = params[\"processed_template_filename\"]\n",
    "\n",
    "# Training dataset used for existing VAE model\n",
    "mapped_compendium_filename = params[\"mapped_compendium_filename\"]\n",
    "\n",
    "# Normalized compendium filename\n",
    "normalized_compendium_filename = params[\"normalized_compendium_filename\"]\n",
    "\n",
    "# Scaler transform used to scale compendium data into 0-1 range for training\n",
    "scaler_filename = params[\"scaler_filename\"]\n",
    "\n",
    "# Test statistic used to rank genes by\n",
    "col_to_rank_genes = params[\"rank_genes_by\"]\n",
    "\n",
    "# Minimum mean count per gene\n",
    "count_threshold = params[\"count_threshold\"]\n",
    "\n",
    "# Column headers to use to make summary statistic table\n",
    "logFC_name = params[\"DE_logFC_name\"]\n",
    "pvalue_name = params[\"DE_pvalue_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata files\n",
    "\n",
    "# Load metadata file with processing information\n",
    "sample_id_metadata_filename = os.path.join(\n",
    "    \"data\", \"metadata\", f\"{project_id}_process_samples.tsv\"\n",
    ")\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "metadata_filename = os.path.join(\"data\", \"metadata\", f\"{project_id}_groups.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output filename\n",
    "gene_summary_filename = f\"generic_gene_summary_{project_id}.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map template experiment to same feature space as training compendium\n",
    "\n",
    "In order to simulate a new gene expression experiment, we will need to encode this experiment into the learned latent space. This requires that the feature space (i.e. genes) in the template experiment match the features in the compendium used to train the VAE model. These cells process the template experiment to be of the expected input format:\n",
    "* Template data is expected to be a matrix that is sample x gene\n",
    "* Template experiment is expected to have the same genes as the compendium experiment. Genes that are in the template experiment but not in the compendium are removed. Genes that are in the compendium but missing in the template experiment are added and the gene expression value is set to the median gene expression value of that gene across the samples in the compendium."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Template experiment needs to be of the form sample x gene\n",
    "template_filename_only = template_filename.split(\"/\")[-1].split(\".\")[0]\n",
    "transposed_template_filename = os.path.join(\n",
    "    local_dir, template_filename_only + \"_transposed.txt\"\n",
    ")\n",
    "\n",
    "new_experiment_process.transpose_save(template_filename, transposed_template_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(72, 58528)\n",
      "(49651, 17755)\n"
     ]
    }
   ],
   "source": [
    "new_experiment_process.process_template_experiment(\n",
    "    transposed_template_filename,\n",
    "    mapped_compendium_filename,\n",
    "    scaler_filename,\n",
    "    mapped_template_filename,\n",
    "    normalized_template_filename,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate experiments based on template experiment\n",
    "\n",
    "Embed template experiment into learned latent space and linearly shift template experiment to different locations of the latent space to create new experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression_linux/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression_linux/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate experiments based on template experiment\n",
    "normalized_compendium_data = pd.read_csv(\n",
    "    normalized_compendium_filename, sep=\"\\t\", index_col=0, header=0\n",
    ")\n",
    "normalized_template_data = pd.read_csv(\n",
    "    normalized_template_filename, sep=\"\\t\", index_col=0, header=0\n",
    ")\n",
    "\n",
    "for run_id in range(num_runs):\n",
    "    new_experiment_process.embed_shift_template_experiment(\n",
    "        normalized_compendium_data,\n",
    "        normalized_template_data,\n",
    "        vae_model_dir,\n",
    "        project_id,\n",
    "        scaler_filename,\n",
    "        local_dir,\n",
    "        latent_dim,\n",
    "        run_id,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process template and simulated experiments\n",
    "\n",
    "* Remove samples not required for comparison\n",
    "* Make sure ordering of samples matches metadata for proper comparison\n",
    "* Make sure values are cast as integers if using DESeq\n",
    "* Filter lowly expressed genes if using DESeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"human_general_analysis\" in vae_model_dir:\n",
    "    method = \"deseq\"\n",
    "else:\n",
    "    method = \"limma\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(sample_id_metadata_filename):\n",
    "    sample_id_metadata_filename = None\n",
    "\n",
    "if method == \"deseq\":\n",
    "    stats.process_samples_for_DESeq(\n",
    "        mapped_template_filename,\n",
    "        metadata_filename,\n",
    "        processed_template_filename,\n",
    "        count_threshold,\n",
    "        sample_id_metadata_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        out_simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}_processed.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_DESeq(\n",
    "            simulated_filename,\n",
    "            metadata_filename,\n",
    "            out_simulated_filename,\n",
    "            count_threshold,\n",
    "            sample_id_metadata_filename,\n",
    "        )\n",
    "else:\n",
    "    stats.process_samples_for_limma(\n",
    "        mapped_template_filename,\n",
    "        metadata_filename,\n",
    "        processed_template_filename,\n",
    "        sample_id_metadata_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_limma(\n",
    "            simulated_filename,\n",
    "            metadata_filename,\n",
    "            None,\n",
    "            sample_id_metadata_filename,\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential expression analysis\n",
    "\n",
    "* If data is RNA-seq then use DESeq2 (using human_general_analysis model)\n",
    "* If data is microarray then use Limma (using human_cancer_analysis, pseudomonas_analysis models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: DelayedArray\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: BiocParallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 1223 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        metadata_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        metadata_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 148 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 3593 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 2362 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 4816 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 937 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 1197 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 1265 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 1147 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 4575 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 1475 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 2469 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 1264 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 832 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 213 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 603 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 1982 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 2291 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 879 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 1945 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 304 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 2628 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 965 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 3288 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 647 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n",
      "R[write to console]: -- replacing outliers and refitting for 6142 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i base_dir -i local_dir -i num_runs -i method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# Files created: \"<local_dir>/DE_stats/DE_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\",\n",
    "        i,\n",
    "        \"_processed.txt\",\n",
    "        sep = \"\"\n",
    "    )\n",
    "    if (method == \"deseq\"){\n",
    "        get_DE_stats_DESeq(\n",
    "            metadata_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "    }\n",
    "    else {\n",
    "        get_DE_stats_limma(\n",
    "            metadata_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rank genes\n",
    "\n",
    "Genes are ranked by their \"generic-ness\" - how frequently these genes are changed across the simulated experiments using user-specific test statistic (i.e. log2 fold change)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression_linux/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (2 levels on the left, 1 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    "    logFC_name,\n",
    "    pvalue_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary table\n",
    "\n",
    "* Gene ID: Gene identifier (hgnc symbols for human data or PA number for *P. aeruginosa* data)\n",
    "* (Real): Statistics for template experiment\n",
    "* (Simulated): Statistics across simulated experiments\n",
    "* Number of experiments: Number of simulated experiments\n",
    "* Z-score: High z-score indicates that gene is more changed in template compared to the null set of simulated experiments (high z-score = highly specific to template experiment)\n",
    "\n",
    "\n",
    "**Note:**\n",
    "* If using DESeq, genes with NaN in only the `Adj P-value (Real)` column are those genes flagged because of the `cooksCutoff` parameter. The cook's distance as a diagnostic to tell if a single sample has a count which has a disproportionate impact on the log fold change and p-values. These genes are flagged with an NA in the pvalue and padj columns of the result table.\n",
    "\n",
    "* If using DESeq with count threshold, some genes may not be present in all simulated experiments (i.e. the `Number of experiments (simulated)` will not equal the number of simulated experiments you specified in the beginning. This pre-filtering will lead to some genes found in few simulated experiments and so the background/null set for that gene is not robust. Thus, the user should sort by both z-score and number of experiments to identify specific expressed genes.\n",
    "\n",
    "* If using DESeq without count threshold, some genes may still not be present in all simulated experiments (i.e. the `Number of experiments (simulated)`  will not equal the number of simulated experiments you specified in the beginning. If the gene is 0 expressed across all samples and thus automatically given an NA in `log fold change, adjusted p-value` columns. Thus, the user should sort by both z-score and number of experiments to identify specific expressed genes.\n",
    "\n",
    "For more information you can read [DESeq FAQs](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RXFP2</th>\n",
       "      <td>RXFP2</td>\n",
       "      <td>2.039775e-09</td>\n",
       "      <td>17159.0</td>\n",
       "      <td>21.707360</td>\n",
       "      <td>-21.707360</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>14996.0</td>\n",
       "      <td>84.459840</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.615232</td>\n",
       "      <td>0.607123</td>\n",
       "      <td>24</td>\n",
       "      <td>34.741114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LPAR6</th>\n",
       "      <td>LPAR6</td>\n",
       "      <td>1.253706e-04</td>\n",
       "      <td>17130.0</td>\n",
       "      <td>4.844192</td>\n",
       "      <td>-4.844192</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>10539.0</td>\n",
       "      <td>59.355638</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.347514</td>\n",
       "      <td>0.236803</td>\n",
       "      <td>25</td>\n",
       "      <td>18.989098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRRC4</th>\n",
       "      <td>LRRC4</td>\n",
       "      <td>2.994721e-05</td>\n",
       "      <td>17151.0</td>\n",
       "      <td>5.863040</td>\n",
       "      <td>-5.863040</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>15758.0</td>\n",
       "      <td>88.751831</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.517340</td>\n",
       "      <td>0.344745</td>\n",
       "      <td>25</td>\n",
       "      <td>15.506263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GPX7</th>\n",
       "      <td>GPX7</td>\n",
       "      <td>6.607610e-03</td>\n",
       "      <td>17153.0</td>\n",
       "      <td>6.082224</td>\n",
       "      <td>6.082224</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>10176.0</td>\n",
       "      <td>57.311029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.402356</td>\n",
       "      <td>0.370654</td>\n",
       "      <td>25</td>\n",
       "      <td>15.323893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH11</th>\n",
       "      <td>CDH11</td>\n",
       "      <td>3.358492e-02</td>\n",
       "      <td>17154.0</td>\n",
       "      <td>6.143647</td>\n",
       "      <td>-6.143647</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>13082.0</td>\n",
       "      <td>73.679171</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.466132</td>\n",
       "      <td>0.422218</td>\n",
       "      <td>25</td>\n",
       "      <td>13.446865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SGCD</th>\n",
       "      <td>SGCD</td>\n",
       "      <td>1.980677e-01</td>\n",
       "      <td>17056.0</td>\n",
       "      <td>3.837687</td>\n",
       "      <td>-3.837687</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>12409.0</td>\n",
       "      <td>69.888476</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.400874</td>\n",
       "      <td>0.295899</td>\n",
       "      <td>25</td>\n",
       "      <td>11.614811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01122</th>\n",
       "      <td>LINC01122</td>\n",
       "      <td>6.332486e-01</td>\n",
       "      <td>16409.0</td>\n",
       "      <td>1.868682</td>\n",
       "      <td>-1.868682</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>7036.0</td>\n",
       "      <td>39.624873</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.255531</td>\n",
       "      <td>0.141522</td>\n",
       "      <td>25</td>\n",
       "      <td>11.398626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNTNAP2</th>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>7.272676e-07</td>\n",
       "      <td>17156.0</td>\n",
       "      <td>7.331360</td>\n",
       "      <td>-7.331360</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>14945.0</td>\n",
       "      <td>84.172581</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.641911</td>\n",
       "      <td>0.598723</td>\n",
       "      <td>25</td>\n",
       "      <td>11.172856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD163L1</th>\n",
       "      <td>CD163L1</td>\n",
       "      <td>4.388168e-03</td>\n",
       "      <td>17021.0</td>\n",
       "      <td>3.532494</td>\n",
       "      <td>3.532494</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>6993.0</td>\n",
       "      <td>39.382674</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.357009</td>\n",
       "      <td>0.298431</td>\n",
       "      <td>25</td>\n",
       "      <td>10.640587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RRM1</th>\n",
       "      <td>RRM1</td>\n",
       "      <td>8.404496e-08</td>\n",
       "      <td>16913.0</td>\n",
       "      <td>2.990056</td>\n",
       "      <td>2.990056</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>6799.0</td>\n",
       "      <td>38.289963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.287038</td>\n",
       "      <td>0.254274</td>\n",
       "      <td>25</td>\n",
       "      <td>10.630326</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "RXFP2          RXFP2        2.039775e-09      17159.0   \n",
       "LPAR6          LPAR6        1.253706e-04      17130.0   \n",
       "LRRC4          LRRC4        2.994721e-05      17151.0   \n",
       "GPX7            GPX7        6.607610e-03      17153.0   \n",
       "CDH11          CDH11        3.358492e-02      17154.0   \n",
       "SGCD            SGCD        1.980677e-01      17056.0   \n",
       "LINC01122  LINC01122        6.332486e-01      16409.0   \n",
       "CNTNAP2      CNTNAP2        7.272676e-07      17156.0   \n",
       "CD163L1      CD163L1        4.388168e-03      17021.0   \n",
       "RRM1            RRM1        8.404496e-08      16913.0   \n",
       "\n",
       "           abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "RXFP2                       21.707360             -21.707360   \n",
       "LPAR6                        4.844192              -4.844192   \n",
       "LRRC4                        5.863040              -5.863040   \n",
       "GPX7                         6.082224               6.082224   \n",
       "CDH11                        6.143647              -6.143647   \n",
       "SGCD                         3.837687              -3.837687   \n",
       "LINC01122                    1.868682              -1.868682   \n",
       "CNTNAP2                      7.331360              -7.331360   \n",
       "CD163L1                      3.532494               3.532494   \n",
       "RRM1                         2.990056               2.990056   \n",
       "\n",
       "           Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "RXFP2                            0.999950           14996.0   \n",
       "LPAR6                            0.999951           10539.0   \n",
       "LRRC4                            0.999951           15758.0   \n",
       "GPX7                             0.999949           10176.0   \n",
       "CDH11                            0.999951           13082.0   \n",
       "SGCD                             0.999951           12409.0   \n",
       "LINC01122                        0.999949            7036.0   \n",
       "CNTNAP2                          0.999951           14945.0   \n",
       "CD163L1                          0.999951            6993.0   \n",
       "RRM1                             0.999949            6799.0   \n",
       "\n",
       "           Percentile (simulated)  Percent DE (simulated)  \\\n",
       "RXFP2                   84.459840                     0.0   \n",
       "LPAR6                   59.355638                     0.0   \n",
       "LRRC4                   88.751831                     0.0   \n",
       "GPX7                    57.311029                     0.0   \n",
       "CDH11                   73.679171                     0.0   \n",
       "SGCD                    69.888476                     0.0   \n",
       "LINC01122               39.624873                     0.0   \n",
       "CNTNAP2                 84.172581                     0.0   \n",
       "CD163L1                 39.382674                     0.0   \n",
       "RRM1                    38.289963                     0.0   \n",
       "\n",
       "           Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "RXFP2                                  0.615232                   0.607123   \n",
       "LPAR6                                  0.347514                   0.236803   \n",
       "LRRC4                                  0.517340                   0.344745   \n",
       "GPX7                                   0.402356                   0.370654   \n",
       "CDH11                                  0.466132                   0.422218   \n",
       "SGCD                                   0.400874                   0.295899   \n",
       "LINC01122                              0.255531                   0.141522   \n",
       "CNTNAP2                                0.641911                   0.598723   \n",
       "CD163L1                                0.357009                   0.298431   \n",
       "RRM1                                   0.287038                   0.254274   \n",
       "\n",
       "           Number of experiments (simulated)    Z score  \n",
       "RXFP2                                     24  34.741114  \n",
       "LPAR6                                     25  18.989098  \n",
       "LRRC4                                     25  15.506263  \n",
       "GPX7                                      25  15.323893  \n",
       "CDH11                                     25  13.446865  \n",
       "SGCD                                      25  11.614811  \n",
       "LINC01122                                 25  11.398626  \n",
       "CNTNAP2                                   25  11.172856  \n",
       "CD163L1                                   25  10.640587  \n",
       "RRM1                                      25  10.630326  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get summary table\n",
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    \"gene\",\n",
    "    params,\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene ID                                 False\n",
       "Adj P-value (Real)                       True\n",
       "Rank (Real)                              True\n",
       "abs(log2FoldChange) (Real)               True\n",
       "log2FoldChange (Real)                    True\n",
       "Median adj p-value (simulated)          False\n",
       "Rank (simulated)                        False\n",
       "Percentile (simulated)                  False\n",
       "Percent DE (simulated)                  False\n",
       "Mean abs(log2FoldChange) (simulated)    False\n",
       "Std deviation (simulated)               False\n",
       "Number of experiments (simulated)       False\n",
       "Z score                                  True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks.isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACOD1</th>\n",
       "      <td>ACOD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>16969.0</td>\n",
       "      <td>95.572829</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.963899</td>\n",
       "      <td>0.797525</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL9</th>\n",
       "      <td>ACTL9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>17211.0</td>\n",
       "      <td>96.935902</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.978766</td>\n",
       "      <td>0.782871</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTRT2</th>\n",
       "      <td>ACTRT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>17069.0</td>\n",
       "      <td>96.136082</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.937594</td>\n",
       "      <td>0.781906</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM30</th>\n",
       "      <td>ADAM30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>13941.0</td>\n",
       "      <td>78.517517</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.636214</td>\n",
       "      <td>0.704730</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM6</th>\n",
       "      <td>ADAM6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>14720.0</td>\n",
       "      <td>82.905261</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.839843</td>\n",
       "      <td>0.854896</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAMDEC1</th>\n",
       "      <td>ADAMDEC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>16772.0</td>\n",
       "      <td>94.463220</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.943490</td>\n",
       "      <td>0.855386</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKAIN1</th>\n",
       "      <td>AKAIN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>15067.0</td>\n",
       "      <td>84.859750</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.773474</td>\n",
       "      <td>0.644060</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALPI</th>\n",
       "      <td>ALPI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999946</td>\n",
       "      <td>15669.0</td>\n",
       "      <td>88.250535</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.459798</td>\n",
       "      <td>1.753978</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMELY</th>\n",
       "      <td>AMELY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999954</td>\n",
       "      <td>12640.0</td>\n",
       "      <td>71.189591</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.653440</td>\n",
       "      <td>0.627023</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANHX</th>\n",
       "      <td>ANHX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>16825.0</td>\n",
       "      <td>94.761744</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.043032</td>\n",
       "      <td>0.953204</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANP32D</th>\n",
       "      <td>ANP32D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>10304.0</td>\n",
       "      <td>58.031993</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.412181</td>\n",
       "      <td>0.400176</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APOA4</th>\n",
       "      <td>APOA4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999910</td>\n",
       "      <td>17506.0</td>\n",
       "      <td>98.597499</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>1.877437</td>\n",
       "      <td>1.960742</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APOBEC1</th>\n",
       "      <td>APOBEC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>16338.0</td>\n",
       "      <td>92.018700</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.962360</td>\n",
       "      <td>0.953836</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APOC3</th>\n",
       "      <td>APOC3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>17523.0</td>\n",
       "      <td>98.693252</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.311955</td>\n",
       "      <td>1.104857</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASB10</th>\n",
       "      <td>ASB10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>17038.0</td>\n",
       "      <td>95.961473</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.890658</td>\n",
       "      <td>0.685825</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASB17</th>\n",
       "      <td>ASB17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>15144.0</td>\n",
       "      <td>85.293455</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.736293</td>\n",
       "      <td>0.695222</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATXN8OS</th>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>15688.0</td>\n",
       "      <td>88.357553</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.899167</td>\n",
       "      <td>0.818626</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AWAT1</th>\n",
       "      <td>AWAT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>11476.0</td>\n",
       "      <td>64.633322</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.665827</td>\n",
       "      <td>0.784592</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCORP1</th>\n",
       "      <td>BCORP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>8875.0</td>\n",
       "      <td>49.983102</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.578825</td>\n",
       "      <td>0.784105</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BGLT3</th>\n",
       "      <td>BGLT3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>16235.0</td>\n",
       "      <td>91.438549</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.929932</td>\n",
       "      <td>0.996257</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHLHE23</th>\n",
       "      <td>BHLHE23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999954</td>\n",
       "      <td>17151.0</td>\n",
       "      <td>96.597950</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.148463</td>\n",
       "      <td>0.851245</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BMP15</th>\n",
       "      <td>BMP15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>17142.0</td>\n",
       "      <td>96.547257</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.097471</td>\n",
       "      <td>0.769280</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BPIFA3</th>\n",
       "      <td>BPIFA3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>16831.0</td>\n",
       "      <td>94.795539</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.929227</td>\n",
       "      <td>0.690446</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BPIFA4P</th>\n",
       "      <td>BPIFA4P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>17379.0</td>\n",
       "      <td>97.882167</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.001450</td>\n",
       "      <td>0.671031</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BPIFB2</th>\n",
       "      <td>BPIFB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999946</td>\n",
       "      <td>17706.0</td>\n",
       "      <td>99.724006</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.486836</td>\n",
       "      <td>1.195091</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BPIFB6</th>\n",
       "      <td>BPIFB6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>15423.0</td>\n",
       "      <td>86.864932</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.699831</td>\n",
       "      <td>0.795452</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10orf53</th>\n",
       "      <td>C10orf53</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>14285.0</td>\n",
       "      <td>80.455109</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.639587</td>\n",
       "      <td>0.539423</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10orf62</th>\n",
       "      <td>C10orf62</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>16284.0</td>\n",
       "      <td>91.714543</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.816509</td>\n",
       "      <td>0.703211</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10orf71</th>\n",
       "      <td>C10orf71</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>17419.0</td>\n",
       "      <td>98.107469</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.081614</td>\n",
       "      <td>0.709021</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10orf71.AS1</th>\n",
       "      <td>C10orf71.AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>16914.0</td>\n",
       "      <td>95.263039</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.871973</td>\n",
       "      <td>0.739904</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM225</th>\n",
       "      <td>TMEM225</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>15418.0</td>\n",
       "      <td>86.836769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.626117</td>\n",
       "      <td>0.628148</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM235</th>\n",
       "      <td>TMEM235</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999954</td>\n",
       "      <td>16373.0</td>\n",
       "      <td>92.215839</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.911884</td>\n",
       "      <td>0.882291</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM35A</th>\n",
       "      <td>TMEM35A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>11499.0</td>\n",
       "      <td>64.762870</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.781146</td>\n",
       "      <td>0.863998</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNP1</th>\n",
       "      <td>TNP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999936</td>\n",
       "      <td>17741.0</td>\n",
       "      <td>99.921145</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.008474</td>\n",
       "      <td>1.342752</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53TG3HP</th>\n",
       "      <td>TP53TG3HP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999946</td>\n",
       "      <td>14957.0</td>\n",
       "      <td>84.240171</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.645933</td>\n",
       "      <td>0.647190</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTLL2</th>\n",
       "      <td>TTLL2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>12308.0</td>\n",
       "      <td>69.319590</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.658707</td>\n",
       "      <td>0.753076</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY10</th>\n",
       "      <td>TTTY10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>13780.0</td>\n",
       "      <td>77.610679</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.646779</td>\n",
       "      <td>0.632708</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY11</th>\n",
       "      <td>TTTY11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>16828.0</td>\n",
       "      <td>94.778641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.720188</td>\n",
       "      <td>0.493526</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY12</th>\n",
       "      <td>TTTY12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>14859.0</td>\n",
       "      <td>83.688183</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.599186</td>\n",
       "      <td>0.552915</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY13</th>\n",
       "      <td>TTTY13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>16729.0</td>\n",
       "      <td>94.221021</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.798044</td>\n",
       "      <td>0.555075</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY5</th>\n",
       "      <td>TTTY5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>14633.0</td>\n",
       "      <td>82.415230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.586428</td>\n",
       "      <td>0.539987</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY7</th>\n",
       "      <td>TTTY7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999946</td>\n",
       "      <td>16288.0</td>\n",
       "      <td>91.737073</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.624759</td>\n",
       "      <td>0.516706</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TYR</th>\n",
       "      <td>TYR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>16848.0</td>\n",
       "      <td>94.891292</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.991401</td>\n",
       "      <td>0.944317</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TYROBP</th>\n",
       "      <td>TYROBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>14143.0</td>\n",
       "      <td>79.655289</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.715698</td>\n",
       "      <td>0.698000</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UCMA</th>\n",
       "      <td>UCMA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>13942.0</td>\n",
       "      <td>78.523150</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.746615</td>\n",
       "      <td>0.890936</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UCN3</th>\n",
       "      <td>UCN3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>13559.0</td>\n",
       "      <td>76.365889</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.756965</td>\n",
       "      <td>0.799008</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UMOD</th>\n",
       "      <td>UMOD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>17675.0</td>\n",
       "      <td>99.549397</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.362513</td>\n",
       "      <td>0.953912</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UMODL1.AS1</th>\n",
       "      <td>UMODL1.AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>14987.0</td>\n",
       "      <td>84.409147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.683292</td>\n",
       "      <td>0.691188</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UST.AS1</th>\n",
       "      <td>UST.AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>14840.0</td>\n",
       "      <td>83.581165</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.724316</td>\n",
       "      <td>0.709263</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UTF1</th>\n",
       "      <td>UTF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>17283.0</td>\n",
       "      <td>97.341444</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.121456</td>\n",
       "      <td>0.867692</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VN1R4</th>\n",
       "      <td>VN1R4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999949</td>\n",
       "      <td>16387.0</td>\n",
       "      <td>92.294694</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.764756</td>\n",
       "      <td>0.632975</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VPREB1</th>\n",
       "      <td>VPREB1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999954</td>\n",
       "      <td>17185.0</td>\n",
       "      <td>96.789456</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.999496</td>\n",
       "      <td>0.740980</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VSIG4</th>\n",
       "      <td>VSIG4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>15642.0</td>\n",
       "      <td>88.098457</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.820308</td>\n",
       "      <td>0.818934</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WFDC9</th>\n",
       "      <td>WFDC9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>17516.0</td>\n",
       "      <td>98.653824</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.336360</td>\n",
       "      <td>0.823860</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WNT8A</th>\n",
       "      <td>WNT8A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>16447.0</td>\n",
       "      <td>92.632646</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.990865</td>\n",
       "      <td>0.958187</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XPNPEP2</th>\n",
       "      <td>XPNPEP2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>13180.0</td>\n",
       "      <td>74.231159</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.779482</td>\n",
       "      <td>0.834996</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC13</th>\n",
       "      <td>ZCCHC13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999946</td>\n",
       "      <td>15141.0</td>\n",
       "      <td>85.276557</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.794019</td>\n",
       "      <td>0.697602</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIC3</th>\n",
       "      <td>ZIC3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999948</td>\n",
       "      <td>15970.0</td>\n",
       "      <td>89.945928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.960722</td>\n",
       "      <td>0.922915</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF831</th>\n",
       "      <td>ZNF831</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999951</td>\n",
       "      <td>10502.0</td>\n",
       "      <td>59.147234</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.538895</td>\n",
       "      <td>0.577708</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZPBP</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>16511.0</td>\n",
       "      <td>92.993128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.962039</td>\n",
       "      <td>0.923555</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>596 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "ACOD1                ACOD1                 NaN          NaN   \n",
       "ACTL9                ACTL9                 NaN          NaN   \n",
       "ACTRT2              ACTRT2                 NaN          NaN   \n",
       "ADAM30              ADAM30                 NaN          NaN   \n",
       "ADAM6                ADAM6                 NaN          NaN   \n",
       "ADAMDEC1          ADAMDEC1                 NaN          NaN   \n",
       "AKAIN1              AKAIN1                 NaN          NaN   \n",
       "ALPI                  ALPI                 NaN          NaN   \n",
       "AMELY                AMELY                 NaN          NaN   \n",
       "ANHX                  ANHX                 NaN          NaN   \n",
       "ANP32D              ANP32D                 NaN          NaN   \n",
       "APOA4                APOA4                 NaN          NaN   \n",
       "APOBEC1            APOBEC1                 NaN          NaN   \n",
       "APOC3                APOC3                 NaN          NaN   \n",
       "ASB10                ASB10                 NaN          NaN   \n",
       "ASB17                ASB17                 NaN          NaN   \n",
       "ATXN8OS            ATXN8OS                 NaN          NaN   \n",
       "AWAT1                AWAT1                 NaN          NaN   \n",
       "BCORP1              BCORP1                 NaN          NaN   \n",
       "BGLT3                BGLT3                 NaN          NaN   \n",
       "BHLHE23            BHLHE23                 NaN          NaN   \n",
       "BMP15                BMP15                 NaN          NaN   \n",
       "BPIFA3              BPIFA3                 NaN          NaN   \n",
       "BPIFA4P            BPIFA4P                 NaN          NaN   \n",
       "BPIFB2              BPIFB2                 NaN          NaN   \n",
       "BPIFB6              BPIFB6                 NaN          NaN   \n",
       "C10orf53          C10orf53                 NaN          NaN   \n",
       "C10orf62          C10orf62                 NaN          NaN   \n",
       "C10orf71          C10orf71                 NaN          NaN   \n",
       "C10orf71.AS1  C10orf71.AS1                 NaN          NaN   \n",
       "...                    ...                 ...          ...   \n",
       "TMEM225            TMEM225                 NaN          NaN   \n",
       "TMEM235            TMEM235                 NaN          NaN   \n",
       "TMEM35A            TMEM35A                 NaN          NaN   \n",
       "TNP1                  TNP1                 NaN          NaN   \n",
       "TP53TG3HP        TP53TG3HP                 NaN          NaN   \n",
       "TTLL2                TTLL2                 NaN          NaN   \n",
       "TTTY10              TTTY10                 NaN          NaN   \n",
       "TTTY11              TTTY11                 NaN          NaN   \n",
       "TTTY12              TTTY12                 NaN          NaN   \n",
       "TTTY13              TTTY13                 NaN          NaN   \n",
       "TTTY5                TTTY5                 NaN          NaN   \n",
       "TTTY7                TTTY7                 NaN          NaN   \n",
       "TYR                    TYR                 NaN          NaN   \n",
       "TYROBP              TYROBP                 NaN          NaN   \n",
       "UCMA                  UCMA                 NaN          NaN   \n",
       "UCN3                  UCN3                 NaN          NaN   \n",
       "UMOD                  UMOD                 NaN          NaN   \n",
       "UMODL1.AS1      UMODL1.AS1                 NaN          NaN   \n",
       "UST.AS1            UST.AS1                 NaN          NaN   \n",
       "UTF1                  UTF1                 NaN          NaN   \n",
       "VN1R4                VN1R4                 NaN          NaN   \n",
       "VPREB1              VPREB1                 NaN          NaN   \n",
       "VSIG4                VSIG4                 NaN          NaN   \n",
       "WFDC9                WFDC9                 NaN          NaN   \n",
       "WNT8A                WNT8A                 NaN          NaN   \n",
       "XPNPEP2            XPNPEP2                 NaN          NaN   \n",
       "ZCCHC13            ZCCHC13                 NaN          NaN   \n",
       "ZIC3                  ZIC3                 NaN          NaN   \n",
       "ZNF831              ZNF831                 NaN          NaN   \n",
       "ZPBP                  ZPBP                 NaN          NaN   \n",
       "\n",
       "              abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "ACOD1                                NaN                    NaN   \n",
       "ACTL9                                NaN                    NaN   \n",
       "ACTRT2                               NaN                    NaN   \n",
       "ADAM30                               NaN                    NaN   \n",
       "ADAM6                                NaN                    NaN   \n",
       "ADAMDEC1                             NaN                    NaN   \n",
       "AKAIN1                               NaN                    NaN   \n",
       "ALPI                                 NaN                    NaN   \n",
       "AMELY                                NaN                    NaN   \n",
       "ANHX                                 NaN                    NaN   \n",
       "ANP32D                               NaN                    NaN   \n",
       "APOA4                                NaN                    NaN   \n",
       "APOBEC1                              NaN                    NaN   \n",
       "APOC3                                NaN                    NaN   \n",
       "ASB10                                NaN                    NaN   \n",
       "ASB17                                NaN                    NaN   \n",
       "ATXN8OS                              NaN                    NaN   \n",
       "AWAT1                                NaN                    NaN   \n",
       "BCORP1                               NaN                    NaN   \n",
       "BGLT3                                NaN                    NaN   \n",
       "BHLHE23                              NaN                    NaN   \n",
       "BMP15                                NaN                    NaN   \n",
       "BPIFA3                               NaN                    NaN   \n",
       "BPIFA4P                              NaN                    NaN   \n",
       "BPIFB2                               NaN                    NaN   \n",
       "BPIFB6                               NaN                    NaN   \n",
       "C10orf53                             NaN                    NaN   \n",
       "C10orf62                             NaN                    NaN   \n",
       "C10orf71                             NaN                    NaN   \n",
       "C10orf71.AS1                         NaN                    NaN   \n",
       "...                                  ...                    ...   \n",
       "TMEM225                              NaN                    NaN   \n",
       "TMEM235                              NaN                    NaN   \n",
       "TMEM35A                              NaN                    NaN   \n",
       "TNP1                                 NaN                    NaN   \n",
       "TP53TG3HP                            NaN                    NaN   \n",
       "TTLL2                                NaN                    NaN   \n",
       "TTTY10                               NaN                    NaN   \n",
       "TTTY11                               NaN                    NaN   \n",
       "TTTY12                               NaN                    NaN   \n",
       "TTTY13                               NaN                    NaN   \n",
       "TTTY5                                NaN                    NaN   \n",
       "TTTY7                                NaN                    NaN   \n",
       "TYR                                  NaN                    NaN   \n",
       "TYROBP                               NaN                    NaN   \n",
       "UCMA                                 NaN                    NaN   \n",
       "UCN3                                 NaN                    NaN   \n",
       "UMOD                                 NaN                    NaN   \n",
       "UMODL1.AS1                           NaN                    NaN   \n",
       "UST.AS1                              NaN                    NaN   \n",
       "UTF1                                 NaN                    NaN   \n",
       "VN1R4                                NaN                    NaN   \n",
       "VPREB1                               NaN                    NaN   \n",
       "VSIG4                                NaN                    NaN   \n",
       "WFDC9                                NaN                    NaN   \n",
       "WNT8A                                NaN                    NaN   \n",
       "XPNPEP2                              NaN                    NaN   \n",
       "ZCCHC13                              NaN                    NaN   \n",
       "ZIC3                                 NaN                    NaN   \n",
       "ZNF831                               NaN                    NaN   \n",
       "ZPBP                                 NaN                    NaN   \n",
       "\n",
       "              Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "ACOD1                               0.999950           16969.0   \n",
       "ACTL9                               0.999948           17211.0   \n",
       "ACTRT2                              0.999950           17069.0   \n",
       "ADAM30                              0.999949           13941.0   \n",
       "ADAM6                               0.999949           14720.0   \n",
       "ADAMDEC1                            0.999951           16772.0   \n",
       "AKAIN1                              0.999950           15067.0   \n",
       "ALPI                                0.999946           15669.0   \n",
       "AMELY                               0.999954           12640.0   \n",
       "ANHX                                0.999950           16825.0   \n",
       "ANP32D                              0.999951           10304.0   \n",
       "APOA4                               0.999910           17506.0   \n",
       "APOBEC1                             0.999950           16338.0   \n",
       "APOC3                               0.999948           17523.0   \n",
       "ASB10                               0.999951           17038.0   \n",
       "ASB17                               0.999949           15144.0   \n",
       "ATXN8OS                             0.999951           15688.0   \n",
       "AWAT1                               0.999950           11476.0   \n",
       "BCORP1                              0.999951            8875.0   \n",
       "BGLT3                               0.999951           16235.0   \n",
       "BHLHE23                             0.999954           17151.0   \n",
       "BMP15                               0.999951           17142.0   \n",
       "BPIFA3                              0.999949           16831.0   \n",
       "BPIFA4P                             0.999950           17379.0   \n",
       "BPIFB2                              0.999946           17706.0   \n",
       "BPIFB6                              0.999951           15423.0   \n",
       "C10orf53                            0.999949           14285.0   \n",
       "C10orf62                            0.999949           16284.0   \n",
       "C10orf71                            0.999949           17419.0   \n",
       "C10orf71.AS1                        0.999951           16914.0   \n",
       "...                                      ...               ...   \n",
       "TMEM225                             0.999949           15418.0   \n",
       "TMEM235                             0.999954           16373.0   \n",
       "TMEM35A                             0.999951           11499.0   \n",
       "TNP1                                0.999936           17741.0   \n",
       "TP53TG3HP                           0.999946           14957.0   \n",
       "TTLL2                               0.999951           12308.0   \n",
       "TTTY10                              0.999951           13780.0   \n",
       "TTTY11                              0.999948           16828.0   \n",
       "TTTY12                              0.999948           14859.0   \n",
       "TTTY13                              0.999949           16729.0   \n",
       "TTTY5                               0.999948           14633.0   \n",
       "TTTY7                               0.999946           16288.0   \n",
       "TYR                                 0.999949           16848.0   \n",
       "TYROBP                              0.999951           14143.0   \n",
       "UCMA                                0.999951           13942.0   \n",
       "UCN3                                0.999950           13559.0   \n",
       "UMOD                                0.999948           17675.0   \n",
       "UMODL1.AS1                          0.999951           14987.0   \n",
       "UST.AS1                             0.999950           14840.0   \n",
       "UTF1                                0.999951           17283.0   \n",
       "VN1R4                               0.999949           16387.0   \n",
       "VPREB1                              0.999954           17185.0   \n",
       "VSIG4                               0.999951           15642.0   \n",
       "WFDC9                               0.999950           17516.0   \n",
       "WNT8A                               0.999950           16447.0   \n",
       "XPNPEP2                             0.999951           13180.0   \n",
       "ZCCHC13                             0.999946           15141.0   \n",
       "ZIC3                                0.999948           15970.0   \n",
       "ZNF831                              0.999951           10502.0   \n",
       "ZPBP                                0.999950           16511.0   \n",
       "\n",
       "              Percentile (simulated)  Percent DE (simulated)  \\\n",
       "ACOD1                      95.572829                0.000000   \n",
       "ACTL9                      96.935902                0.000000   \n",
       "ACTRT2                     96.136082                0.000000   \n",
       "ADAM30                     78.517517                0.000000   \n",
       "ADAM6                      82.905261                0.000000   \n",
       "ADAMDEC1                   94.463220                0.000000   \n",
       "AKAIN1                     84.859750                0.000000   \n",
       "ALPI                       88.250535                0.000000   \n",
       "AMELY                      71.189591                0.000000   \n",
       "ANHX                       94.761744                0.000000   \n",
       "ANP32D                     58.031993                0.000000   \n",
       "APOA4                      98.597499                0.045455   \n",
       "APOBEC1                    92.018700                0.000000   \n",
       "APOC3                      98.693252                0.000000   \n",
       "ASB10                      95.961473                0.000000   \n",
       "ASB17                      85.293455                0.000000   \n",
       "ATXN8OS                    88.357553                0.040000   \n",
       "AWAT1                      64.633322                0.000000   \n",
       "BCORP1                     49.983102                0.000000   \n",
       "BGLT3                      91.438549                0.000000   \n",
       "BHLHE23                    96.597950                0.000000   \n",
       "BMP15                      96.547257                0.000000   \n",
       "BPIFA3                     94.795539                0.000000   \n",
       "BPIFA4P                    97.882167                0.000000   \n",
       "BPIFB2                     99.724006                0.000000   \n",
       "BPIFB6                     86.864932                0.000000   \n",
       "C10orf53                   80.455109                0.000000   \n",
       "C10orf62                   91.714543                0.000000   \n",
       "C10orf71                   98.107469                0.000000   \n",
       "C10orf71.AS1               95.263039                0.000000   \n",
       "...                              ...                     ...   \n",
       "TMEM225                    86.836769                0.000000   \n",
       "TMEM235                    92.215839                0.000000   \n",
       "TMEM35A                    64.762870                0.000000   \n",
       "TNP1                       99.921145                0.000000   \n",
       "TP53TG3HP                  84.240171                0.000000   \n",
       "TTLL2                      69.319590                0.000000   \n",
       "TTTY10                     77.610679                0.000000   \n",
       "TTTY11                     94.778641                0.000000   \n",
       "TTTY12                     83.688183                0.000000   \n",
       "TTTY13                     94.221021                0.000000   \n",
       "TTTY5                      82.415230                0.000000   \n",
       "TTTY7                      91.737073                0.000000   \n",
       "TYR                        94.891292                0.000000   \n",
       "TYROBP                     79.655289                0.000000   \n",
       "UCMA                       78.523150                0.000000   \n",
       "UCN3                       76.365889                0.000000   \n",
       "UMOD                       99.549397                0.000000   \n",
       "UMODL1.AS1                 84.409147                0.000000   \n",
       "UST.AS1                    83.581165                0.000000   \n",
       "UTF1                       97.341444                0.000000   \n",
       "VN1R4                      92.294694                0.000000   \n",
       "VPREB1                     96.789456                0.000000   \n",
       "VSIG4                      88.098457                0.000000   \n",
       "WFDC9                      98.653824                0.000000   \n",
       "WNT8A                      92.632646                0.000000   \n",
       "XPNPEP2                    74.231159                0.000000   \n",
       "ZCCHC13                    85.276557                0.000000   \n",
       "ZIC3                       89.945928                0.000000   \n",
       "ZNF831                     59.147234                0.000000   \n",
       "ZPBP                       92.993128                0.000000   \n",
       "\n",
       "              Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "ACOD1                                     0.963899                   0.797525   \n",
       "ACTL9                                     0.978766                   0.782871   \n",
       "ACTRT2                                    0.937594                   0.781906   \n",
       "ADAM30                                    0.636214                   0.704730   \n",
       "ADAM6                                     0.839843                   0.854896   \n",
       "ADAMDEC1                                  0.943490                   0.855386   \n",
       "AKAIN1                                    0.773474                   0.644060   \n",
       "ALPI                                      1.459798                   1.753978   \n",
       "AMELY                                     0.653440                   0.627023   \n",
       "ANHX                                      1.043032                   0.953204   \n",
       "ANP32D                                    0.412181                   0.400176   \n",
       "APOA4                                     1.877437                   1.960742   \n",
       "APOBEC1                                   0.962360                   0.953836   \n",
       "APOC3                                     1.311955                   1.104857   \n",
       "ASB10                                     0.890658                   0.685825   \n",
       "ASB17                                     0.736293                   0.695222   \n",
       "ATXN8OS                                   0.899167                   0.818626   \n",
       "AWAT1                                     0.665827                   0.784592   \n",
       "BCORP1                                    0.578825                   0.784105   \n",
       "BGLT3                                     0.929932                   0.996257   \n",
       "BHLHE23                                   1.148463                   0.851245   \n",
       "BMP15                                     1.097471                   0.769280   \n",
       "BPIFA3                                    0.929227                   0.690446   \n",
       "BPIFA4P                                   1.001450                   0.671031   \n",
       "BPIFB2                                    1.486836                   1.195091   \n",
       "BPIFB6                                    0.699831                   0.795452   \n",
       "C10orf53                                  0.639587                   0.539423   \n",
       "C10orf62                                  0.816509                   0.703211   \n",
       "C10orf71                                  1.081614                   0.709021   \n",
       "C10orf71.AS1                              0.871973                   0.739904   \n",
       "...                                            ...                        ...   \n",
       "TMEM225                                   0.626117                   0.628148   \n",
       "TMEM235                                   0.911884                   0.882291   \n",
       "TMEM35A                                   0.781146                   0.863998   \n",
       "TNP1                                      2.008474                   1.342752   \n",
       "TP53TG3HP                                 0.645933                   0.647190   \n",
       "TTLL2                                     0.658707                   0.753076   \n",
       "TTTY10                                    0.646779                   0.632708   \n",
       "TTTY11                                    0.720188                   0.493526   \n",
       "TTTY12                                    0.599186                   0.552915   \n",
       "TTTY13                                    0.798044                   0.555075   \n",
       "TTTY5                                     0.586428                   0.539987   \n",
       "TTTY7                                     0.624759                   0.516706   \n",
       "TYR                                       0.991401                   0.944317   \n",
       "TYROBP                                    0.715698                   0.698000   \n",
       "UCMA                                      0.746615                   0.890936   \n",
       "UCN3                                      0.756965                   0.799008   \n",
       "UMOD                                      1.362513                   0.953912   \n",
       "UMODL1.AS1                                0.683292                   0.691188   \n",
       "UST.AS1                                   0.724316                   0.709263   \n",
       "UTF1                                      1.121456                   0.867692   \n",
       "VN1R4                                     0.764756                   0.632975   \n",
       "VPREB1                                    0.999496                   0.740980   \n",
       "VSIG4                                     0.820308                   0.818934   \n",
       "WFDC9                                     1.336360                   0.823860   \n",
       "WNT8A                                     0.990865                   0.958187   \n",
       "XPNPEP2                                   0.779482                   0.834996   \n",
       "ZCCHC13                                   0.794019                   0.697602   \n",
       "ZIC3                                      0.960722                   0.922915   \n",
       "ZNF831                                    0.538895                   0.577708   \n",
       "ZPBP                                      0.962039                   0.923555   \n",
       "\n",
       "              Number of experiments (simulated)  Z score  \n",
       "ACOD1                                        24      NaN  \n",
       "ACTL9                                        22      NaN  \n",
       "ACTRT2                                       20      NaN  \n",
       "ADAM30                                       19      NaN  \n",
       "ADAM6                                        21      NaN  \n",
       "ADAMDEC1                                     25      NaN  \n",
       "AKAIN1                                       22      NaN  \n",
       "ALPI                                         23      NaN  \n",
       "AMELY                                        22      NaN  \n",
       "ANHX                                         22      NaN  \n",
       "ANP32D                                       25      NaN  \n",
       "APOA4                                        22      NaN  \n",
       "APOBEC1                                      24      NaN  \n",
       "APOC3                                        24      NaN  \n",
       "ASB10                                        23      NaN  \n",
       "ASB17                                        19      NaN  \n",
       "ATXN8OS                                      25      NaN  \n",
       "AWAT1                                        24      NaN  \n",
       "BCORP1                                       25      NaN  \n",
       "BGLT3                                        21      NaN  \n",
       "BHLHE23                                      24      NaN  \n",
       "BMP15                                        21      NaN  \n",
       "BPIFA3                                       19      NaN  \n",
       "BPIFA4P                                      22      NaN  \n",
       "BPIFB2                                       23      NaN  \n",
       "BPIFB6                                       21      NaN  \n",
       "C10orf53                                     23      NaN  \n",
       "C10orf62                                     23      NaN  \n",
       "C10orf71                                     25      NaN  \n",
       "C10orf71.AS1                                 23      NaN  \n",
       "...                                         ...      ...  \n",
       "TMEM225                                      19      NaN  \n",
       "TMEM235                                      24      NaN  \n",
       "TMEM35A                                      25      NaN  \n",
       "TNP1                                         18      NaN  \n",
       "TP53TG3HP                                    17      NaN  \n",
       "TTLL2                                        25      NaN  \n",
       "TTTY10                                       25      NaN  \n",
       "TTTY11                                       14      NaN  \n",
       "TTTY12                                       14      NaN  \n",
       "TTTY13                                       15      NaN  \n",
       "TTTY5                                        12      NaN  \n",
       "TTTY7                                        15      NaN  \n",
       "TYR                                          23      NaN  \n",
       "TYROBP                                       25      NaN  \n",
       "UCMA                                         23      NaN  \n",
       "UCN3                                         24      NaN  \n",
       "UMOD                                         20      NaN  \n",
       "UMODL1.AS1                                   25      NaN  \n",
       "UST.AS1                                      22      NaN  \n",
       "UTF1                                         23      NaN  \n",
       "VN1R4                                        19      NaN  \n",
       "VPREB1                                       24      NaN  \n",
       "VSIG4                                        25      NaN  \n",
       "WFDC9                                        16      NaN  \n",
       "WNT8A                                        24      NaN  \n",
       "XPNPEP2                                      25      NaN  \n",
       "ZCCHC13                                      21      NaN  \n",
       "ZIC3                                         24      NaN  \n",
       "ZNF831                                       25      NaN  \n",
       "ZPBP                                         24      NaN  \n",
       "\n",
       "[596 rows x 13 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks[summary_gene_ranks.isna().any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "summary_gene_ranks.to_csv(gene_summary_filename, sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression_linux] *",
   "language": "python",
   "name": "conda-env-generic_expression_linux-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
