{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Application: new experiment\n",
    "\n",
    "This notebook allows users to find specific genes in their experiment of interest using an existing VAE model\n",
    "\n",
    "This notebook will generate a `generic_gene_summary_<experiment id>.tsv` file that contains a z-score per gene that indicates how specific a gene is the experiment in question."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from ponyo import utils\n",
    "from generic_expression_patterns_modules import (\n",
    "    process,\n",
    "    new_experiment_process,\n",
    "    stats,\n",
    "    ranking,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User input\n",
    "\n",
    "User needs to define the following in the [config file](../configs/config_new_experiment.tsv):\n",
    "\n",
    "1. Template experiment (`raw_template_filename`). This is the experiment you are interested in studying.\n",
    "2. Training compendium used to train VAE, including unnormalized gene mapped version (`mapped_compendium_filename`) and normalized version (`normalized_compendium_filename`). Links to these datasets can be found in the README.\n",
    "3. Scaler transform (`scaler_filename`) used to normalize the training compendium. This can be found in the `data/` directory within the analysis folder.\n",
    "4. Directory (`vae_model_dir`) containing trained VAE model.\n",
    "5. Experiment id (`project_id`) to label newly create simulated experiments.\n",
    "6. The number of experiments to simulate (`num_simulated`)\n",
    "7. Minimum average read count to filter data by (`count_threshold`)\n",
    "8. Size of existing models latent dimension (`latent_dim`). Should be 30.\n",
    "9. Name of column header (`rank_genes_by`) from DE association statistic results. This column will be use to rank genes. Select \"logFC\", \"P.Value\", \"adj.P.Val\", \"t\" if using Limma. Select \"log2FoldChange\", \"pvalue\", \"padj\" if using DESeq.\n",
    "10. `DE_logFC_name` is either \"logFC\" or \"log2FoldChange\". This is used for plotting volcano plots.\n",
    "11. `DE_pvalue_name` is either \"adj.P.Val\" or \"padj\". This is used for plotting volcano plots.\n",
    "\n",
    "The remaining parameters specify filenames that intermediate data files will be written to.\n",
    "\n",
    "The user also needs to provide metadata files:\n",
    "1. `<experiment id>_process_samples.tsv` contains 2 columns (sample ids, label that indicates if the sample is kept or removed). See [example](data/metadata/cis-gem-par-KU1919_process_samples.tsv). **Note: This file is not required if the user wishes to use all the samples in the template experiment file.**\n",
    "2. `<experiment id>_groups.tsv` contains 2 columns: sample ids, group label to perform DE analysis. See [example](data/metadata/cis-gem-par-KU1919_groups.tsv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = os.path.abspath(\n",
    "    os.path.join(base_dir, \"configs\", \"config_new_experiment.tsv\")\n",
    ")\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load config params\n",
    "\n",
    "# Local directory to store intermediate files\n",
    "local_dir = params[\"local_dir\"]\n",
    "\n",
    "# Number of simulated experiments to generate\n",
    "num_runs = params[\"num_simulated\"]\n",
    "\n",
    "# Directory containing trained VAE model\n",
    "vae_model_dir = params[\"vae_model_dir\"]\n",
    "\n",
    "# Dimension of latent space used in VAE model\n",
    "latent_dim = params[\"latent_dim\"]\n",
    "\n",
    "# ID for template experiment\n",
    "# This ID will be used to label new simulated experiments\n",
    "project_id = params[\"project_id\"]\n",
    "\n",
    "# Template experiment filename\n",
    "template_filename = params[\"raw_template_filename\"]\n",
    "mapped_template_filename = params[\"mapped_template_filename\"]\n",
    "normalized_template_filename = params[\"normalized_template_filename\"]\n",
    "processed_template_filename = params[\"processed_template_filename\"]\n",
    "\n",
    "# Training dataset used for existing VAE model\n",
    "mapped_compendium_filename = params[\"mapped_compendium_filename\"]\n",
    "\n",
    "# Normalized compendium filename\n",
    "normalized_compendium_filename = params[\"normalized_compendium_filename\"]\n",
    "\n",
    "# Scaler transform used to scale compendium data into 0-1 range for training\n",
    "scaler_filename = params[\"scaler_filename\"]\n",
    "\n",
    "# Test statistic used to rank genes by\n",
    "col_to_rank_genes = params[\"rank_genes_by\"]\n",
    "\n",
    "# Minimum mean count per gene\n",
    "count_threshold = params[\"count_threshold\"]\n",
    "\n",
    "# Column headers to use to make summary statistic table\n",
    "logFC_name = params[\"DE_logFC_name\"]\n",
    "pvalue_name = params[\"DE_pvalue_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata files\n",
    "\n",
    "# Load metadata file with processing information\n",
    "sample_id_metadata_filename = os.path.join(\n",
    "    \"data\", \"metadata\", f\"{project_id}_process_samples.tsv\"\n",
    ")\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "metadata_filename = os.path.join(\"data\", \"metadata\", f\"{project_id}_groups.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output filename\n",
    "gene_summary_filename = f\"generic_gene_summary_{project_id}.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map template experiment to same feature space as training compendium\n",
    "\n",
    "In order to simulate a new gene expression experiment, we will need to encode this experiment into the learned latent space. This requires that the feature space (i.e. genes) in the template experiment match the features in the compendium used to train the VAE model. These cells process the template experiment to be of the expected input format:\n",
    "* Template data is expected to be a matrix that is sample x gene\n",
    "* Template experiment is expected to have the same genes as the compendium experiment. Genes that are in the template experiment but not in the compendium are removed. Genes that are in the compendium but missing in the template experiment are added and the gene expression value is set to the median gene expression value of that gene across the samples in the compendium."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Template experiment needs to be of the form sample x gene\n",
    "template_filename_only = template_filename.split(\"/\")[-1].split(\".\")[0]\n",
    "transposed_template_filename = os.path.join(\n",
    "    local_dir, template_filename_only + \"_transposed.txt\"\n",
    ")\n",
    "\n",
    "new_experiment_process.transpose_save(template_filename, transposed_template_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(72, 58528)\n",
      "(49651, 17755)\n"
     ]
    }
   ],
   "source": [
    "new_experiment_process.process_template_experiment(\n",
    "    transposed_template_filename,\n",
    "    mapped_compendium_filename,\n",
    "    scaler_filename,\n",
    "    mapped_template_filename,\n",
    "    normalized_template_filename,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate experiments based on template experiment\n",
    "\n",
    "Embed template experiment into learned latent space and linearly shift template experiment to different locations of the latent space to create new experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate experiments based on template experiment\n",
    "normalized_compendium_data = pd.read_csv(\n",
    "    normalized_compendium_filename, sep=\"\\t\", index_col=0, header=0\n",
    ")\n",
    "normalized_template_data = pd.read_csv(\n",
    "    normalized_template_filename, sep=\"\\t\", index_col=0, header=0\n",
    ")\n",
    "\n",
    "for run_id in range(num_runs):\n",
    "    new_experiment_process.embed_shift_template_experiment(\n",
    "        normalized_compendium_data,\n",
    "        normalized_template_data,\n",
    "        vae_model_dir,\n",
    "        project_id,\n",
    "        scaler_filename,\n",
    "        local_dir,\n",
    "        latent_dim,\n",
    "        run_id,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process template and simulated experiments\n",
    "\n",
    "* Remove samples not required for comparison\n",
    "* Make sure ordering of samples matches metadata for proper comparison\n",
    "* Make sure values are cast as integers if using DESeq\n",
    "* Filter lowly expressed genes if using DESeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"human_general_analysis\" in vae_model_dir:\n",
    "    method = \"deseq\"\n",
    "else:\n",
    "    method = \"limma\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(sample_id_metadata_filename):\n",
    "    sample_id_metadata_filename = None\n",
    "\n",
    "if method == \"deseq\":\n",
    "    stats.process_samples_for_DESeq(\n",
    "        mapped_template_filename,\n",
    "        metadata_filename,\n",
    "        processed_template_filename,\n",
    "        count_threshold,\n",
    "        sample_id_metadata_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        out_simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}_processed.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_DESeq(\n",
    "            simulated_filename,\n",
    "            metadata_filename,\n",
    "            out_simulated_filename,\n",
    "            count_threshold,\n",
    "            sample_id_metadata_filename,\n",
    "        )\n",
    "else:\n",
    "    stats.process_samples_for_limma(\n",
    "        mapped_template_filename,\n",
    "        metadata_filename,\n",
    "        processed_template_filename,\n",
    "        sample_id_metadata_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_limma(\n",
    "            simulated_filename,\n",
    "            metadata_filename,\n",
    "            None,\n",
    "            sample_id_metadata_filename,\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential expression analysis\n",
    "\n",
    "* If data is RNA-seq then use DESeq2 (using human_general_analysis model)\n",
    "* If data is microarray then use Limma (using human_cancer_analysis, pseudomonas_analysis models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: S4Vectors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: stats4\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocGenerics\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: parallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: IRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomicRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomeInfoDb\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: SummarizedExperiment\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Biobase\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: DelayedArray\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: matrixStats\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocParallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: estimating size factors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: estimating dispersions\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: gene-wise dispersion estimates\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: mean-dispersion relationship\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: final dispersion estimates\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: fitting model and testing\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 1223 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        metadata_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        metadata_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 74 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 220 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 1837 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 534 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 967 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 4666 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 855 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 3258 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 85 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 791 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 331 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 2124 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 1696 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 1273 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 621 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 2968 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 819 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 886 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 322 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 2078 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 109 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 1638 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 1860 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 3521 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- replacing outliers and refitting for 968 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i base_dir -i local_dir -i num_runs -i method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# Files created: \"<local_dir>/DE_stats/DE_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\",\n",
    "        i,\n",
    "        \"_processed.txt\",\n",
    "        sep = \"\"\n",
    "    )\n",
    "    if (method == \"deseq\"){\n",
    "        get_DE_stats_DESeq(\n",
    "            metadata_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "    }\n",
    "    else {\n",
    "        get_DE_stats_limma(\n",
    "            metadata_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rank genes\n",
    "\n",
    "Genes are ranked by their \"generic-ness\" - how frequently these genes are changed across the simulated experiments using user-specific test statistic (i.e. log2 fold change)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    "    logFC_name,\n",
    "    pvalue_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary table\n",
    "\n",
    "* Gene ID: Gene identifier (hgnc symbols for human data or PA number for *P. aeruginosa* data)\n",
    "* (Real): Statistics for template experiment\n",
    "* (Simulated): Statistics across simulated experiments\n",
    "* Number of experiments: Number of simulated experiments\n",
    "* Z-score: High z-score indicates that gene is more changed in template compared to the null set of simulated experiments (high z-score = highly specific to template experiment)\n",
    "\n",
    "\n",
    "**Note:**\n",
    "* If using DESeq, genes with NaN in only the `Adj P-value (Real)` column are those genes flagged because of the `cooksCutoff` parameter. The cook's distance as a diagnostic to tell if a single sample has a count which has a disproportionate impact on the log fold change and p-values. These genes are flagged with an NA in the pvalue and padj columns of the result table.\n",
    "\n",
    "* If using DESeq with count threshold, some genes may not be present in all simulated experiments (i.e. the `Number of experiments (simulated)` will not equal the number of simulated experiments you specified in the beginning. This pre-filtering will lead to some genes found in few simulated experiments and so the background/null set for that gene is not robust. Thus, the user should sort by both z-score and number of experiments to identify specific expressed genes.\n",
    "\n",
    "* If using DESeq without count threshold, some genes may still not be present in all simulated experiments (i.e. the `Number of experiments (simulated)`  will not equal the number of simulated experiments you specified in the beginning. If the gene is 0 expressed across all samples and thus automatically given an NA in `log fold change, adjusted p-value` columns. Thus, the user should sort by both z-score and number of experiments to identify specific expressed genes.\n",
    "\n",
    "For more information you can read [DESeq FAQs](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (1 levels on the left, 2 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RXFP2</th>\n",
       "      <td>RXFP2</td>\n",
       "      <td>2.039775e-09</td>\n",
       "      <td>17159.0</td>\n",
       "      <td>21.707360</td>\n",
       "      <td>-21.707360</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>11427.0</td>\n",
       "      <td>0.404230</td>\n",
       "      <td>0.406799</td>\n",
       "      <td>24</td>\n",
       "      <td>52.367643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GPX7</th>\n",
       "      <td>GPX7</td>\n",
       "      <td>6.607610e-03</td>\n",
       "      <td>17153.0</td>\n",
       "      <td>6.082224</td>\n",
       "      <td>6.082224</td>\n",
       "      <td>0.999880</td>\n",
       "      <td>12541.0</td>\n",
       "      <td>0.318692</td>\n",
       "      <td>0.218875</td>\n",
       "      <td>25</td>\n",
       "      <td>26.332482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRRC4</th>\n",
       "      <td>LRRC4</td>\n",
       "      <td>2.994721e-05</td>\n",
       "      <td>17151.0</td>\n",
       "      <td>5.863040</td>\n",
       "      <td>-5.863040</td>\n",
       "      <td>0.999841</td>\n",
       "      <td>12175.0</td>\n",
       "      <td>0.342337</td>\n",
       "      <td>0.229571</td>\n",
       "      <td>25</td>\n",
       "      <td>24.047880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DCN</th>\n",
       "      <td>DCN</td>\n",
       "      <td>1.660303e-06</td>\n",
       "      <td>17158.0</td>\n",
       "      <td>8.829010</td>\n",
       "      <td>-8.829010</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>6224.0</td>\n",
       "      <td>0.316071</td>\n",
       "      <td>0.408809</td>\n",
       "      <td>25</td>\n",
       "      <td>20.823779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH11</th>\n",
       "      <td>CDH11</td>\n",
       "      <td>3.358492e-02</td>\n",
       "      <td>17154.0</td>\n",
       "      <td>6.143647</td>\n",
       "      <td>-6.143647</td>\n",
       "      <td>0.999880</td>\n",
       "      <td>8694.0</td>\n",
       "      <td>0.342905</td>\n",
       "      <td>0.358618</td>\n",
       "      <td>25</td>\n",
       "      <td>16.175274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EYA4</th>\n",
       "      <td>EYA4</td>\n",
       "      <td>3.603901e-03</td>\n",
       "      <td>17140.0</td>\n",
       "      <td>5.136665</td>\n",
       "      <td>-5.136665</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>11319.0</td>\n",
       "      <td>0.338946</td>\n",
       "      <td>0.310174</td>\n",
       "      <td>25</td>\n",
       "      <td>15.467823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RRM1</th>\n",
       "      <td>RRM1</td>\n",
       "      <td>8.404496e-08</td>\n",
       "      <td>16913.0</td>\n",
       "      <td>2.990056</td>\n",
       "      <td>2.990056</td>\n",
       "      <td>0.999880</td>\n",
       "      <td>10521.0</td>\n",
       "      <td>0.263583</td>\n",
       "      <td>0.177253</td>\n",
       "      <td>25</td>\n",
       "      <td>15.381820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF366</th>\n",
       "      <td>ZNF366</td>\n",
       "      <td>5.178133e-02</td>\n",
       "      <td>16996.0</td>\n",
       "      <td>3.428502</td>\n",
       "      <td>3.428502</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>10878.0</td>\n",
       "      <td>0.278472</td>\n",
       "      <td>0.208072</td>\n",
       "      <td>25</td>\n",
       "      <td>15.139151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTRA3</th>\n",
       "      <td>HTRA3</td>\n",
       "      <td>2.245523e-02</td>\n",
       "      <td>17016.0</td>\n",
       "      <td>3.489279</td>\n",
       "      <td>-3.489279</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>2486.0</td>\n",
       "      <td>0.187404</td>\n",
       "      <td>0.218743</td>\n",
       "      <td>25</td>\n",
       "      <td>15.094767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XPC</th>\n",
       "      <td>XPC</td>\n",
       "      <td>3.529001e-02</td>\n",
       "      <td>13369.0</td>\n",
       "      <td>0.577698</td>\n",
       "      <td>0.577698</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>135.0</td>\n",
       "      <td>0.069185</td>\n",
       "      <td>0.035642</td>\n",
       "      <td>25</td>\n",
       "      <td>14.267395</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "RXFP2    RXFP2        2.039775e-09      17159.0                   21.707360   \n",
       "GPX7      GPX7        6.607610e-03      17153.0                    6.082224   \n",
       "LRRC4    LRRC4        2.994721e-05      17151.0                    5.863040   \n",
       "DCN        DCN        1.660303e-06      17158.0                    8.829010   \n",
       "CDH11    CDH11        3.358492e-02      17154.0                    6.143647   \n",
       "EYA4      EYA4        3.603901e-03      17140.0                    5.136665   \n",
       "RRM1      RRM1        8.404496e-08      16913.0                    2.990056   \n",
       "ZNF366  ZNF366        5.178133e-02      16996.0                    3.428502   \n",
       "HTRA3    HTRA3        2.245523e-02      17016.0                    3.489279   \n",
       "XPC        XPC        3.529001e-02      13369.0                    0.577698   \n",
       "\n",
       "        log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "RXFP2              -21.707360                        0.999885   \n",
       "GPX7                 6.082224                        0.999880   \n",
       "LRRC4               -5.863040                        0.999841   \n",
       "DCN                 -8.829010                        0.999889   \n",
       "CDH11               -6.143647                        0.999880   \n",
       "EYA4                -5.136665                        0.999889   \n",
       "RRM1                 2.990056                        0.999880   \n",
       "ZNF366               3.428502                        0.999889   \n",
       "HTRA3               -3.489279                        0.999889   \n",
       "XPC                  0.577698                        0.999889   \n",
       "\n",
       "        Rank (simulated)  Mean abs(log2FoldChange) (simulated)  \\\n",
       "RXFP2            11427.0                              0.404230   \n",
       "GPX7             12541.0                              0.318692   \n",
       "LRRC4            12175.0                              0.342337   \n",
       "DCN               6224.0                              0.316071   \n",
       "CDH11             8694.0                              0.342905   \n",
       "EYA4             11319.0                              0.338946   \n",
       "RRM1             10521.0                              0.263583   \n",
       "ZNF366           10878.0                              0.278472   \n",
       "HTRA3             2486.0                              0.187404   \n",
       "XPC                135.0                              0.069185   \n",
       "\n",
       "        Std deviation (simulated)  Number of experiments (simulated)  \\\n",
       "RXFP2                    0.406799                                 24   \n",
       "GPX7                     0.218875                                 25   \n",
       "LRRC4                    0.229571                                 25   \n",
       "DCN                      0.408809                                 25   \n",
       "CDH11                    0.358618                                 25   \n",
       "EYA4                     0.310174                                 25   \n",
       "RRM1                     0.177253                                 25   \n",
       "ZNF366                   0.208072                                 25   \n",
       "HTRA3                    0.218743                                 25   \n",
       "XPC                      0.035642                                 25   \n",
       "\n",
       "          Z score  \n",
       "RXFP2   52.367643  \n",
       "GPX7    26.332482  \n",
       "LRRC4   24.047880  \n",
       "DCN     20.823779  \n",
       "CDH11   16.175274  \n",
       "EYA4    15.467823  \n",
       "RRM1    15.381820  \n",
       "ZNF366  15.139151  \n",
       "HTRA3   15.094767  \n",
       "XPC     14.267395  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get summary table\n",
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    \"gene\",\n",
    "    params,\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene ID                                 False\n",
       "Adj P-value (Real)                       True\n",
       "Rank (Real)                              True\n",
       "abs(log2FoldChange) (Real)               True\n",
       "log2FoldChange (Real)                    True\n",
       "Median adj p-value (simulated)          False\n",
       "Rank (simulated)                        False\n",
       "Mean abs(log2FoldChange) (simulated)    False\n",
       "Std deviation (simulated)               False\n",
       "Number of experiments (simulated)       False\n",
       "Z score                                  True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks.isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACOD1</th>\n",
       "      <td>ACOD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>13655.0</td>\n",
       "      <td>0.758958</td>\n",
       "      <td>0.925459</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL9</th>\n",
       "      <td>ACTL9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>17555.0</td>\n",
       "      <td>0.955850</td>\n",
       "      <td>0.769786</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTRT2</th>\n",
       "      <td>ACTRT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>17524.0</td>\n",
       "      <td>0.908983</td>\n",
       "      <td>0.582880</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM30</th>\n",
       "      <td>ADAM30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>14966.0</td>\n",
       "      <td>0.632421</td>\n",
       "      <td>0.651123</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM6</th>\n",
       "      <td>ADAM6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16311.0</td>\n",
       "      <td>0.659662</td>\n",
       "      <td>0.597932</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAMDEC1</th>\n",
       "      <td>ADAMDEC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999880</td>\n",
       "      <td>15939.0</td>\n",
       "      <td>0.786723</td>\n",
       "      <td>1.042967</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKAIN1</th>\n",
       "      <td>AKAIN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999833</td>\n",
       "      <td>17533.0</td>\n",
       "      <td>0.966889</td>\n",
       "      <td>0.747688</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALPI</th>\n",
       "      <td>ALPI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999880</td>\n",
       "      <td>16805.0</td>\n",
       "      <td>1.008875</td>\n",
       "      <td>1.346828</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMELY</th>\n",
       "      <td>AMELY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>15012.0</td>\n",
       "      <td>0.476614</td>\n",
       "      <td>0.419563</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANHX</th>\n",
       "      <td>ANHX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16978.0</td>\n",
       "      <td>0.846781</td>\n",
       "      <td>1.046244</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANP32D</th>\n",
       "      <td>ANP32D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>10191.0</td>\n",
       "      <td>0.319034</td>\n",
       "      <td>0.480157</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APOA4</th>\n",
       "      <td>APOA4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>17303.0</td>\n",
       "      <td>0.965051</td>\n",
       "      <td>0.881272</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APOBEC1</th>\n",
       "      <td>APOBEC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>16012.0</td>\n",
       "      <td>0.733672</td>\n",
       "      <td>0.680885</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APOC3</th>\n",
       "      <td>APOC3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>17552.0</td>\n",
       "      <td>1.072260</td>\n",
       "      <td>0.977913</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASB10</th>\n",
       "      <td>ASB10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16364.0</td>\n",
       "      <td>0.673198</td>\n",
       "      <td>0.621873</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASB17</th>\n",
       "      <td>ASB17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16294.0</td>\n",
       "      <td>0.603400</td>\n",
       "      <td>0.493900</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATXN8OS</th>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>13575.0</td>\n",
       "      <td>0.455794</td>\n",
       "      <td>0.441975</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AWAT1</th>\n",
       "      <td>AWAT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>12413.0</td>\n",
       "      <td>0.410261</td>\n",
       "      <td>0.385483</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCORP1</th>\n",
       "      <td>BCORP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999880</td>\n",
       "      <td>12307.0</td>\n",
       "      <td>0.519897</td>\n",
       "      <td>0.601908</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BGLT3</th>\n",
       "      <td>BGLT3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16831.0</td>\n",
       "      <td>0.672061</td>\n",
       "      <td>0.610435</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHLHE23</th>\n",
       "      <td>BHLHE23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>15949.0</td>\n",
       "      <td>0.768898</td>\n",
       "      <td>0.667933</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BMP15</th>\n",
       "      <td>BMP15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>17217.0</td>\n",
       "      <td>0.978697</td>\n",
       "      <td>0.702113</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BPIFA3</th>\n",
       "      <td>BPIFA3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16659.0</td>\n",
       "      <td>0.757522</td>\n",
       "      <td>0.624267</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BPIFA4P</th>\n",
       "      <td>BPIFA4P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999880</td>\n",
       "      <td>17631.0</td>\n",
       "      <td>1.344693</td>\n",
       "      <td>0.964002</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BPIFB2</th>\n",
       "      <td>BPIFB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999815</td>\n",
       "      <td>17701.0</td>\n",
       "      <td>1.561908</td>\n",
       "      <td>1.574638</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BPIFB6</th>\n",
       "      <td>BPIFB6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>14996.0</td>\n",
       "      <td>0.493344</td>\n",
       "      <td>0.417897</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10orf53</th>\n",
       "      <td>C10orf53</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>13456.0</td>\n",
       "      <td>0.481401</td>\n",
       "      <td>0.499069</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10orf62</th>\n",
       "      <td>C10orf62</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>15013.0</td>\n",
       "      <td>0.695336</td>\n",
       "      <td>0.638164</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10orf71</th>\n",
       "      <td>C10orf71</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>16660.0</td>\n",
       "      <td>0.793753</td>\n",
       "      <td>0.795268</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10orf71.AS1</th>\n",
       "      <td>C10orf71.AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16750.0</td>\n",
       "      <td>0.705796</td>\n",
       "      <td>0.626105</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM225</th>\n",
       "      <td>TMEM225</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16731.0</td>\n",
       "      <td>0.718176</td>\n",
       "      <td>0.621040</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM235</th>\n",
       "      <td>TMEM235</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16852.0</td>\n",
       "      <td>0.567180</td>\n",
       "      <td>0.346753</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM35A</th>\n",
       "      <td>TMEM35A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>5953.0</td>\n",
       "      <td>0.380043</td>\n",
       "      <td>0.636526</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNP1</th>\n",
       "      <td>TNP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>17258.0</td>\n",
       "      <td>1.073763</td>\n",
       "      <td>0.918441</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53TG3HP</th>\n",
       "      <td>TP53TG3HP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>16804.0</td>\n",
       "      <td>0.727115</td>\n",
       "      <td>0.693699</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTLL2</th>\n",
       "      <td>TTLL2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>9044.0</td>\n",
       "      <td>0.486287</td>\n",
       "      <td>0.621651</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY10</th>\n",
       "      <td>TTTY10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>13410.0</td>\n",
       "      <td>0.479833</td>\n",
       "      <td>0.440272</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY11</th>\n",
       "      <td>TTTY11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16489.0</td>\n",
       "      <td>0.678352</td>\n",
       "      <td>0.631756</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY12</th>\n",
       "      <td>TTTY12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>15445.0</td>\n",
       "      <td>0.663548</td>\n",
       "      <td>0.598549</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY13</th>\n",
       "      <td>TTTY13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>17447.0</td>\n",
       "      <td>0.769720</td>\n",
       "      <td>0.545728</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY5</th>\n",
       "      <td>TTTY5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>17275.0</td>\n",
       "      <td>0.717564</td>\n",
       "      <td>0.593104</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY7</th>\n",
       "      <td>TTTY7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>15556.0</td>\n",
       "      <td>0.649802</td>\n",
       "      <td>0.545826</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TYR</th>\n",
       "      <td>TYR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>16757.0</td>\n",
       "      <td>0.795919</td>\n",
       "      <td>0.722871</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TYROBP</th>\n",
       "      <td>TYROBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>14269.0</td>\n",
       "      <td>0.628838</td>\n",
       "      <td>0.682847</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UCMA</th>\n",
       "      <td>UCMA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999861</td>\n",
       "      <td>16544.0</td>\n",
       "      <td>0.758329</td>\n",
       "      <td>0.693053</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UCN3</th>\n",
       "      <td>UCN3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>15827.0</td>\n",
       "      <td>0.670912</td>\n",
       "      <td>0.531386</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UMOD</th>\n",
       "      <td>UMOD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>17579.0</td>\n",
       "      <td>1.299608</td>\n",
       "      <td>1.322679</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UMODL1.AS1</th>\n",
       "      <td>UMODL1.AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999861</td>\n",
       "      <td>13485.0</td>\n",
       "      <td>0.438783</td>\n",
       "      <td>0.387876</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UST.AS1</th>\n",
       "      <td>UST.AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999861</td>\n",
       "      <td>16886.0</td>\n",
       "      <td>0.679047</td>\n",
       "      <td>0.473761</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UTF1</th>\n",
       "      <td>UTF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>17601.0</td>\n",
       "      <td>1.077227</td>\n",
       "      <td>0.664902</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VN1R4</th>\n",
       "      <td>VN1R4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>17382.0</td>\n",
       "      <td>0.741433</td>\n",
       "      <td>0.509311</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VPREB1</th>\n",
       "      <td>VPREB1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16863.0</td>\n",
       "      <td>0.810462</td>\n",
       "      <td>0.661579</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VSIG4</th>\n",
       "      <td>VSIG4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>15183.0</td>\n",
       "      <td>0.642416</td>\n",
       "      <td>0.645679</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WFDC9</th>\n",
       "      <td>WFDC9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999861</td>\n",
       "      <td>17665.0</td>\n",
       "      <td>1.375993</td>\n",
       "      <td>1.085689</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WNT8A</th>\n",
       "      <td>WNT8A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>13845.0</td>\n",
       "      <td>0.598120</td>\n",
       "      <td>0.533356</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XPNPEP2</th>\n",
       "      <td>XPNPEP2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999889</td>\n",
       "      <td>13955.0</td>\n",
       "      <td>0.597607</td>\n",
       "      <td>0.703951</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC13</th>\n",
       "      <td>ZCCHC13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>16347.0</td>\n",
       "      <td>0.624157</td>\n",
       "      <td>0.465855</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIC3</th>\n",
       "      <td>ZIC3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999880</td>\n",
       "      <td>16062.0</td>\n",
       "      <td>0.739692</td>\n",
       "      <td>0.765171</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF831</th>\n",
       "      <td>ZNF831</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999880</td>\n",
       "      <td>13403.0</td>\n",
       "      <td>0.513317</td>\n",
       "      <td>0.617612</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZPBP</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999904</td>\n",
       "      <td>14842.0</td>\n",
       "      <td>0.676494</td>\n",
       "      <td>0.688041</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>596 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "ACOD1                ACOD1                 NaN          NaN   \n",
       "ACTL9                ACTL9                 NaN          NaN   \n",
       "ACTRT2              ACTRT2                 NaN          NaN   \n",
       "ADAM30              ADAM30                 NaN          NaN   \n",
       "ADAM6                ADAM6                 NaN          NaN   \n",
       "ADAMDEC1          ADAMDEC1                 NaN          NaN   \n",
       "AKAIN1              AKAIN1                 NaN          NaN   \n",
       "ALPI                  ALPI                 NaN          NaN   \n",
       "AMELY                AMELY                 NaN          NaN   \n",
       "ANHX                  ANHX                 NaN          NaN   \n",
       "ANP32D              ANP32D                 NaN          NaN   \n",
       "APOA4                APOA4                 NaN          NaN   \n",
       "APOBEC1            APOBEC1                 NaN          NaN   \n",
       "APOC3                APOC3                 NaN          NaN   \n",
       "ASB10                ASB10                 NaN          NaN   \n",
       "ASB17                ASB17                 NaN          NaN   \n",
       "ATXN8OS            ATXN8OS                 NaN          NaN   \n",
       "AWAT1                AWAT1                 NaN          NaN   \n",
       "BCORP1              BCORP1                 NaN          NaN   \n",
       "BGLT3                BGLT3                 NaN          NaN   \n",
       "BHLHE23            BHLHE23                 NaN          NaN   \n",
       "BMP15                BMP15                 NaN          NaN   \n",
       "BPIFA3              BPIFA3                 NaN          NaN   \n",
       "BPIFA4P            BPIFA4P                 NaN          NaN   \n",
       "BPIFB2              BPIFB2                 NaN          NaN   \n",
       "BPIFB6              BPIFB6                 NaN          NaN   \n",
       "C10orf53          C10orf53                 NaN          NaN   \n",
       "C10orf62          C10orf62                 NaN          NaN   \n",
       "C10orf71          C10orf71                 NaN          NaN   \n",
       "C10orf71.AS1  C10orf71.AS1                 NaN          NaN   \n",
       "...                    ...                 ...          ...   \n",
       "TMEM225            TMEM225                 NaN          NaN   \n",
       "TMEM235            TMEM235                 NaN          NaN   \n",
       "TMEM35A            TMEM35A                 NaN          NaN   \n",
       "TNP1                  TNP1                 NaN          NaN   \n",
       "TP53TG3HP        TP53TG3HP                 NaN          NaN   \n",
       "TTLL2                TTLL2                 NaN          NaN   \n",
       "TTTY10              TTTY10                 NaN          NaN   \n",
       "TTTY11              TTTY11                 NaN          NaN   \n",
       "TTTY12              TTTY12                 NaN          NaN   \n",
       "TTTY13              TTTY13                 NaN          NaN   \n",
       "TTTY5                TTTY5                 NaN          NaN   \n",
       "TTTY7                TTTY7                 NaN          NaN   \n",
       "TYR                    TYR                 NaN          NaN   \n",
       "TYROBP              TYROBP                 NaN          NaN   \n",
       "UCMA                  UCMA                 NaN          NaN   \n",
       "UCN3                  UCN3                 NaN          NaN   \n",
       "UMOD                  UMOD                 NaN          NaN   \n",
       "UMODL1.AS1      UMODL1.AS1                 NaN          NaN   \n",
       "UST.AS1            UST.AS1                 NaN          NaN   \n",
       "UTF1                  UTF1                 NaN          NaN   \n",
       "VN1R4                VN1R4                 NaN          NaN   \n",
       "VPREB1              VPREB1                 NaN          NaN   \n",
       "VSIG4                VSIG4                 NaN          NaN   \n",
       "WFDC9                WFDC9                 NaN          NaN   \n",
       "WNT8A                WNT8A                 NaN          NaN   \n",
       "XPNPEP2            XPNPEP2                 NaN          NaN   \n",
       "ZCCHC13            ZCCHC13                 NaN          NaN   \n",
       "ZIC3                  ZIC3                 NaN          NaN   \n",
       "ZNF831              ZNF831                 NaN          NaN   \n",
       "ZPBP                  ZPBP                 NaN          NaN   \n",
       "\n",
       "              abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "ACOD1                                NaN                    NaN   \n",
       "ACTL9                                NaN                    NaN   \n",
       "ACTRT2                               NaN                    NaN   \n",
       "ADAM30                               NaN                    NaN   \n",
       "ADAM6                                NaN                    NaN   \n",
       "ADAMDEC1                             NaN                    NaN   \n",
       "AKAIN1                               NaN                    NaN   \n",
       "ALPI                                 NaN                    NaN   \n",
       "AMELY                                NaN                    NaN   \n",
       "ANHX                                 NaN                    NaN   \n",
       "ANP32D                               NaN                    NaN   \n",
       "APOA4                                NaN                    NaN   \n",
       "APOBEC1                              NaN                    NaN   \n",
       "APOC3                                NaN                    NaN   \n",
       "ASB10                                NaN                    NaN   \n",
       "ASB17                                NaN                    NaN   \n",
       "ATXN8OS                              NaN                    NaN   \n",
       "AWAT1                                NaN                    NaN   \n",
       "BCORP1                               NaN                    NaN   \n",
       "BGLT3                                NaN                    NaN   \n",
       "BHLHE23                              NaN                    NaN   \n",
       "BMP15                                NaN                    NaN   \n",
       "BPIFA3                               NaN                    NaN   \n",
       "BPIFA4P                              NaN                    NaN   \n",
       "BPIFB2                               NaN                    NaN   \n",
       "BPIFB6                               NaN                    NaN   \n",
       "C10orf53                             NaN                    NaN   \n",
       "C10orf62                             NaN                    NaN   \n",
       "C10orf71                             NaN                    NaN   \n",
       "C10orf71.AS1                         NaN                    NaN   \n",
       "...                                  ...                    ...   \n",
       "TMEM225                              NaN                    NaN   \n",
       "TMEM235                              NaN                    NaN   \n",
       "TMEM35A                              NaN                    NaN   \n",
       "TNP1                                 NaN                    NaN   \n",
       "TP53TG3HP                            NaN                    NaN   \n",
       "TTLL2                                NaN                    NaN   \n",
       "TTTY10                               NaN                    NaN   \n",
       "TTTY11                               NaN                    NaN   \n",
       "TTTY12                               NaN                    NaN   \n",
       "TTTY13                               NaN                    NaN   \n",
       "TTTY5                                NaN                    NaN   \n",
       "TTTY7                                NaN                    NaN   \n",
       "TYR                                  NaN                    NaN   \n",
       "TYROBP                               NaN                    NaN   \n",
       "UCMA                                 NaN                    NaN   \n",
       "UCN3                                 NaN                    NaN   \n",
       "UMOD                                 NaN                    NaN   \n",
       "UMODL1.AS1                           NaN                    NaN   \n",
       "UST.AS1                              NaN                    NaN   \n",
       "UTF1                                 NaN                    NaN   \n",
       "VN1R4                                NaN                    NaN   \n",
       "VPREB1                               NaN                    NaN   \n",
       "VSIG4                                NaN                    NaN   \n",
       "WFDC9                                NaN                    NaN   \n",
       "WNT8A                                NaN                    NaN   \n",
       "XPNPEP2                              NaN                    NaN   \n",
       "ZCCHC13                              NaN                    NaN   \n",
       "ZIC3                                 NaN                    NaN   \n",
       "ZNF831                               NaN                    NaN   \n",
       "ZPBP                                 NaN                    NaN   \n",
       "\n",
       "              Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "ACOD1                               0.999889           13655.0   \n",
       "ACTL9                               0.999885           17555.0   \n",
       "ACTRT2                              0.999904           17524.0   \n",
       "ADAM30                              0.999885           14966.0   \n",
       "ADAM6                               0.999904           16311.0   \n",
       "ADAMDEC1                            0.999880           15939.0   \n",
       "AKAIN1                              0.999833           17533.0   \n",
       "ALPI                                0.999880           16805.0   \n",
       "AMELY                               0.999889           15012.0   \n",
       "ANHX                                0.999904           16978.0   \n",
       "ANP32D                              0.999889           10191.0   \n",
       "APOA4                               0.999889           17303.0   \n",
       "APOBEC1                             0.999889           16012.0   \n",
       "APOC3                               0.999904           17552.0   \n",
       "ASB10                               0.999904           16364.0   \n",
       "ASB17                               0.999904           16294.0   \n",
       "ATXN8OS                             0.999889           13575.0   \n",
       "AWAT1                               0.999904           12413.0   \n",
       "BCORP1                              0.999880           12307.0   \n",
       "BGLT3                               0.999904           16831.0   \n",
       "BHLHE23                             0.999904           15949.0   \n",
       "BMP15                               0.999889           17217.0   \n",
       "BPIFA3                              0.999904           16659.0   \n",
       "BPIFA4P                             0.999880           17631.0   \n",
       "BPIFB2                              0.999815           17701.0   \n",
       "BPIFB6                              0.999904           14996.0   \n",
       "C10orf53                            0.999904           13456.0   \n",
       "C10orf62                            0.999885           15013.0   \n",
       "C10orf71                            0.999885           16660.0   \n",
       "C10orf71.AS1                        0.999904           16750.0   \n",
       "...                                      ...               ...   \n",
       "TMEM225                             0.999904           16731.0   \n",
       "TMEM235                             0.999904           16852.0   \n",
       "TMEM35A                             0.999889            5953.0   \n",
       "TNP1                                0.999889           17258.0   \n",
       "TP53TG3HP                           0.999885           16804.0   \n",
       "TTLL2                               0.999889            9044.0   \n",
       "TTTY10                              0.999904           13410.0   \n",
       "TTTY11                              0.999904           16489.0   \n",
       "TTTY12                              0.999904           15445.0   \n",
       "TTTY13                              0.999885           17447.0   \n",
       "TTTY5                               0.999904           17275.0   \n",
       "TTTY7                               0.999889           15556.0   \n",
       "TYR                                 0.999889           16757.0   \n",
       "TYROBP                              0.999889           14269.0   \n",
       "UCMA                                0.999861           16544.0   \n",
       "UCN3                                0.999885           15827.0   \n",
       "UMOD                                0.999885           17579.0   \n",
       "UMODL1.AS1                          0.999861           13485.0   \n",
       "UST.AS1                             0.999861           16886.0   \n",
       "UTF1                                0.999885           17601.0   \n",
       "VN1R4                               0.999904           17382.0   \n",
       "VPREB1                              0.999904           16863.0   \n",
       "VSIG4                               0.999889           15183.0   \n",
       "WFDC9                               0.999861           17665.0   \n",
       "WNT8A                               0.999904           13845.0   \n",
       "XPNPEP2                             0.999889           13955.0   \n",
       "ZCCHC13                             0.999904           16347.0   \n",
       "ZIC3                                0.999880           16062.0   \n",
       "ZNF831                              0.999880           13403.0   \n",
       "ZPBP                                0.999904           14842.0   \n",
       "\n",
       "              Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "ACOD1                                     0.758958                   0.925459   \n",
       "ACTL9                                     0.955850                   0.769786   \n",
       "ACTRT2                                    0.908983                   0.582880   \n",
       "ADAM30                                    0.632421                   0.651123   \n",
       "ADAM6                                     0.659662                   0.597932   \n",
       "ADAMDEC1                                  0.786723                   1.042967   \n",
       "AKAIN1                                    0.966889                   0.747688   \n",
       "ALPI                                      1.008875                   1.346828   \n",
       "AMELY                                     0.476614                   0.419563   \n",
       "ANHX                                      0.846781                   1.046244   \n",
       "ANP32D                                    0.319034                   0.480157   \n",
       "APOA4                                     0.965051                   0.881272   \n",
       "APOBEC1                                   0.733672                   0.680885   \n",
       "APOC3                                     1.072260                   0.977913   \n",
       "ASB10                                     0.673198                   0.621873   \n",
       "ASB17                                     0.603400                   0.493900   \n",
       "ATXN8OS                                   0.455794                   0.441975   \n",
       "AWAT1                                     0.410261                   0.385483   \n",
       "BCORP1                                    0.519897                   0.601908   \n",
       "BGLT3                                     0.672061                   0.610435   \n",
       "BHLHE23                                   0.768898                   0.667933   \n",
       "BMP15                                     0.978697                   0.702113   \n",
       "BPIFA3                                    0.757522                   0.624267   \n",
       "BPIFA4P                                   1.344693                   0.964002   \n",
       "BPIFB2                                    1.561908                   1.574638   \n",
       "BPIFB6                                    0.493344                   0.417897   \n",
       "C10orf53                                  0.481401                   0.499069   \n",
       "C10orf62                                  0.695336                   0.638164   \n",
       "C10orf71                                  0.793753                   0.795268   \n",
       "C10orf71.AS1                              0.705796                   0.626105   \n",
       "...                                            ...                        ...   \n",
       "TMEM225                                   0.718176                   0.621040   \n",
       "TMEM235                                   0.567180                   0.346753   \n",
       "TMEM35A                                   0.380043                   0.636526   \n",
       "TNP1                                      1.073763                   0.918441   \n",
       "TP53TG3HP                                 0.727115                   0.693699   \n",
       "TTLL2                                     0.486287                   0.621651   \n",
       "TTTY10                                    0.479833                   0.440272   \n",
       "TTTY11                                    0.678352                   0.631756   \n",
       "TTTY12                                    0.663548                   0.598549   \n",
       "TTTY13                                    0.769720                   0.545728   \n",
       "TTTY5                                     0.717564                   0.593104   \n",
       "TTTY7                                     0.649802                   0.545826   \n",
       "TYR                                       0.795919                   0.722871   \n",
       "TYROBP                                    0.628838                   0.682847   \n",
       "UCMA                                      0.758329                   0.693053   \n",
       "UCN3                                      0.670912                   0.531386   \n",
       "UMOD                                      1.299608                   1.322679   \n",
       "UMODL1.AS1                                0.438783                   0.387876   \n",
       "UST.AS1                                   0.679047                   0.473761   \n",
       "UTF1                                      1.077227                   0.664902   \n",
       "VN1R4                                     0.741433                   0.509311   \n",
       "VPREB1                                    0.810462                   0.661579   \n",
       "VSIG4                                     0.642416                   0.645679   \n",
       "WFDC9                                     1.375993                   1.085689   \n",
       "WNT8A                                     0.598120                   0.533356   \n",
       "XPNPEP2                                   0.597607                   0.703951   \n",
       "ZCCHC13                                   0.624157                   0.465855   \n",
       "ZIC3                                      0.739692                   0.765171   \n",
       "ZNF831                                    0.513317                   0.617612   \n",
       "ZPBP                                      0.676494                   0.688041   \n",
       "\n",
       "              Number of experiments (simulated)  Z score  \n",
       "ACOD1                                        25      NaN  \n",
       "ACTL9                                        24      NaN  \n",
       "ACTRT2                                       24      NaN  \n",
       "ADAM30                                       24      NaN  \n",
       "ADAM6                                        24      NaN  \n",
       "ADAMDEC1                                     25      NaN  \n",
       "AKAIN1                                       24      NaN  \n",
       "ALPI                                         25      NaN  \n",
       "AMELY                                        23      NaN  \n",
       "ANHX                                         24      NaN  \n",
       "ANP32D                                       25      NaN  \n",
       "APOA4                                        25      NaN  \n",
       "APOBEC1                                      25      NaN  \n",
       "APOC3                                        24      NaN  \n",
       "ASB10                                        24      NaN  \n",
       "ASB17                                        24      NaN  \n",
       "ATXN8OS                                      25      NaN  \n",
       "AWAT1                                        24      NaN  \n",
       "BCORP1                                       25      NaN  \n",
       "BGLT3                                        24      NaN  \n",
       "BHLHE23                                      24      NaN  \n",
       "BMP15                                        23      NaN  \n",
       "BPIFA3                                       24      NaN  \n",
       "BPIFA4P                                      25      NaN  \n",
       "BPIFB2                                       24      NaN  \n",
       "BPIFB6                                       24      NaN  \n",
       "C10orf53                                     24      NaN  \n",
       "C10orf62                                     24      NaN  \n",
       "C10orf71                                     24      NaN  \n",
       "C10orf71.AS1                                 24      NaN  \n",
       "...                                         ...      ...  \n",
       "TMEM225                                      24      NaN  \n",
       "TMEM235                                      24      NaN  \n",
       "TMEM35A                                      25      NaN  \n",
       "TNP1                                         25      NaN  \n",
       "TP53TG3HP                                    24      NaN  \n",
       "TTLL2                                        25      NaN  \n",
       "TTTY10                                       24      NaN  \n",
       "TTTY11                                       20      NaN  \n",
       "TTTY12                                       20      NaN  \n",
       "TTTY13                                       22      NaN  \n",
       "TTTY5                                        20      NaN  \n",
       "TTTY7                                        21      NaN  \n",
       "TYR                                          25      NaN  \n",
       "TYROBP                                       25      NaN  \n",
       "UCMA                                         24      NaN  \n",
       "UCN3                                         24      NaN  \n",
       "UMOD                                         24      NaN  \n",
       "UMODL1.AS1                                   24      NaN  \n",
       "UST.AS1                                      24      NaN  \n",
       "UTF1                                         24      NaN  \n",
       "VN1R4                                        22      NaN  \n",
       "VPREB1                                       24      NaN  \n",
       "VSIG4                                        25      NaN  \n",
       "WFDC9                                        22      NaN  \n",
       "WNT8A                                        24      NaN  \n",
       "XPNPEP2                                      25      NaN  \n",
       "ZCCHC13                                      24      NaN  \n",
       "ZIC3                                         25      NaN  \n",
       "ZNF831                                       25      NaN  \n",
       "ZPBP                                         24      NaN  \n",
       "\n",
       "[596 rows x 11 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks[summary_gene_ranks.isna().any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "summary_gene_ranks.to_csv(gene_summary_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
