{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Application: new experiment\n",
    "\n",
    "This notebook allows users to find specific genes in their experiment of interest using an existing VAE model\n",
    "\n",
    "This notebook will generate a `generic_gene_summary_<experiment id>.tsv` file that contains a z-score per gene that indicates how specific a gene is the experiment in question."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from ponyo import utils\n",
    "from generic_expression_patterns_modules import process, new_experiment_process, stats, ranking"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User input\n",
    "\n",
    "User needs to define the following in the [config file](../configs/config_new_experiment.tsv):\n",
    "\n",
    "1. Template experiment. This is the experiment you are interested in studying\n",
    "2. Training compendium used to train VAE, including unnormalized gene mapped version and normalized version\n",
    "3. Scaler transform used to normalize the training compendium\n",
    "4. Directory containing trained VAE model\n",
    "5. Experiment id to label newly create simulated experiments\n",
    "\n",
    "The user also needs to provide metadata files:\n",
    "1. `<experiment id>_process_samples.tsv` contains 2 columns (sample ids, label that indicates if the sample is kept or removed). See [example](data/metadata/cis-gem-par-KU1919_process_samples.tsv). **Note: This file is not required if the user wishes to use all the samples in the template experiment file.**\n",
    "2. `<experiment id>_groups.tsv` contains 2 columns: sample ids, group label to perform DE analysis. See [example](data/metadata/cis-gem-par-KU1919_groups.tsv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = os.path.abspath(\n",
    "    os.path.join(base_dir, \"configs\", \"config_new_experiment.tsv\")\n",
    ")\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load config params\n",
    "\n",
    "# Local directory to store intermediate files\n",
    "local_dir = params['local_dir']\n",
    "\n",
    "# Number of simulated experiments to generate\n",
    "num_runs = params['num_simulated']\n",
    "\n",
    "# Directory containing trained VAE model\n",
    "vae_model_dir = params['vae_model_dir']\n",
    "\n",
    "# Dimension of latent space used in VAE model\n",
    "latent_dim = params['latent_dim']\n",
    "\n",
    "# ID for template experiment\n",
    "# This ID will be used to label new simulated experiments\n",
    "project_id = params['project_id']\n",
    "\n",
    "# Template experiment filename\n",
    "template_filename = params['raw_template_filename']\n",
    "mapped_template_filename = params['mapped_template_filename']\n",
    "normalized_template_filename = params['normalized_template_filename']\n",
    "processed_template_filename = params['processed_template_filename']\n",
    "\n",
    "# Training dataset used for existing VAE model\n",
    "mapped_compendium_filename = params['mapped_compendium_filename']\n",
    "\n",
    "# Normalized compendium filename\n",
    "normalized_compendium_filename = params['normalized_compendium_filename']\n",
    "\n",
    "# Scaler transform used to scale compendium data into 0-1 range for training\n",
    "scaler_filename = params['scaler_filename']\n",
    "\n",
    "# Test statistic used to rank genes by\n",
    "col_to_rank_genes = params['rank_genes_by']\n",
    "\n",
    "# Minimum mean count per gene\n",
    "count_threshold = params['count_threshold']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata files\n",
    "\n",
    "# Load metadata file with processing information\n",
    "sample_id_metadata_filename = os.path.join(\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    f\"{project_id}_process_samples.tsv\"\n",
    ")\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "metadata_filename = os.path.join(\n",
    "    \"data\",\n",
    "    \"metadata\",\n",
    "    f\"{project_id}_groups.tsv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output filename\n",
    "gene_summary_filename = f\"generic_gene_summary_{project_id}.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map template experiment to same feature space as training compendium\n",
    "\n",
    "In order to simulate a new gene expression experiment, we will need to encode this experiment into the learned latent space. This requires that the feature space (i.e. genes) in the template experiment match the features in the compendium used to train the VAE model. These cells process the template experiment to be of the expected input format:\n",
    "* Template data is expected to be a matrix that is sample x gene\n",
    "* Template experiment is expected to have the same genes as the compendium experiment. Genes that are in the template experiment but not in the compendium are removed. Genes that are in the compendium but missing in the template experiment are added and the gene expression value is set to the median gene expression value of that gene across the samples in the compendium."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Template experiment needs to be of the form sample x gene\n",
    "template_filename_only = template_filename.split(\"/\")[-1].split(\".\")[0]\n",
    "transposed_template_filename = os.path.join(local_dir, template_filename_only+\"_transposed.txt\")\n",
    "\n",
    "new_experiment_process.transpose_save(template_filename, transposed_template_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(72, 58528)\n",
      "(49651, 17755)\n"
     ]
    }
   ],
   "source": [
    "new_experiment_process.process_template_experiment(\n",
    "    transposed_template_filename,\n",
    "    mapped_compendium_filename,\n",
    "    scaler_filename,\n",
    "    mapped_template_filename,\n",
    "    normalized_template_filename,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate experiments based on template experiment\n",
    "\n",
    "Embed template experiment into learned latent space and linearly shift template experiment to different locations of the latent space to create new experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate experiments based on template experiment\n",
    "normalized_compendium_data = pd.read_csv(normalized_compendium_filename, sep=\"\\t\", index_col=0, header=0)\n",
    "normalized_template_data = pd.read_csv(normalized_template_filename, sep=\"\\t\", index_col=0, header=0)\n",
    "\n",
    "for run_id in range(num_runs):\n",
    "    new_experiment_process.embed_shift_template_experiment(\n",
    "        normalized_compendium_data,\n",
    "        normalized_template_data,\n",
    "        vae_model_dir,\n",
    "        project_id,\n",
    "        scaler_filename,\n",
    "        local_dir,\n",
    "        latent_dim,\n",
    "        run_id\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process template and simulated experiments\n",
    "\n",
    "* Remove samples not required for comparison\n",
    "* Make sure ordering of samples matches metadata for proper comparison\n",
    "* Make sure values are cast as integers if using DESeq\n",
    "* Filter lowly expressed genes if using DESeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"human_general_analysis\" in vae_model_dir:\n",
    "    method = \"deseq\"\n",
    "else:\n",
    "    method = \"limma\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(sample_id_metadata_filename):\n",
    "    sample_id_metadata_filename = None\n",
    "    \n",
    "if method == \"deseq\":\n",
    "    stats.process_samples_for_DESeq(\n",
    "        mapped_template_filename,\n",
    "        metadata_filename,\n",
    "        processed_template_filename,\n",
    "        count_threshold,\n",
    "        sample_id_metadata_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\"\n",
    "        )\n",
    "        out_simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}_processed.txt\"\n",
    "        )\n",
    "        stats.process_samples_for_DESeq(\n",
    "            simulated_filename,\n",
    "            metadata_filename,\n",
    "            out_simulated_filename,\n",
    "            count_threshold,\n",
    "            sample_id_metadata_filename,\n",
    "    )\n",
    "else:\n",
    "    stats.process_samples_for_limma(\n",
    "        mapped_template_filename,\n",
    "        metadata_filename,\n",
    "        processed_template_filename,\n",
    "        sample_id_metadata_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\"\n",
    "        )\n",
    "        stats.process_samples_for_limma(\n",
    "            simulated_filename,\n",
    "            metadata_filename,\n",
    "            None,\n",
    "            sample_id_metadata_filename,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential expression analysis\n",
    "\n",
    "* If data is RNA-seq then use DESeq2 (using human_general_analysis model)\n",
    "* If data is microarray then use Limma (using human_cancer_analysis, pseudomonas_analysis models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: S4Vectors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: stats4\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocGenerics\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: parallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: IRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomicRanges\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: GenomeInfoDb\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: SummarizedExperiment\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: Biobase\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: DelayedArray\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: matrixStats\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: Loading required package: BiocParallel\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: estimating size factors\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: estimating dispersions\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: gene-wise dispersion estimates\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: mean-dispersion relationship\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: final dispersion estimates\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: fitting model and testing\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        metadata_filename,\n",
    "        project_id, \n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        metadata_filename,\n",
    "        project_id, \n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    ) \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/rpy2/rinterface/__init__.py:146: RRuntimeWarning: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n",
       "[1] \"Checking sample ordering...\"\n",
       "[1] TRUE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i base_dir -i local_dir -i num_runs -i method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# Files created: \"<local_dir>/DE_stats/DE_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir, \n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\", \n",
    "        i,\n",
    "        \"_processed.txt\",\n",
    "        sep = \"\"\n",
    "    )\n",
    "    if (method == \"deseq\"){\n",
    "        get_DE_stats_DESeq(\n",
    "            metadata_filename,\n",
    "            project_id, \n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "    }\n",
    "    else {\n",
    "        get_DE_stats_limma(\n",
    "            metadata_filename,\n",
    "            project_id, \n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rank genes\n",
    "\n",
    "Genes are ranked by their \"generic-ness\" - how frequently these genes are changed across the simulated experiments using user-specific test statistic (i.e. log2 fold change)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir,\n",
    "    \"DE_stats\",\n",
    "    f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary table\n",
    "\n",
    "* Gene ID: Gene identifier (hgnc symbols for human data or PA number for *P. aeruginosa* data)\n",
    "* (Real): Statistics for template experiment\n",
    "* (Simulated): Statistics across simulated experiments\n",
    "* Number of experiments: Number of simulated experiments\n",
    "* Z-score: High z-score indicates that gene is more changed in template compared to the null set of simulated experiments (high z-score = highly specific to template experiment)\n",
    "\n",
    "\n",
    "**Note:** \n",
    "* If using DESeq, genes with NaN in `Adj P-value (Real)` column are those genes flagged because of the `cooksCutoff` parameter. The cook's distance as a diagnostic to tell if a single sample has a count which has a disproportionate impact on the log fold change and p-values. These genes are flagged with an NA in the pvalue and padj columns of the result table. \n",
    "\n",
    "* If using DESeq with count threshold, some genes may not be present in all simulated experiments (i.e. the `Number of experiments (simulated)` will not equal the number of simulated experiments you specified in the beginning. This pre-filtering will lead to some genes found in few simulated experiments and so the background/null set for that gene is not robust. Thus, the user should sort by both z-score and number of experiments to identify specific expressed genes.\n",
    "\n",
    "* If using DESeq without count threshold, some genes may still not be present in all simulated experiments (i.e. the `Number of experiments (simulated)`  will not equal the number of simulated experiments you specified in the beginning. If the gene is 0 expressed across all samples and thus automatically given an NA in log fold change, adjusted p-value columns. Thus, the user should sort by both z-score and number of experiments to identify specific expressed genes.\n",
    "\n",
    "For more information you can read [DESeq FAQs](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pvaluesNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (1 levels on the left, 2 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DCK</th>\n",
       "      <td>DCK</td>\n",
       "      <td>1.882191e-61</td>\n",
       "      <td>15957.0</td>\n",
       "      <td>8.213785</td>\n",
       "      <td>-8.213785</td>\n",
       "      <td>0.999050</td>\n",
       "      <td>5754.0</td>\n",
       "      <td>0.486371</td>\n",
       "      <td>0.372840</td>\n",
       "      <td>25</td>\n",
       "      <td>20.725828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGST2</th>\n",
       "      <td>MGST2</td>\n",
       "      <td>5.249609e-26</td>\n",
       "      <td>15933.0</td>\n",
       "      <td>6.904599</td>\n",
       "      <td>-6.904599</td>\n",
       "      <td>0.998963</td>\n",
       "      <td>723.0</td>\n",
       "      <td>0.362911</td>\n",
       "      <td>0.363182</td>\n",
       "      <td>25</td>\n",
       "      <td>18.012164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SFT2D3</th>\n",
       "      <td>SFT2D3</td>\n",
       "      <td>1.883872e-06</td>\n",
       "      <td>15973.0</td>\n",
       "      <td>10.966455</td>\n",
       "      <td>10.966455</td>\n",
       "      <td>0.999050</td>\n",
       "      <td>4242.0</td>\n",
       "      <td>0.508671</td>\n",
       "      <td>0.616672</td>\n",
       "      <td>25</td>\n",
       "      <td>16.958425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EPHA5</th>\n",
       "      <td>EPHA5</td>\n",
       "      <td>7.162665e-24</td>\n",
       "      <td>15974.0</td>\n",
       "      <td>12.505744</td>\n",
       "      <td>-12.505744</td>\n",
       "      <td>0.994858</td>\n",
       "      <td>9697.0</td>\n",
       "      <td>0.851190</td>\n",
       "      <td>0.753435</td>\n",
       "      <td>23</td>\n",
       "      <td>15.468551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM71E1</th>\n",
       "      <td>FAM71E1</td>\n",
       "      <td>5.392281e-08</td>\n",
       "      <td>15948.0</td>\n",
       "      <td>7.544186</td>\n",
       "      <td>-7.544186</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>5549.0</td>\n",
       "      <td>0.451948</td>\n",
       "      <td>0.468341</td>\n",
       "      <td>25</td>\n",
       "      <td>15.143315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NME5</th>\n",
       "      <td>NME5</td>\n",
       "      <td>2.531529e-10</td>\n",
       "      <td>15962.0</td>\n",
       "      <td>8.449430</td>\n",
       "      <td>-8.449430</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>7985.0</td>\n",
       "      <td>0.742114</td>\n",
       "      <td>0.536800</td>\n",
       "      <td>23</td>\n",
       "      <td>14.357880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HERC3</th>\n",
       "      <td>HERC3</td>\n",
       "      <td>2.630223e-19</td>\n",
       "      <td>15913.0</td>\n",
       "      <td>6.164117</td>\n",
       "      <td>-6.164117</td>\n",
       "      <td>0.998963</td>\n",
       "      <td>3903.0</td>\n",
       "      <td>0.492526</td>\n",
       "      <td>0.398572</td>\n",
       "      <td>25</td>\n",
       "      <td>14.229787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC2A10</th>\n",
       "      <td>SLC2A10</td>\n",
       "      <td>2.912661e-17</td>\n",
       "      <td>15969.0</td>\n",
       "      <td>10.696798</td>\n",
       "      <td>-10.696798</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>8523.0</td>\n",
       "      <td>0.763058</td>\n",
       "      <td>0.791810</td>\n",
       "      <td>24</td>\n",
       "      <td>12.545604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KL</th>\n",
       "      <td>KL</td>\n",
       "      <td>1.058591e-17</td>\n",
       "      <td>15971.0</td>\n",
       "      <td>10.787192</td>\n",
       "      <td>-10.787192</td>\n",
       "      <td>0.999564</td>\n",
       "      <td>10067.0</td>\n",
       "      <td>0.904635</td>\n",
       "      <td>0.838714</td>\n",
       "      <td>24</td>\n",
       "      <td>11.782981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRIA3</th>\n",
       "      <td>GRIA3</td>\n",
       "      <td>3.355769e-12</td>\n",
       "      <td>15967.0</td>\n",
       "      <td>9.179724</td>\n",
       "      <td>-9.179724</td>\n",
       "      <td>0.999280</td>\n",
       "      <td>14319.0</td>\n",
       "      <td>0.923363</td>\n",
       "      <td>0.706004</td>\n",
       "      <td>22</td>\n",
       "      <td>11.694490</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "DCK          DCK        1.882191e-61      15957.0                    8.213785   \n",
       "MGST2      MGST2        5.249609e-26      15933.0                    6.904599   \n",
       "SFT2D3    SFT2D3        1.883872e-06      15973.0                   10.966455   \n",
       "EPHA5      EPHA5        7.162665e-24      15974.0                   12.505744   \n",
       "FAM71E1  FAM71E1        5.392281e-08      15948.0                    7.544186   \n",
       "NME5        NME5        2.531529e-10      15962.0                    8.449430   \n",
       "HERC3      HERC3        2.630223e-19      15913.0                    6.164117   \n",
       "SLC2A10  SLC2A10        2.912661e-17      15969.0                   10.696798   \n",
       "KL            KL        1.058591e-17      15971.0                   10.787192   \n",
       "GRIA3      GRIA3        3.355769e-12      15967.0                    9.179724   \n",
       "\n",
       "         log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "DCK                  -8.213785                        0.999050   \n",
       "MGST2                -6.904599                        0.998963   \n",
       "SFT2D3               10.966455                        0.999050   \n",
       "EPHA5               -12.505744                        0.994858   \n",
       "FAM71E1              -7.544186                        0.998997   \n",
       "NME5                 -8.449430                        0.998997   \n",
       "HERC3                -6.164117                        0.998963   \n",
       "SLC2A10             -10.696798                        0.998997   \n",
       "KL                  -10.787192                        0.999564   \n",
       "GRIA3                -9.179724                        0.999280   \n",
       "\n",
       "         Rank (simulated)  Mean abs(log2FoldChange) (simulated)  \\\n",
       "DCK                5754.0                              0.486371   \n",
       "MGST2               723.0                              0.362911   \n",
       "SFT2D3             4242.0                              0.508671   \n",
       "EPHA5              9697.0                              0.851190   \n",
       "FAM71E1            5549.0                              0.451948   \n",
       "NME5               7985.0                              0.742114   \n",
       "HERC3              3903.0                              0.492526   \n",
       "SLC2A10            8523.0                              0.763058   \n",
       "KL                10067.0                              0.904635   \n",
       "GRIA3             14319.0                              0.923363   \n",
       "\n",
       "         Std deviation (simulated)  Number of experiments (simulated)  \\\n",
       "DCK                       0.372840                                 25   \n",
       "MGST2                     0.363182                                 25   \n",
       "SFT2D3                    0.616672                                 25   \n",
       "EPHA5                     0.753435                                 23   \n",
       "FAM71E1                   0.468341                                 25   \n",
       "NME5                      0.536800                                 23   \n",
       "HERC3                     0.398572                                 25   \n",
       "SLC2A10                   0.791810                                 24   \n",
       "KL                        0.838714                                 24   \n",
       "GRIA3                     0.706004                                 22   \n",
       "\n",
       "           Z score  \n",
       "DCK      20.725828  \n",
       "MGST2    18.012164  \n",
       "SFT2D3   16.958425  \n",
       "EPHA5    15.468551  \n",
       "FAM71E1  15.143315  \n",
       "NME5     14.357880  \n",
       "HERC3    14.229787  \n",
       "SLC2A10  12.545604  \n",
       "KL       11.782981  \n",
       "GRIA3    11.694490  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get summary table\n",
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    'gene',\n",
    "    params\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene ID                                 False\n",
       "Adj P-value (Real)                       True\n",
       "Rank (Real)                              True\n",
       "abs(log2FoldChange) (Real)               True\n",
       "log2FoldChange (Real)                    True\n",
       "Median adj p-value (simulated)          False\n",
       "Rank (simulated)                        False\n",
       "Mean abs(log2FoldChange) (simulated)    False\n",
       "Std deviation (simulated)               False\n",
       "Number of experiments (simulated)       False\n",
       "Z score                                  True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks.isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KRT5</th>\n",
       "      <td>KRT5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14648.0</td>\n",
       "      <td>1.605915</td>\n",
       "      <td>1.605915</td>\n",
       "      <td>0.999564</td>\n",
       "      <td>16329.0</td>\n",
       "      <td>1.166528</td>\n",
       "      <td>0.796012</td>\n",
       "      <td>20</td>\n",
       "      <td>0.551985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAC</th>\n",
       "      <td>AADAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999333</td>\n",
       "      <td>15738.0</td>\n",
       "      <td>1.136864</td>\n",
       "      <td>0.817906</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADACL2</th>\n",
       "      <td>AADACL2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999564</td>\n",
       "      <td>10048.0</td>\n",
       "      <td>0.823911</td>\n",
       "      <td>0.891710</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADACP1</th>\n",
       "      <td>AADACP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999564</td>\n",
       "      <td>9363.0</td>\n",
       "      <td>0.864613</td>\n",
       "      <td>0.822658</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC13</th>\n",
       "      <td>ABCC13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>13770.5</td>\n",
       "      <td>0.872969</td>\n",
       "      <td>0.611306</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P1</th>\n",
       "      <td>ABCC6P1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999333</td>\n",
       "      <td>11669.0</td>\n",
       "      <td>0.963358</td>\n",
       "      <td>0.840906</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACKR1</th>\n",
       "      <td>ACKR1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999590</td>\n",
       "      <td>15420.0</td>\n",
       "      <td>1.226246</td>\n",
       "      <td>0.905918</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACOD1</th>\n",
       "      <td>ACOD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999564</td>\n",
       "      <td>17534.0</td>\n",
       "      <td>1.569318</td>\n",
       "      <td>0.951284</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACOT12</th>\n",
       "      <td>ACOT12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>13073.0</td>\n",
       "      <td>0.840446</td>\n",
       "      <td>0.846915</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACSM1</th>\n",
       "      <td>ACSM1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>14947.0</td>\n",
       "      <td>1.081291</td>\n",
       "      <td>0.953081</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL7A</th>\n",
       "      <td>ACTL7A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>14333.0</td>\n",
       "      <td>1.095669</td>\n",
       "      <td>1.034700</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTL9</th>\n",
       "      <td>ACTL9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>16108.0</td>\n",
       "      <td>1.326044</td>\n",
       "      <td>1.017599</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTRT2</th>\n",
       "      <td>ACTRT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.992743</td>\n",
       "      <td>10599.0</td>\n",
       "      <td>0.984813</td>\n",
       "      <td>1.110716</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAD1</th>\n",
       "      <td>ADAD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999218</td>\n",
       "      <td>11347.0</td>\n",
       "      <td>1.064397</td>\n",
       "      <td>0.986995</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAD2</th>\n",
       "      <td>ADAD2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>11564.0</td>\n",
       "      <td>0.816135</td>\n",
       "      <td>0.586090</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM18</th>\n",
       "      <td>ADAM18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999050</td>\n",
       "      <td>4349.0</td>\n",
       "      <td>0.741704</td>\n",
       "      <td>0.909907</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM2</th>\n",
       "      <td>ADAM2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>11995.0</td>\n",
       "      <td>0.928309</td>\n",
       "      <td>0.783897</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM5</th>\n",
       "      <td>ADAM5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999346</td>\n",
       "      <td>10837.0</td>\n",
       "      <td>0.859152</td>\n",
       "      <td>0.725460</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM6</th>\n",
       "      <td>ADAM6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998076</td>\n",
       "      <td>4393.0</td>\n",
       "      <td>0.711498</td>\n",
       "      <td>0.734518</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAMDEC1</th>\n",
       "      <td>ADAMDEC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999333</td>\n",
       "      <td>17035.0</td>\n",
       "      <td>1.319993</td>\n",
       "      <td>0.932814</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAMTS16</th>\n",
       "      <td>ADAMTS16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>13457.0</td>\n",
       "      <td>0.884074</td>\n",
       "      <td>0.945483</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAMTS20</th>\n",
       "      <td>ADAMTS20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>10231.0</td>\n",
       "      <td>0.832813</td>\n",
       "      <td>0.848892</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADARB2</th>\n",
       "      <td>ADARB2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999333</td>\n",
       "      <td>15739.0</td>\n",
       "      <td>0.906310</td>\n",
       "      <td>0.663126</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADARB2.AS1</th>\n",
       "      <td>ADARB2.AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>4513.0</td>\n",
       "      <td>0.686892</td>\n",
       "      <td>1.013882</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADCYAP1</th>\n",
       "      <td>ADCYAP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999333</td>\n",
       "      <td>16470.0</td>\n",
       "      <td>1.095572</td>\n",
       "      <td>0.768004</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADIG</th>\n",
       "      <td>ADIG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>7549.5</td>\n",
       "      <td>0.950571</td>\n",
       "      <td>1.089967</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADORA3</th>\n",
       "      <td>ADORA3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>14396.0</td>\n",
       "      <td>0.968147</td>\n",
       "      <td>0.728322</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADRB3</th>\n",
       "      <td>ADRB3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999050</td>\n",
       "      <td>12815.0</td>\n",
       "      <td>0.942848</td>\n",
       "      <td>0.826491</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP11</th>\n",
       "      <td>AGAP11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>5801.0</td>\n",
       "      <td>0.630398</td>\n",
       "      <td>0.803771</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTR2</th>\n",
       "      <td>AGTR2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999564</td>\n",
       "      <td>13626.0</td>\n",
       "      <td>1.005858</td>\n",
       "      <td>0.809055</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZAR1</th>\n",
       "      <td>ZAR1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>8977.0</td>\n",
       "      <td>0.669929</td>\n",
       "      <td>0.381470</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZBTB20.AS1</th>\n",
       "      <td>ZBTB20.AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>17659.0</td>\n",
       "      <td>1.420848</td>\n",
       "      <td>0.888232</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZBTB46.AS1</th>\n",
       "      <td>ZBTB46.AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>9189.0</td>\n",
       "      <td>0.821911</td>\n",
       "      <td>0.888074</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZC2HC1B</th>\n",
       "      <td>ZC2HC1B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999050</td>\n",
       "      <td>9610.0</td>\n",
       "      <td>0.902370</td>\n",
       "      <td>0.943740</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC13</th>\n",
       "      <td>ZCCHC13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.992743</td>\n",
       "      <td>6275.0</td>\n",
       "      <td>0.683683</td>\n",
       "      <td>0.519412</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFY</th>\n",
       "      <td>ZFY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999050</td>\n",
       "      <td>8213.0</td>\n",
       "      <td>0.633449</td>\n",
       "      <td>0.580128</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIC1</th>\n",
       "      <td>ZIC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>16957.0</td>\n",
       "      <td>1.176521</td>\n",
       "      <td>0.619447</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIC3</th>\n",
       "      <td>ZIC3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>14457.0</td>\n",
       "      <td>0.976478</td>\n",
       "      <td>0.715314</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIC4</th>\n",
       "      <td>ZIC4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>16545.0</td>\n",
       "      <td>1.060821</td>\n",
       "      <td>0.685805</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIM2</th>\n",
       "      <td>ZIM2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998963</td>\n",
       "      <td>13176.0</td>\n",
       "      <td>1.011619</td>\n",
       "      <td>0.586075</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIM3</th>\n",
       "      <td>ZIM3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998630</td>\n",
       "      <td>11365.0</td>\n",
       "      <td>1.113619</td>\n",
       "      <td>1.120134</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZMAT4</th>\n",
       "      <td>ZMAT4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999333</td>\n",
       "      <td>15477.0</td>\n",
       "      <td>0.822516</td>\n",
       "      <td>0.507937</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF135</th>\n",
       "      <td>ZNF135</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>7662.0</td>\n",
       "      <td>0.659440</td>\n",
       "      <td>0.843000</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF229</th>\n",
       "      <td>ZNF229</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998963</td>\n",
       "      <td>6258.0</td>\n",
       "      <td>0.620232</td>\n",
       "      <td>0.523254</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF32.AS3</th>\n",
       "      <td>ZNF32.AS3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.992743</td>\n",
       "      <td>17576.0</td>\n",
       "      <td>1.404103</td>\n",
       "      <td>0.779750</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF418</th>\n",
       "      <td>ZNF418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>8202.0</td>\n",
       "      <td>0.851773</td>\n",
       "      <td>1.041167</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF547</th>\n",
       "      <td>ZNF547</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>1519.0</td>\n",
       "      <td>0.463075</td>\n",
       "      <td>0.534501</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF560</th>\n",
       "      <td>ZNF560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>13334.0</td>\n",
       "      <td>0.901327</td>\n",
       "      <td>0.771061</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF577</th>\n",
       "      <td>ZNF577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998929</td>\n",
       "      <td>8633.0</td>\n",
       "      <td>0.921354</td>\n",
       "      <td>1.094319</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF649.AS1</th>\n",
       "      <td>ZNF649.AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998997</td>\n",
       "      <td>8148.0</td>\n",
       "      <td>0.795041</td>\n",
       "      <td>0.950439</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF683</th>\n",
       "      <td>ZNF683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999577</td>\n",
       "      <td>17648.0</td>\n",
       "      <td>1.480176</td>\n",
       "      <td>0.845569</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF831</th>\n",
       "      <td>ZNF831</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999564</td>\n",
       "      <td>17108.0</td>\n",
       "      <td>1.235556</td>\n",
       "      <td>0.767020</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRF4</th>\n",
       "      <td>ZNRF4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.996050</td>\n",
       "      <td>11858.0</td>\n",
       "      <td>1.015976</td>\n",
       "      <td>0.812906</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZP1</th>\n",
       "      <td>ZP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998813</td>\n",
       "      <td>15717.0</td>\n",
       "      <td>1.083469</td>\n",
       "      <td>0.884521</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZP4</th>\n",
       "      <td>ZP4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.986855</td>\n",
       "      <td>17122.5</td>\n",
       "      <td>1.432252</td>\n",
       "      <td>1.143826</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZPBP</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999110</td>\n",
       "      <td>16977.0</td>\n",
       "      <td>1.139833</td>\n",
       "      <td>1.094336</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZPBP2</th>\n",
       "      <td>ZPBP2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.998866</td>\n",
       "      <td>17241.0</td>\n",
       "      <td>1.294443</td>\n",
       "      <td>1.042031</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN1</th>\n",
       "      <td>ZSCAN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999282</td>\n",
       "      <td>6414.0</td>\n",
       "      <td>0.682119</td>\n",
       "      <td>0.693831</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN4</th>\n",
       "      <td>ZSCAN4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999346</td>\n",
       "      <td>16231.0</td>\n",
       "      <td>1.153264</td>\n",
       "      <td>0.694575</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM2</th>\n",
       "      <td>ZSWIM2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.999103</td>\n",
       "      <td>9345.5</td>\n",
       "      <td>0.680654</td>\n",
       "      <td>0.448030</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1782 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "KRT5              KRT5                 NaN      14648.0   \n",
       "AADAC            AADAC                 NaN          NaN   \n",
       "AADACL2        AADACL2                 NaN          NaN   \n",
       "AADACP1        AADACP1                 NaN          NaN   \n",
       "ABCC13          ABCC13                 NaN          NaN   \n",
       "ABCC6P1        ABCC6P1                 NaN          NaN   \n",
       "ACKR1            ACKR1                 NaN          NaN   \n",
       "ACOD1            ACOD1                 NaN          NaN   \n",
       "ACOT12          ACOT12                 NaN          NaN   \n",
       "ACSM1            ACSM1                 NaN          NaN   \n",
       "ACTL7A          ACTL7A                 NaN          NaN   \n",
       "ACTL9            ACTL9                 NaN          NaN   \n",
       "ACTRT2          ACTRT2                 NaN          NaN   \n",
       "ADAD1            ADAD1                 NaN          NaN   \n",
       "ADAD2            ADAD2                 NaN          NaN   \n",
       "ADAM18          ADAM18                 NaN          NaN   \n",
       "ADAM2            ADAM2                 NaN          NaN   \n",
       "ADAM5            ADAM5                 NaN          NaN   \n",
       "ADAM6            ADAM6                 NaN          NaN   \n",
       "ADAMDEC1      ADAMDEC1                 NaN          NaN   \n",
       "ADAMTS16      ADAMTS16                 NaN          NaN   \n",
       "ADAMTS20      ADAMTS20                 NaN          NaN   \n",
       "ADARB2          ADARB2                 NaN          NaN   \n",
       "ADARB2.AS1  ADARB2.AS1                 NaN          NaN   \n",
       "ADCYAP1        ADCYAP1                 NaN          NaN   \n",
       "ADIG              ADIG                 NaN          NaN   \n",
       "ADORA3          ADORA3                 NaN          NaN   \n",
       "ADRB3            ADRB3                 NaN          NaN   \n",
       "AGAP11          AGAP11                 NaN          NaN   \n",
       "AGTR2            AGTR2                 NaN          NaN   \n",
       "...                ...                 ...          ...   \n",
       "ZAR1              ZAR1                 NaN          NaN   \n",
       "ZBTB20.AS1  ZBTB20.AS1                 NaN          NaN   \n",
       "ZBTB46.AS1  ZBTB46.AS1                 NaN          NaN   \n",
       "ZC2HC1B        ZC2HC1B                 NaN          NaN   \n",
       "ZCCHC13        ZCCHC13                 NaN          NaN   \n",
       "ZFY                ZFY                 NaN          NaN   \n",
       "ZIC1              ZIC1                 NaN          NaN   \n",
       "ZIC3              ZIC3                 NaN          NaN   \n",
       "ZIC4              ZIC4                 NaN          NaN   \n",
       "ZIM2              ZIM2                 NaN          NaN   \n",
       "ZIM3              ZIM3                 NaN          NaN   \n",
       "ZMAT4            ZMAT4                 NaN          NaN   \n",
       "ZNF135          ZNF135                 NaN          NaN   \n",
       "ZNF229          ZNF229                 NaN          NaN   \n",
       "ZNF32.AS3    ZNF32.AS3                 NaN          NaN   \n",
       "ZNF418          ZNF418                 NaN          NaN   \n",
       "ZNF547          ZNF547                 NaN          NaN   \n",
       "ZNF560          ZNF560                 NaN          NaN   \n",
       "ZNF577          ZNF577                 NaN          NaN   \n",
       "ZNF649.AS1  ZNF649.AS1                 NaN          NaN   \n",
       "ZNF683          ZNF683                 NaN          NaN   \n",
       "ZNF831          ZNF831                 NaN          NaN   \n",
       "ZNRF4            ZNRF4                 NaN          NaN   \n",
       "ZP1                ZP1                 NaN          NaN   \n",
       "ZP4                ZP4                 NaN          NaN   \n",
       "ZPBP              ZPBP                 NaN          NaN   \n",
       "ZPBP2            ZPBP2                 NaN          NaN   \n",
       "ZSCAN1          ZSCAN1                 NaN          NaN   \n",
       "ZSCAN4          ZSCAN4                 NaN          NaN   \n",
       "ZSWIM2          ZSWIM2                 NaN          NaN   \n",
       "\n",
       "            abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "KRT5                          1.605915               1.605915   \n",
       "AADAC                              NaN                    NaN   \n",
       "AADACL2                            NaN                    NaN   \n",
       "AADACP1                            NaN                    NaN   \n",
       "ABCC13                             NaN                    NaN   \n",
       "ABCC6P1                            NaN                    NaN   \n",
       "ACKR1                              NaN                    NaN   \n",
       "ACOD1                              NaN                    NaN   \n",
       "ACOT12                             NaN                    NaN   \n",
       "ACSM1                              NaN                    NaN   \n",
       "ACTL7A                             NaN                    NaN   \n",
       "ACTL9                              NaN                    NaN   \n",
       "ACTRT2                             NaN                    NaN   \n",
       "ADAD1                              NaN                    NaN   \n",
       "ADAD2                              NaN                    NaN   \n",
       "ADAM18                             NaN                    NaN   \n",
       "ADAM2                              NaN                    NaN   \n",
       "ADAM5                              NaN                    NaN   \n",
       "ADAM6                              NaN                    NaN   \n",
       "ADAMDEC1                           NaN                    NaN   \n",
       "ADAMTS16                           NaN                    NaN   \n",
       "ADAMTS20                           NaN                    NaN   \n",
       "ADARB2                             NaN                    NaN   \n",
       "ADARB2.AS1                         NaN                    NaN   \n",
       "ADCYAP1                            NaN                    NaN   \n",
       "ADIG                               NaN                    NaN   \n",
       "ADORA3                             NaN                    NaN   \n",
       "ADRB3                              NaN                    NaN   \n",
       "AGAP11                             NaN                    NaN   \n",
       "AGTR2                              NaN                    NaN   \n",
       "...                                ...                    ...   \n",
       "ZAR1                               NaN                    NaN   \n",
       "ZBTB20.AS1                         NaN                    NaN   \n",
       "ZBTB46.AS1                         NaN                    NaN   \n",
       "ZC2HC1B                            NaN                    NaN   \n",
       "ZCCHC13                            NaN                    NaN   \n",
       "ZFY                                NaN                    NaN   \n",
       "ZIC1                               NaN                    NaN   \n",
       "ZIC3                               NaN                    NaN   \n",
       "ZIC4                               NaN                    NaN   \n",
       "ZIM2                               NaN                    NaN   \n",
       "ZIM3                               NaN                    NaN   \n",
       "ZMAT4                              NaN                    NaN   \n",
       "ZNF135                             NaN                    NaN   \n",
       "ZNF229                             NaN                    NaN   \n",
       "ZNF32.AS3                          NaN                    NaN   \n",
       "ZNF418                             NaN                    NaN   \n",
       "ZNF547                             NaN                    NaN   \n",
       "ZNF560                             NaN                    NaN   \n",
       "ZNF577                             NaN                    NaN   \n",
       "ZNF649.AS1                         NaN                    NaN   \n",
       "ZNF683                             NaN                    NaN   \n",
       "ZNF831                             NaN                    NaN   \n",
       "ZNRF4                              NaN                    NaN   \n",
       "ZP1                                NaN                    NaN   \n",
       "ZP4                                NaN                    NaN   \n",
       "ZPBP                               NaN                    NaN   \n",
       "ZPBP2                              NaN                    NaN   \n",
       "ZSCAN1                             NaN                    NaN   \n",
       "ZSCAN4                             NaN                    NaN   \n",
       "ZSWIM2                             NaN                    NaN   \n",
       "\n",
       "            Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "KRT5                              0.999564           16329.0   \n",
       "AADAC                             0.999333           15738.0   \n",
       "AADACL2                           0.999564           10048.0   \n",
       "AADACP1                           0.999564            9363.0   \n",
       "ABCC13                            0.999103           13770.5   \n",
       "ABCC6P1                           0.999333           11669.0   \n",
       "ACKR1                             0.999590           15420.0   \n",
       "ACOD1                             0.999564           17534.0   \n",
       "ACOT12                            0.999103           13073.0   \n",
       "ACSM1                             0.999103           14947.0   \n",
       "ACTL7A                            0.998630           14333.0   \n",
       "ACTL9                             0.998630           16108.0   \n",
       "ACTRT2                            0.992743           10599.0   \n",
       "ADAD1                             0.999218           11347.0   \n",
       "ADAD2                             0.999103           11564.0   \n",
       "ADAM18                            0.999050            4349.0   \n",
       "ADAM2                             0.999103           11995.0   \n",
       "ADAM5                             0.999346           10837.0   \n",
       "ADAM6                             0.998076            4393.0   \n",
       "ADAMDEC1                          0.999333           17035.0   \n",
       "ADAMTS16                          0.998997           13457.0   \n",
       "ADAMTS20                          0.998997           10231.0   \n",
       "ADARB2                            0.999333           15739.0   \n",
       "ADARB2.AS1                        0.998630            4513.0   \n",
       "ADCYAP1                           0.999333           16470.0   \n",
       "ADIG                              0.998630            7549.5   \n",
       "ADORA3                            0.999103           14396.0   \n",
       "ADRB3                             0.999050           12815.0   \n",
       "AGAP11                            0.998997            5801.0   \n",
       "AGTR2                             0.999564           13626.0   \n",
       "...                                    ...               ...   \n",
       "ZAR1                              0.998997            8977.0   \n",
       "ZBTB20.AS1                        0.999103           17659.0   \n",
       "ZBTB46.AS1                        0.998997            9189.0   \n",
       "ZC2HC1B                           0.999050            9610.0   \n",
       "ZCCHC13                           0.992743            6275.0   \n",
       "ZFY                               0.999050            8213.0   \n",
       "ZIC1                              0.999103           16957.0   \n",
       "ZIC3                              0.998630           14457.0   \n",
       "ZIC4                              0.999103           16545.0   \n",
       "ZIM2                              0.998963           13176.0   \n",
       "ZIM3                              0.998630           11365.0   \n",
       "ZMAT4                             0.999333           15477.0   \n",
       "ZNF135                            0.999103            7662.0   \n",
       "ZNF229                            0.998963            6258.0   \n",
       "ZNF32.AS3                         0.992743           17576.0   \n",
       "ZNF418                            0.998997            8202.0   \n",
       "ZNF547                            0.998997            1519.0   \n",
       "ZNF560                            0.999103           13334.0   \n",
       "ZNF577                            0.998929            8633.0   \n",
       "ZNF649.AS1                        0.998997            8148.0   \n",
       "ZNF683                            0.999577           17648.0   \n",
       "ZNF831                            0.999564           17108.0   \n",
       "ZNRF4                             0.996050           11858.0   \n",
       "ZP1                               0.998813           15717.0   \n",
       "ZP4                               0.986855           17122.5   \n",
       "ZPBP                              0.999110           16977.0   \n",
       "ZPBP2                             0.998866           17241.0   \n",
       "ZSCAN1                            0.999282            6414.0   \n",
       "ZSCAN4                            0.999346           16231.0   \n",
       "ZSWIM2                            0.999103            9345.5   \n",
       "\n",
       "            Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "KRT5                                    1.166528                   0.796012   \n",
       "AADAC                                   1.136864                   0.817906   \n",
       "AADACL2                                 0.823911                   0.891710   \n",
       "AADACP1                                 0.864613                   0.822658   \n",
       "ABCC13                                  0.872969                   0.611306   \n",
       "ABCC6P1                                 0.963358                   0.840906   \n",
       "ACKR1                                   1.226246                   0.905918   \n",
       "ACOD1                                   1.569318                   0.951284   \n",
       "ACOT12                                  0.840446                   0.846915   \n",
       "ACSM1                                   1.081291                   0.953081   \n",
       "ACTL7A                                  1.095669                   1.034700   \n",
       "ACTL9                                   1.326044                   1.017599   \n",
       "ACTRT2                                  0.984813                   1.110716   \n",
       "ADAD1                                   1.064397                   0.986995   \n",
       "ADAD2                                   0.816135                   0.586090   \n",
       "ADAM18                                  0.741704                   0.909907   \n",
       "ADAM2                                   0.928309                   0.783897   \n",
       "ADAM5                                   0.859152                   0.725460   \n",
       "ADAM6                                   0.711498                   0.734518   \n",
       "ADAMDEC1                                1.319993                   0.932814   \n",
       "ADAMTS16                                0.884074                   0.945483   \n",
       "ADAMTS20                                0.832813                   0.848892   \n",
       "ADARB2                                  0.906310                   0.663126   \n",
       "ADARB2.AS1                              0.686892                   1.013882   \n",
       "ADCYAP1                                 1.095572                   0.768004   \n",
       "ADIG                                    0.950571                   1.089967   \n",
       "ADORA3                                  0.968147                   0.728322   \n",
       "ADRB3                                   0.942848                   0.826491   \n",
       "AGAP11                                  0.630398                   0.803771   \n",
       "AGTR2                                   1.005858                   0.809055   \n",
       "...                                          ...                        ...   \n",
       "ZAR1                                    0.669929                   0.381470   \n",
       "ZBTB20.AS1                              1.420848                   0.888232   \n",
       "ZBTB46.AS1                              0.821911                   0.888074   \n",
       "ZC2HC1B                                 0.902370                   0.943740   \n",
       "ZCCHC13                                 0.683683                   0.519412   \n",
       "ZFY                                     0.633449                   0.580128   \n",
       "ZIC1                                    1.176521                   0.619447   \n",
       "ZIC3                                    0.976478                   0.715314   \n",
       "ZIC4                                    1.060821                   0.685805   \n",
       "ZIM2                                    1.011619                   0.586075   \n",
       "ZIM3                                    1.113619                   1.120134   \n",
       "ZMAT4                                   0.822516                   0.507937   \n",
       "ZNF135                                  0.659440                   0.843000   \n",
       "ZNF229                                  0.620232                   0.523254   \n",
       "ZNF32.AS3                               1.404103                   0.779750   \n",
       "ZNF418                                  0.851773                   1.041167   \n",
       "ZNF547                                  0.463075                   0.534501   \n",
       "ZNF560                                  0.901327                   0.771061   \n",
       "ZNF577                                  0.921354                   1.094319   \n",
       "ZNF649.AS1                              0.795041                   0.950439   \n",
       "ZNF683                                  1.480176                   0.845569   \n",
       "ZNF831                                  1.235556                   0.767020   \n",
       "ZNRF4                                   1.015976                   0.812906   \n",
       "ZP1                                     1.083469                   0.884521   \n",
       "ZP4                                     1.432252                   1.143826   \n",
       "ZPBP                                    1.139833                   1.094336   \n",
       "ZPBP2                                   1.294443                   1.042031   \n",
       "ZSCAN1                                  0.682119                   0.693831   \n",
       "ZSCAN4                                  1.153264                   0.694575   \n",
       "ZSWIM2                                  0.680654                   0.448030   \n",
       "\n",
       "            Number of experiments (simulated)   Z score  \n",
       "KRT5                                       20  0.551985  \n",
       "AADAC                                      23       NaN  \n",
       "AADACL2                                    18       NaN  \n",
       "AADACP1                                    23       NaN  \n",
       "ABCC13                                     21       NaN  \n",
       "ABCC6P1                                    23       NaN  \n",
       "ACKR1                                      24       NaN  \n",
       "ACOD1                                      14       NaN  \n",
       "ACOT12                                     20       NaN  \n",
       "ACSM1                                      23       NaN  \n",
       "ACTL7A                                     14       NaN  \n",
       "ACTL9                                      13       NaN  \n",
       "ACTRT2                                     10       NaN  \n",
       "ADAD1                                      11       NaN  \n",
       "ADAD2                                      21       NaN  \n",
       "ADAM18                                     16       NaN  \n",
       "ADAM2                                      15       NaN  \n",
       "ADAM5                                      13       NaN  \n",
       "ADAM6                                      10       NaN  \n",
       "ADAMDEC1                                   19       NaN  \n",
       "ADAMTS16                                   24       NaN  \n",
       "ADAMTS20                                   21       NaN  \n",
       "ADARB2                                     24       NaN  \n",
       "ADARB2.AS1                                 13       NaN  \n",
       "ADCYAP1                                    21       NaN  \n",
       "ADIG                                       13       NaN  \n",
       "ADORA3                                     24       NaN  \n",
       "ADRB3                                      22       NaN  \n",
       "AGAP11                                     25       NaN  \n",
       "AGTR2                                      19       NaN  \n",
       "...                                       ...       ...  \n",
       "ZAR1                                       13       NaN  \n",
       "ZBTB20.AS1                                 23       NaN  \n",
       "ZBTB46.AS1                                 15       NaN  \n",
       "ZC2HC1B                                    16       NaN  \n",
       "ZCCHC13                                    10       NaN  \n",
       "ZFY                                        24       NaN  \n",
       "ZIC1                                       23       NaN  \n",
       "ZIC3                                       18       NaN  \n",
       "ZIC4                                       23       NaN  \n",
       "ZIM2                                       22       NaN  \n",
       "ZIM3                                       11       NaN  \n",
       "ZMAT4                                      21       NaN  \n",
       "ZNF135                                     25       NaN  \n",
       "ZNF229                                     25       NaN  \n",
       "ZNF32.AS3                                  10       NaN  \n",
       "ZNF418                                     25       NaN  \n",
       "ZNF547                                     25       NaN  \n",
       "ZNF560                                     19       NaN  \n",
       "ZNF577                                     25       NaN  \n",
       "ZNF649.AS1                                 25       NaN  \n",
       "ZNF683                                     20       NaN  \n",
       "ZNF831                                     20       NaN  \n",
       "ZNRF4                                      12       NaN  \n",
       "ZP1                                        19       NaN  \n",
       "ZP4                                        11       NaN  \n",
       "ZPBP                                       13       NaN  \n",
       "ZPBP2                                      13       NaN  \n",
       "ZSCAN1                                     23       NaN  \n",
       "ZSCAN4                                     13       NaN  \n",
       "ZSWIM2                                     15       NaN  \n",
       "\n",
       "[1782 rows x 11 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks[summary_gene_ranks.isna().any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "summary_gene_ranks.to_csv(gene_summary_filename, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
