{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify generic genes and pathways\n",
    "\n",
    "This notebook is meant to identify common differentially expressed genes in the PAO1 and PA14 compendia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "import scipy.stats as ss\n",
    "from keras.models import load_model\n",
    "from rpy2.robjects import pandas2ri\n",
    "from ponyo import utils, simulate_expression_data\n",
    "from generic_expression_patterns_modules import process, stats, ranking\n",
    "\n",
    "pandas2ri.activate()\n",
    "\n",
    "np.random.seed(123)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = os.path.abspath(\n",
    "    os.path.join(base_dir, \"configs\", \"config_pseudomonas_pa14_rnaseq.tsv\")\n",
    ")\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load params\n",
    "local_dir = params[\"local_dir\"]\n",
    "dataset_name = params[\"dataset_name\"]\n",
    "NN_architecture = params[\"NN_architecture\"]\n",
    "num_runs = params[\"num_simulated\"]\n",
    "project_id = params[\"project_id\"]\n",
    "metadata_col_id = params[\"metadata_colname\"]\n",
    "raw_template_filename = params[\"raw_template_filename\"]\n",
    "processed_template_filename = params[\"processed_template_filename\"]\n",
    "normalized_compendium_filename = params[\"normalized_compendium_filename\"]\n",
    "scaler_filename = params[\"scaler_filename\"]\n",
    "col_to_rank_genes = params[\"rank_genes_by\"]\n",
    "logFC_name = params[\"DE_logFC_name\"]\n",
    "pvalue_name = params[\"DE_pvalue_name\"]\n",
    "latent_dim = params[\"latent_dim\"]\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "sample_id_metadata_filename = os.path.join(\n",
    "    base_dir, dataset_name, \"data\", \"metadata\", f\"{project_id}_process_samples.tsv\"\n",
    ")\n",
    "\n",
    "# Load metadata file with grouping assignments for samples\n",
    "metadata_filename = os.path.join(\n",
    "    base_dir, dataset_name, \"data\", \"metadata\", f\"{project_id}_groups.tsv\"\n",
    ")\n",
    "\n",
    "# Load pickled file\n",
    "scaler = pickle.load(open(scaler_filename, \"rb\"))\n",
    "\n",
    "# Percentile threshold to identify generic genes\n",
    "percentile_threshold = 80.0\n",
    "\n",
    "metadata_simulate_filename = \"data/metadata/SraRunTable.csv\"\n",
    "metadata_delimiter = \",\"\n",
    "experiment_id_colname = \"SRA_study\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output files\n",
    "gene_summary_filename = os.path.join(\n",
    "    base_dir, dataset_name, f\"generic_gene_summary_{project_id}.tsv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Need to customize code from ponyo\n",
    "\n",
    "The current simulation-related function in ponyo, `get_sample_ids` assumes that the user is using one of two different metadata files (one associated with the pseudomonas compendium and another associated with recount2). The compendium dataset we are using here has a slightly different format for their metadata file.\n",
    "\n",
    "Here we are temporarily writing our own function customized for this Pa RNA-seq compendia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sample_ids(\n",
    "    metadata_filename, delimiter, experiment_colname, experiment_id, sample_id_colname\n",
    "):\n",
    "    \"\"\"\n",
    "    Returns sample ids (found in gene expression df) associated with\n",
    "    a given list of experiment ids (found in the metadata)\n",
    "\n",
    "    Arguments\n",
    "    ----------\n",
    "    metadata_filename: str\n",
    "        Metadata file path. An example metadata file can be found\n",
    "        here: https://github.com/greenelab/ponyo/blob/master/human_tests/data/metadata/recount2_metadata.tsv\n",
    "\n",
    "    delimiter: str\n",
    "        Delimiter for metadata file\n",
    "\n",
    "    experiment_colname: str\n",
    "        Column header that contains the experiment ids\n",
    "\n",
    "    experiment_id: str\n",
    "        Experiment id selected to retrieve sample ids for\n",
    "\n",
    "    sample_id_colname: str\n",
    "        Column header that contains sample id that maps expression data\n",
    "        and metadata\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Read in metadata\n",
    "    metadata = pd.read_csv(metadata_filename, header=0, sep=delimiter, index_col=None)\n",
    "\n",
    "    # Set index column to experiment id column\n",
    "    metadata.set_index(experiment_colname, inplace=True)\n",
    "\n",
    "    # Select samples associated with experiment id\n",
    "    selected_metadata = metadata.loc[experiment_id]\n",
    "    sample_ids = list(selected_metadata[sample_id_colname])\n",
    "\n",
    "    return sample_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shift_template_experiment_custom(\n",
    "    normalized_data_filename,\n",
    "    NN_architecture,\n",
    "    latent_dim,\n",
    "    dataset_name,\n",
    "    scaler,\n",
    "    metadata_filename,\n",
    "    metadata_delimiter,\n",
    "    experiment_id_colname,\n",
    "    sample_id_colname,\n",
    "    selected_experiment_id,\n",
    "    local_dir,\n",
    "    base_dir,\n",
    "    num_runs,\n",
    "):\n",
    "    \"\"\"\n",
    "    Generate new simulated experiment using the selected_experiment_id as a template\n",
    "    experiment using the same workflow as `simulate_by_latent_transform`\n",
    "\n",
    "    This will return a file with a single simulated experiment following the workflow mentioned.\n",
    "    This function can be run multiple times to generate multiple simulated experiments from a\n",
    "    single selected_experiment_id.\n",
    "\n",
    "    Arguments\n",
    "    ----------\n",
    "    normalized_data_filename: str\n",
    "        File containing normalized gene expression data\n",
    "\n",
    "        ------------------------------| PA0001 | PA0002 |...\n",
    "        05_PA14000-4-2_5-10-07_S2.CEL | 0.8533 | 0.7252 |...\n",
    "        54375-4-05.CEL                | 0.7789 | 0.7678 |...\n",
    "        ...                           | ...    | ...    |...\n",
    "\n",
    "    NN_architecture: str\n",
    "        Name of neural network architecture to use.\n",
    "        Format 'NN_<intermediate layer>_<latent layer>'\n",
    "\n",
    "    latent_dim: int\n",
    "        The number of dimensions in the latent space\n",
    "\n",
    "    dataset_name: str\n",
    "        Name for analysis directory. Either \"Human\" or \"Pseudomonas\"\n",
    "\n",
    "    scaler: minmax model\n",
    "        Model used to transform data into a different range\n",
    "\n",
    "    metadata_filename: str\n",
    "        Metadata file path. Note: The format of this metadata file\n",
    "        requires the index column to contain experiment ids.\n",
    "\n",
    "    metadata_delimiter: str\n",
    "        Delimiter for metadata file\n",
    "\n",
    "    experiment_colname: str\n",
    "        Column header that contains the experiment ids\n",
    "\n",
    "    sample_id_colname: str\n",
    "        Column header that contains sample id that maps expression data\n",
    "        and metadata\n",
    "\n",
    "    selected_experiment_id: str\n",
    "        Experiment id selected as template\n",
    "\n",
    "    local_dir: str\n",
    "        Parent directory on local machine to store intermediate results\n",
    "\n",
    "    base_dir: str\n",
    "        Root directory containing analysis subdirectories\n",
    "\n",
    "    num_runs: int\n",
    "        Number of experiments to simulate\n",
    "\n",
    "    Returns\n",
    "    --------\n",
    "    simulated_data_filename: str\n",
    "        File containing simulated gene expression data\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Files\n",
    "    NN_dir = os.path.join(base_dir, dataset_name, \"models\", NN_architecture)\n",
    "\n",
    "    model_encoder_filename = glob.glob(os.path.join(NN_dir, \"*_encoder_model.h5\"))[0]\n",
    "\n",
    "    weights_encoder_filename = glob.glob(os.path.join(NN_dir, \"*_encoder_weights.h5\"))[\n",
    "        0\n",
    "    ]\n",
    "\n",
    "    model_decoder_filename = glob.glob(os.path.join(NN_dir, \"*_decoder_model.h5\"))[0]\n",
    "\n",
    "    weights_decoder_filename = glob.glob(os.path.join(NN_dir, \"*_decoder_weights.h5\"))[\n",
    "        0\n",
    "    ]\n",
    "\n",
    "    # Load saved models\n",
    "    loaded_model = load_model(model_encoder_filename, compile=False)\n",
    "    loaded_decode_model = load_model(model_decoder_filename, compile=False)\n",
    "\n",
    "    loaded_model.load_weights(weights_encoder_filename)\n",
    "    loaded_decode_model.load_weights(weights_decoder_filename)\n",
    "\n",
    "    # Read data\n",
    "    normalized_data = pd.read_csv(\n",
    "        normalized_data_filename, header=0, sep=\"\\t\", index_col=0\n",
    "    )\n",
    "\n",
    "    # Get corresponding sample ids\n",
    "    sample_ids = get_sample_ids(\n",
    "        metadata_filename,\n",
    "        metadata_delimiter,\n",
    "        experiment_id_colname,\n",
    "        selected_experiment_id,\n",
    "        sample_id_colname,\n",
    "    )\n",
    "\n",
    "    # Gene expression data for selected samples\n",
    "    selected_data_df = normalized_data.loc[sample_ids]\n",
    "\n",
    "    for run in range(num_runs):\n",
    "        simulated_data_decoded_df, simulated_data_encoded_df = run_shift_template(\n",
    "            loaded_model,\n",
    "            loaded_decode_model,\n",
    "            normalized_data,\n",
    "            selected_data_df,\n",
    "            latent_dim,\n",
    "        )\n",
    "\n",
    "        # Un-normalize the data in order to run DE analysis downstream\n",
    "        simulated_data_scaled = scaler.inverse_transform(simulated_data_decoded_df)\n",
    "\n",
    "        simulated_data_scaled_df = pd.DataFrame(\n",
    "            simulated_data_scaled,\n",
    "            columns=simulated_data_decoded_df.columns,\n",
    "            index=simulated_data_decoded_df.index,\n",
    "        )\n",
    "\n",
    "        # Save\n",
    "        out_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            \"selected_simulated_data_\"\n",
    "            + selected_experiment_id\n",
    "            + \"_\"\n",
    "            + str(run)\n",
    "            + \".txt\",\n",
    "        )\n",
    "        simulated_data_scaled_df.to_csv(out_filename, float_format=\"%.3f\", sep=\"\\t\")\n",
    "\n",
    "        out_encoded_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_encoded_data_{selected_experiment_id}_{run}.txt\",\n",
    "        )\n",
    "\n",
    "        simulated_data_encoded_df.to_csv(\n",
    "            out_encoded_filename, float_format=\"%.3f\", sep=\"\\t\"\n",
    "        )\n",
    "\n",
    "    # Save template data for visualization validation\n",
    "    test_filename = os.path.join(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment\",\n",
    "        \"template_normalized_data_\" + selected_experiment_id + \"_test.txt\",\n",
    "    )\n",
    "    selected_data_df.to_csv(test_filename, float_format=\"%.3f\", sep=\"\\t\")\n",
    "\n",
    "\n",
    "def run_shift_template(encoder, decoder, normalized_data, selected_data_df, latent_dim):\n",
    "    \"\"\"\n",
    "    This function does the template shifting used in `shift_template_experiment`.\n",
    "\n",
    "    Arguments\n",
    "    ---------\n",
    "    encoder: keras.models.Model\n",
    "        The encoder half of the VAE. `encoder` takes in a (samples x genes) dataframe of\n",
    "        gene expression data and encodes it into a latent space\n",
    "\n",
    "    decoder: keras.models.Model\n",
    "        The decoder half of the VAE. `decoder` takes a dataframe of means and standard deviations\n",
    "        and uses them to simulate gene expression data close to the distribution of normalized_data\n",
    "\n",
    "    normalized_data: pd.DataFrame\n",
    "        The data to be used to train the VAE\n",
    "\n",
    "    selected_data_df: pd.DataFrame\n",
    "        The samples to be shifted in the latent space\n",
    "\n",
    "    latent_dim: int\n",
    "        The dimension of the latent space the samples will be shifted in\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    simulated_data_decoded_df: pd.DataFrame\n",
    "        The simulated data created by shifting the samples in the latent space\n",
    "\n",
    "    simulated_data_encoded_df: pd.DataFrame\n",
    "        The latent means and standard deviations in the latent space used to simulate the data\n",
    "    \"\"\"\n",
    "    # Encode selected experiment into latent space\n",
    "    data_encoded = encoder.predict_on_batch(selected_data_df)\n",
    "    data_encoded_df = pd.DataFrame(data_encoded, index=selected_data_df.index)\n",
    "\n",
    "    # Get centroid of original data\n",
    "    centroid = data_encoded_df.mean(axis=0)\n",
    "\n",
    "    # Add individual vectors(centroid, sample point) to new_centroid\n",
    "\n",
    "    # Encode original gene expression data into latent space\n",
    "    data_encoded_all = encoder.predict_on_batch(normalized_data)\n",
    "    data_encoded_all_df = pd.DataFrame(data_encoded_all, index=normalized_data.index)\n",
    "\n",
    "    data_encoded_all_df.head()\n",
    "\n",
    "    # Find a new location in the latent space by sampling from the latent space\n",
    "    encoded_means = data_encoded_all_df.mean(axis=0)\n",
    "    encoded_stds = data_encoded_all_df.std(axis=0)\n",
    "\n",
    "    latent_dim = int(latent_dim)\n",
    "    new_centroid = np.zeros(latent_dim)\n",
    "\n",
    "    for j in range(latent_dim):\n",
    "        new_centroid[j] = np.random.normal(encoded_means[j], encoded_stds[j])\n",
    "\n",
    "    shift_vec_df = new_centroid - centroid\n",
    "\n",
    "    simulated_data_encoded_df = data_encoded_df.apply(\n",
    "        lambda x: x + shift_vec_df, axis=1\n",
    "    )\n",
    "\n",
    "    # Decode simulated data into raw gene space\n",
    "    simulated_data_decoded = decoder.predict_on_batch(simulated_data_encoded_df)\n",
    "\n",
    "    simulated_data_decoded_df = pd.DataFrame(\n",
    "        simulated_data_decoded,\n",
    "        index=simulated_data_encoded_df.index,\n",
    "        columns=selected_data_df.columns,\n",
    "    )\n",
    "\n",
    "    return simulated_data_decoded_df, simulated_data_encoded_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulate experiments using selected template experiment\n",
    "Workflow:\n",
    "\n",
    "1. Get the gene expression data for the selected template experiment\n",
    "2. Encode this experiment into a latent space using the trained VAE model\n",
    "3. Linearly shift the encoded template experiment in the latent space\n",
    "4. Decode the samples. This results in a new experiment\n",
    "5. Repeat steps 1-4 to get multiple simulated experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate multiple experiments\n",
    "# This step creates the following files in \"<local_dir>/pseudo_experiment/\" directory:\n",
    "#   - selected_simulated_data_SRP012656_<n>.txt\n",
    "#   - selected_simulated_encoded_data_SRP012656_<n>.txt\n",
    "#   - template_normalized_data_SRP012656_test.txt\n",
    "# in which \"<n>\" is an integer in the range of [0, num_runs-1]\n",
    "os.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)\n",
    "shift_template_experiment_custom(\n",
    "    normalized_compendium_filename,\n",
    "    NN_architecture,\n",
    "    latent_dim,\n",
    "    dataset_name,\n",
    "    scaler,\n",
    "    metadata_simulate_filename,\n",
    "    metadata_delimiter,\n",
    "    experiment_id_colname,\n",
    "    metadata_col_id,\n",
    "    project_id,\n",
    "    local_dir,\n",
    "    base_dir,\n",
    "    num_runs,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "simulated_filename = os.path.join(\n",
    "    local_dir, \"pseudo_experiment\", f\"selected_simulated_data_{project_id}_1.txt\"\n",
    ")\n",
    "\n",
    "test = pd.read_csv(simulated_filename, sep=\"\\t\", index_col=0, header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PA14_55610</th>\n",
       "      <th>PA14_55600</th>\n",
       "      <th>PA14_55590</th>\n",
       "      <th>PA14_55580</th>\n",
       "      <th>PA14_55570</th>\n",
       "      <th>PA14_55560</th>\n",
       "      <th>PA14_55550</th>\n",
       "      <th>PA14_55540</th>\n",
       "      <th>PA14_55530</th>\n",
       "      <th>PA14_55520</th>\n",
       "      <th>...</th>\n",
       "      <th>PA14_19205</th>\n",
       "      <th>PA14_17675</th>\n",
       "      <th>PA14_67975</th>\n",
       "      <th>PA14_36345</th>\n",
       "      <th>PA14_43405</th>\n",
       "      <th>PA14_38825</th>\n",
       "      <th>PA14_24245</th>\n",
       "      <th>PA14_28895</th>\n",
       "      <th>PA14_55117</th>\n",
       "      <th>PA14_59845</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRX1740204</th>\n",
       "      <td>111.777</td>\n",
       "      <td>50.537</td>\n",
       "      <td>10.080</td>\n",
       "      <td>19.057</td>\n",
       "      <td>21.094</td>\n",
       "      <td>19.976</td>\n",
       "      <td>17.336</td>\n",
       "      <td>18.764</td>\n",
       "      <td>16.352</td>\n",
       "      <td>4.021</td>\n",
       "      <td>...</td>\n",
       "      <td>44.504</td>\n",
       "      <td>162.106</td>\n",
       "      <td>633.776</td>\n",
       "      <td>144.577</td>\n",
       "      <td>0.171</td>\n",
       "      <td>29.336</td>\n",
       "      <td>137.091</td>\n",
       "      <td>683.270</td>\n",
       "      <td>155.982</td>\n",
       "      <td>209.853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740205</th>\n",
       "      <td>103.666</td>\n",
       "      <td>51.088</td>\n",
       "      <td>10.955</td>\n",
       "      <td>26.069</td>\n",
       "      <td>16.420</td>\n",
       "      <td>16.012</td>\n",
       "      <td>12.438</td>\n",
       "      <td>15.369</td>\n",
       "      <td>10.092</td>\n",
       "      <td>2.846</td>\n",
       "      <td>...</td>\n",
       "      <td>41.855</td>\n",
       "      <td>160.967</td>\n",
       "      <td>799.159</td>\n",
       "      <td>103.454</td>\n",
       "      <td>0.135</td>\n",
       "      <td>18.419</td>\n",
       "      <td>116.171</td>\n",
       "      <td>550.327</td>\n",
       "      <td>152.655</td>\n",
       "      <td>87.486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740206</th>\n",
       "      <td>102.711</td>\n",
       "      <td>49.355</td>\n",
       "      <td>9.203</td>\n",
       "      <td>15.811</td>\n",
       "      <td>18.975</td>\n",
       "      <td>19.545</td>\n",
       "      <td>14.986</td>\n",
       "      <td>18.000</td>\n",
       "      <td>18.599</td>\n",
       "      <td>4.126</td>\n",
       "      <td>...</td>\n",
       "      <td>48.051</td>\n",
       "      <td>176.501</td>\n",
       "      <td>683.735</td>\n",
       "      <td>174.970</td>\n",
       "      <td>0.236</td>\n",
       "      <td>43.901</td>\n",
       "      <td>134.203</td>\n",
       "      <td>682.758</td>\n",
       "      <td>140.616</td>\n",
       "      <td>211.644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740207</th>\n",
       "      <td>123.578</td>\n",
       "      <td>25.647</td>\n",
       "      <td>10.035</td>\n",
       "      <td>12.350</td>\n",
       "      <td>37.381</td>\n",
       "      <td>10.266</td>\n",
       "      <td>13.395</td>\n",
       "      <td>21.082</td>\n",
       "      <td>54.470</td>\n",
       "      <td>3.969</td>\n",
       "      <td>...</td>\n",
       "      <td>37.466</td>\n",
       "      <td>64.821</td>\n",
       "      <td>314.995</td>\n",
       "      <td>250.590</td>\n",
       "      <td>0.176</td>\n",
       "      <td>38.048</td>\n",
       "      <td>202.545</td>\n",
       "      <td>432.848</td>\n",
       "      <td>146.042</td>\n",
       "      <td>801.247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740208</th>\n",
       "      <td>127.446</td>\n",
       "      <td>41.818</td>\n",
       "      <td>11.924</td>\n",
       "      <td>28.756</td>\n",
       "      <td>43.823</td>\n",
       "      <td>21.700</td>\n",
       "      <td>17.510</td>\n",
       "      <td>27.530</td>\n",
       "      <td>37.405</td>\n",
       "      <td>6.179</td>\n",
       "      <td>...</td>\n",
       "      <td>44.970</td>\n",
       "      <td>102.648</td>\n",
       "      <td>465.611</td>\n",
       "      <td>257.010</td>\n",
       "      <td>0.473</td>\n",
       "      <td>56.969</td>\n",
       "      <td>189.951</td>\n",
       "      <td>570.901</td>\n",
       "      <td>139.930</td>\n",
       "      <td>400.680</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 5891 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PA14_55610  PA14_55600  PA14_55590  PA14_55580  PA14_55570  \\\n",
       "SRX1740204     111.777      50.537      10.080      19.057      21.094   \n",
       "SRX1740205     103.666      51.088      10.955      26.069      16.420   \n",
       "SRX1740206     102.711      49.355       9.203      15.811      18.975   \n",
       "SRX1740207     123.578      25.647      10.035      12.350      37.381   \n",
       "SRX1740208     127.446      41.818      11.924      28.756      43.823   \n",
       "\n",
       "            PA14_55560  PA14_55550  PA14_55540  PA14_55530  PA14_55520  \\\n",
       "SRX1740204      19.976      17.336      18.764      16.352       4.021   \n",
       "SRX1740205      16.012      12.438      15.369      10.092       2.846   \n",
       "SRX1740206      19.545      14.986      18.000      18.599       4.126   \n",
       "SRX1740207      10.266      13.395      21.082      54.470       3.969   \n",
       "SRX1740208      21.700      17.510      27.530      37.405       6.179   \n",
       "\n",
       "               ...      PA14_19205  PA14_17675  PA14_67975  PA14_36345  \\\n",
       "SRX1740204     ...          44.504     162.106     633.776     144.577   \n",
       "SRX1740205     ...          41.855     160.967     799.159     103.454   \n",
       "SRX1740206     ...          48.051     176.501     683.735     174.970   \n",
       "SRX1740207     ...          37.466      64.821     314.995     250.590   \n",
       "SRX1740208     ...          44.970     102.648     465.611     257.010   \n",
       "\n",
       "            PA14_43405  PA14_38825  PA14_24245  PA14_28895  PA14_55117  \\\n",
       "SRX1740204       0.171      29.336     137.091     683.270     155.982   \n",
       "SRX1740205       0.135      18.419     116.171     550.327     152.655   \n",
       "SRX1740206       0.236      43.901     134.203     682.758     140.616   \n",
       "SRX1740207       0.176      38.048     202.545     432.848     146.042   \n",
       "SRX1740208       0.473      56.969     189.951     570.901     139.930   \n",
       "\n",
       "            PA14_59845  \n",
       "SRX1740204     209.853  \n",
       "SRX1740205      87.486  \n",
       "SRX1740206     211.644  \n",
       "SRX1740207     801.247  \n",
       "SRX1740208     400.680  \n",
       "\n",
       "[5 rows x 5891 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process template and simulated data\n",
    "\n",
    "* Remove samples not required for comparison.\n",
    "* Make sure ordering of samples matches metadata for proper comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(sample_id_metadata_filename):\n",
    "    sample_id_metadata_filename = None\n",
    "\n",
    "stats.process_samples_for_DESeq(\n",
    "    raw_template_filename,\n",
    "    metadata_filename,\n",
    "    processed_template_filename,\n",
    "    sample_id_metadata_filename,\n",
    ")\n",
    "\n",
    "for i in range(num_runs):\n",
    "    simulated_filename = os.path.join(\n",
    "        local_dir, \"pseudo_experiment\", f\"selected_simulated_data_{project_id}_{i}.txt\"\n",
    "    )\n",
    "    stats.process_samples_for_DESeq(\n",
    "        simulated_filename,\n",
    "        metadata_filename,\n",
    "        None,\n",
    "        sample_id_metadata_filename,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quick check\n",
    "template_data = pd.read_csv(\n",
    "    processed_template_filename, header=0, index_col=0, sep=\"\\t\"\n",
    ")\n",
    "\n",
    "assert template_data.shape[0] == 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PA14_55610</th>\n",
       "      <th>PA14_55600</th>\n",
       "      <th>PA14_55590</th>\n",
       "      <th>PA14_55580</th>\n",
       "      <th>PA14_55570</th>\n",
       "      <th>PA14_55560</th>\n",
       "      <th>PA14_55550</th>\n",
       "      <th>PA14_55540</th>\n",
       "      <th>PA14_55530</th>\n",
       "      <th>PA14_55520</th>\n",
       "      <th>...</th>\n",
       "      <th>PA14_19205</th>\n",
       "      <th>PA14_17675</th>\n",
       "      <th>PA14_67975</th>\n",
       "      <th>PA14_36345</th>\n",
       "      <th>PA14_43405</th>\n",
       "      <th>PA14_38825</th>\n",
       "      <th>PA14_24245</th>\n",
       "      <th>PA14_28895</th>\n",
       "      <th>PA14_55117</th>\n",
       "      <th>PA14_59845</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRX1740204</th>\n",
       "      <td>179</td>\n",
       "      <td>76</td>\n",
       "      <td>33</td>\n",
       "      <td>1142</td>\n",
       "      <td>356</td>\n",
       "      <td>104</td>\n",
       "      <td>31</td>\n",
       "      <td>85</td>\n",
       "      <td>75</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>48</td>\n",
       "      <td>19</td>\n",
       "      <td>261</td>\n",
       "      <td>275</td>\n",
       "      <td>0</td>\n",
       "      <td>56</td>\n",
       "      <td>135</td>\n",
       "      <td>156</td>\n",
       "      <td>63</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740205</th>\n",
       "      <td>160</td>\n",
       "      <td>65</td>\n",
       "      <td>31</td>\n",
       "      <td>1395</td>\n",
       "      <td>435</td>\n",
       "      <td>115</td>\n",
       "      <td>37</td>\n",
       "      <td>101</td>\n",
       "      <td>58</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>45</td>\n",
       "      <td>25</td>\n",
       "      <td>389</td>\n",
       "      <td>251</td>\n",
       "      <td>0</td>\n",
       "      <td>116</td>\n",
       "      <td>125</td>\n",
       "      <td>164</td>\n",
       "      <td>58</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740206</th>\n",
       "      <td>156</td>\n",
       "      <td>69</td>\n",
       "      <td>34</td>\n",
       "      <td>695</td>\n",
       "      <td>207</td>\n",
       "      <td>56</td>\n",
       "      <td>20</td>\n",
       "      <td>52</td>\n",
       "      <td>55</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>38</td>\n",
       "      <td>23</td>\n",
       "      <td>253</td>\n",
       "      <td>341</td>\n",
       "      <td>1</td>\n",
       "      <td>126</td>\n",
       "      <td>133</td>\n",
       "      <td>136</td>\n",
       "      <td>56</td>\n",
       "      <td>306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740207</th>\n",
       "      <td>230</td>\n",
       "      <td>54</td>\n",
       "      <td>20</td>\n",
       "      <td>474</td>\n",
       "      <td>174</td>\n",
       "      <td>39</td>\n",
       "      <td>15</td>\n",
       "      <td>47</td>\n",
       "      <td>74</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>24</td>\n",
       "      <td>173</td>\n",
       "      <td>483</td>\n",
       "      <td>0</td>\n",
       "      <td>113</td>\n",
       "      <td>197</td>\n",
       "      <td>102</td>\n",
       "      <td>72</td>\n",
       "      <td>390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX1740208</th>\n",
       "      <td>199</td>\n",
       "      <td>64</td>\n",
       "      <td>32</td>\n",
       "      <td>562</td>\n",
       "      <td>169</td>\n",
       "      <td>59</td>\n",
       "      <td>19</td>\n",
       "      <td>52</td>\n",
       "      <td>86</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>21</td>\n",
       "      <td>165</td>\n",
       "      <td>368</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>143</td>\n",
       "      <td>125</td>\n",
       "      <td>71</td>\n",
       "      <td>330</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 5891 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PA14_55610  PA14_55600  PA14_55590  PA14_55580  PA14_55570  \\\n",
       "SRX1740204         179          76          33        1142         356   \n",
       "SRX1740205         160          65          31        1395         435   \n",
       "SRX1740206         156          69          34         695         207   \n",
       "SRX1740207         230          54          20         474         174   \n",
       "SRX1740208         199          64          32         562         169   \n",
       "\n",
       "            PA14_55560  PA14_55550  PA14_55540  PA14_55530  PA14_55520  \\\n",
       "SRX1740204         104          31          85          75           5   \n",
       "SRX1740205         115          37         101          58           4   \n",
       "SRX1740206          56          20          52          55           4   \n",
       "SRX1740207          39          15          47          74           5   \n",
       "SRX1740208          59          19          52          86           8   \n",
       "\n",
       "               ...      PA14_19205  PA14_17675  PA14_67975  PA14_36345  \\\n",
       "SRX1740204     ...              48          19         261         275   \n",
       "SRX1740205     ...              45          25         389         251   \n",
       "SRX1740206     ...              38          23         253         341   \n",
       "SRX1740207     ...              20          24         173         483   \n",
       "SRX1740208     ...              27          21         165         368   \n",
       "\n",
       "            PA14_43405  PA14_38825  PA14_24245  PA14_28895  PA14_55117  \\\n",
       "SRX1740204           0          56         135         156          63   \n",
       "SRX1740205           0         116         125         164          58   \n",
       "SRX1740206           1         126         133         136          56   \n",
       "SRX1740207           0         113         197         102          72   \n",
       "SRX1740208           1          95         143         125          71   \n",
       "\n",
       "            PA14_59845  \n",
       "SRX1740204         261  \n",
       "SRX1740205         296  \n",
       "SRX1740206         306  \n",
       "SRX1740207         390  \n",
       "SRX1740208         330  \n",
       "\n",
       "[5 rows x 5891 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: DelayedArray\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: BiocParallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i processed_template_filename -i local_dir -i base_dir\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "get_DE_stats_DESeq(metadata_filename,\n",
    "                   project_id,\n",
    "                   processed_template_filename,\n",
    "                   \"template\",\n",
    "                   local_dir,\n",
    "                   \"real\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i metadata_filename -i project_id -i base_dir -i local_dir -i num_runs -o num_sign_DEGs_simulated\n",
    "\n",
    "source(paste0(base_dir,'/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "num_sign_DEGs_simulated <- c()\n",
    "\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\",\n",
    "        i,\n",
    "        \".txt\",\n",
    "        sep=\"\"\n",
    "    )\n",
    "\n",
    "    run_output <- get_DE_stats_DESeq(\n",
    "        metadata_filename,\n",
    "        project_id,\n",
    "        simulated_data_filename,\n",
    "        \"simulated\",\n",
    "        local_dir,\n",
    "        i\n",
    "    )\n",
    "    num_sign_DEGs_simulated <- c(num_sign_DEGs_simulated, run_output)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rank genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (2 levels on the left, 1 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    "    logFC_name,\n",
    "    pvalue_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA14_53300</th>\n",
       "      <td>PA14_53300</td>\n",
       "      <td>3.277490e-44</td>\n",
       "      <td>5890.0</td>\n",
       "      <td>3.820634</td>\n",
       "      <td>-3.820634</td>\n",
       "      <td>0.615595</td>\n",
       "      <td>4244.0</td>\n",
       "      <td>72.037351</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.249700</td>\n",
       "      <td>0.194925</td>\n",
       "      <td>25</td>\n",
       "      <td>18.319543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_21530</th>\n",
       "      <td>PA14_21530</td>\n",
       "      <td>1.617143e-28</td>\n",
       "      <td>5887.0</td>\n",
       "      <td>3.273077</td>\n",
       "      <td>-3.273077</td>\n",
       "      <td>0.757436</td>\n",
       "      <td>4379.0</td>\n",
       "      <td>74.329372</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.222476</td>\n",
       "      <td>0.172918</td>\n",
       "      <td>25</td>\n",
       "      <td>17.641922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_66480</th>\n",
       "      <td>PA14_66480</td>\n",
       "      <td>5.056304e-07</td>\n",
       "      <td>5700.0</td>\n",
       "      <td>1.299171</td>\n",
       "      <td>1.299171</td>\n",
       "      <td>0.646066</td>\n",
       "      <td>1318.0</td>\n",
       "      <td>22.359932</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.127432</td>\n",
       "      <td>0.066487</td>\n",
       "      <td>25</td>\n",
       "      <td>17.623670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_12590</th>\n",
       "      <td>PA14_12590</td>\n",
       "      <td>2.669583e-01</td>\n",
       "      <td>5791.0</td>\n",
       "      <td>1.517689</td>\n",
       "      <td>1.517689</td>\n",
       "      <td>0.949002</td>\n",
       "      <td>1069.0</td>\n",
       "      <td>18.132428</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.115920</td>\n",
       "      <td>0.080223</td>\n",
       "      <td>25</td>\n",
       "      <td>17.473402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_10500</th>\n",
       "      <td>PA14_10500</td>\n",
       "      <td>2.400867e-30</td>\n",
       "      <td>5884.0</td>\n",
       "      <td>3.173688</td>\n",
       "      <td>-3.173688</td>\n",
       "      <td>0.286293</td>\n",
       "      <td>4426.0</td>\n",
       "      <td>75.127334</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.254018</td>\n",
       "      <td>0.173930</td>\n",
       "      <td>25</td>\n",
       "      <td>16.786449</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "PA14_53300  PA14_53300        3.277490e-44       5890.0   \n",
       "PA14_21530  PA14_21530        1.617143e-28       5887.0   \n",
       "PA14_66480  PA14_66480        5.056304e-07       5700.0   \n",
       "PA14_12590  PA14_12590        2.669583e-01       5791.0   \n",
       "PA14_10500  PA14_10500        2.400867e-30       5884.0   \n",
       "\n",
       "            abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "PA14_53300                    3.820634              -3.820634   \n",
       "PA14_21530                    3.273077              -3.273077   \n",
       "PA14_66480                    1.299171               1.299171   \n",
       "PA14_12590                    1.517689               1.517689   \n",
       "PA14_10500                    3.173688              -3.173688   \n",
       "\n",
       "            Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "PA14_53300                        0.615595            4244.0   \n",
       "PA14_21530                        0.757436            4379.0   \n",
       "PA14_66480                        0.646066            1318.0   \n",
       "PA14_12590                        0.949002            1069.0   \n",
       "PA14_10500                        0.286293            4426.0   \n",
       "\n",
       "            Percentile (simulated)  Percent DE (simulated)  \\\n",
       "PA14_53300               72.037351                     0.0   \n",
       "PA14_21530               74.329372                     0.0   \n",
       "PA14_66480               22.359932                     0.0   \n",
       "PA14_12590               18.132428                     0.0   \n",
       "PA14_10500               75.127334                     0.0   \n",
       "\n",
       "            Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "PA14_53300                              0.249700                   0.194925   \n",
       "PA14_21530                              0.222476                   0.172918   \n",
       "PA14_66480                              0.127432                   0.066487   \n",
       "PA14_12590                              0.115920                   0.080223   \n",
       "PA14_10500                              0.254018                   0.173930   \n",
       "\n",
       "            Number of experiments (simulated)    Z score  \n",
       "PA14_53300                                 25  18.319543  \n",
       "PA14_21530                                 25  17.641922  \n",
       "PA14_66480                                 25  17.623670  \n",
       "PA14_12590                                 25  17.473402  \n",
       "PA14_10500                                 25  16.786449  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    \"gene\",\n",
    "    params,\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "      <th>Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA14_53300</th>\n",
       "      <td>PA14_53300</td>\n",
       "      <td>3.277490e-44</td>\n",
       "      <td>5890.0</td>\n",
       "      <td>3.820634</td>\n",
       "      <td>-3.820634</td>\n",
       "      <td>0.615595</td>\n",
       "      <td>4244.0</td>\n",
       "      <td>72.037351</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.249700</td>\n",
       "      <td>0.194925</td>\n",
       "      <td>25</td>\n",
       "      <td>18.319543</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_21530</th>\n",
       "      <td>PA14_21530</td>\n",
       "      <td>1.617143e-28</td>\n",
       "      <td>5887.0</td>\n",
       "      <td>3.273077</td>\n",
       "      <td>-3.273077</td>\n",
       "      <td>0.757436</td>\n",
       "      <td>4379.0</td>\n",
       "      <td>74.329372</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.222476</td>\n",
       "      <td>0.172918</td>\n",
       "      <td>25</td>\n",
       "      <td>17.641922</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_66480</th>\n",
       "      <td>PA14_66480</td>\n",
       "      <td>5.056304e-07</td>\n",
       "      <td>5700.0</td>\n",
       "      <td>1.299171</td>\n",
       "      <td>1.299171</td>\n",
       "      <td>0.646066</td>\n",
       "      <td>1318.0</td>\n",
       "      <td>22.359932</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.127432</td>\n",
       "      <td>0.066487</td>\n",
       "      <td>25</td>\n",
       "      <td>17.623670</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_12590</th>\n",
       "      <td>PA14_12590</td>\n",
       "      <td>2.669583e-01</td>\n",
       "      <td>5791.0</td>\n",
       "      <td>1.517689</td>\n",
       "      <td>1.517689</td>\n",
       "      <td>0.949002</td>\n",
       "      <td>1069.0</td>\n",
       "      <td>18.132428</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.115920</td>\n",
       "      <td>0.080223</td>\n",
       "      <td>25</td>\n",
       "      <td>17.473402</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_10500</th>\n",
       "      <td>PA14_10500</td>\n",
       "      <td>2.400867e-30</td>\n",
       "      <td>5884.0</td>\n",
       "      <td>3.173688</td>\n",
       "      <td>-3.173688</td>\n",
       "      <td>0.286293</td>\n",
       "      <td>4426.0</td>\n",
       "      <td>75.127334</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.254018</td>\n",
       "      <td>0.173930</td>\n",
       "      <td>25</td>\n",
       "      <td>16.786449</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "PA14_53300  PA14_53300        3.277490e-44       5890.0   \n",
       "PA14_21530  PA14_21530        1.617143e-28       5887.0   \n",
       "PA14_66480  PA14_66480        5.056304e-07       5700.0   \n",
       "PA14_12590  PA14_12590        2.669583e-01       5791.0   \n",
       "PA14_10500  PA14_10500        2.400867e-30       5884.0   \n",
       "\n",
       "            abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "PA14_53300                    3.820634              -3.820634   \n",
       "PA14_21530                    3.273077              -3.273077   \n",
       "PA14_66480                    1.299171               1.299171   \n",
       "PA14_12590                    1.517689               1.517689   \n",
       "PA14_10500                    3.173688              -3.173688   \n",
       "\n",
       "            Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "PA14_53300                        0.615595            4244.0   \n",
       "PA14_21530                        0.757436            4379.0   \n",
       "PA14_66480                        0.646066            1318.0   \n",
       "PA14_12590                        0.949002            1069.0   \n",
       "PA14_10500                        0.286293            4426.0   \n",
       "\n",
       "            Percentile (simulated)  Percent DE (simulated)  \\\n",
       "PA14_53300               72.037351                     0.0   \n",
       "PA14_21530               74.329372                     0.0   \n",
       "PA14_66480               22.359932                     0.0   \n",
       "PA14_12590               18.132428                     0.0   \n",
       "PA14_10500               75.127334                     0.0   \n",
       "\n",
       "            Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "PA14_53300                              0.249700                   0.194925   \n",
       "PA14_21530                              0.222476                   0.172918   \n",
       "PA14_66480                              0.127432                   0.066487   \n",
       "PA14_12590                              0.115920                   0.080223   \n",
       "PA14_10500                              0.254018                   0.173930   \n",
       "\n",
       "            Number of experiments (simulated)    Z score Gene Name  \n",
       "PA14_53300                                 25  18.319543       NaN  \n",
       "PA14_21530                                 25  17.641922       NaN  \n",
       "PA14_66480                                 25  17.623670       NaN  \n",
       "PA14_12590                                 25  17.473402       NaN  \n",
       "PA14_10500                                 25  16.786449       NaN  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add gene name as column to summary dataframe\n",
    "summary_gene_ranks = ranking.add_pseudomonas_gene_name_col(summary_gene_ranks, base_dir)\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "      <th>Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PA14_06940</th>\n",
       "      <td>PA14_06940</td>\n",
       "      <td>9.948748e-01</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0.005601</td>\n",
       "      <td>0.005601</td>\n",
       "      <td>0.000785</td>\n",
       "      <td>5891.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.44000</td>\n",
       "      <td>1.127401</td>\n",
       "      <td>0.700321</td>\n",
       "      <td>25</td>\n",
       "      <td>-1.601835</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_33800</th>\n",
       "      <td>PA14_33800</td>\n",
       "      <td>9.161133e-01</td>\n",
       "      <td>525.0</td>\n",
       "      <td>0.051189</td>\n",
       "      <td>0.051189</td>\n",
       "      <td>0.160410</td>\n",
       "      <td>5890.0</td>\n",
       "      <td>99.983022</td>\n",
       "      <td>0.24000</td>\n",
       "      <td>0.900162</td>\n",
       "      <td>0.557125</td>\n",
       "      <td>25</td>\n",
       "      <td>-1.523848</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_17700</th>\n",
       "      <td>PA14_17700</td>\n",
       "      <td>2.629764e-52</td>\n",
       "      <td>5883.0</td>\n",
       "      <td>3.131006</td>\n",
       "      <td>3.131006</td>\n",
       "      <td>0.001289</td>\n",
       "      <td>5889.0</td>\n",
       "      <td>99.966044</td>\n",
       "      <td>0.40000</td>\n",
       "      <td>0.947139</td>\n",
       "      <td>0.650671</td>\n",
       "      <td>25</td>\n",
       "      <td>3.356331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_13280</th>\n",
       "      <td>PA14_13280</td>\n",
       "      <td>5.929881e-02</td>\n",
       "      <td>5008.0</td>\n",
       "      <td>0.728369</td>\n",
       "      <td>0.728369</td>\n",
       "      <td>0.038319</td>\n",
       "      <td>5888.0</td>\n",
       "      <td>99.949066</td>\n",
       "      <td>0.20000</td>\n",
       "      <td>0.751504</td>\n",
       "      <td>0.399483</td>\n",
       "      <td>25</td>\n",
       "      <td>-0.057912</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_27690</th>\n",
       "      <td>PA14_27690</td>\n",
       "      <td>6.656786e-01</td>\n",
       "      <td>1929.0</td>\n",
       "      <td>0.192381</td>\n",
       "      <td>0.192381</td>\n",
       "      <td>0.461541</td>\n",
       "      <td>5887.0</td>\n",
       "      <td>99.932088</td>\n",
       "      <td>0.26087</td>\n",
       "      <td>1.025154</td>\n",
       "      <td>0.780311</td>\n",
       "      <td>23</td>\n",
       "      <td>-1.067233</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Gene ID  Adj P-value (Real)  Rank (Real)  \\\n",
       "PA14_06940  PA14_06940        9.948748e-01         57.0   \n",
       "PA14_33800  PA14_33800        9.161133e-01        525.0   \n",
       "PA14_17700  PA14_17700        2.629764e-52       5883.0   \n",
       "PA14_13280  PA14_13280        5.929881e-02       5008.0   \n",
       "PA14_27690  PA14_27690        6.656786e-01       1929.0   \n",
       "\n",
       "            abs(log2FoldChange) (Real)  log2FoldChange (Real)  \\\n",
       "PA14_06940                    0.005601               0.005601   \n",
       "PA14_33800                    0.051189               0.051189   \n",
       "PA14_17700                    3.131006               3.131006   \n",
       "PA14_13280                    0.728369               0.728369   \n",
       "PA14_27690                    0.192381               0.192381   \n",
       "\n",
       "            Median adj p-value (simulated)  Rank (simulated)  \\\n",
       "PA14_06940                        0.000785            5891.0   \n",
       "PA14_33800                        0.160410            5890.0   \n",
       "PA14_17700                        0.001289            5889.0   \n",
       "PA14_13280                        0.038319            5888.0   \n",
       "PA14_27690                        0.461541            5887.0   \n",
       "\n",
       "            Percentile (simulated)  Percent DE (simulated)  \\\n",
       "PA14_06940              100.000000                 0.44000   \n",
       "PA14_33800               99.983022                 0.24000   \n",
       "PA14_17700               99.966044                 0.40000   \n",
       "PA14_13280               99.949066                 0.20000   \n",
       "PA14_27690               99.932088                 0.26087   \n",
       "\n",
       "            Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "PA14_06940                              1.127401                   0.700321   \n",
       "PA14_33800                              0.900162                   0.557125   \n",
       "PA14_17700                              0.947139                   0.650671   \n",
       "PA14_13280                              0.751504                   0.399483   \n",
       "PA14_27690                              1.025154                   0.780311   \n",
       "\n",
       "            Number of experiments (simulated)   Z score Gene Name  \n",
       "PA14_06940                                 25 -1.601835       NaN  \n",
       "PA14_33800                                 25 -1.523848       NaN  \n",
       "PA14_17700                                 25  3.356331       NaN  \n",
       "PA14_13280                                 25 -0.057912       NaN  \n",
       "PA14_27690                                 23 -1.067233       NaN  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks.sort_values(by=\"Percentile (simulated)\", ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene ID                                 False\n",
       "Adj P-value (Real)                       True\n",
       "Rank (Real)                              True\n",
       "abs(log2FoldChange) (Real)               True\n",
       "log2FoldChange (Real)                    True\n",
       "Median adj p-value (simulated)          False\n",
       "Rank (simulated)                        False\n",
       "Percentile (simulated)                  False\n",
       "Percent DE (simulated)                  False\n",
       "Mean abs(log2FoldChange) (simulated)    False\n",
       "Std deviation (simulated)               False\n",
       "Number of experiments (simulated)       False\n",
       "Z score                                  True\n",
       "Gene Name                                True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if there is an NaN values, there should not be\n",
    "summary_gene_ranks.isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create `gene_summary_filename`\n",
    "summary_gene_ranks.to_csv(gene_summary_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare gene ranking\n",
    "\n",
    "We can only compare the ranking between the PAO1 RNA-seq compendium vs GAPE, where we still see good concordance as expected.\n",
    "\n",
    "When we look for common genes, we do this based on percentiles generated by SOPHIE for both the PAO1 and PA14 compendia to be consistent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "if \"pao1\" in config_filename:\n",
    "    # Get generic genes identified by Crow et. al.\n",
    "    GAPE_filename = params[\"reference_gene_filename\"]\n",
    "    ref_gene_col = params[\"reference_gene_name_col\"]\n",
    "    ref_rank_col = params[\"reference_rank_col\"]\n",
    "\n",
    "    figure_filename = f\"gene_ranking_{col_to_rank_genes}.svg\"\n",
    "\n",
    "    corr, shared_ranking = ranking.compare_gene_ranking(\n",
    "        summary_gene_ranks, GAPE_filename, ref_gene_col, ref_rank_col, figure_filename\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "encoding": "# -*- coding: utf-8 -*-",
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
