{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process Crow et al. data\n",
    "This notebook does the following:\n",
    "\n",
    "1. Select template experiment. This template experiment will be used in the next [notebook](2_identify_generic_genes_pathways.ipynb) to simulate experiments with the same experimental design but testing a different biological process.\n",
    "\n",
    "2. Crow et al. data was downloaded using `explore_RNAseq_only_generic_genes/download_Crow_data.R` script that downloads expression data from https://github.com/PavlidisLab/gemmaAPI.R\n",
    "\n",
    "Note: For the analysis exploring the RNA-seq only common DEGs we used the union of genes per experiment, which resulted in some samples having NaNs for some samples. For this analysis we are taking the intersection so that we can remove all NaNs to train.\n",
    "\n",
    "3. Train VAE on processed data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:21: The name tf.ConfigProto is deprecated. Please use tf.compat.v1.ConfigProto instead.\n",
      "\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:25: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.\n",
      "\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:25: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n",
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext rpy2.ipython\n",
    "%autoreload 2\n",
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "from ponyo import utils, train_vae_modules\n",
    "from generic_expression_patterns_modules import process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/Documents/Repos/generic-expression-patterns/generic_expression_patterns_modules/process.py:57: The name tf.set_random_seed is deprecated. Please use tf.compat.v1.set_random_seed instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Set seeds to get reproducible VAE trained models\n",
    "process.set_all_seeds()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set parameters for data processing\n",
    "\n",
    "Most parameters are read from `config_filename`. We manually selected bioproject [GSE11352](https://www.ncbi.nlm.nih.gov/gds/?term=GSE11352[Accession]) as the template experiment, which contains breast cell lines treated with estradiol at 12H, 24H and 48H."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "# Read in config variables\n",
    "config_filename = os.path.abspath(\n",
    "    os.path.join(base_dir, \"configs\", \"config_human_Crow.tsv\")\n",
    ")\n",
    "\n",
    "params = utils.read_config(config_filename)\n",
    "\n",
    "local_dir = params[\"local_dir\"]\n",
    "dataset_name = params[\"dataset_name\"]\n",
    "\n",
    "# File that contains gene ranks identified by Crow et. al.\n",
    "DE_prior_filename = params[\"reference_gene_filename\"]\n",
    "\n",
    "# Template experiment ID\n",
    "project_id = params[\"project_id\"]\n",
    "\n",
    "# Column name containing sample ids\n",
    "metadata_colname = params[\"metadata_colname\"]\n",
    "\n",
    "# Output file: pickled list of shared genes(generated during gene ID mapping)\n",
    "shared_genes_filename = params[\"shared_genes_filename\"]\n",
    "\n",
    "# Output files of template experiment data\n",
    "mapped_template_filename = params[\"mapped_template_filename\"]\n",
    "processed_template_filename = params[\"processed_template_filename\"]\n",
    "\n",
    "# Output files of Rani's compendium data\n",
    "raw_compendium_filename = params[\"raw_compendium_filename\"]\n",
    "mapped_compendium_filename = params[\"mapped_compendium_filename\"]\n",
    "normalized_compendium_filename = params[\"normalized_compendium_filename\"]\n",
    "\n",
    "# Output file: pickled scaler (generated during compendium normalization)\n",
    "scaler_filename = params[\"scaler_filename\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load compendium data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32476, 14487)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HSPA6|HSPA7</th>\n",
       "      <th>UBA7</th>\n",
       "      <th>ESRRA</th>\n",
       "      <th>CYP2A7|CYP2A6</th>\n",
       "      <th>ADAM32</th>\n",
       "      <th>CORO6</th>\n",
       "      <th>CCDC65</th>\n",
       "      <th>LEAP2</th>\n",
       "      <th>CTCFL</th>\n",
       "      <th>C4orf33</th>\n",
       "      <th>...</th>\n",
       "      <th>TRIM15</th>\n",
       "      <th>NBL1|MICOS10-NBL1</th>\n",
       "      <th>GRK2</th>\n",
       "      <th>ATN1</th>\n",
       "      <th>FBXO41</th>\n",
       "      <th>KCTD13</th>\n",
       "      <th>TFEB</th>\n",
       "      <th>DND1</th>\n",
       "      <th>CASKIN2</th>\n",
       "      <th>GIMAP1-GIMAP5|GIMAP5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>L_IB38___BioAssayId=72907Name=050705MJA_U133_2.0_IB08.CEL</th>\n",
       "      <td>8.96564</td>\n",
       "      <td>83.32761</td>\n",
       "      <td>48.94783</td>\n",
       "      <td>6.05333</td>\n",
       "      <td>17.74886</td>\n",
       "      <td>177.35550</td>\n",
       "      <td>22.08059</td>\n",
       "      <td>24.29061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57.44873</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L_IB40___BioAssayId=72909Name=050705MJA_U133_2.0_IB10.CEL</th>\n",
       "      <td>769.30450</td>\n",
       "      <td>61.90878</td>\n",
       "      <td>119.33116</td>\n",
       "      <td>17.42643</td>\n",
       "      <td>11.58205</td>\n",
       "      <td>117.40292</td>\n",
       "      <td>46.72446</td>\n",
       "      <td>50.81827</td>\n",
       "      <td>4.21085</td>\n",
       "      <td>29.51632</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L_IB41___BioAssayId=72910Name=050705MJA_U133_2.0_IB11.CEL</th>\n",
       "      <td>23.69639</td>\n",
       "      <td>93.37774</td>\n",
       "      <td>80.67838</td>\n",
       "      <td>27.73698</td>\n",
       "      <td>18.56551</td>\n",
       "      <td>175.07208</td>\n",
       "      <td>21.01308</td>\n",
       "      <td>56.90577</td>\n",
       "      <td>4.52370</td>\n",
       "      <td>70.19458</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L_IB33___BioAssayId=72902Name=050705MJA_U133_2.0_IB03.CEL</th>\n",
       "      <td>10.37628</td>\n",
       "      <td>56.58948</td>\n",
       "      <td>64.83450</td>\n",
       "      <td>14.86978</td>\n",
       "      <td>8.71849</td>\n",
       "      <td>41.07041</td>\n",
       "      <td>24.34623</td>\n",
       "      <td>62.00211</td>\n",
       "      <td>12.89886</td>\n",
       "      <td>38.07436</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L_IB36___BioAssayId=72905Name=050705MJA_U133_2.0_IB06.CEL</th>\n",
       "      <td>48.51214</td>\n",
       "      <td>41.29277</td>\n",
       "      <td>76.53021</td>\n",
       "      <td>15.66673</td>\n",
       "      <td>5.35315</td>\n",
       "      <td>27.70762</td>\n",
       "      <td>4.19627</td>\n",
       "      <td>13.33500</td>\n",
       "      <td>0.54027</td>\n",
       "      <td>14.19567</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 14487 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    HSPA6|HSPA7      UBA7  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....      8.96564  83.32761   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....    769.30450  61.90878   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....     23.69639  93.37774   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....     10.37628  56.58948   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....     48.51214  41.29277   \n",
       "\n",
       "                                                        ESRRA  CYP2A7|CYP2A6  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....   48.94783        6.05333   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....  119.33116       17.42643   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....   80.67838       27.73698   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....   64.83450       14.86978   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....   76.53021       15.66673   \n",
       "\n",
       "                                                      ADAM32      CORO6  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....  17.74886  177.35550   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....  11.58205  117.40292   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  18.56551  175.07208   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....   8.71849   41.07041   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....   5.35315   27.70762   \n",
       "\n",
       "                                                      CCDC65     LEAP2  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....  22.08059  24.29061   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....  46.72446  50.81827   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  21.01308  56.90577   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....  24.34623  62.00211   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....   4.19627  13.33500   \n",
       "\n",
       "                                                       CTCFL   C4orf33  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....       NaN  57.44873   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....   4.21085  29.51632   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....   4.52370  70.19458   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....  12.89886  38.07436   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....   0.54027  14.19567   \n",
       "\n",
       "                                                            ...           \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....          ...            \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....          ...            \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....          ...            \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....          ...            \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....          ...            \n",
       "\n",
       "                                                    TRIM15  NBL1|MICOS10-NBL1  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....     NaN                NaN   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....     NaN                NaN   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....     NaN                NaN   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....     NaN                NaN   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....     NaN                NaN   \n",
       "\n",
       "                                                    GRK2  ATN1  FBXO41  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....   NaN   NaN     NaN   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....   NaN   NaN     NaN   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....   NaN   NaN     NaN   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....   NaN   NaN     NaN   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....   NaN   NaN     NaN   \n",
       "\n",
       "                                                    KCTD13  TFEB  DND1  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....     NaN   NaN   NaN   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....     NaN   NaN   NaN   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....     NaN   NaN   NaN   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....     NaN   NaN   NaN   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....     NaN   NaN   NaN   \n",
       "\n",
       "                                                    CASKIN2  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....      NaN   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....      NaN   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....      NaN   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....      NaN   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....      NaN   \n",
       "\n",
       "                                                    GIMAP1-GIMAP5|GIMAP5  \n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....                   NaN  \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....                   NaN  \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....                   NaN  \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....                   NaN  \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....                   NaN  \n",
       "\n",
       "[5 rows x 14487 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_compendium = pd.read_csv(raw_compendium_filename, sep=\"\\t\", header=0, index_col=0)\n",
    "print(raw_compendium.shape)\n",
    "raw_compendium.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process compendium data\n",
    "\n",
    "1. Drop probe column\n",
    "2. Transpose\n",
    "3. Get only shared genes from Crow et. al.\n",
    "4. Normalize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32476, 9240)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AA06</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM2</th>\n",
       "      <th>ZSWIM3</th>\n",
       "      <th>ZSWIM5</th>\n",
       "      <th>ZSWIM6</th>\n",
       "      <th>ZSWIM8-AS1</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZYG11A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>L_IB38___BioAssayId=72907Name=050705MJA_U133_2.0_IB08.CEL</th>\n",
       "      <td>11.63699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>676.43811</td>\n",
       "      <td>18.94505</td>\n",
       "      <td>8.59358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.44703</td>\n",
       "      <td>24.31123</td>\n",
       "      <td>50.73169</td>\n",
       "      <td>141.09604</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.83527</td>\n",
       "      <td>67.54955</td>\n",
       "      <td>470.20654</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.50054</td>\n",
       "      <td>59.74241</td>\n",
       "      <td>37.28952</td>\n",
       "      <td>37.95333</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L_IB40___BioAssayId=72909Name=050705MJA_U133_2.0_IB10.CEL</th>\n",
       "      <td>20.67487</td>\n",
       "      <td>NaN</td>\n",
       "      <td>591.69562</td>\n",
       "      <td>23.72454</td>\n",
       "      <td>9.03128</td>\n",
       "      <td>NaN</td>\n",
       "      <td>47.68120</td>\n",
       "      <td>14.11284</td>\n",
       "      <td>49.37021</td>\n",
       "      <td>62.37345</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.71398</td>\n",
       "      <td>19.14305</td>\n",
       "      <td>325.37585</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68.06551</td>\n",
       "      <td>27.87321</td>\n",
       "      <td>32.73832</td>\n",
       "      <td>38.87195</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L_IB41___BioAssayId=72910Name=050705MJA_U133_2.0_IB11.CEL</th>\n",
       "      <td>40.67805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1069.66284</td>\n",
       "      <td>50.84652</td>\n",
       "      <td>8.31981</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.64032</td>\n",
       "      <td>14.81776</td>\n",
       "      <td>109.17731</td>\n",
       "      <td>237.93060</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.07349</td>\n",
       "      <td>66.29779</td>\n",
       "      <td>855.58441</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88.50397</td>\n",
       "      <td>60.85935</td>\n",
       "      <td>66.90208</td>\n",
       "      <td>84.84532</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L_IB33___BioAssayId=72902Name=050705MJA_U133_2.0_IB03.CEL</th>\n",
       "      <td>13.79470</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1358.15430</td>\n",
       "      <td>45.66821</td>\n",
       "      <td>8.50909</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.39715</td>\n",
       "      <td>18.87131</td>\n",
       "      <td>40.65568</td>\n",
       "      <td>99.47049</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57.62479</td>\n",
       "      <td>840.18787</td>\n",
       "      <td>NaN</td>\n",
       "      <td>90.49266</td>\n",
       "      <td>51.56433</td>\n",
       "      <td>62.14825</td>\n",
       "      <td>80.57075</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L_IB36___BioAssayId=72905Name=050705MJA_U133_2.0_IB06.CEL</th>\n",
       "      <td>13.70568</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1152.48254</td>\n",
       "      <td>46.30679</td>\n",
       "      <td>9.29537</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.76678</td>\n",
       "      <td>14.42896</td>\n",
       "      <td>0.36647</td>\n",
       "      <td>69.15773</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.52186</td>\n",
       "      <td>43.80604</td>\n",
       "      <td>389.94690</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53.26638</td>\n",
       "      <td>13.28645</td>\n",
       "      <td>42.16255</td>\n",
       "      <td>59.41403</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 9240 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        A1BG  A1BG-AS1  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....  11.63699       NaN   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....  20.67487       NaN   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  40.67805       NaN   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....  13.79470       NaN   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....  13.70568       NaN   \n",
       "\n",
       "                                                           A2M   A2M-AS1  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....   676.43811  18.94505   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....   591.69562  23.72454   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  1069.66284  50.84652   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....  1358.15430  45.66821   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....  1152.48254  46.30679   \n",
       "\n",
       "                                                      A2ML1  A4GALT     A4GNT  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....  8.59358     NaN  19.44703   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....  9.03128     NaN  47.68120   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  8.31981     NaN  12.64032   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....  8.50909     NaN  17.39715   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....  9.29537     NaN  22.76678   \n",
       "\n",
       "                                                        AA06       AAAS  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....  24.31123   50.73169   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....  14.11284   49.37021   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  14.81776  109.17731   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....  18.87131   40.65568   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....  14.42896    0.36647   \n",
       "\n",
       "                                                         AACS   ...    ZSWIM2  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....  141.09604   ...       NaN   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....   62.37345   ...       NaN   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  237.93060   ...       NaN   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....   99.47049   ...       NaN   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....   69.15773   ...       NaN   \n",
       "\n",
       "                                                      ZSWIM3    ZSWIM5  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....   4.83527  67.54955   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....  16.71398  19.14305   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  24.07349  66.29779   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....       NaN  57.62479   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....  17.52186  43.80604   \n",
       "\n",
       "                                                       ZSWIM6  ZSWIM8-AS1  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....  470.20654         NaN   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....  325.37585         NaN   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  855.58441         NaN   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....  840.18787         NaN   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....  389.94690         NaN   \n",
       "\n",
       "                                                        ZW10     ZWINT  \\\n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....  71.50054  59.74241   \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....  68.06551  27.87321   \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  88.50397  60.85935   \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....  90.49266  51.56433   \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....  53.26638  13.28645   \n",
       "\n",
       "                                                        ZXDA      ZXDB  ZYG11A  \n",
       "L_IB38___BioAssayId=72907Name=050705MJA_U133_2....  37.28952  37.95333     NaN  \n",
       "L_IB40___BioAssayId=72909Name=050705MJA_U133_2....  32.73832  38.87195     NaN  \n",
       "L_IB41___BioAssayId=72910Name=050705MJA_U133_2....  66.90208  84.84532     NaN  \n",
       "L_IB33___BioAssayId=72902Name=050705MJA_U133_2....  62.14825  80.57075     NaN  \n",
       "L_IB36___BioAssayId=72905Name=050705MJA_U133_2....  42.16255  59.41403     NaN  \n",
       "\n",
       "[5 rows x 9240 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Matrix is sample x gene\n",
    "# Note, there are NaNs in the matrix. I'm not exactly sure what is causing this since the data\n",
    "# downloaded was based on what Crow et al used in their analysis and therefore should have been\n",
    "# filtered by platform.\n",
    "# All genes have at least 1 NaN so dropping all genes with NaN removes all the data\n",
    "# Instead we will move genes if they are NaN in _most_ samples (>90%)\n",
    "x = raw_compendium.isna().sum()\n",
    "\n",
    "processed_compendium = raw_compendium[x.index[x < 3000]]\n",
    "\n",
    "# Get only gene expression data for genes in Crow et. al.\n",
    "our_gene_ids_hgnc = list(processed_compendium.columns)\n",
    "\n",
    "published_generic_genes = process.get_published_generic_genes(DE_prior_filename)\n",
    "shared_genes_hgnc = list(set(our_gene_ids_hgnc).intersection(published_generic_genes))\n",
    "\n",
    "# In Python, the order of elements in a list that is converted from a set\n",
    "# is non-deterministic, so it is sorted here to have reproducible result.\n",
    "shared_genes_hgnc.sort()\n",
    "\n",
    "# Pickle `shared_genes_hgnc` and save as `shared_genes_filename`\n",
    "if not os.path.exists(shared_genes_filename):\n",
    "    with open(shared_genes_filename, \"wb\") as pkl_fh:\n",
    "        pickle.dump(shared_genes_hgnc, pkl_fh)\n",
    "\n",
    "mapped_compendium = processed_compendium[shared_genes_hgnc]\n",
    "print(mapped_compendium.shape)\n",
    "mapped_compendium.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "mapped_compendium.to_csv(mapped_compendium_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize data\n",
    "process.normalize_compendium(\n",
    "    mapped_compendium_filename, normalized_compendium_filename, scaler_filename\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select and process template data\n",
    "\n",
    "1. Get gene expression associated with `project_id`, which was manually selected by the user and specified in the config file.\n",
    "\n",
    "Note: The data is not normalized so that we can perform DE analysis in next notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36, 9240)\n"
     ]
    }
   ],
   "source": [
    "# Note: This is the only notebook using this function, so for now it is included here\n",
    "# Get sample ids associated with selected project id\n",
    "def get_sample_ids(experiment_id, mapping_filename):\n",
    "    \"\"\"\n",
    "    Return sample ids for a given experiment id\n",
    "\n",
    "    \"\"\"\n",
    "    # Read in metadata\n",
    "    metadata = pd.read_csv(mapping_filename, header=0)\n",
    "    metadata.set_index(\"Experiment id\", inplace=True)\n",
    "\n",
    "    selected_metadata = metadata.loc[experiment_id]\n",
    "    sample_ids = list(selected_metadata[metadata_colname])\n",
    "\n",
    "    return sample_ids\n",
    "\n",
    "\n",
    "# metadata file with mapping from experiment to sample\n",
    "experiment_to_sample_metadata_filename = os.path.join(\n",
    "    base_dir, dataset_name, \"data\", \"metadata\", \"experiment_sample_annotations.csv\"\n",
    ")\n",
    "\n",
    "sample_ids = get_sample_ids(project_id, experiment_to_sample_metadata_filename)\n",
    "\n",
    "# Get expression data\n",
    "template_mapped = mapped_compendium.loc[sample_ids]\n",
    "print(template_mapped.shape)\n",
    "\n",
    "# Save\n",
    "template_mapped.to_csv(mapped_template_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train VAE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create VAE directories if needed\n",
    "output_dirs = [\n",
    "    os.path.join(base_dir, dataset_name, \"models\"),\n",
    "    os.path.join(base_dir, dataset_name, \"logs\"),\n",
    "]\n",
    "\n",
    "NN_architecture = params[\"NN_architecture\"]\n",
    "\n",
    "# Check if NN architecture directory exist otherwise create\n",
    "for each_dir in output_dirs:\n",
    "    sub_dir = os.path.join(each_dir, NN_architecture)\n",
    "    os.makedirs(sub_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input dataset contains 32476 samples and 9240 genes\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "tracking <tf.Variable 'Variable:0' shape=() dtype=float32> beta\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/nn_impl.py:183: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Use tf.where in 2.0, which has the same broadcast rule as np.where\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/engine/training_utils.py:819: UserWarning: Output custom_variational_layer_1 missing from loss dictionary. We assume this was done on purpose. The fit and evaluate APIs will not be expecting any data to be passed to custom_variational_layer_1.\n",
      "  'be expecting any data to be passed to {0}.'.format(name))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n",
      "Train on 24357 samples, validate on 8119 samples\n",
      "Epoch 1/40\n",
      "24357/24357 [==============================] - 1935s 79ms/step - loss: nan - val_loss: nan\n",
      "Epoch 2/40\n",
      "24357/24357 [==============================] - 1941s 80ms/step - loss: nan - val_loss: nan\n",
      "Epoch 3/40\n",
      "  560/24357 [..............................] - ETA: 31:52 - loss: nan"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-9638032f4528>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Train VAE on new compendium data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mtrain_vae_modules\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtrain_vae\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconfig_filename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnormalized_compendium_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/train_vae_modules.py\u001b[0m in \u001b[0;36mtrain_vae\u001b[0;34m(config_file, input_data_file)\u001b[0m\n\u001b[1;32m    135\u001b[0m         \u001b[0mdataset_name\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    136\u001b[0m         \u001b[0mtrain_architecture\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 137\u001b[0;31m         \u001b[0mvalidation_frac\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    138\u001b[0m     )\n",
      "\u001b[0;32m~/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/vae.py\u001b[0m in \u001b[0;36mtybalt_2layer_model\u001b[0;34m(learning_rate, batch_size, epochs, kappa, intermediate_dim, latent_dim, epsilon_std, rnaseq, base_dir, dataset_name, NN_name, validation_frac)\u001b[0m\n\u001b[1;32m    268\u001b[0m         \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbatch_size\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    269\u001b[0m         \u001b[0mvalidation_data\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrnaseq_test_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 270\u001b[0;31m         \u001b[0mcallbacks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mWarmUpCallback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbeta\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkappa\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    271\u001b[0m     )\n\u001b[1;32m    272\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/engine/training.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_freq, max_queue_size, workers, use_multiprocessing, **kwargs)\u001b[0m\n\u001b[1;32m   1237\u001b[0m                                         \u001b[0msteps_per_epoch\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msteps_per_epoch\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1238\u001b[0m                                         \u001b[0mvalidation_steps\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvalidation_steps\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1239\u001b[0;31m                                         validation_freq=validation_freq)\n\u001b[0m\u001b[1;32m   1240\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1241\u001b[0m     def evaluate(self,\n",
      "\u001b[0;32m~/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/engine/training_arrays.py\u001b[0m in \u001b[0;36mfit_loop\u001b[0;34m(model, fit_function, fit_inputs, out_labels, batch_size, epochs, verbose, callbacks, val_function, val_inputs, shuffle, initial_epoch, steps_per_epoch, validation_steps, validation_freq)\u001b[0m\n\u001b[1;32m    194\u001b[0m                     \u001b[0mins_batch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mins_batch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtoarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 196\u001b[0;31m                 \u001b[0mouts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfit_function\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mins_batch\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    197\u001b[0m                 \u001b[0mouts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mto_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mouts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0ml\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mo\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mout_labels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mouts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/keras/backend.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, inputs)\u001b[0m\n\u001b[1;32m   3474\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3475\u001b[0m     fetched = self._callable_fn(*array_vals,\n\u001b[0;32m-> 3476\u001b[0;31m                                 run_metadata=self.run_metadata)\n\u001b[0m\u001b[1;32m   3477\u001b[0m     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_fetch_callbacks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfetched\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_fetches\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3478\u001b[0m     output_structure = nest.pack_sequence_as(\n",
      "\u001b[0;32m~/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/client/session.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1470\u001b[0m         ret = tf_session.TF_SessionRunCallable(self._session._session,\n\u001b[1;32m   1471\u001b[0m                                                \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_handle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1472\u001b[0;31m                                                run_metadata_ptr)\n\u001b[0m\u001b[1;32m   1473\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mrun_metadata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1474\u001b[0m           \u001b[0mproto_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtf_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTF_GetBuffer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrun_metadata_ptr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Train VAE on new compendium data\n",
    "train_vae_modules.train_vae(config_filename, normalized_compendium_filename)"
   ]
  }
 ],
 "metadata": {
  "finalized": {
   "timestamp": 1592246126078,
   "trusted": true
  },
  "jupytext": {
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
