{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0b54cfe2",
   "metadata": {},
   "source": [
    "# Evaluation\n",
    "\n",
    "How often do regular differential expression analysis vs sophie prioritize the specific vs generic genes?\n",
    "\n",
    "1. Simulate 1 template perturbation experiment using the technique above\n",
    "2. Apply SOPHIE to get ranking of specific and generic genes based on their z-score.\n",
    "3. Apply traditional DE analysis and get ranking of specific and generic genes based on their log fold change value\n",
    "4. Compare the difference in ranking between specific and generic genes using SOPHIE vs traditional metrics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b68f083",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:21: The name tf.ConfigProto is deprecated. Please use tf.compat.v1.ConfigProto instead.\n",
      "\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:25: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.\n",
      "\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:25: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext rpy2.ipython\n",
    "import os\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from rpy2.robjects import pandas2ri\n",
    "from ponyo import utils, train_vae_modules, simulate_expression_data\n",
    "from generic_expression_patterns_modules import (\n",
    "    process,\n",
    "    new_experiment_process,  # REMOVE\n",
    "    stats,\n",
    "    ranking,\n",
    ")\n",
    "\n",
    "np.random.seed(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4fa09de1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = \"config_sophie_vs_trad.tsv\"\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6e9a766",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load config params\n",
    "\n",
    "# Local directory to store intermediate files\n",
    "local_dir = params[\"local_dir\"]\n",
    "\n",
    "#\n",
    "dataset_name = params[\"dataset_name\"]\n",
    "\n",
    "# File containing un-normalized template experiment\n",
    "raw_template_filename = params[\"raw_template_filename\"]\n",
    "\n",
    "# Un-normalized compendium filename\n",
    "raw_compendium_filename = params[\"raw_compendium_filename\"]\n",
    "\n",
    "# Normalized compendium filename\n",
    "normalized_compendium_filename = params[\"normalized_compendium_filename\"]\n",
    "\n",
    "# ID for template experiment to be selected\n",
    "project_id = params[\"project_id\"]\n",
    "\n",
    "# Number of simulated experiments to generate\n",
    "num_runs = params[\"num_simulated\"]\n",
    "\n",
    "# Directory containing trained VAE model\n",
    "vae_model_dir = params[\"vae_model_dir\"]\n",
    "\n",
    "# Size of the latent dimension\n",
    "latent_dim = params[\"latent_dim\"]\n",
    "\n",
    "# Scaler transform used to scale compendium data into 0-1 range for training\n",
    "scaler_filename = params[\"scaler_filename\"]\n",
    "\n",
    "# Which DE method to use\n",
    "# We recommend that if data is RNA-seq then use DESeq2\n",
    "# If data is microarray then use Limma\n",
    "de_method = params[\"DE_method\"]\n",
    "\n",
    "# If using DE-seq, setting this parameter will\n",
    "# remove genes below a certain threshold\n",
    "count_threshold = params[\"count_threshold\"]\n",
    "\n",
    "# Metadata file that specifies which samples to keep for DE analysis (Optional)\n",
    "template_process_samples_filename = params[\"template_process_samples_filename\"]\n",
    "\n",
    "# Metadata file that specifies sample grouping for DE analysis\n",
    "template_DE_grouping_filename = params[\"template_DE_grouping_filename\"]\n",
    "\n",
    "# Statistic to use to rank genes or pathways by\n",
    "# Choices are {} FILL IN\n",
    "col_to_rank_genes = params[\"rank_genes_by\"]\n",
    "\n",
    "# Pickle files saving specific and generic gene ids\n",
    "template_specific_gene_ids_filename = params[\"template_specific_gene_ids_filename\"]\n",
    "generic_gene_ids_filename = \"generic_gene_ids.pickle\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5fccae1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files generated by this notebook\n",
    "\n",
    "# File storing template experiment with gene ids mapped to compendium gene ids\n",
    "mapped_template_filename = params[\"mapped_template_filename\"]\n",
    "\n",
    "# File storing normalized template experiment\n",
    "normalized_template_filename = params[\"normalized_template_filename\"]\n",
    "\n",
    "# File storing processed template experiment,\n",
    "# after samples have been selected for comparison in DE analysis\n",
    "processed_template_filename = params[\"processed_template_filename\"]\n",
    "\n",
    "# Output summary file\n",
    "output_filename = params[\"output_filename\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f10a5605",
   "metadata": {},
   "source": [
    "## SOPHIE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "69f51a6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4, 1000)\n",
      "(360, 1000)\n"
     ]
    }
   ],
   "source": [
    "# Process template\n",
    "new_experiment_process.process_template_experiment(\n",
    "    raw_template_filename,\n",
    "    raw_compendium_filename,\n",
    "    scaler_filename,\n",
    "    mapped_template_filename,\n",
    "    normalized_template_filename,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0e48cdb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate multiple experiments UPDATE COMMENT\n",
    "# This step creates the following files in \"<local_dir>/pseudo_experiment/\" directory:\n",
    "#   - selected_simulated_data_SRP012656_<n>.txt\n",
    "#   - selected_simulated_encoded_data_SRP012656_<n>.txt\n",
    "#   - template_normalized_data_SRP012656_test.txt\n",
    "# in which \"<n>\" is an integer in the range of [0, num_runs-1]\n",
    "\n",
    "# REMOVE LATER\n",
    "# dataset_name = \"pre_model_unseen_template\"\n",
    "# Load pickled file\n",
    "scaler = pickle.load(open(scaler_filename, \"rb\"))\n",
    "\n",
    "# Update simulated dir\n",
    "os.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)\n",
    "\n",
    "# Update to take in file to be consisten\n",
    "normalized_compendium = pd.read_csv(\n",
    "    normalized_compendium_filename, header=0, sep=\"\\t\", index_col=0\n",
    ")\n",
    "normalized_template = pd.read_csv(\n",
    "    normalized_template_filename, header=0, sep=\"\\t\", index_col=0\n",
    ")\n",
    "# ------------\n",
    "# Update call when new version of ponyo\n",
    "for run_id in range(num_runs):\n",
    "    new_experiment_process.embed_shift_template_experiment(\n",
    "        normalized_compendium,\n",
    "        normalized_template,\n",
    "        vae_model_dir,\n",
    "        project_id,\n",
    "        scaler_filename,\n",
    "        local_dir,\n",
    "        latent_dim,\n",
    "        run_id,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "09bf6c28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "## Update simulated dir\n",
    "if not os.path.exists(template_process_samples_filename):\n",
    "    template_process_samples_filename = None\n",
    "\n",
    "if de_method == \"deseq\":\n",
    "    # Process template data\n",
    "    stats.process_samples_for_DESeq(\n",
    "        raw_template_filename,\n",
    "        template_DE_grouping_filename,\n",
    "        processed_template_filename,\n",
    "        count_threshold,\n",
    "        template_process_samples_filename,\n",
    "    )\n",
    "\n",
    "    # Process simulated data\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        out_simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}_processed.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_DESeq(\n",
    "            simulated_filename,\n",
    "            template_DE_grouping_filename,\n",
    "            out_simulated_filename,\n",
    "            count_threshold,\n",
    "            template_process_samples_filename,\n",
    "        )\n",
    "else:\n",
    "    stats.process_samples_for_limma(\n",
    "        raw_template_filename,\n",
    "        template_DE_grouping_filename,\n",
    "        processed_template_filename,\n",
    "        template_process_samples_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_limma(\n",
    "            simulated_filename,\n",
    "            template_DE_grouping_filename,\n",
    "            None,\n",
    "            template_process_samples_filename,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c56b91df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ee7b0585",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: DelayedArray\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: BiocParallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (de_method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0dee119f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i base_dir -i local_dir -i num_runs -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# Files created: \"<local_dir>/DE_stats/DE_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\",\n",
    "        i,\n",
    "        \"_processed.txt\",\n",
    "        sep = \"\"\n",
    "    )\n",
    "    if (de_method == \"deseq\"){\n",
    "        get_DE_stats_DESeq(\n",
    "            template_DE_grouping_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "    }\n",
    "    else {\n",
    "        get_DE_stats_limma(\n",
    "            template_DE_grouping_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2c0e853b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (2 levels on the left, 1 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "# Added\n",
    "if de_method == \"deseq\":\n",
    "    logFC_name = \"log2FoldChange\"\n",
    "    pvalue_name = \"padj\"\n",
    "else:\n",
    "    logFC_name = \"logFC\"\n",
    "    pvalue_name = \"adj.P.Val\"\n",
    "\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    "    logFC_name,\n",
    "    pvalue_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a9ea6ba3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_304</th>\n",
       "      <td>G_304</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>634.0</td>\n",
       "      <td>0.241100</td>\n",
       "      <td>0.241100</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>509.0</td>\n",
       "      <td>50.850851</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.027875</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>2</td>\n",
       "      <td>149949.993472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_409</th>\n",
       "      <td>G_409</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>821.0</td>\n",
       "      <td>0.372228</td>\n",
       "      <td>-0.372228</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>226.0</td>\n",
       "      <td>22.522523</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015319</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>2</td>\n",
       "      <td>12726.785644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_718</th>\n",
       "      <td>G_718</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>989.0</td>\n",
       "      <td>0.850727</td>\n",
       "      <td>0.850727</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>62.0</td>\n",
       "      <td>6.106106</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007334</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>2</td>\n",
       "      <td>4455.762233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_425</th>\n",
       "      <td>G_425</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>823.0</td>\n",
       "      <td>0.373348</td>\n",
       "      <td>-0.373348</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>348.0</td>\n",
       "      <td>34.734735</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019537</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>2</td>\n",
       "      <td>3286.667271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_72</th>\n",
       "      <td>G_72</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>399.0</td>\n",
       "      <td>0.134655</td>\n",
       "      <td>-0.134655</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>296.0</td>\n",
       "      <td>29.529530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017578</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>2</td>\n",
       "      <td>3105.611932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_650</th>\n",
       "      <td>G_650</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>644.0</td>\n",
       "      <td>0.250344</td>\n",
       "      <td>0.250344</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>188.0</td>\n",
       "      <td>18.718719</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.013463</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>2</td>\n",
       "      <td>2342.390442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_638</th>\n",
       "      <td>G_638</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>957.0</td>\n",
       "      <td>0.637908</td>\n",
       "      <td>-0.637908</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>947.0</td>\n",
       "      <td>94.694695</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.077306</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>2</td>\n",
       "      <td>1871.061021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_454</th>\n",
       "      <td>G_454</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>963.0</td>\n",
       "      <td>0.674718</td>\n",
       "      <td>-0.674718</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>932.0</td>\n",
       "      <td>93.193193</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.071221</td>\n",
       "      <td>0.000349</td>\n",
       "      <td>2</td>\n",
       "      <td>1727.039845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_822</th>\n",
       "      <td>G_822</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>513.0</td>\n",
       "      <td>0.179884</td>\n",
       "      <td>-0.179884</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.100100</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001759</td>\n",
       "      <td>0.000136</td>\n",
       "      <td>2</td>\n",
       "      <td>1313.896804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_829</th>\n",
       "      <td>G_829</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>994.0</td>\n",
       "      <td>0.938991</td>\n",
       "      <td>-0.938991</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>613.0</td>\n",
       "      <td>61.261261</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.032964</td>\n",
       "      <td>0.000724</td>\n",
       "      <td>2</td>\n",
       "      <td>1251.903427</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "G_304   G_304            0.998894        634.0                    0.241100   \n",
       "G_409   G_409            0.998894        821.0                    0.372228   \n",
       "G_718   G_718            0.998894        989.0                    0.850727   \n",
       "G_425   G_425            0.998894        823.0                    0.373348   \n",
       "G_72     G_72            0.998894        399.0                    0.134655   \n",
       "G_650   G_650            0.998894        644.0                    0.250344   \n",
       "G_638   G_638            0.998894        957.0                    0.637908   \n",
       "G_454   G_454            0.998894        963.0                    0.674718   \n",
       "G_822   G_822            0.998894        513.0                    0.179884   \n",
       "G_829   G_829            0.998894        994.0                    0.938991   \n",
       "\n",
       "       log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "G_304               0.241100                        0.999851   \n",
       "G_409              -0.372228                        0.999851   \n",
       "G_718               0.850727                        0.999851   \n",
       "G_425              -0.373348                        0.999851   \n",
       "G_72               -0.134655                        0.999851   \n",
       "G_650               0.250344                        0.999851   \n",
       "G_638              -0.637908                        0.999851   \n",
       "G_454              -0.674718                        0.999851   \n",
       "G_822              -0.179884                        0.999851   \n",
       "G_829              -0.938991                        0.999851   \n",
       "\n",
       "       Rank (simulated)  Percentile (simulated)  Percent DE (simulated)  \\\n",
       "G_304             509.0               50.850851                     0.0   \n",
       "G_409             226.0               22.522523                     0.0   \n",
       "G_718              62.0                6.106106                     0.0   \n",
       "G_425             348.0               34.734735                     0.0   \n",
       "G_72              296.0               29.529530                     0.0   \n",
       "G_650             188.0               18.718719                     0.0   \n",
       "G_638             947.0               94.694695                     0.0   \n",
       "G_454             932.0               93.193193                     0.0   \n",
       "G_822               2.0                0.100100                     0.0   \n",
       "G_829             613.0               61.261261                     0.0   \n",
       "\n",
       "       Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "G_304                              0.027875                   0.000001   \n",
       "G_409                              0.015319                   0.000028   \n",
       "G_718                              0.007334                   0.000189   \n",
       "G_425                              0.019537                   0.000108   \n",
       "G_72                               0.017578                   0.000038   \n",
       "G_650                              0.013463                   0.000101   \n",
       "G_638                              0.077306                   0.000300   \n",
       "G_454                              0.071221                   0.000349   \n",
       "G_822                              0.001759                   0.000136   \n",
       "G_829                              0.032964                   0.000724   \n",
       "\n",
       "       Number of experiments (simulated)        Z score  \n",
       "G_304                                  2  149949.993472  \n",
       "G_409                                  2   12726.785644  \n",
       "G_718                                  2    4455.762233  \n",
       "G_425                                  2    3286.667271  \n",
       "G_72                                   2    3105.611932  \n",
       "G_650                                  2    2342.390442  \n",
       "G_638                                  2    1871.061021  \n",
       "G_454                                  2    1727.039845  \n",
       "G_822                                  2    1313.896804  \n",
       "G_829                                  2    1251.903427  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get summary table\n",
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    \"gene\",\n",
    "    params,\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eb3fcde8",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_gene_ranks_sorted = summary_gene_ranks.sort_values(\n",
    "    by=\"Z score\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a69d2399",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ranking based on Z-score\n",
    "summary_gene_ranks_sorted[\"rank\"] = summary_gene_ranks_sorted[\"Z score\"].rank(\n",
    "    ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7e668e37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_304</th>\n",
       "      <td>G_304</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>634.0</td>\n",
       "      <td>0.241100</td>\n",
       "      <td>0.241100</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>509.0</td>\n",
       "      <td>50.850851</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.027875</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>2</td>\n",
       "      <td>149949.993472</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_409</th>\n",
       "      <td>G_409</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>821.0</td>\n",
       "      <td>0.372228</td>\n",
       "      <td>-0.372228</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>226.0</td>\n",
       "      <td>22.522523</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015319</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>2</td>\n",
       "      <td>12726.785644</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_718</th>\n",
       "      <td>G_718</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>989.0</td>\n",
       "      <td>0.850727</td>\n",
       "      <td>0.850727</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>62.0</td>\n",
       "      <td>6.106106</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007334</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>2</td>\n",
       "      <td>4455.762233</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_425</th>\n",
       "      <td>G_425</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>823.0</td>\n",
       "      <td>0.373348</td>\n",
       "      <td>-0.373348</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>348.0</td>\n",
       "      <td>34.734735</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019537</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>2</td>\n",
       "      <td>3286.667271</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_72</th>\n",
       "      <td>G_72</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>399.0</td>\n",
       "      <td>0.134655</td>\n",
       "      <td>-0.134655</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>296.0</td>\n",
       "      <td>29.529530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017578</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>2</td>\n",
       "      <td>3105.611932</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_650</th>\n",
       "      <td>G_650</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>644.0</td>\n",
       "      <td>0.250344</td>\n",
       "      <td>0.250344</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>188.0</td>\n",
       "      <td>18.718719</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.013463</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>2</td>\n",
       "      <td>2342.390442</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_638</th>\n",
       "      <td>G_638</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>957.0</td>\n",
       "      <td>0.637908</td>\n",
       "      <td>-0.637908</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>947.0</td>\n",
       "      <td>94.694695</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.077306</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>2</td>\n",
       "      <td>1871.061021</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_454</th>\n",
       "      <td>G_454</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>963.0</td>\n",
       "      <td>0.674718</td>\n",
       "      <td>-0.674718</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>932.0</td>\n",
       "      <td>93.193193</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.071221</td>\n",
       "      <td>0.000349</td>\n",
       "      <td>2</td>\n",
       "      <td>1727.039845</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_822</th>\n",
       "      <td>G_822</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>513.0</td>\n",
       "      <td>0.179884</td>\n",
       "      <td>-0.179884</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.100100</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001759</td>\n",
       "      <td>0.000136</td>\n",
       "      <td>2</td>\n",
       "      <td>1313.896804</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_829</th>\n",
       "      <td>G_829</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>994.0</td>\n",
       "      <td>0.938991</td>\n",
       "      <td>-0.938991</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>613.0</td>\n",
       "      <td>61.261261</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.032964</td>\n",
       "      <td>0.000724</td>\n",
       "      <td>2</td>\n",
       "      <td>1251.903427</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "G_304   G_304            0.998894        634.0                    0.241100   \n",
       "G_409   G_409            0.998894        821.0                    0.372228   \n",
       "G_718   G_718            0.998894        989.0                    0.850727   \n",
       "G_425   G_425            0.998894        823.0                    0.373348   \n",
       "G_72     G_72            0.998894        399.0                    0.134655   \n",
       "G_650   G_650            0.998894        644.0                    0.250344   \n",
       "G_638   G_638            0.998894        957.0                    0.637908   \n",
       "G_454   G_454            0.998894        963.0                    0.674718   \n",
       "G_822   G_822            0.998894        513.0                    0.179884   \n",
       "G_829   G_829            0.998894        994.0                    0.938991   \n",
       "\n",
       "       log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "G_304               0.241100                        0.999851   \n",
       "G_409              -0.372228                        0.999851   \n",
       "G_718               0.850727                        0.999851   \n",
       "G_425              -0.373348                        0.999851   \n",
       "G_72               -0.134655                        0.999851   \n",
       "G_650               0.250344                        0.999851   \n",
       "G_638              -0.637908                        0.999851   \n",
       "G_454              -0.674718                        0.999851   \n",
       "G_822              -0.179884                        0.999851   \n",
       "G_829              -0.938991                        0.999851   \n",
       "\n",
       "       Rank (simulated)  Percentile (simulated)  Percent DE (simulated)  \\\n",
       "G_304             509.0               50.850851                     0.0   \n",
       "G_409             226.0               22.522523                     0.0   \n",
       "G_718              62.0                6.106106                     0.0   \n",
       "G_425             348.0               34.734735                     0.0   \n",
       "G_72              296.0               29.529530                     0.0   \n",
       "G_650             188.0               18.718719                     0.0   \n",
       "G_638             947.0               94.694695                     0.0   \n",
       "G_454             932.0               93.193193                     0.0   \n",
       "G_822               2.0                0.100100                     0.0   \n",
       "G_829             613.0               61.261261                     0.0   \n",
       "\n",
       "       Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "G_304                              0.027875                   0.000001   \n",
       "G_409                              0.015319                   0.000028   \n",
       "G_718                              0.007334                   0.000189   \n",
       "G_425                              0.019537                   0.000108   \n",
       "G_72                               0.017578                   0.000038   \n",
       "G_650                              0.013463                   0.000101   \n",
       "G_638                              0.077306                   0.000300   \n",
       "G_454                              0.071221                   0.000349   \n",
       "G_822                              0.001759                   0.000136   \n",
       "G_829                              0.032964                   0.000724   \n",
       "\n",
       "       Number of experiments (simulated)        Z score  rank  \n",
       "G_304                                  2  149949.993472   1.0  \n",
       "G_409                                  2   12726.785644   2.0  \n",
       "G_718                                  2    4455.762233   3.0  \n",
       "G_425                                  2    3286.667271   4.0  \n",
       "G_72                                   2    3105.611932   5.0  \n",
       "G_650                                  2    2342.390442   6.0  \n",
       "G_638                                  2    1871.061021   7.0  \n",
       "G_454                                  2    1727.039845   8.0  \n",
       "G_822                                  2    1313.896804   9.0  \n",
       "G_829                                  2    1251.903427  10.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31f20d47",
   "metadata": {},
   "source": [
    "## Traditional DE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c9fad72e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (de_method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9be7e0bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load DE statistics file\n",
    "trad_de_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "trad_de_stats = pd.read_csv(trad_de_stats_filename, sep=\"\\t\", index_col=0, header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e7df4273",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort by log fold change\n",
    "trad_de_stats_sorted = trad_de_stats.sort_values(by=\"log2FoldChange\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c4c22916",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ranking based on log2FoldChange\n",
    "trad_de_stats_sorted[\"rank\"] = trad_de_stats_sorted[\"log2FoldChange\"].rank(\n",
    "    ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b13768f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_513</th>\n",
       "      <td>68.425319</td>\n",
       "      <td>1.030356</td>\n",
       "      <td>1.286037</td>\n",
       "      <td>0.801187</td>\n",
       "      <td>0.423023</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_753</th>\n",
       "      <td>68.934742</td>\n",
       "      <td>0.857123</td>\n",
       "      <td>1.261278</td>\n",
       "      <td>0.679567</td>\n",
       "      <td>0.496779</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_718</th>\n",
       "      <td>71.891717</td>\n",
       "      <td>0.850727</td>\n",
       "      <td>1.145588</td>\n",
       "      <td>0.742612</td>\n",
       "      <td>0.457717</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_135</th>\n",
       "      <td>79.934183</td>\n",
       "      <td>0.832334</td>\n",
       "      <td>0.907726</td>\n",
       "      <td>0.916944</td>\n",
       "      <td>0.359172</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_86</th>\n",
       "      <td>67.862238</td>\n",
       "      <td>0.801011</td>\n",
       "      <td>1.305851</td>\n",
       "      <td>0.613401</td>\n",
       "      <td>0.539611</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_633</th>\n",
       "      <td>76.890754</td>\n",
       "      <td>0.797921</td>\n",
       "      <td>0.985482</td>\n",
       "      <td>0.809676</td>\n",
       "      <td>0.418126</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_705</th>\n",
       "      <td>68.785660</td>\n",
       "      <td>0.787412</td>\n",
       "      <td>1.264556</td>\n",
       "      <td>0.622678</td>\n",
       "      <td>0.533496</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_545</th>\n",
       "      <td>74.512134</td>\n",
       "      <td>0.778864</td>\n",
       "      <td>1.057420</td>\n",
       "      <td>0.736570</td>\n",
       "      <td>0.461384</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_16</th>\n",
       "      <td>74.315754</td>\n",
       "      <td>0.757656</td>\n",
       "      <td>1.061977</td>\n",
       "      <td>0.713439</td>\n",
       "      <td>0.475574</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_555</th>\n",
       "      <td>68.282549</td>\n",
       "      <td>0.747342</td>\n",
       "      <td>1.285940</td>\n",
       "      <td>0.581164</td>\n",
       "      <td>0.561130</td>\n",
       "      <td>0.998894</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        baseMean  log2FoldChange     lfcSE      stat    pvalue      padj  rank\n",
       "G_513  68.425319        1.030356  1.286037  0.801187  0.423023  0.998894   1.0\n",
       "G_753  68.934742        0.857123  1.261278  0.679567  0.496779  0.998894   2.0\n",
       "G_718  71.891717        0.850727  1.145588  0.742612  0.457717  0.998894   3.0\n",
       "G_135  79.934183        0.832334  0.907726  0.916944  0.359172  0.998894   4.0\n",
       "G_86   67.862238        0.801011  1.305851  0.613401  0.539611  0.998894   5.0\n",
       "G_633  76.890754        0.797921  0.985482  0.809676  0.418126  0.998894   6.0\n",
       "G_705  68.785660        0.787412  1.264556  0.622678  0.533496  0.998894   7.0\n",
       "G_545  74.512134        0.778864  1.057420  0.736570  0.461384  0.998894   8.0\n",
       "G_16   74.315754        0.757656  1.061977  0.713439  0.475574  0.998894   9.0\n",
       "G_555  68.282549        0.747342  1.285940  0.581164  0.561130  0.998894  10.0"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcc042c5",
   "metadata": {},
   "source": [
    "## Compare\n",
    "\n",
    "1. mean rank of specific genes - mean rank of generic genes for template experiment\n",
    "2. We will need to re-run this notebook for each template experiment and then plot the distribution of difference scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3c68482d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load pickled file\n",
    "with open(template_specific_gene_ids_filename, \"rb\") as specific_fh:\n",
    "    specific_gene_ids = pickle.load(specific_fh)\n",
    "\n",
    "with open(generic_gene_ids_filename, \"rb\") as generic_fh:\n",
    "    generic_gene_ids = pickle.load(generic_fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bc3907e3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get mean of specific gene ranks\n",
    "sophie_specific_mean = summary_gene_ranks_sorted.loc[specific_gene_ids, \"rank\"].mean()\n",
    "trad_specific_mean = trad_de_stats_sorted.loc[specific_gene_ids, \"rank\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3067e55e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_456    233.0\n",
       "G_692    487.0\n",
       "G_800     16.0\n",
       "G_504    710.0\n",
       "G_846    229.0\n",
       "G_833     23.0\n",
       "G_269    786.0\n",
       "G_19     227.0\n",
       "G_387    784.0\n",
       "G_476    242.0\n",
       "Name: rank, dtype: float64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.loc[specific_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "735497d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_456     14.0\n",
       "G_692    745.0\n",
       "G_800     62.0\n",
       "G_504    321.0\n",
       "G_846    729.0\n",
       "G_833    836.0\n",
       "G_269    367.0\n",
       "G_19     306.0\n",
       "G_387    580.0\n",
       "G_476    900.0\n",
       "Name: rank, dtype: float64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.loc[specific_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "10b4c68d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get mean of generic gene ranks\n",
    "sophie_generic_mean = summary_gene_ranks_sorted.loc[generic_gene_ids, \"rank\"].mean()\n",
    "trad_generic_mean = trad_de_stats_sorted.loc[generic_gene_ids, \"rank\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d50a8569",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_368    872.0\n",
       "G_809    597.0\n",
       "G_961    744.0\n",
       "G_664    102.0\n",
       "G_750    507.0\n",
       "G_785    614.0\n",
       "G_554     19.0\n",
       "G_585    666.0\n",
       "G_287    336.0\n",
       "G_746    273.0\n",
       "G_636    602.0\n",
       "G_592    539.0\n",
       "G_494    652.0\n",
       "G_916     35.0\n",
       "G_830    904.0\n",
       "G_169    352.0\n",
       "G_463    470.0\n",
       "G_787    819.0\n",
       "G_466    434.0\n",
       "G_967    277.0\n",
       "G_956    157.0\n",
       "G_61     303.0\n",
       "G_469    278.0\n",
       "G_210    632.0\n",
       "G_983    237.0\n",
       "G_426    555.0\n",
       "G_189    781.0\n",
       "G_706    716.0\n",
       "G_662    885.0\n",
       "G_23     354.0\n",
       "         ...  \n",
       "G_667    215.0\n",
       "G_721    275.0\n",
       "G_910    358.0\n",
       "G_874    919.0\n",
       "G_252    702.0\n",
       "G_179    768.0\n",
       "G_259    844.0\n",
       "G_717    505.0\n",
       "G_515    953.0\n",
       "G_461    587.0\n",
       "G_121    579.0\n",
       "G_251    601.0\n",
       "G_812    635.0\n",
       "G_999    812.0\n",
       "G_219    461.0\n",
       "G_363    385.0\n",
       "G_425      4.0\n",
       "G_875    825.0\n",
       "G_489    307.0\n",
       "G_539    570.0\n",
       "G_938    980.0\n",
       "G_253    261.0\n",
       "G_583    322.0\n",
       "G_770     38.0\n",
       "G_280    508.0\n",
       "G_864    894.0\n",
       "G_724    588.0\n",
       "G_949    177.0\n",
       "G_274    967.0\n",
       "G_352     75.0\n",
       "Name: rank, Length: 100, dtype: float64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.loc[generic_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "839f4a81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_368    477.0\n",
       "G_809    932.0\n",
       "G_961    578.0\n",
       "G_664    121.0\n",
       "G_750    287.0\n",
       "G_785    730.0\n",
       "G_554    122.0\n",
       "G_585    717.0\n",
       "G_287    156.0\n",
       "G_746    195.0\n",
       "G_636    180.0\n",
       "G_592    201.0\n",
       "G_494    718.0\n",
       "G_916    123.0\n",
       "G_830    465.0\n",
       "G_169    961.0\n",
       "G_463    225.0\n",
       "G_787    629.0\n",
       "G_466    840.0\n",
       "G_967    731.0\n",
       "G_956     41.0\n",
       "G_61     679.0\n",
       "G_469     84.0\n",
       "G_210    301.0\n",
       "G_983    105.0\n",
       "G_426    851.0\n",
       "G_189    277.0\n",
       "G_706    204.0\n",
       "G_662    438.0\n",
       "G_23     959.0\n",
       "         ...  \n",
       "G_667    310.0\n",
       "G_721     16.0\n",
       "G_910    960.0\n",
       "G_874    483.0\n",
       "G_252    598.0\n",
       "G_179    622.0\n",
       "G_259    364.0\n",
       "G_717    390.0\n",
       "G_515    593.0\n",
       "G_461    194.0\n",
       "G_121    777.0\n",
       "G_251    243.0\n",
       "G_812    738.0\n",
       "G_999    482.0\n",
       "G_219     93.0\n",
       "G_363     73.0\n",
       "G_425    914.0\n",
       "G_875    387.0\n",
       "G_489    241.0\n",
       "G_539    240.0\n",
       "G_938    334.0\n",
       "G_253    974.0\n",
       "G_583    165.0\n",
       "G_770    969.0\n",
       "G_280    271.0\n",
       "G_864    564.0\n",
       "G_724    305.0\n",
       "G_949    200.0\n",
       "G_274    490.0\n",
       "G_352     98.0\n",
       "Name: rank, Length: 100, dtype: float64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.loc[generic_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8c08e343",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sophie difference:  -124.15000000000003\n",
      "traditional difference:  24.319999999999993\n"
     ]
    }
   ],
   "source": [
    "# Difference\n",
    "diff_sophie = sophie_specific_mean - sophie_generic_mean\n",
    "diff_trad = trad_specific_mean - trad_generic_mean\n",
    "\n",
    "print(\"sophie difference: \", diff_sophie)\n",
    "print(\"traditional difference: \", diff_trad)"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
