{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0b54cfe2",
   "metadata": {},
   "source": [
    "# Evaluation\n",
    "\n",
    "How often do regular differential expression analysis vs sophie prioritize the specific vs generic genes?\n",
    "\n",
    "1. Simulate 1 template perturbation experiment using the technique above\n",
    "2. Apply SOPHIE to get ranking of specific and generic genes based on their z-score.\n",
    "3. Apply traditional DE analysis and get ranking of specific and generic genes based on their log fold change value\n",
    "4. Compare the difference in ranking between specific and generic genes using SOPHIE vs traditional metrics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b68f083",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:21: The name tf.ConfigProto is deprecated. Please use tf.compat.v1.ConfigProto instead.\n",
      "\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:25: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.\n",
      "\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:25: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext rpy2.ipython\n",
    "import os\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from rpy2.robjects import pandas2ri\n",
    "from ponyo import utils, train_vae_modules, simulate_expression_data\n",
    "from generic_expression_patterns_modules import (\n",
    "    process,\n",
    "    new_experiment_process,  # REMOVE\n",
    "    stats,\n",
    "    ranking,\n",
    ")\n",
    "\n",
    "np.random.seed(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4fa09de1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = \"config_sophie_vs_trad.tsv\"\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6e9a766",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load config params\n",
    "\n",
    "# Local directory to store intermediate files\n",
    "local_dir = params[\"local_dir\"]\n",
    "\n",
    "#\n",
    "dataset_name = params[\"dataset_name\"]\n",
    "\n",
    "# File containing un-normalized template experiment\n",
    "raw_template_filename = params[\"raw_template_filename\"]\n",
    "\n",
    "# Un-normalized compendium filename\n",
    "raw_compendium_filename = params[\"raw_compendium_filename\"]\n",
    "\n",
    "# Normalized compendium filename\n",
    "normalized_compendium_filename = params[\"normalized_compendium_filename\"]\n",
    "\n",
    "# ID for template experiment to be selected\n",
    "project_id = params[\"project_id\"]\n",
    "\n",
    "# Number of simulated experiments to generate\n",
    "num_runs = params[\"num_simulated\"]\n",
    "\n",
    "# Directory containing trained VAE model\n",
    "vae_model_dir = params[\"vae_model_dir\"]\n",
    "\n",
    "# Size of the latent dimension\n",
    "latent_dim = params[\"latent_dim\"]\n",
    "\n",
    "# Scaler transform used to scale compendium data into 0-1 range for training\n",
    "scaler_filename = params[\"scaler_filename\"]\n",
    "\n",
    "# Which DE method to use\n",
    "# We recommend that if data is RNA-seq then use DESeq2\n",
    "# If data is microarray then use Limma\n",
    "de_method = params[\"DE_method\"]\n",
    "\n",
    "# If using DE-seq, setting this parameter will\n",
    "# remove genes below a certain threshold\n",
    "count_threshold = params[\"count_threshold\"]\n",
    "\n",
    "# Metadata file that specifies which samples to keep for DE analysis (Optional)\n",
    "template_process_samples_filename = params[\"template_process_samples_filename\"]\n",
    "\n",
    "# Metadata file that specifies sample grouping for DE analysis\n",
    "template_DE_grouping_filename = params[\"template_DE_grouping_filename\"]\n",
    "\n",
    "# Statistic to use to rank genes or pathways by\n",
    "# Choices are {} FILL IN\n",
    "col_to_rank_genes = params[\"rank_genes_by\"]\n",
    "\n",
    "# Pickle files saving specific and generic gene ids\n",
    "template_specific_gene_ids_filename = params[\"template_specific_gene_ids_filename\"]\n",
    "generic_gene_ids_filename = \"generic_gene_ids.pickle\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5fccae1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files generated by this notebook\n",
    "\n",
    "# File storing template experiment with gene ids mapped to compendium gene ids\n",
    "mapped_template_filename = params[\"mapped_template_filename\"]\n",
    "\n",
    "# File storing normalized template experiment\n",
    "normalized_template_filename = params[\"normalized_template_filename\"]\n",
    "\n",
    "# File storing processed template experiment,\n",
    "# after samples have been selected for comparison in DE analysis\n",
    "processed_template_filename = params[\"processed_template_filename\"]\n",
    "\n",
    "# Output summary file\n",
    "output_filename = params[\"output_filename\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f10a5605",
   "metadata": {},
   "source": [
    "## SOPHIE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "69f51a6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4, 1000)\n",
      "(360, 1000)\n"
     ]
    }
   ],
   "source": [
    "# Process template\n",
    "new_experiment_process.process_template_experiment(\n",
    "    raw_template_filename,\n",
    "    raw_compendium_filename,\n",
    "    scaler_filename,\n",
    "    mapped_template_filename,\n",
    "    normalized_template_filename,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0e48cdb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate multiple experiments UPDATE COMMENT\n",
    "# This step creates the following files in \"<local_dir>/pseudo_experiment/\" directory:\n",
    "#   - selected_simulated_data_SRP012656_<n>.txt\n",
    "#   - selected_simulated_encoded_data_SRP012656_<n>.txt\n",
    "#   - template_normalized_data_SRP012656_test.txt\n",
    "# in which \"<n>\" is an integer in the range of [0, num_runs-1]\n",
    "\n",
    "# REMOVE LATER\n",
    "# dataset_name = \"pre_model_unseen_template\"\n",
    "# Load pickled file\n",
    "scaler = pickle.load(open(scaler_filename, \"rb\"))\n",
    "\n",
    "# Update simulated dir\n",
    "os.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)\n",
    "\n",
    "# Update to take in file to be consisten\n",
    "normalized_compendium = pd.read_csv(\n",
    "    normalized_compendium_filename, header=0, sep=\"\\t\", index_col=0\n",
    ")\n",
    "normalized_template = pd.read_csv(\n",
    "    normalized_template_filename, header=0, sep=\"\\t\", index_col=0\n",
    ")\n",
    "# ------------\n",
    "# Update call when new version of ponyo\n",
    "for run_id in range(num_runs):\n",
    "    new_experiment_process.embed_shift_template_experiment(\n",
    "        normalized_compendium,\n",
    "        normalized_template,\n",
    "        vae_model_dir,\n",
    "        project_id,\n",
    "        scaler_filename,\n",
    "        local_dir,\n",
    "        latent_dim,\n",
    "        run_id,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "09bf6c28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "## Update simulated dir\n",
    "if not os.path.exists(template_process_samples_filename):\n",
    "    template_process_samples_filename = None\n",
    "\n",
    "if de_method == \"deseq\":\n",
    "    # Process template data\n",
    "    stats.process_samples_for_DESeq(\n",
    "        raw_template_filename,\n",
    "        template_DE_grouping_filename,\n",
    "        processed_template_filename,\n",
    "        count_threshold,\n",
    "        template_process_samples_filename,\n",
    "    )\n",
    "\n",
    "    # Process simulated data\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        out_simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}_processed.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_DESeq(\n",
    "            simulated_filename,\n",
    "            template_DE_grouping_filename,\n",
    "            out_simulated_filename,\n",
    "            count_threshold,\n",
    "            template_process_samples_filename,\n",
    "        )\n",
    "else:\n",
    "    stats.process_samples_for_limma(\n",
    "        raw_template_filename,\n",
    "        template_DE_grouping_filename,\n",
    "        processed_template_filename,\n",
    "        template_process_samples_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_limma(\n",
    "            simulated_filename,\n",
    "            template_DE_grouping_filename,\n",
    "            None,\n",
    "            template_process_samples_filename,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c56b91df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ee7b0585",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: DelayedArray\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: BiocParallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (de_method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0dee119f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i base_dir -i local_dir -i num_runs -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# Files created: \"<local_dir>/DE_stats/DE_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\",\n",
    "        i,\n",
    "        \"_processed.txt\",\n",
    "        sep = \"\"\n",
    "    )\n",
    "    if (de_method == \"deseq\"){\n",
    "        get_DE_stats_DESeq(\n",
    "            template_DE_grouping_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "    }\n",
    "    else {\n",
    "        get_DE_stats_limma(\n",
    "            template_DE_grouping_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2c0e853b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (2 levels on the left, 1 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "# Added\n",
    "if de_method == \"deseq\":\n",
    "    logFC_name = \"log2FoldChange\"\n",
    "    pvalue_name = \"padj\"\n",
    "else:\n",
    "    logFC_name = \"logFC\"\n",
    "    pvalue_name = \"adj.P.Val\"\n",
    "\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    "    logFC_name,\n",
    "    pvalue_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a9ea6ba3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_137</th>\n",
       "      <td>G_137</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>998.0</td>\n",
       "      <td>0.183802</td>\n",
       "      <td>-0.183802</td>\n",
       "      <td>0.942993</td>\n",
       "      <td>267.0</td>\n",
       "      <td>26.626627</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.045722</td>\n",
       "      <td>0.030321</td>\n",
       "      <td>10</td>\n",
       "      <td>4.553959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_696</th>\n",
       "      <td>G_696</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>999.0</td>\n",
       "      <td>0.194072</td>\n",
       "      <td>-0.194072</td>\n",
       "      <td>0.928404</td>\n",
       "      <td>166.0</td>\n",
       "      <td>16.516517</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.060837</td>\n",
       "      <td>0.033908</td>\n",
       "      <td>10</td>\n",
       "      <td>3.929277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_564</th>\n",
       "      <td>G_564</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>961.0</td>\n",
       "      <td>0.129410</td>\n",
       "      <td>0.129410</td>\n",
       "      <td>0.970573</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.901902</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.031790</td>\n",
       "      <td>0.025275</td>\n",
       "      <td>10</td>\n",
       "      <td>3.862334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_328</th>\n",
       "      <td>G_328</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>980.0</td>\n",
       "      <td>0.145465</td>\n",
       "      <td>-0.145465</td>\n",
       "      <td>0.936806</td>\n",
       "      <td>58.0</td>\n",
       "      <td>5.705706</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.039655</td>\n",
       "      <td>0.029553</td>\n",
       "      <td>10</td>\n",
       "      <td>3.580394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_286</th>\n",
       "      <td>G_286</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>984.0</td>\n",
       "      <td>0.148703</td>\n",
       "      <td>-0.148703</td>\n",
       "      <td>0.891810</td>\n",
       "      <td>231.0</td>\n",
       "      <td>23.023023</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.054301</td>\n",
       "      <td>0.032582</td>\n",
       "      <td>10</td>\n",
       "      <td>2.897330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_875</th>\n",
       "      <td>G_875</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>964.0</td>\n",
       "      <td>0.133251</td>\n",
       "      <td>-0.133251</td>\n",
       "      <td>0.899590</td>\n",
       "      <td>191.0</td>\n",
       "      <td>19.019019</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.053604</td>\n",
       "      <td>0.028587</td>\n",
       "      <td>10</td>\n",
       "      <td>2.786105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_521</th>\n",
       "      <td>G_521</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>983.0</td>\n",
       "      <td>0.148425</td>\n",
       "      <td>0.148425</td>\n",
       "      <td>0.944842</td>\n",
       "      <td>292.0</td>\n",
       "      <td>29.129129</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.047234</td>\n",
       "      <td>0.036612</td>\n",
       "      <td>10</td>\n",
       "      <td>2.763869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_618</th>\n",
       "      <td>G_618</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>977.0</td>\n",
       "      <td>0.142026</td>\n",
       "      <td>-0.142026</td>\n",
       "      <td>0.922467</td>\n",
       "      <td>306.0</td>\n",
       "      <td>30.530531</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.056194</td>\n",
       "      <td>0.033217</td>\n",
       "      <td>10</td>\n",
       "      <td>2.583978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_8</th>\n",
       "      <td>G_8</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>968.0</td>\n",
       "      <td>0.134471</td>\n",
       "      <td>0.134471</td>\n",
       "      <td>0.900976</td>\n",
       "      <td>125.0</td>\n",
       "      <td>12.412412</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.053907</td>\n",
       "      <td>0.031843</td>\n",
       "      <td>10</td>\n",
       "      <td>2.530033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_913</th>\n",
       "      <td>G_913</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>987.0</td>\n",
       "      <td>0.151699</td>\n",
       "      <td>0.151699</td>\n",
       "      <td>0.946122</td>\n",
       "      <td>107.0</td>\n",
       "      <td>10.610611</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.052862</td>\n",
       "      <td>0.039358</td>\n",
       "      <td>10</td>\n",
       "      <td>2.511202</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "G_137   G_137            0.998503        998.0                    0.183802   \n",
       "G_696   G_696            0.998503        999.0                    0.194072   \n",
       "G_564   G_564            0.998503        961.0                    0.129410   \n",
       "G_328   G_328            0.998503        980.0                    0.145465   \n",
       "G_286   G_286            0.998503        984.0                    0.148703   \n",
       "G_875   G_875            0.998503        964.0                    0.133251   \n",
       "G_521   G_521            0.998503        983.0                    0.148425   \n",
       "G_618   G_618            0.998503        977.0                    0.142026   \n",
       "G_8       G_8            0.998503        968.0                    0.134471   \n",
       "G_913   G_913            0.998503        987.0                    0.151699   \n",
       "\n",
       "       log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "G_137              -0.183802                        0.942993   \n",
       "G_696              -0.194072                        0.928404   \n",
       "G_564               0.129410                        0.970573   \n",
       "G_328              -0.145465                        0.936806   \n",
       "G_286              -0.148703                        0.891810   \n",
       "G_875              -0.133251                        0.899590   \n",
       "G_521               0.148425                        0.944842   \n",
       "G_618              -0.142026                        0.922467   \n",
       "G_8                 0.134471                        0.900976   \n",
       "G_913               0.151699                        0.946122   \n",
       "\n",
       "       Rank (simulated)  Percentile (simulated)  Percent DE (simulated)  \\\n",
       "G_137             267.0               26.626627                     0.0   \n",
       "G_696             166.0               16.516517                     0.0   \n",
       "G_564              20.0                1.901902                     0.0   \n",
       "G_328              58.0                5.705706                     0.0   \n",
       "G_286             231.0               23.023023                     0.0   \n",
       "G_875             191.0               19.019019                     0.0   \n",
       "G_521             292.0               29.129129                     0.0   \n",
       "G_618             306.0               30.530531                     0.0   \n",
       "G_8               125.0               12.412412                     0.0   \n",
       "G_913             107.0               10.610611                     0.0   \n",
       "\n",
       "       Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "G_137                              0.045722                   0.030321   \n",
       "G_696                              0.060837                   0.033908   \n",
       "G_564                              0.031790                   0.025275   \n",
       "G_328                              0.039655                   0.029553   \n",
       "G_286                              0.054301                   0.032582   \n",
       "G_875                              0.053604                   0.028587   \n",
       "G_521                              0.047234                   0.036612   \n",
       "G_618                              0.056194                   0.033217   \n",
       "G_8                                0.053907                   0.031843   \n",
       "G_913                              0.052862                   0.039358   \n",
       "\n",
       "       Number of experiments (simulated)   Z score  \n",
       "G_137                                 10  4.553959  \n",
       "G_696                                 10  3.929277  \n",
       "G_564                                 10  3.862334  \n",
       "G_328                                 10  3.580394  \n",
       "G_286                                 10  2.897330  \n",
       "G_875                                 10  2.786105  \n",
       "G_521                                 10  2.763869  \n",
       "G_618                                 10  2.583978  \n",
       "G_8                                   10  2.530033  \n",
       "G_913                                 10  2.511202  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get summary table\n",
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    \"gene\",\n",
    "    params,\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eb3fcde8",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_gene_ranks_sorted = summary_gene_ranks.sort_values(\n",
    "    by=\"Z score\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a69d2399",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ranking based on Z-score\n",
    "summary_gene_ranks_sorted[\"rank\"] = summary_gene_ranks_sorted[\"Z score\"].rank(\n",
    "    ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7e668e37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_137</th>\n",
       "      <td>G_137</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>998.0</td>\n",
       "      <td>0.183802</td>\n",
       "      <td>-0.183802</td>\n",
       "      <td>0.942993</td>\n",
       "      <td>267.0</td>\n",
       "      <td>26.626627</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.045722</td>\n",
       "      <td>0.030321</td>\n",
       "      <td>10</td>\n",
       "      <td>4.553959</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_696</th>\n",
       "      <td>G_696</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>999.0</td>\n",
       "      <td>0.194072</td>\n",
       "      <td>-0.194072</td>\n",
       "      <td>0.928404</td>\n",
       "      <td>166.0</td>\n",
       "      <td>16.516517</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.060837</td>\n",
       "      <td>0.033908</td>\n",
       "      <td>10</td>\n",
       "      <td>3.929277</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_564</th>\n",
       "      <td>G_564</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>961.0</td>\n",
       "      <td>0.129410</td>\n",
       "      <td>0.129410</td>\n",
       "      <td>0.970573</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.901902</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.031790</td>\n",
       "      <td>0.025275</td>\n",
       "      <td>10</td>\n",
       "      <td>3.862334</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_328</th>\n",
       "      <td>G_328</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>980.0</td>\n",
       "      <td>0.145465</td>\n",
       "      <td>-0.145465</td>\n",
       "      <td>0.936806</td>\n",
       "      <td>58.0</td>\n",
       "      <td>5.705706</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.039655</td>\n",
       "      <td>0.029553</td>\n",
       "      <td>10</td>\n",
       "      <td>3.580394</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_286</th>\n",
       "      <td>G_286</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>984.0</td>\n",
       "      <td>0.148703</td>\n",
       "      <td>-0.148703</td>\n",
       "      <td>0.891810</td>\n",
       "      <td>231.0</td>\n",
       "      <td>23.023023</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.054301</td>\n",
       "      <td>0.032582</td>\n",
       "      <td>10</td>\n",
       "      <td>2.897330</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_875</th>\n",
       "      <td>G_875</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>964.0</td>\n",
       "      <td>0.133251</td>\n",
       "      <td>-0.133251</td>\n",
       "      <td>0.899590</td>\n",
       "      <td>191.0</td>\n",
       "      <td>19.019019</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.053604</td>\n",
       "      <td>0.028587</td>\n",
       "      <td>10</td>\n",
       "      <td>2.786105</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_521</th>\n",
       "      <td>G_521</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>983.0</td>\n",
       "      <td>0.148425</td>\n",
       "      <td>0.148425</td>\n",
       "      <td>0.944842</td>\n",
       "      <td>292.0</td>\n",
       "      <td>29.129129</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.047234</td>\n",
       "      <td>0.036612</td>\n",
       "      <td>10</td>\n",
       "      <td>2.763869</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_618</th>\n",
       "      <td>G_618</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>977.0</td>\n",
       "      <td>0.142026</td>\n",
       "      <td>-0.142026</td>\n",
       "      <td>0.922467</td>\n",
       "      <td>306.0</td>\n",
       "      <td>30.530531</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.056194</td>\n",
       "      <td>0.033217</td>\n",
       "      <td>10</td>\n",
       "      <td>2.583978</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_8</th>\n",
       "      <td>G_8</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>968.0</td>\n",
       "      <td>0.134471</td>\n",
       "      <td>0.134471</td>\n",
       "      <td>0.900976</td>\n",
       "      <td>125.0</td>\n",
       "      <td>12.412412</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.053907</td>\n",
       "      <td>0.031843</td>\n",
       "      <td>10</td>\n",
       "      <td>2.530033</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_913</th>\n",
       "      <td>G_913</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>987.0</td>\n",
       "      <td>0.151699</td>\n",
       "      <td>0.151699</td>\n",
       "      <td>0.946122</td>\n",
       "      <td>107.0</td>\n",
       "      <td>10.610611</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.052862</td>\n",
       "      <td>0.039358</td>\n",
       "      <td>10</td>\n",
       "      <td>2.511202</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "G_137   G_137            0.998503        998.0                    0.183802   \n",
       "G_696   G_696            0.998503        999.0                    0.194072   \n",
       "G_564   G_564            0.998503        961.0                    0.129410   \n",
       "G_328   G_328            0.998503        980.0                    0.145465   \n",
       "G_286   G_286            0.998503        984.0                    0.148703   \n",
       "G_875   G_875            0.998503        964.0                    0.133251   \n",
       "G_521   G_521            0.998503        983.0                    0.148425   \n",
       "G_618   G_618            0.998503        977.0                    0.142026   \n",
       "G_8       G_8            0.998503        968.0                    0.134471   \n",
       "G_913   G_913            0.998503        987.0                    0.151699   \n",
       "\n",
       "       log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "G_137              -0.183802                        0.942993   \n",
       "G_696              -0.194072                        0.928404   \n",
       "G_564               0.129410                        0.970573   \n",
       "G_328              -0.145465                        0.936806   \n",
       "G_286              -0.148703                        0.891810   \n",
       "G_875              -0.133251                        0.899590   \n",
       "G_521               0.148425                        0.944842   \n",
       "G_618              -0.142026                        0.922467   \n",
       "G_8                 0.134471                        0.900976   \n",
       "G_913               0.151699                        0.946122   \n",
       "\n",
       "       Rank (simulated)  Percentile (simulated)  Percent DE (simulated)  \\\n",
       "G_137             267.0               26.626627                     0.0   \n",
       "G_696             166.0               16.516517                     0.0   \n",
       "G_564              20.0                1.901902                     0.0   \n",
       "G_328              58.0                5.705706                     0.0   \n",
       "G_286             231.0               23.023023                     0.0   \n",
       "G_875             191.0               19.019019                     0.0   \n",
       "G_521             292.0               29.129129                     0.0   \n",
       "G_618             306.0               30.530531                     0.0   \n",
       "G_8               125.0               12.412412                     0.0   \n",
       "G_913             107.0               10.610611                     0.0   \n",
       "\n",
       "       Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "G_137                              0.045722                   0.030321   \n",
       "G_696                              0.060837                   0.033908   \n",
       "G_564                              0.031790                   0.025275   \n",
       "G_328                              0.039655                   0.029553   \n",
       "G_286                              0.054301                   0.032582   \n",
       "G_875                              0.053604                   0.028587   \n",
       "G_521                              0.047234                   0.036612   \n",
       "G_618                              0.056194                   0.033217   \n",
       "G_8                                0.053907                   0.031843   \n",
       "G_913                              0.052862                   0.039358   \n",
       "\n",
       "       Number of experiments (simulated)   Z score  rank  \n",
       "G_137                                 10  4.553959   1.0  \n",
       "G_696                                 10  3.929277   2.0  \n",
       "G_564                                 10  3.862334   3.0  \n",
       "G_328                                 10  3.580394   4.0  \n",
       "G_286                                 10  2.897330   5.0  \n",
       "G_875                                 10  2.786105   6.0  \n",
       "G_521                                 10  2.763869   7.0  \n",
       "G_618                                 10  2.583978   8.0  \n",
       "G_8                                   10  2.530033   9.0  \n",
       "G_913                                 10  2.511202  10.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31f20d47",
   "metadata": {},
   "source": [
    "## Traditional DE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c9fad72e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (de_method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9be7e0bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load DE statistics file\n",
    "trad_de_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "trad_de_stats = pd.read_csv(trad_de_stats_filename, sep=\"\\t\", index_col=0, header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e7df4273",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort by log fold change\n",
    "trad_de_stats_sorted = trad_de_stats.sort_values(by=\"log2FoldChange\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c4c22916",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ranking based on log2FoldChange\n",
    "trad_de_stats_sorted[\"rank\"] = trad_de_stats_sorted[\"log2FoldChange\"].rank(\n",
    "    ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b13768f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_862</th>\n",
       "      <td>2816.870147</td>\n",
       "      <td>0.173619</td>\n",
       "      <td>0.101177</td>\n",
       "      <td>1.715991</td>\n",
       "      <td>0.086164</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_187</th>\n",
       "      <td>2913.237654</td>\n",
       "      <td>0.165636</td>\n",
       "      <td>0.087628</td>\n",
       "      <td>1.890223</td>\n",
       "      <td>0.058728</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_548</th>\n",
       "      <td>2890.561620</td>\n",
       "      <td>0.162476</td>\n",
       "      <td>0.088060</td>\n",
       "      <td>1.845048</td>\n",
       "      <td>0.065031</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_938</th>\n",
       "      <td>2894.179951</td>\n",
       "      <td>0.159485</td>\n",
       "      <td>0.093224</td>\n",
       "      <td>1.710759</td>\n",
       "      <td>0.087126</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_537</th>\n",
       "      <td>2815.139770</td>\n",
       "      <td>0.155824</td>\n",
       "      <td>0.104766</td>\n",
       "      <td>1.487351</td>\n",
       "      <td>0.136922</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_200</th>\n",
       "      <td>2849.108244</td>\n",
       "      <td>0.152376</td>\n",
       "      <td>0.202281</td>\n",
       "      <td>0.753289</td>\n",
       "      <td>0.451276</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_913</th>\n",
       "      <td>2857.709181</td>\n",
       "      <td>0.151699</td>\n",
       "      <td>0.099102</td>\n",
       "      <td>1.530730</td>\n",
       "      <td>0.125836</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_322</th>\n",
       "      <td>2737.129054</td>\n",
       "      <td>0.150670</td>\n",
       "      <td>0.127496</td>\n",
       "      <td>1.181759</td>\n",
       "      <td>0.237301</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_521</th>\n",
       "      <td>2685.154759</td>\n",
       "      <td>0.148425</td>\n",
       "      <td>0.151785</td>\n",
       "      <td>0.977858</td>\n",
       "      <td>0.328145</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_807</th>\n",
       "      <td>2814.465233</td>\n",
       "      <td>0.147657</td>\n",
       "      <td>0.108098</td>\n",
       "      <td>1.365951</td>\n",
       "      <td>0.171954</td>\n",
       "      <td>0.998503</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          baseMean  log2FoldChange     lfcSE      stat    pvalue      padj  \\\n",
       "G_862  2816.870147        0.173619  0.101177  1.715991  0.086164  0.998503   \n",
       "G_187  2913.237654        0.165636  0.087628  1.890223  0.058728  0.998503   \n",
       "G_548  2890.561620        0.162476  0.088060  1.845048  0.065031  0.998503   \n",
       "G_938  2894.179951        0.159485  0.093224  1.710759  0.087126  0.998503   \n",
       "G_537  2815.139770        0.155824  0.104766  1.487351  0.136922  0.998503   \n",
       "G_200  2849.108244        0.152376  0.202281  0.753289  0.451276  0.998503   \n",
       "G_913  2857.709181        0.151699  0.099102  1.530730  0.125836  0.998503   \n",
       "G_322  2737.129054        0.150670  0.127496  1.181759  0.237301  0.998503   \n",
       "G_521  2685.154759        0.148425  0.151785  0.977858  0.328145  0.998503   \n",
       "G_807  2814.465233        0.147657  0.108098  1.365951  0.171954  0.998503   \n",
       "\n",
       "       rank  \n",
       "G_862   1.0  \n",
       "G_187   2.0  \n",
       "G_548   3.0  \n",
       "G_938   4.0  \n",
       "G_537   5.0  \n",
       "G_200   6.0  \n",
       "G_913   7.0  \n",
       "G_322   8.0  \n",
       "G_521   9.0  \n",
       "G_807  10.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcc042c5",
   "metadata": {},
   "source": [
    "## Compare\n",
    "\n",
    "1. mean rank of specific genes - mean rank of generic genes for template experiment\n",
    "2. We will need to re-run this notebook for each template experiment and then plot the distribution of difference scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3c68482d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load pickled file\n",
    "with open(template_specific_gene_ids_filename, \"rb\") as specific_fh:\n",
    "    specific_gene_ids = pickle.load(specific_fh)\n",
    "\n",
    "with open(generic_gene_ids_filename, \"rb\") as generic_fh:\n",
    "    generic_gene_ids = pickle.load(generic_fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bc3907e3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get mean of specific gene ranks\n",
    "sophie_specific_mean = summary_gene_ranks_sorted.loc[specific_gene_ids, \"rank\"].mean()\n",
    "trad_specific_mean = trad_de_stats_sorted.loc[specific_gene_ids, \"rank\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8fe89cb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "486.6\n",
      "624.6\n"
     ]
    }
   ],
   "source": [
    "print(sophie_specific_mean)\n",
    "print(trad_specific_mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3067e55e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_248     247.0\n",
       "G_584     626.0\n",
       "G_894     130.0\n",
       "G_455    1000.0\n",
       "G_462     846.0\n",
       "G_841     284.0\n",
       "G_277     275.0\n",
       "G_885     945.0\n",
       "G_590     210.0\n",
       "G_848     303.0\n",
       "Name: rank, dtype: float64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.loc[specific_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "735497d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_248    868.0\n",
       "G_584    592.0\n",
       "G_894    950.0\n",
       "G_455    566.0\n",
       "G_462    682.0\n",
       "G_841    232.0\n",
       "G_277    149.0\n",
       "G_885    607.0\n",
       "G_590    940.0\n",
       "G_848    660.0\n",
       "Name: rank, dtype: float64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.loc[specific_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "10b4c68d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get mean of generic gene ranks\n",
    "sophie_generic_mean = summary_gene_ranks_sorted.loc[generic_gene_ids, \"rank\"].mean()\n",
    "trad_generic_mean = trad_de_stats_sorted.loc[generic_gene_ids, \"rank\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "491c6519",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "498.06\n",
      "483.69\n"
     ]
    }
   ],
   "source": [
    "print(sophie_generic_mean)\n",
    "print(trad_generic_mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d50a8569",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_668    793.0\n",
       "G_229    925.0\n",
       "G_341    752.0\n",
       "G_15     294.0\n",
       "G_314    900.0\n",
       "G_925    331.0\n",
       "G_57     540.0\n",
       "G_198    264.0\n",
       "G_87      14.0\n",
       "G_47     559.0\n",
       "G_567     31.0\n",
       "G_86     335.0\n",
       "G_75     614.0\n",
       "G_246    126.0\n",
       "G_974    771.0\n",
       "G_659    315.0\n",
       "G_322    112.0\n",
       "G_495    489.0\n",
       "G_203     60.0\n",
       "G_458    324.0\n",
       "G_0      795.0\n",
       "G_192     65.0\n",
       "G_954    168.0\n",
       "G_296    334.0\n",
       "G_490    114.0\n",
       "G_62     138.0\n",
       "G_843    106.0\n",
       "G_480    441.0\n",
       "G_810    591.0\n",
       "G_872    670.0\n",
       "         ...  \n",
       "G_200     16.0\n",
       "G_861    586.0\n",
       "G_918    532.0\n",
       "G_806    884.0\n",
       "G_541    177.0\n",
       "G_646    722.0\n",
       "G_466    153.0\n",
       "G_811    281.0\n",
       "G_213    768.0\n",
       "G_821     32.0\n",
       "G_238    680.0\n",
       "G_621    630.0\n",
       "G_76     715.0\n",
       "G_180     80.0\n",
       "G_570    178.0\n",
       "G_261    914.0\n",
       "G_558    952.0\n",
       "G_574    799.0\n",
       "G_698    595.0\n",
       "G_662    851.0\n",
       "G_712    505.0\n",
       "G_794    475.0\n",
       "G_818    224.0\n",
       "G_516     38.0\n",
       "G_329    689.0\n",
       "G_728    855.0\n",
       "G_210    868.0\n",
       "G_840    813.0\n",
       "G_223    568.0\n",
       "G_823    109.0\n",
       "Name: rank, Length: 100, dtype: float64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.loc[generic_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "839f4a81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_668    606.0\n",
       "G_229    565.0\n",
       "G_341    649.0\n",
       "G_15     901.0\n",
       "G_314    435.0\n",
       "G_925    852.0\n",
       "G_57     270.0\n",
       "G_198    881.0\n",
       "G_87     995.0\n",
       "G_47     259.0\n",
       "G_567    993.0\n",
       "G_86      92.0\n",
       "G_75     604.0\n",
       "G_246    952.0\n",
       "G_974    792.0\n",
       "G_659    889.0\n",
       "G_322      8.0\n",
       "G_495    758.0\n",
       "G_203     33.0\n",
       "G_458    237.0\n",
       "G_0      468.0\n",
       "G_192     43.0\n",
       "G_954    105.0\n",
       "G_296    305.0\n",
       "G_490     61.0\n",
       "G_62      14.0\n",
       "G_843    922.0\n",
       "G_480    775.0\n",
       "G_810    665.0\n",
       "G_872    274.0\n",
       "         ...  \n",
       "G_200      6.0\n",
       "G_861    224.0\n",
       "G_918    208.0\n",
       "G_806    362.0\n",
       "G_541     76.0\n",
       "G_646    368.0\n",
       "G_466    898.0\n",
       "G_811    779.0\n",
       "G_213    371.0\n",
       "G_821     58.0\n",
       "G_238    560.0\n",
       "G_621    342.0\n",
       "G_76     316.0\n",
       "G_180     73.0\n",
       "G_570    236.0\n",
       "G_261    633.0\n",
       "G_558    537.0\n",
       "G_574    357.0\n",
       "G_698    180.0\n",
       "G_662    937.0\n",
       "G_712    117.0\n",
       "G_794    827.0\n",
       "G_818    896.0\n",
       "G_516     17.0\n",
       "G_329    553.0\n",
       "G_728    746.0\n",
       "G_210    514.0\n",
       "G_840    458.0\n",
       "G_223    701.0\n",
       "G_823     86.0\n",
       "Name: rank, Length: 100, dtype: float64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.loc[generic_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8c08e343",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sophie difference:  -11.45999999999998\n",
      "traditional difference:  140.91000000000003\n"
     ]
    }
   ],
   "source": [
    "# Difference\n",
    "diff_sophie = sophie_specific_mean - sophie_generic_mean\n",
    "diff_trad = trad_specific_mean - trad_generic_mean\n",
    "\n",
    "print(\"sophie difference: \", diff_sophie)\n",
    "print(\"traditional difference: \", diff_trad)"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
