{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0b54cfe2",
   "metadata": {},
   "source": [
    "# Evaluation\n",
    "\n",
    "How often do regular differential expression analysis vs sophie prioritize the specific vs generic genes?\n",
    "\n",
    "1. Simulate 1 template perturbation experiment using the technique above\n",
    "2. Apply SOPHIE to get ranking of specific and generic genes based on their z-score.\n",
    "3. Apply traditional DE analysis and get ranking of specific and generic genes based on their log fold change value\n",
    "4. Compare the difference in ranking between specific and generic genes using SOPHIE vs traditional metrics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b68f083",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:21: The name tf.ConfigProto is deprecated. Please use tf.compat.v1.ConfigProto instead.\n",
      "\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:25: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.\n",
      "\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:25: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext rpy2.ipython\n",
    "import os\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from rpy2.robjects import pandas2ri\n",
    "from ponyo import utils, train_vae_modules, simulate_expression_data\n",
    "from generic_expression_patterns_modules import (\n",
    "    process,\n",
    "    new_experiment_process,  # REMOVE\n",
    "    stats,\n",
    "    ranking,\n",
    ")\n",
    "\n",
    "np.random.seed(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4fa09de1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = \"config_sophie_vs_trad.tsv\"\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6e9a766",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load config params\n",
    "\n",
    "# Local directory to store intermediate files\n",
    "local_dir = params[\"local_dir\"]\n",
    "\n",
    "#\n",
    "dataset_name = params[\"dataset_name\"]\n",
    "\n",
    "# File containing un-normalized template experiment\n",
    "raw_template_filename = params[\"raw_template_filename\"]\n",
    "\n",
    "# Un-normalized compendium filename\n",
    "raw_compendium_filename = params[\"raw_compendium_filename\"]\n",
    "\n",
    "# Normalized compendium filename\n",
    "normalized_compendium_filename = params[\"normalized_compendium_filename\"]\n",
    "\n",
    "# ID for template experiment to be selected\n",
    "project_id = params[\"project_id\"]\n",
    "\n",
    "# Number of simulated experiments to generate\n",
    "num_runs = params[\"num_simulated\"]\n",
    "\n",
    "# Directory containing trained VAE model\n",
    "vae_model_dir = params[\"vae_model_dir\"]\n",
    "\n",
    "# Size of the latent dimension\n",
    "latent_dim = params[\"latent_dim\"]\n",
    "\n",
    "# Scaler transform used to scale compendium data into 0-1 range for training\n",
    "scaler_filename = params[\"scaler_filename\"]\n",
    "\n",
    "# Which DE method to use\n",
    "# We recommend that if data is RNA-seq then use DESeq2\n",
    "# If data is microarray then use Limma\n",
    "de_method = params[\"DE_method\"]\n",
    "\n",
    "# If using DE-seq, setting this parameter will\n",
    "# remove genes below a certain threshold\n",
    "count_threshold = params[\"count_threshold\"]\n",
    "\n",
    "# Metadata file that specifies which samples to keep for DE analysis (Optional)\n",
    "template_process_samples_filename = params[\"template_process_samples_filename\"]\n",
    "\n",
    "# Metadata file that specifies sample grouping for DE analysis\n",
    "template_DE_grouping_filename = params[\"template_DE_grouping_filename\"]\n",
    "\n",
    "# Statistic to use to rank genes or pathways by\n",
    "# Choices are {} FILL IN\n",
    "col_to_rank_genes = params[\"rank_genes_by\"]\n",
    "\n",
    "# Pickle files saving specific and generic gene ids\n",
    "template_specific_gene_ids_filename = params[\"template_specific_gene_ids_filename\"]\n",
    "generic_gene_ids_filename = \"generic_gene_ids.pickle\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5fccae1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files generated by this notebook\n",
    "\n",
    "# File storing template experiment with gene ids mapped to compendium gene ids\n",
    "mapped_template_filename = params[\"mapped_template_filename\"]\n",
    "\n",
    "# File storing normalized template experiment\n",
    "normalized_template_filename = params[\"normalized_template_filename\"]\n",
    "\n",
    "# File storing processed template experiment,\n",
    "# after samples have been selected for comparison in DE analysis\n",
    "processed_template_filename = params[\"processed_template_filename\"]\n",
    "\n",
    "# Output summary file\n",
    "output_filename = params[\"output_filename\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f10a5605",
   "metadata": {},
   "source": [
    "## SOPHIE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "69f51a6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8, 1000)\n",
      "(720, 1000)\n"
     ]
    }
   ],
   "source": [
    "# Process template\n",
    "new_experiment_process.process_template_experiment(\n",
    "    raw_template_filename,\n",
    "    raw_compendium_filename,\n",
    "    scaler_filename,\n",
    "    mapped_template_filename,\n",
    "    normalized_template_filename,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0e48cdb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate multiple experiments UPDATE COMMENT\n",
    "# This step creates the following files in \"<local_dir>/pseudo_experiment/\" directory:\n",
    "#   - selected_simulated_data_SRP012656_<n>.txt\n",
    "#   - selected_simulated_encoded_data_SRP012656_<n>.txt\n",
    "#   - template_normalized_data_SRP012656_test.txt\n",
    "# in which \"<n>\" is an integer in the range of [0, num_runs-1]\n",
    "\n",
    "# REMOVE LATER\n",
    "# dataset_name = \"pre_model_unseen_template\"\n",
    "# Load pickled file\n",
    "scaler = pickle.load(open(scaler_filename, \"rb\"))\n",
    "\n",
    "# Update simulated dir\n",
    "os.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)\n",
    "\n",
    "# Update to take in file to be consisten\n",
    "normalized_compendium = pd.read_csv(\n",
    "    normalized_compendium_filename, header=0, sep=\"\\t\", index_col=0\n",
    ")\n",
    "normalized_template = pd.read_csv(\n",
    "    normalized_template_filename, header=0, sep=\"\\t\", index_col=0\n",
    ")\n",
    "# ------------\n",
    "# Update call when new version of ponyo\n",
    "for run_id in range(num_runs):\n",
    "    new_experiment_process.embed_shift_template_experiment(\n",
    "        normalized_compendium,\n",
    "        normalized_template,\n",
    "        vae_model_dir,\n",
    "        project_id,\n",
    "        scaler_filename,\n",
    "        local_dir,\n",
    "        latent_dim,\n",
    "        run_id,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "09bf6c28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "## Update simulated dir\n",
    "if not os.path.exists(template_process_samples_filename):\n",
    "    template_process_samples_filename = None\n",
    "\n",
    "if de_method == \"deseq\":\n",
    "    # Process template data\n",
    "    stats.process_samples_for_DESeq(\n",
    "        raw_template_filename,\n",
    "        template_DE_grouping_filename,\n",
    "        processed_template_filename,\n",
    "        count_threshold,\n",
    "        template_process_samples_filename,\n",
    "    )\n",
    "\n",
    "    # Process simulated data\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        out_simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}_processed.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_DESeq(\n",
    "            simulated_filename,\n",
    "            template_DE_grouping_filename,\n",
    "            out_simulated_filename,\n",
    "            count_threshold,\n",
    "            template_process_samples_filename,\n",
    "        )\n",
    "else:\n",
    "    stats.process_samples_for_limma(\n",
    "        raw_template_filename,\n",
    "        template_DE_grouping_filename,\n",
    "        processed_template_filename,\n",
    "        template_process_samples_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_limma(\n",
    "            simulated_filename,\n",
    "            template_DE_grouping_filename,\n",
    "            None,\n",
    "            template_process_samples_filename,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c56b91df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ee7b0585",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: DelayedArray\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: BiocParallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (de_method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0dee119f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i base_dir -i local_dir -i num_runs -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# Files created: \"<local_dir>/DE_stats/DE_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\",\n",
    "        i,\n",
    "        \"_processed.txt\",\n",
    "        sep = \"\"\n",
    "    )\n",
    "    if (de_method == \"deseq\"){\n",
    "        get_DE_stats_DESeq(\n",
    "            template_DE_grouping_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "    }\n",
    "    else {\n",
    "        get_DE_stats_limma(\n",
    "            template_DE_grouping_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2c0e853b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (2 levels on the left, 1 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "# Added\n",
    "if de_method == \"deseq\":\n",
    "    logFC_name = \"log2FoldChange\"\n",
    "    pvalue_name = \"padj\"\n",
    "else:\n",
    "    logFC_name = \"logFC\"\n",
    "    pvalue_name = \"adj.P.Val\"\n",
    "\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    "    logFC_name,\n",
    "    pvalue_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a9ea6ba3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_937</th>\n",
       "      <td>G_937</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>988.0</td>\n",
       "      <td>5.570529</td>\n",
       "      <td>5.570529</td>\n",
       "      <td>0.712998</td>\n",
       "      <td>316.0</td>\n",
       "      <td>31.531532</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.147347</td>\n",
       "      <td>0.289673</td>\n",
       "      <td>10</td>\n",
       "      <td>15.269569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_336</th>\n",
       "      <td>G_336</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>986.0</td>\n",
       "      <td>5.084454</td>\n",
       "      <td>5.084454</td>\n",
       "      <td>0.857904</td>\n",
       "      <td>174.0</td>\n",
       "      <td>17.317317</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.867031</td>\n",
       "      <td>0.355073</td>\n",
       "      <td>10</td>\n",
       "      <td>11.877627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_243</th>\n",
       "      <td>G_243</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>990.5</td>\n",
       "      <td>6.244824</td>\n",
       "      <td>6.244824</td>\n",
       "      <td>0.707890</td>\n",
       "      <td>694.0</td>\n",
       "      <td>69.369369</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.469970</td>\n",
       "      <td>0.409972</td>\n",
       "      <td>10</td>\n",
       "      <td>11.646779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_815</th>\n",
       "      <td>G_815</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>987.0</td>\n",
       "      <td>5.523477</td>\n",
       "      <td>5.523477</td>\n",
       "      <td>0.660331</td>\n",
       "      <td>201.0</td>\n",
       "      <td>20.020020</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.030100</td>\n",
       "      <td>0.423226</td>\n",
       "      <td>10</td>\n",
       "      <td>10.616967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_881</th>\n",
       "      <td>G_881</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>982.0</td>\n",
       "      <td>4.479740</td>\n",
       "      <td>4.479740</td>\n",
       "      <td>0.835015</td>\n",
       "      <td>567.0</td>\n",
       "      <td>56.656657</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.275472</td>\n",
       "      <td>0.318806</td>\n",
       "      <td>10</td>\n",
       "      <td>10.050831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_644</th>\n",
       "      <td>G_644</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>980.0</td>\n",
       "      <td>3.941796</td>\n",
       "      <td>3.941796</td>\n",
       "      <td>0.710424</td>\n",
       "      <td>504.0</td>\n",
       "      <td>50.350350</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.241366</td>\n",
       "      <td>0.283321</td>\n",
       "      <td>10</td>\n",
       "      <td>9.531341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_391</th>\n",
       "      <td>G_391</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>992.0</td>\n",
       "      <td>6.268838</td>\n",
       "      <td>6.268838</td>\n",
       "      <td>0.707890</td>\n",
       "      <td>971.0</td>\n",
       "      <td>97.097097</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.304053</td>\n",
       "      <td>0.416849</td>\n",
       "      <td>10</td>\n",
       "      <td>9.511322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_488</th>\n",
       "      <td>G_488</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>990.5</td>\n",
       "      <td>6.244824</td>\n",
       "      <td>6.244824</td>\n",
       "      <td>0.707890</td>\n",
       "      <td>732.0</td>\n",
       "      <td>73.173173</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.645245</td>\n",
       "      <td>0.637893</td>\n",
       "      <td>10</td>\n",
       "      <td>7.210577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_236</th>\n",
       "      <td>G_236</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>989.0</td>\n",
       "      <td>5.902202</td>\n",
       "      <td>5.902202</td>\n",
       "      <td>0.782481</td>\n",
       "      <td>870.0</td>\n",
       "      <td>86.986987</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.544682</td>\n",
       "      <td>0.619812</td>\n",
       "      <td>10</td>\n",
       "      <td>7.030395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_645</th>\n",
       "      <td>G_645</td>\n",
       "      <td>0.000417</td>\n",
       "      <td>983.0</td>\n",
       "      <td>4.826379</td>\n",
       "      <td>4.826379</td>\n",
       "      <td>0.782481</td>\n",
       "      <td>948.0</td>\n",
       "      <td>94.794795</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.010571</td>\n",
       "      <td>0.413573</td>\n",
       "      <td>10</td>\n",
       "      <td>6.808484</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "G_937   G_937            0.000110        988.0                    5.570529   \n",
       "G_336   G_336            0.000189        986.0                    5.084454   \n",
       "G_243   G_243            0.000007        990.5                    6.244824   \n",
       "G_815   G_815            0.000119        987.0                    5.523477   \n",
       "G_881   G_881            0.000162        982.0                    4.479740   \n",
       "G_644   G_644            0.000163        980.0                    3.941796   \n",
       "G_391   G_391            0.000007        992.0                    6.268838   \n",
       "G_488   G_488            0.000007        990.5                    6.244824   \n",
       "G_236   G_236            0.000020        989.0                    5.902202   \n",
       "G_645   G_645            0.000417        983.0                    4.826379   \n",
       "\n",
       "       log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "G_937               5.570529                        0.712998   \n",
       "G_336               5.084454                        0.857904   \n",
       "G_243               6.244824                        0.707890   \n",
       "G_815               5.523477                        0.660331   \n",
       "G_881               4.479740                        0.835015   \n",
       "G_644               3.941796                        0.710424   \n",
       "G_391               6.268838                        0.707890   \n",
       "G_488               6.244824                        0.707890   \n",
       "G_236               5.902202                        0.782481   \n",
       "G_645               4.826379                        0.782481   \n",
       "\n",
       "       Rank (simulated)  Percentile (simulated)  Percent DE (simulated)  \\\n",
       "G_937             316.0               31.531532                     0.0   \n",
       "G_336             174.0               17.317317                     0.0   \n",
       "G_243             694.0               69.369369                     0.0   \n",
       "G_815             201.0               20.020020                     0.0   \n",
       "G_881             567.0               56.656657                     0.0   \n",
       "G_644             504.0               50.350350                     0.0   \n",
       "G_391             971.0               97.097097                     0.0   \n",
       "G_488             732.0               73.173173                     0.0   \n",
       "G_236             870.0               86.986987                     0.0   \n",
       "G_645             948.0               94.794795                     0.0   \n",
       "\n",
       "       Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "G_937                              1.147347                   0.289673   \n",
       "G_336                              0.867031                   0.355073   \n",
       "G_243                              1.469970                   0.409972   \n",
       "G_815                              1.030100                   0.423226   \n",
       "G_881                              1.275472                   0.318806   \n",
       "G_644                              1.241366                   0.283321   \n",
       "G_391                              2.304053                   0.416849   \n",
       "G_488                              1.645245                   0.637893   \n",
       "G_236                              1.544682                   0.619812   \n",
       "G_645                              2.010571                   0.413573   \n",
       "\n",
       "       Number of experiments (simulated)    Z score  \n",
       "G_937                                 10  15.269569  \n",
       "G_336                                 10  11.877627  \n",
       "G_243                                 10  11.646779  \n",
       "G_815                                 10  10.616967  \n",
       "G_881                                 10  10.050831  \n",
       "G_644                                 10   9.531341  \n",
       "G_391                                 10   9.511322  \n",
       "G_488                                 10   7.210577  \n",
       "G_236                                 10   7.030395  \n",
       "G_645                                 10   6.808484  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get summary table\n",
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    \"gene\",\n",
    "    params,\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eb3fcde8",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_gene_ranks_sorted = summary_gene_ranks.sort_values(\n",
    "    by=\"Z score\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a69d2399",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ranking based on Z-score\n",
    "summary_gene_ranks_sorted[\"rank\"] = summary_gene_ranks_sorted[\"Z score\"].rank(\n",
    "    ascending=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7e668e37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_937</th>\n",
       "      <td>G_937</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>988.0</td>\n",
       "      <td>5.570529</td>\n",
       "      <td>5.570529</td>\n",
       "      <td>0.712998</td>\n",
       "      <td>316.0</td>\n",
       "      <td>31.531532</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.147347</td>\n",
       "      <td>0.289673</td>\n",
       "      <td>10</td>\n",
       "      <td>15.269569</td>\n",
       "      <td>992.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_336</th>\n",
       "      <td>G_336</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>986.0</td>\n",
       "      <td>5.084454</td>\n",
       "      <td>5.084454</td>\n",
       "      <td>0.857904</td>\n",
       "      <td>174.0</td>\n",
       "      <td>17.317317</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.867031</td>\n",
       "      <td>0.355073</td>\n",
       "      <td>10</td>\n",
       "      <td>11.877627</td>\n",
       "      <td>991.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_243</th>\n",
       "      <td>G_243</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>990.5</td>\n",
       "      <td>6.244824</td>\n",
       "      <td>6.244824</td>\n",
       "      <td>0.707890</td>\n",
       "      <td>694.0</td>\n",
       "      <td>69.369369</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.469970</td>\n",
       "      <td>0.409972</td>\n",
       "      <td>10</td>\n",
       "      <td>11.646779</td>\n",
       "      <td>990.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_815</th>\n",
       "      <td>G_815</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>987.0</td>\n",
       "      <td>5.523477</td>\n",
       "      <td>5.523477</td>\n",
       "      <td>0.660331</td>\n",
       "      <td>201.0</td>\n",
       "      <td>20.020020</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.030100</td>\n",
       "      <td>0.423226</td>\n",
       "      <td>10</td>\n",
       "      <td>10.616967</td>\n",
       "      <td>989.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_881</th>\n",
       "      <td>G_881</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>982.0</td>\n",
       "      <td>4.479740</td>\n",
       "      <td>4.479740</td>\n",
       "      <td>0.835015</td>\n",
       "      <td>567.0</td>\n",
       "      <td>56.656657</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.275472</td>\n",
       "      <td>0.318806</td>\n",
       "      <td>10</td>\n",
       "      <td>10.050831</td>\n",
       "      <td>988.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_644</th>\n",
       "      <td>G_644</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>980.0</td>\n",
       "      <td>3.941796</td>\n",
       "      <td>3.941796</td>\n",
       "      <td>0.710424</td>\n",
       "      <td>504.0</td>\n",
       "      <td>50.350350</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.241366</td>\n",
       "      <td>0.283321</td>\n",
       "      <td>10</td>\n",
       "      <td>9.531341</td>\n",
       "      <td>987.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_391</th>\n",
       "      <td>G_391</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>992.0</td>\n",
       "      <td>6.268838</td>\n",
       "      <td>6.268838</td>\n",
       "      <td>0.707890</td>\n",
       "      <td>971.0</td>\n",
       "      <td>97.097097</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.304053</td>\n",
       "      <td>0.416849</td>\n",
       "      <td>10</td>\n",
       "      <td>9.511322</td>\n",
       "      <td>986.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_488</th>\n",
       "      <td>G_488</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>990.5</td>\n",
       "      <td>6.244824</td>\n",
       "      <td>6.244824</td>\n",
       "      <td>0.707890</td>\n",
       "      <td>732.0</td>\n",
       "      <td>73.173173</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.645245</td>\n",
       "      <td>0.637893</td>\n",
       "      <td>10</td>\n",
       "      <td>7.210577</td>\n",
       "      <td>985.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_236</th>\n",
       "      <td>G_236</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>989.0</td>\n",
       "      <td>5.902202</td>\n",
       "      <td>5.902202</td>\n",
       "      <td>0.782481</td>\n",
       "      <td>870.0</td>\n",
       "      <td>86.986987</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.544682</td>\n",
       "      <td>0.619812</td>\n",
       "      <td>10</td>\n",
       "      <td>7.030395</td>\n",
       "      <td>984.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_645</th>\n",
       "      <td>G_645</td>\n",
       "      <td>0.000417</td>\n",
       "      <td>983.0</td>\n",
       "      <td>4.826379</td>\n",
       "      <td>4.826379</td>\n",
       "      <td>0.782481</td>\n",
       "      <td>948.0</td>\n",
       "      <td>94.794795</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.010571</td>\n",
       "      <td>0.413573</td>\n",
       "      <td>10</td>\n",
       "      <td>6.808484</td>\n",
       "      <td>983.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "G_937   G_937            0.000110        988.0                    5.570529   \n",
       "G_336   G_336            0.000189        986.0                    5.084454   \n",
       "G_243   G_243            0.000007        990.5                    6.244824   \n",
       "G_815   G_815            0.000119        987.0                    5.523477   \n",
       "G_881   G_881            0.000162        982.0                    4.479740   \n",
       "G_644   G_644            0.000163        980.0                    3.941796   \n",
       "G_391   G_391            0.000007        992.0                    6.268838   \n",
       "G_488   G_488            0.000007        990.5                    6.244824   \n",
       "G_236   G_236            0.000020        989.0                    5.902202   \n",
       "G_645   G_645            0.000417        983.0                    4.826379   \n",
       "\n",
       "       log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "G_937               5.570529                        0.712998   \n",
       "G_336               5.084454                        0.857904   \n",
       "G_243               6.244824                        0.707890   \n",
       "G_815               5.523477                        0.660331   \n",
       "G_881               4.479740                        0.835015   \n",
       "G_644               3.941796                        0.710424   \n",
       "G_391               6.268838                        0.707890   \n",
       "G_488               6.244824                        0.707890   \n",
       "G_236               5.902202                        0.782481   \n",
       "G_645               4.826379                        0.782481   \n",
       "\n",
       "       Rank (simulated)  Percentile (simulated)  Percent DE (simulated)  \\\n",
       "G_937             316.0               31.531532                     0.0   \n",
       "G_336             174.0               17.317317                     0.0   \n",
       "G_243             694.0               69.369369                     0.0   \n",
       "G_815             201.0               20.020020                     0.0   \n",
       "G_881             567.0               56.656657                     0.0   \n",
       "G_644             504.0               50.350350                     0.0   \n",
       "G_391             971.0               97.097097                     0.0   \n",
       "G_488             732.0               73.173173                     0.0   \n",
       "G_236             870.0               86.986987                     0.0   \n",
       "G_645             948.0               94.794795                     0.0   \n",
       "\n",
       "       Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "G_937                              1.147347                   0.289673   \n",
       "G_336                              0.867031                   0.355073   \n",
       "G_243                              1.469970                   0.409972   \n",
       "G_815                              1.030100                   0.423226   \n",
       "G_881                              1.275472                   0.318806   \n",
       "G_644                              1.241366                   0.283321   \n",
       "G_391                              2.304053                   0.416849   \n",
       "G_488                              1.645245                   0.637893   \n",
       "G_236                              1.544682                   0.619812   \n",
       "G_645                              2.010571                   0.413573   \n",
       "\n",
       "       Number of experiments (simulated)    Z score   rank  \n",
       "G_937                                 10  15.269569  992.0  \n",
       "G_336                                 10  11.877627  991.0  \n",
       "G_243                                 10  11.646779  990.0  \n",
       "G_815                                 10  10.616967  989.0  \n",
       "G_881                                 10  10.050831  988.0  \n",
       "G_644                                 10   9.531341  987.0  \n",
       "G_391                                 10   9.511322  986.0  \n",
       "G_488                                 10   7.210577  985.0  \n",
       "G_236                                 10   7.030395  984.0  \n",
       "G_645                                 10   6.808484  983.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31f20d47",
   "metadata": {},
   "source": [
    "## Traditional DE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c9fad72e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (de_method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9be7e0bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load DE statistics file\n",
    "trad_de_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "trad_de_stats = pd.read_csv(trad_de_stats_filename, sep=\"\\t\", index_col=0, header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e7df4273",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort by log fold change\n",
    "trad_de_stats_sorted = trad_de_stats.sort_values(by=\"log2FoldChange\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c4c22916",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ranking based on log2FoldChange\n",
    "trad_de_stats_sorted[\"rank\"] = trad_de_stats_sorted[\"log2FoldChange\"].rank(\n",
    "    ascending=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b13768f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_391</th>\n",
       "      <td>7.125108</td>\n",
       "      <td>6.268838</td>\n",
       "      <td>1.115622</td>\n",
       "      <td>5.619143</td>\n",
       "      <td>1.919067e-08</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>992.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_488</th>\n",
       "      <td>7.007174</td>\n",
       "      <td>6.244824</td>\n",
       "      <td>1.116568</td>\n",
       "      <td>5.592876</td>\n",
       "      <td>2.233383e-08</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>990.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_243</th>\n",
       "      <td>7.007174</td>\n",
       "      <td>6.244824</td>\n",
       "      <td>1.116568</td>\n",
       "      <td>5.592876</td>\n",
       "      <td>2.233383e-08</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>990.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_236</th>\n",
       "      <td>9.241785</td>\n",
       "      <td>5.902202</td>\n",
       "      <td>1.100117</td>\n",
       "      <td>5.365066</td>\n",
       "      <td>8.091953e-08</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>989.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_937</th>\n",
       "      <td>7.362413</td>\n",
       "      <td>5.570529</td>\n",
       "      <td>1.112559</td>\n",
       "      <td>5.006952</td>\n",
       "      <td>5.529870e-07</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>988.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_815</th>\n",
       "      <td>7.137844</td>\n",
       "      <td>5.523477</td>\n",
       "      <td>1.114443</td>\n",
       "      <td>4.956265</td>\n",
       "      <td>7.186114e-07</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>987.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_336</th>\n",
       "      <td>8.766909</td>\n",
       "      <td>5.084454</td>\n",
       "      <td>1.097677</td>\n",
       "      <td>4.632012</td>\n",
       "      <td>3.621293e-06</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>986.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_649</th>\n",
       "      <td>7.616165</td>\n",
       "      <td>4.872302</td>\n",
       "      <td>1.105285</td>\n",
       "      <td>4.408185</td>\n",
       "      <td>1.042404e-05</td>\n",
       "      <td>0.000383</td>\n",
       "      <td>985.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_383</th>\n",
       "      <td>7.489930</td>\n",
       "      <td>4.850577</td>\n",
       "      <td>1.106579</td>\n",
       "      <td>4.383400</td>\n",
       "      <td>1.168413e-05</td>\n",
       "      <td>0.000386</td>\n",
       "      <td>984.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_645</th>\n",
       "      <td>7.373417</td>\n",
       "      <td>4.826379</td>\n",
       "      <td>1.107106</td>\n",
       "      <td>4.359456</td>\n",
       "      <td>1.303862e-05</td>\n",
       "      <td>0.000417</td>\n",
       "      <td>983.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       baseMean  log2FoldChange     lfcSE      stat        pvalue      padj  \\\n",
       "G_391  7.125108        6.268838  1.115622  5.619143  1.919067e-08  0.000007   \n",
       "G_488  7.007174        6.244824  1.116568  5.592876  2.233383e-08  0.000007   \n",
       "G_243  7.007174        6.244824  1.116568  5.592876  2.233383e-08  0.000007   \n",
       "G_236  9.241785        5.902202  1.100117  5.365066  8.091953e-08  0.000020   \n",
       "G_937  7.362413        5.570529  1.112559  5.006952  5.529870e-07  0.000110   \n",
       "G_815  7.137844        5.523477  1.114443  4.956265  7.186114e-07  0.000119   \n",
       "G_336  8.766909        5.084454  1.097677  4.632012  3.621293e-06  0.000189   \n",
       "G_649  7.616165        4.872302  1.105285  4.408185  1.042404e-05  0.000383   \n",
       "G_383  7.489930        4.850577  1.106579  4.383400  1.168413e-05  0.000386   \n",
       "G_645  7.373417        4.826379  1.107106  4.359456  1.303862e-05  0.000417   \n",
       "\n",
       "        rank  \n",
       "G_391  992.0  \n",
       "G_488  990.5  \n",
       "G_243  990.5  \n",
       "G_236  989.0  \n",
       "G_937  988.0  \n",
       "G_815  987.0  \n",
       "G_336  986.0  \n",
       "G_649  985.0  \n",
       "G_383  984.0  \n",
       "G_645  983.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcc042c5",
   "metadata": {},
   "source": [
    "## Compare\n",
    "\n",
    "1. mean rank of specific genes - mean rank of generic genes for template experiment\n",
    "2. We will need to re-run this notebook for each template experiment and then plot the distribution of difference scores\n",
    "\n",
    "We want to compare the mean ranking of specific genes vs the mean ranking of generic genes. If the mean difference is large then yes, that would indicate that there is a difference between the specific and generic genes that we can detect. In addition to the difference we want the specific genes to be higher ranked compared to the generic ones, so we want to see a large positive value if the method is performing better."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3c68482d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load pickled file\n",
    "with open(template_specific_gene_ids_filename, \"rb\") as specific_fh:\n",
    "    specific_gene_ids = pickle.load(specific_fh)\n",
    "\n",
    "with open(generic_gene_ids_filename, \"rb\") as generic_fh:\n",
    "    generic_gene_ids = pickle.load(generic_fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bc3907e3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get mean of specific gene ranks\n",
    "sophie_specific_mean = summary_gene_ranks_sorted.loc[specific_gene_ids, \"rank\"].mean()\n",
    "trad_specific_mean = trad_de_stats_sorted.loc[specific_gene_ids, \"rank\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8fe89cb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "842.1\n",
      "893.9\n"
     ]
    }
   ],
   "source": [
    "print(sophie_specific_mean)\n",
    "print(trad_specific_mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3067e55e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_487    967.0\n",
       "G_453    177.0\n",
       "G_236    984.0\n",
       "G_833    914.0\n",
       "G_6      927.0\n",
       "G_755    903.0\n",
       "G_936    981.0\n",
       "G_914    915.0\n",
       "G_582    695.0\n",
       "G_186    958.0\n",
       "Name: rank, dtype: float64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.loc[specific_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "735497d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_487    972.0\n",
       "G_453    531.0\n",
       "G_236    989.0\n",
       "G_833    949.0\n",
       "G_6      930.0\n",
       "G_755    929.0\n",
       "G_936    964.0\n",
       "G_914    941.0\n",
       "G_582    795.0\n",
       "G_186    939.0\n",
       "Name: rank, dtype: float64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.loc[specific_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "10b4c68d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get mean of generic gene ranks\n",
    "sophie_generic_mean = summary_gene_ranks_sorted.loc[generic_gene_ids, \"rank\"].mean()\n",
    "trad_generic_mean = trad_de_stats_sorted.loc[generic_gene_ids, \"rank\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "491c6519",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "655.05\n",
      "839.85\n"
     ]
    }
   ],
   "source": [
    "print(sophie_generic_mean)\n",
    "print(trad_generic_mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d50a8569",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_716    895.0\n",
       "G_146    960.0\n",
       "G_351    800.0\n",
       "G_886     27.0\n",
       "G_318     55.0\n",
       "G_336    991.0\n",
       "G_389    757.0\n",
       "G_472    922.0\n",
       "G_371    970.0\n",
       "G_315    844.0\n",
       "G_288    187.0\n",
       "G_113     26.0\n",
       "G_109     34.0\n",
       "G_858    972.0\n",
       "G_542    787.0\n",
       "G_70     943.0\n",
       "G_903    976.0\n",
       "G_645    983.0\n",
       "G_649    980.0\n",
       "G_20     901.0\n",
       "G_313    332.0\n",
       "G_918    929.0\n",
       "G_270    968.0\n",
       "G_671    969.0\n",
       "G_489    912.0\n",
       "G_259    932.0\n",
       "G_338    161.0\n",
       "G_269    160.0\n",
       "G_200    838.0\n",
       "G_209    532.0\n",
       "         ...  \n",
       "G_878     91.0\n",
       "G_127    535.0\n",
       "G_937    992.0\n",
       "G_214    595.0\n",
       "G_357     50.0\n",
       "G_419    872.0\n",
       "G_34     722.0\n",
       "G_108    657.0\n",
       "G_383    982.0\n",
       "G_780    211.0\n",
       "G_228    973.0\n",
       "G_815    989.0\n",
       "G_901    444.0\n",
       "G_506    979.0\n",
       "G_102    674.0\n",
       "G_905    894.0\n",
       "G_992     36.0\n",
       "G_890    971.0\n",
       "G_666    687.0\n",
       "G_792     12.0\n",
       "G_813     33.0\n",
       "G_860    952.0\n",
       "G_317    849.0\n",
       "G_539    199.0\n",
       "G_644    987.0\n",
       "G_996    563.0\n",
       "G_414    966.0\n",
       "G_942    951.0\n",
       "G_965    881.0\n",
       "G_359    168.0\n",
       "Name: rank, Length: 100, dtype: float64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.loc[generic_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "839f4a81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_716    947.0\n",
       "G_146    950.0\n",
       "G_351    876.0\n",
       "G_886    903.0\n",
       "G_318    634.0\n",
       "G_336    986.0\n",
       "G_389    890.0\n",
       "G_472    960.0\n",
       "G_371    952.0\n",
       "G_315    920.0\n",
       "G_288    452.0\n",
       "G_113    340.0\n",
       "G_109    747.0\n",
       "G_858    975.0\n",
       "G_542    878.0\n",
       "G_70     954.0\n",
       "G_903    974.0\n",
       "G_645    983.0\n",
       "G_649    985.0\n",
       "G_20     961.0\n",
       "G_313    410.0\n",
       "G_918    940.0\n",
       "G_270    969.0\n",
       "G_671    979.0\n",
       "G_489    948.0\n",
       "G_259    946.0\n",
       "G_338    843.0\n",
       "G_269    702.0\n",
       "G_200    902.0\n",
       "G_209    866.0\n",
       "         ...  \n",
       "G_878    678.0\n",
       "G_127    806.0\n",
       "G_937    988.0\n",
       "G_214    778.0\n",
       "G_357    861.0\n",
       "G_419    935.0\n",
       "G_34     918.0\n",
       "G_108    816.0\n",
       "G_383    984.0\n",
       "G_780    277.0\n",
       "G_228    978.0\n",
       "G_815    987.0\n",
       "G_901    647.0\n",
       "G_506    981.0\n",
       "G_102    936.0\n",
       "G_905    910.0\n",
       "G_992    641.0\n",
       "G_890    976.0\n",
       "G_666    895.0\n",
       "G_792    767.0\n",
       "G_813    326.0\n",
       "G_860    968.0\n",
       "G_317    922.0\n",
       "G_539    363.0\n",
       "G_644    980.0\n",
       "G_996    856.0\n",
       "G_414    958.0\n",
       "G_942    963.0\n",
       "G_965    932.0\n",
       "G_359    280.0\n",
       "Name: rank, Length: 100, dtype: float64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.loc[generic_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8c08e343",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sophie difference:  187.05000000000007\n",
      "traditional difference:  54.049999999999955\n"
     ]
    }
   ],
   "source": [
    "# Difference\n",
    "diff_sophie = sophie_specific_mean - sophie_generic_mean\n",
    "diff_trad = trad_specific_mean - trad_generic_mean\n",
    "\n",
    "print(\"sophie difference: \", diff_sophie)\n",
    "print(\"traditional difference: \", diff_trad)"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
