{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0b54cfe2",
   "metadata": {},
   "source": [
    "# Evaluation\n",
    "\n",
    "How often do regular differential expression analysis vs sophie prioritize the specific vs generic genes?\n",
    "\n",
    "1. Simulate 1 template perturbation experiment using the technique above\n",
    "2. Apply SOPHIE to get ranking of specific and generic genes based on their z-score.\n",
    "3. Apply traditional DE analysis and get ranking of specific and generic genes based on their log fold change value\n",
    "4. Compare the difference in ranking between specific and generic genes using SOPHIE vs traditional metrics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b68f083",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: \n",
      "examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.\n",
      "  \"found relative to the 'datapath' directory.\".format(key))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:21: The name tf.ConfigProto is deprecated. Please use tf.compat.v1.ConfigProto instead.\n",
      "\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:25: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.\n",
      "\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/ponyo/helper_vae.py:25: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext rpy2.ipython\n",
    "import os\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from rpy2.robjects import pandas2ri\n",
    "from ponyo import utils, train_vae_modules, simulate_expression_data\n",
    "from generic_expression_patterns_modules import (\n",
    "    process,\n",
    "    new_experiment_process,  # REMOVE\n",
    "    stats,\n",
    "    ranking,\n",
    ")\n",
    "\n",
    "np.random.seed(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4fa09de1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in config variables\n",
    "base_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "\n",
    "config_filename = \"config_sophie_vs_trad.tsv\"\n",
    "\n",
    "params = utils.read_config(config_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6e9a766",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load config params\n",
    "\n",
    "# Local directory to store intermediate files\n",
    "local_dir = params[\"local_dir\"]\n",
    "\n",
    "#\n",
    "dataset_name = params[\"dataset_name\"]\n",
    "\n",
    "# File containing un-normalized template experiment\n",
    "raw_template_filename = params[\"raw_template_filename\"]\n",
    "\n",
    "# Un-normalized compendium filename\n",
    "raw_compendium_filename = params[\"raw_compendium_filename\"]\n",
    "\n",
    "# Normalized compendium filename\n",
    "normalized_compendium_filename = params[\"normalized_compendium_filename\"]\n",
    "\n",
    "# ID for template experiment to be selected\n",
    "project_id = params[\"project_id\"]\n",
    "\n",
    "# Number of simulated experiments to generate\n",
    "num_runs = params[\"num_simulated\"]\n",
    "\n",
    "# Directory containing trained VAE model\n",
    "vae_model_dir = params[\"vae_model_dir\"]\n",
    "\n",
    "# Size of the latent dimension\n",
    "latent_dim = params[\"latent_dim\"]\n",
    "\n",
    "# Scaler transform used to scale compendium data into 0-1 range for training\n",
    "scaler_filename = params[\"scaler_filename\"]\n",
    "\n",
    "# Which DE method to use\n",
    "# We recommend that if data is RNA-seq then use DESeq2\n",
    "# If data is microarray then use Limma\n",
    "de_method = params[\"DE_method\"]\n",
    "\n",
    "# If using DE-seq, setting this parameter will\n",
    "# remove genes below a certain threshold\n",
    "count_threshold = params[\"count_threshold\"]\n",
    "\n",
    "# Metadata file that specifies which samples to keep for DE analysis (Optional)\n",
    "template_process_samples_filename = params[\"template_process_samples_filename\"]\n",
    "\n",
    "# Metadata file that specifies sample grouping for DE analysis\n",
    "template_DE_grouping_filename = params[\"template_DE_grouping_filename\"]\n",
    "\n",
    "# Statistic to use to rank genes or pathways by\n",
    "# Choices are {} FILL IN\n",
    "col_to_rank_genes = params[\"rank_genes_by\"]\n",
    "\n",
    "# Pickle files saving specific and generic gene ids\n",
    "template_specific_gene_ids_filename = params[\"template_specific_gene_ids_filename\"]\n",
    "generic_gene_ids_filename = \"generic_gene_ids.pickle\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5fccae1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files generated by this notebook\n",
    "\n",
    "# File storing template experiment with gene ids mapped to compendium gene ids\n",
    "mapped_template_filename = params[\"mapped_template_filename\"]\n",
    "\n",
    "# File storing normalized template experiment\n",
    "normalized_template_filename = params[\"normalized_template_filename\"]\n",
    "\n",
    "# File storing processed template experiment,\n",
    "# after samples have been selected for comparison in DE analysis\n",
    "processed_template_filename = params[\"processed_template_filename\"]\n",
    "\n",
    "# Output summary file\n",
    "output_filename = params[\"output_filename\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f10a5605",
   "metadata": {},
   "source": [
    "## SOPHIE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "69f51a6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4, 1000)\n",
      "(360, 1000)\n"
     ]
    }
   ],
   "source": [
    "# Process template\n",
    "new_experiment_process.process_template_experiment(\n",
    "    raw_template_filename,\n",
    "    raw_compendium_filename,\n",
    "    scaler_filename,\n",
    "    mapped_template_filename,\n",
    "    normalized_template_filename,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0e48cdb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:From /home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Simulate multiple experiments UPDATE COMMENT\n",
    "# This step creates the following files in \"<local_dir>/pseudo_experiment/\" directory:\n",
    "#   - selected_simulated_data_SRP012656_<n>.txt\n",
    "#   - selected_simulated_encoded_data_SRP012656_<n>.txt\n",
    "#   - template_normalized_data_SRP012656_test.txt\n",
    "# in which \"<n>\" is an integer in the range of [0, num_runs-1]\n",
    "\n",
    "# REMOVE LATER\n",
    "# dataset_name = \"pre_model_unseen_template\"\n",
    "# Load pickled file\n",
    "scaler = pickle.load(open(scaler_filename, \"rb\"))\n",
    "\n",
    "# Update simulated dir\n",
    "os.makedirs(os.path.join(local_dir, \"pseudo_experiment\"), exist_ok=True)\n",
    "\n",
    "# Update to take in file to be consisten\n",
    "normalized_compendium = pd.read_csv(\n",
    "    normalized_compendium_filename, header=0, sep=\"\\t\", index_col=0\n",
    ")\n",
    "normalized_template = pd.read_csv(\n",
    "    normalized_template_filename, header=0, sep=\"\\t\", index_col=0\n",
    ")\n",
    "# ------------\n",
    "# Update call when new version of ponyo\n",
    "for run_id in range(num_runs):\n",
    "    new_experiment_process.embed_shift_template_experiment(\n",
    "        normalized_compendium,\n",
    "        normalized_template,\n",
    "        vae_model_dir,\n",
    "        project_id,\n",
    "        scaler_filename,\n",
    "        local_dir,\n",
    "        latent_dim,\n",
    "        run_id,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "09bf6c28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n",
      "sample ids are ordered correctly\n"
     ]
    }
   ],
   "source": [
    "## Update simulated dir\n",
    "if not os.path.exists(template_process_samples_filename):\n",
    "    template_process_samples_filename = None\n",
    "\n",
    "if de_method == \"deseq\":\n",
    "    # Process template data\n",
    "    stats.process_samples_for_DESeq(\n",
    "        raw_template_filename,\n",
    "        template_DE_grouping_filename,\n",
    "        processed_template_filename,\n",
    "        count_threshold,\n",
    "        template_process_samples_filename,\n",
    "    )\n",
    "\n",
    "    # Process simulated data\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        out_simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}_processed.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_DESeq(\n",
    "            simulated_filename,\n",
    "            template_DE_grouping_filename,\n",
    "            out_simulated_filename,\n",
    "            count_threshold,\n",
    "            template_process_samples_filename,\n",
    "        )\n",
    "else:\n",
    "    stats.process_samples_for_limma(\n",
    "        raw_template_filename,\n",
    "        template_DE_grouping_filename,\n",
    "        processed_template_filename,\n",
    "        template_process_samples_filename,\n",
    "    )\n",
    "\n",
    "    for i in range(num_runs):\n",
    "        simulated_filename = os.path.join(\n",
    "            local_dir,\n",
    "            \"pseudo_experiment\",\n",
    "            f\"selected_simulated_data_{project_id}_{i}.txt\",\n",
    "        )\n",
    "        stats.process_samples_for_limma(\n",
    "            simulated_filename,\n",
    "            template_DE_grouping_filename,\n",
    "            None,\n",
    "            template_process_samples_filename,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c56b91df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subdirectory: \"<local_dir>/DE_stats/\"\n",
    "os.makedirs(os.path.join(local_dir, \"DE_stats\"), exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ee7b0585",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: DelayedArray\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: BiocParallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (de_method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0dee119f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: There were 12 warnings (use warnings() to see them)\n",
      "R[write to console]: \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Error in estimateDispersionsFit(object, fitType = fitType, quiet = quiet) : \n",
      "  all gene-wise dispersion estimates are within 2 orders of magnitude\n",
      "  from the minimum value, and so the standard curve fitting techniques will not work.\n",
      "  One can instead use the gene-wise estimates as final estimates:\n",
      "  dds <- estimateDispersionsGeneEst(dds)\n",
      "  dispersions(dds) <- mcols(dds)$dispGeneEst\n",
      "  ...then continue with testing using nbinomWaldTest or nbinomLRT\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i base_dir -i local_dir -i num_runs -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# Files created: \"<local_dir>/DE_stats/DE_stats_simulated_data_<project_id>_<n>.txt\"\n",
    "for (i in 0:(num_runs-1)){\n",
    "    simulated_data_filename <- paste(\n",
    "        local_dir,\n",
    "        \"pseudo_experiment/selected_simulated_data_\",\n",
    "        project_id,\n",
    "        \"_\",\n",
    "        i,\n",
    "        \"_processed.txt\",\n",
    "        sep = \"\"\n",
    "    )\n",
    "    if (de_method == \"deseq\"){\n",
    "        get_DE_stats_DESeq(\n",
    "            template_DE_grouping_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "    }\n",
    "    else {\n",
    "        get_DE_stats_limma(\n",
    "            template_DE_grouping_filename,\n",
    "            project_id,\n",
    "            simulated_data_filename,\n",
    "            \"simulated\",\n",
    "            local_dir,\n",
    "            i\n",
    "            )\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2c0e853b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alexandra/anaconda3/envs/generic_expression/lib/python3.7/site-packages/pandas/core/reshape/merge.py:544: UserWarning: merging between different levels can give an unintended result (2 levels on the left, 1 on the right)\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "analysis_type = \"DE\"\n",
    "template_DE_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "# Added\n",
    "if de_method == \"deseq\":\n",
    "    logFC_name = \"log2FoldChange\"\n",
    "    pvalue_name = \"padj\"\n",
    "else:\n",
    "    logFC_name = \"logFC\"\n",
    "    pvalue_name = \"adj.P.Val\"\n",
    "\n",
    "template_DE_stats, simulated_DE_summary_stats = ranking.process_and_rank_genes_pathways(\n",
    "    template_DE_stats_filename,\n",
    "    local_dir,\n",
    "    num_runs,\n",
    "    project_id,\n",
    "    analysis_type,\n",
    "    col_to_rank_genes,\n",
    "    logFC_name,\n",
    "    pvalue_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a9ea6ba3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_94</th>\n",
       "      <td>G_94</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>972.0</td>\n",
       "      <td>0.242436</td>\n",
       "      <td>0.242436</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>46.0</td>\n",
       "      <td>4.504505</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008432</td>\n",
       "      <td>0.005742</td>\n",
       "      <td>10</td>\n",
       "      <td>40.750666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_677</th>\n",
       "      <td>G_677</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>980.0</td>\n",
       "      <td>0.261352</td>\n",
       "      <td>0.261352</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.300300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007409</td>\n",
       "      <td>0.008002</td>\n",
       "      <td>10</td>\n",
       "      <td>31.733043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_452</th>\n",
       "      <td>G_452</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>765.0</td>\n",
       "      <td>0.123959</td>\n",
       "      <td>0.123959</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>61.0</td>\n",
       "      <td>6.006006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007558</td>\n",
       "      <td>0.003923</td>\n",
       "      <td>10</td>\n",
       "      <td>29.673908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_481</th>\n",
       "      <td>G_481</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>962.0</td>\n",
       "      <td>0.226160</td>\n",
       "      <td>0.226160</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>342.0</td>\n",
       "      <td>34.134134</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014111</td>\n",
       "      <td>0.007596</td>\n",
       "      <td>10</td>\n",
       "      <td>27.914393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_299</th>\n",
       "      <td>G_299</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>953.0</td>\n",
       "      <td>0.218870</td>\n",
       "      <td>0.218870</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>388.0</td>\n",
       "      <td>38.738739</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017358</td>\n",
       "      <td>0.007378</td>\n",
       "      <td>10</td>\n",
       "      <td>27.311680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_780</th>\n",
       "      <td>G_780</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>971.0</td>\n",
       "      <td>0.239619</td>\n",
       "      <td>0.239619</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>263.0</td>\n",
       "      <td>26.226226</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015758</td>\n",
       "      <td>0.008373</td>\n",
       "      <td>10</td>\n",
       "      <td>26.734516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_27</th>\n",
       "      <td>G_27</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>998.0</td>\n",
       "      <td>0.394053</td>\n",
       "      <td>0.394053</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>827.0</td>\n",
       "      <td>82.682683</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.023328</td>\n",
       "      <td>0.014050</td>\n",
       "      <td>10</td>\n",
       "      <td>26.386273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_674</th>\n",
       "      <td>G_674</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>990.0</td>\n",
       "      <td>0.293388</td>\n",
       "      <td>-0.293388</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>537.0</td>\n",
       "      <td>53.653654</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019042</td>\n",
       "      <td>0.010638</td>\n",
       "      <td>10</td>\n",
       "      <td>25.788839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_461</th>\n",
       "      <td>G_461</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>986.0</td>\n",
       "      <td>0.276100</td>\n",
       "      <td>-0.276100</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>59.0</td>\n",
       "      <td>5.805806</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011628</td>\n",
       "      <td>0.010431</td>\n",
       "      <td>10</td>\n",
       "      <td>25.355455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_986</th>\n",
       "      <td>G_986</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>976.0</td>\n",
       "      <td>0.251112</td>\n",
       "      <td>-0.251112</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>88.0</td>\n",
       "      <td>8.708709</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010601</td>\n",
       "      <td>0.009685</td>\n",
       "      <td>10</td>\n",
       "      <td>24.833270</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "G_94     G_94            0.999304        972.0                    0.242436   \n",
       "G_677   G_677            0.999304        980.0                    0.261352   \n",
       "G_452   G_452            0.999304        765.0                    0.123959   \n",
       "G_481   G_481            0.999304        962.0                    0.226160   \n",
       "G_299   G_299            0.999304        953.0                    0.218870   \n",
       "G_780   G_780            0.999304        971.0                    0.239619   \n",
       "G_27     G_27            0.999304        998.0                    0.394053   \n",
       "G_674   G_674            0.999304        990.0                    0.293388   \n",
       "G_461   G_461            0.999304        986.0                    0.276100   \n",
       "G_986   G_986            0.999304        976.0                    0.251112   \n",
       "\n",
       "       log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "G_94                0.242436                        0.999405   \n",
       "G_677               0.261352                        0.999405   \n",
       "G_452               0.123959                        0.999405   \n",
       "G_481               0.226160                        0.999405   \n",
       "G_299               0.218870                        0.999405   \n",
       "G_780               0.239619                        0.999405   \n",
       "G_27                0.394053                        0.999405   \n",
       "G_674              -0.293388                        0.999405   \n",
       "G_461              -0.276100                        0.999405   \n",
       "G_986              -0.251112                        0.999405   \n",
       "\n",
       "       Rank (simulated)  Percentile (simulated)  Percent DE (simulated)  \\\n",
       "G_94               46.0                4.504505                     0.0   \n",
       "G_677               4.0                0.300300                     0.0   \n",
       "G_452              61.0                6.006006                     0.0   \n",
       "G_481             342.0               34.134134                     0.0   \n",
       "G_299             388.0               38.738739                     0.0   \n",
       "G_780             263.0               26.226226                     0.0   \n",
       "G_27              827.0               82.682683                     0.0   \n",
       "G_674             537.0               53.653654                     0.0   \n",
       "G_461              59.0                5.805806                     0.0   \n",
       "G_986              88.0                8.708709                     0.0   \n",
       "\n",
       "       Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "G_94                               0.008432                   0.005742   \n",
       "G_677                              0.007409                   0.008002   \n",
       "G_452                              0.007558                   0.003923   \n",
       "G_481                              0.014111                   0.007596   \n",
       "G_299                              0.017358                   0.007378   \n",
       "G_780                              0.015758                   0.008373   \n",
       "G_27                               0.023328                   0.014050   \n",
       "G_674                              0.019042                   0.010638   \n",
       "G_461                              0.011628                   0.010431   \n",
       "G_986                              0.010601                   0.009685   \n",
       "\n",
       "       Number of experiments (simulated)    Z score  \n",
       "G_94                                  10  40.750666  \n",
       "G_677                                 10  31.733043  \n",
       "G_452                                 10  29.673908  \n",
       "G_481                                 10  27.914393  \n",
       "G_299                                 10  27.311680  \n",
       "G_780                                 10  26.734516  \n",
       "G_27                                  10  26.386273  \n",
       "G_674                                 10  25.788839  \n",
       "G_461                                 10  25.355455  \n",
       "G_986                                 10  24.833270  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get summary table\n",
    "summary_gene_ranks = ranking.generate_summary_table(\n",
    "    template_DE_stats_filename,\n",
    "    template_DE_stats,\n",
    "    simulated_DE_summary_stats,\n",
    "    col_to_rank_genes,\n",
    "    local_dir,\n",
    "    \"gene\",\n",
    "    params,\n",
    ")\n",
    "\n",
    "summary_gene_ranks.sort_values(by=\"Z score\", ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eb3fcde8",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_gene_ranks_sorted = summary_gene_ranks.sort_values(\n",
    "    by=\"Z score\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a69d2399",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ranking based on Z-score\n",
    "summary_gene_ranks_sorted[\"rank\"] = summary_gene_ranks_sorted[\"Z score\"].rank(\n",
    "    ascending=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7e668e37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Adj P-value (Real)</th>\n",
       "      <th>Rank (Real)</th>\n",
       "      <th>abs(log2FoldChange) (Real)</th>\n",
       "      <th>log2FoldChange (Real)</th>\n",
       "      <th>Median adj p-value (simulated)</th>\n",
       "      <th>Rank (simulated)</th>\n",
       "      <th>Percentile (simulated)</th>\n",
       "      <th>Percent DE (simulated)</th>\n",
       "      <th>Mean abs(log2FoldChange) (simulated)</th>\n",
       "      <th>Std deviation (simulated)</th>\n",
       "      <th>Number of experiments (simulated)</th>\n",
       "      <th>Z score</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_94</th>\n",
       "      <td>G_94</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>972.0</td>\n",
       "      <td>0.242436</td>\n",
       "      <td>0.242436</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>46.0</td>\n",
       "      <td>4.504505</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008432</td>\n",
       "      <td>0.005742</td>\n",
       "      <td>10</td>\n",
       "      <td>40.750666</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_677</th>\n",
       "      <td>G_677</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>980.0</td>\n",
       "      <td>0.261352</td>\n",
       "      <td>0.261352</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.300300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007409</td>\n",
       "      <td>0.008002</td>\n",
       "      <td>10</td>\n",
       "      <td>31.733043</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_452</th>\n",
       "      <td>G_452</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>765.0</td>\n",
       "      <td>0.123959</td>\n",
       "      <td>0.123959</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>61.0</td>\n",
       "      <td>6.006006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007558</td>\n",
       "      <td>0.003923</td>\n",
       "      <td>10</td>\n",
       "      <td>29.673908</td>\n",
       "      <td>998.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_481</th>\n",
       "      <td>G_481</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>962.0</td>\n",
       "      <td>0.226160</td>\n",
       "      <td>0.226160</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>342.0</td>\n",
       "      <td>34.134134</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014111</td>\n",
       "      <td>0.007596</td>\n",
       "      <td>10</td>\n",
       "      <td>27.914393</td>\n",
       "      <td>997.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_299</th>\n",
       "      <td>G_299</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>953.0</td>\n",
       "      <td>0.218870</td>\n",
       "      <td>0.218870</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>388.0</td>\n",
       "      <td>38.738739</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017358</td>\n",
       "      <td>0.007378</td>\n",
       "      <td>10</td>\n",
       "      <td>27.311680</td>\n",
       "      <td>996.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_780</th>\n",
       "      <td>G_780</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>971.0</td>\n",
       "      <td>0.239619</td>\n",
       "      <td>0.239619</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>263.0</td>\n",
       "      <td>26.226226</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015758</td>\n",
       "      <td>0.008373</td>\n",
       "      <td>10</td>\n",
       "      <td>26.734516</td>\n",
       "      <td>995.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_27</th>\n",
       "      <td>G_27</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>998.0</td>\n",
       "      <td>0.394053</td>\n",
       "      <td>0.394053</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>827.0</td>\n",
       "      <td>82.682683</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.023328</td>\n",
       "      <td>0.014050</td>\n",
       "      <td>10</td>\n",
       "      <td>26.386273</td>\n",
       "      <td>994.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_674</th>\n",
       "      <td>G_674</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>990.0</td>\n",
       "      <td>0.293388</td>\n",
       "      <td>-0.293388</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>537.0</td>\n",
       "      <td>53.653654</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019042</td>\n",
       "      <td>0.010638</td>\n",
       "      <td>10</td>\n",
       "      <td>25.788839</td>\n",
       "      <td>993.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_461</th>\n",
       "      <td>G_461</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>986.0</td>\n",
       "      <td>0.276100</td>\n",
       "      <td>-0.276100</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>59.0</td>\n",
       "      <td>5.805806</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011628</td>\n",
       "      <td>0.010431</td>\n",
       "      <td>10</td>\n",
       "      <td>25.355455</td>\n",
       "      <td>992.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_986</th>\n",
       "      <td>G_986</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>976.0</td>\n",
       "      <td>0.251112</td>\n",
       "      <td>-0.251112</td>\n",
       "      <td>0.999405</td>\n",
       "      <td>88.0</td>\n",
       "      <td>8.708709</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010601</td>\n",
       "      <td>0.009685</td>\n",
       "      <td>10</td>\n",
       "      <td>24.833270</td>\n",
       "      <td>991.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene ID  Adj P-value (Real)  Rank (Real)  abs(log2FoldChange) (Real)  \\\n",
       "G_94     G_94            0.999304        972.0                    0.242436   \n",
       "G_677   G_677            0.999304        980.0                    0.261352   \n",
       "G_452   G_452            0.999304        765.0                    0.123959   \n",
       "G_481   G_481            0.999304        962.0                    0.226160   \n",
       "G_299   G_299            0.999304        953.0                    0.218870   \n",
       "G_780   G_780            0.999304        971.0                    0.239619   \n",
       "G_27     G_27            0.999304        998.0                    0.394053   \n",
       "G_674   G_674            0.999304        990.0                    0.293388   \n",
       "G_461   G_461            0.999304        986.0                    0.276100   \n",
       "G_986   G_986            0.999304        976.0                    0.251112   \n",
       "\n",
       "       log2FoldChange (Real)  Median adj p-value (simulated)  \\\n",
       "G_94                0.242436                        0.999405   \n",
       "G_677               0.261352                        0.999405   \n",
       "G_452               0.123959                        0.999405   \n",
       "G_481               0.226160                        0.999405   \n",
       "G_299               0.218870                        0.999405   \n",
       "G_780               0.239619                        0.999405   \n",
       "G_27                0.394053                        0.999405   \n",
       "G_674              -0.293388                        0.999405   \n",
       "G_461              -0.276100                        0.999405   \n",
       "G_986              -0.251112                        0.999405   \n",
       "\n",
       "       Rank (simulated)  Percentile (simulated)  Percent DE (simulated)  \\\n",
       "G_94               46.0                4.504505                     0.0   \n",
       "G_677               4.0                0.300300                     0.0   \n",
       "G_452              61.0                6.006006                     0.0   \n",
       "G_481             342.0               34.134134                     0.0   \n",
       "G_299             388.0               38.738739                     0.0   \n",
       "G_780             263.0               26.226226                     0.0   \n",
       "G_27              827.0               82.682683                     0.0   \n",
       "G_674             537.0               53.653654                     0.0   \n",
       "G_461              59.0                5.805806                     0.0   \n",
       "G_986              88.0                8.708709                     0.0   \n",
       "\n",
       "       Mean abs(log2FoldChange) (simulated)  Std deviation (simulated)  \\\n",
       "G_94                               0.008432                   0.005742   \n",
       "G_677                              0.007409                   0.008002   \n",
       "G_452                              0.007558                   0.003923   \n",
       "G_481                              0.014111                   0.007596   \n",
       "G_299                              0.017358                   0.007378   \n",
       "G_780                              0.015758                   0.008373   \n",
       "G_27                               0.023328                   0.014050   \n",
       "G_674                              0.019042                   0.010638   \n",
       "G_461                              0.011628                   0.010431   \n",
       "G_986                              0.010601                   0.009685   \n",
       "\n",
       "       Number of experiments (simulated)    Z score    rank  \n",
       "G_94                                  10  40.750666  1000.0  \n",
       "G_677                                 10  31.733043   999.0  \n",
       "G_452                                 10  29.673908   998.0  \n",
       "G_481                                 10  27.914393   997.0  \n",
       "G_299                                 10  27.311680   996.0  \n",
       "G_780                                 10  26.734516   995.0  \n",
       "G_27                                  10  26.386273   994.0  \n",
       "G_674                                 10  25.788839   993.0  \n",
       "G_461                                 10  25.355455   992.0  \n",
       "G_986                                 10  24.833270   991.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31f20d47",
   "metadata": {},
   "source": [
    "## Traditional DE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c9fad72e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Checking sample ordering...\"\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:   the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i template_DE_grouping_filename -i project_id -i processed_template_filename -i local_dir -i base_dir -i de_method\n",
    "\n",
    "source(paste0(base_dir, '/generic_expression_patterns_modules/DE_analysis.R'))\n",
    "\n",
    "# File created: \"<local_dir>/DE_stats/DE_stats_template_data_<project_id>_real.txt\"\n",
    "if (de_method == \"deseq\"){\n",
    "    get_DE_stats_DESeq(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}\n",
    "else{\n",
    "    get_DE_stats_limma(\n",
    "        template_DE_grouping_filename,\n",
    "        project_id,\n",
    "        processed_template_filename,\n",
    "        \"template\",\n",
    "        local_dir,\n",
    "        \"real\"\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9be7e0bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load DE statistics file\n",
    "trad_de_stats_filename = os.path.join(\n",
    "    local_dir, \"DE_stats\", f\"DE_stats_template_data_{project_id}_real.txt\"\n",
    ")\n",
    "\n",
    "trad_de_stats = pd.read_csv(trad_de_stats_filename, sep=\"\\t\", index_col=0, header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e7df4273",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort by log fold change\n",
    "trad_de_stats_sorted = trad_de_stats.sort_values(by=\"log2FoldChange\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c4c22916",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ranking based on log2FoldChange\n",
    "trad_de_stats_sorted[\"rank\"] = trad_de_stats_sorted[\"log2FoldChange\"].rank(\n",
    "    ascending=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b13768f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G_409</th>\n",
       "      <td>247.896269</td>\n",
       "      <td>0.426834</td>\n",
       "      <td>0.341703</td>\n",
       "      <td>1.249135</td>\n",
       "      <td>0.211616</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_455</th>\n",
       "      <td>256.032597</td>\n",
       "      <td>0.422385</td>\n",
       "      <td>0.497494</td>\n",
       "      <td>0.849026</td>\n",
       "      <td>0.395867</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_27</th>\n",
       "      <td>244.078791</td>\n",
       "      <td>0.394053</td>\n",
       "      <td>0.408895</td>\n",
       "      <td>0.963701</td>\n",
       "      <td>0.335196</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>998.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_225</th>\n",
       "      <td>254.563331</td>\n",
       "      <td>0.375690</td>\n",
       "      <td>0.511622</td>\n",
       "      <td>0.734312</td>\n",
       "      <td>0.462758</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>997.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_775</th>\n",
       "      <td>247.954304</td>\n",
       "      <td>0.348207</td>\n",
       "      <td>0.322448</td>\n",
       "      <td>1.079887</td>\n",
       "      <td>0.280193</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>996.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_732</th>\n",
       "      <td>268.965494</td>\n",
       "      <td>0.307362</td>\n",
       "      <td>0.204365</td>\n",
       "      <td>1.503986</td>\n",
       "      <td>0.132585</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>995.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_526</th>\n",
       "      <td>270.449262</td>\n",
       "      <td>0.294261</td>\n",
       "      <td>0.205304</td>\n",
       "      <td>1.433293</td>\n",
       "      <td>0.151774</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>994.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_423</th>\n",
       "      <td>246.780815</td>\n",
       "      <td>0.285675</td>\n",
       "      <td>0.349865</td>\n",
       "      <td>0.816530</td>\n",
       "      <td>0.414197</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>993.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_243</th>\n",
       "      <td>254.108202</td>\n",
       "      <td>0.283231</td>\n",
       "      <td>0.282700</td>\n",
       "      <td>1.001875</td>\n",
       "      <td>0.316404</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>992.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_786</th>\n",
       "      <td>271.956177</td>\n",
       "      <td>0.279859</td>\n",
       "      <td>0.455698</td>\n",
       "      <td>0.614133</td>\n",
       "      <td>0.539127</td>\n",
       "      <td>0.999304</td>\n",
       "      <td>991.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         baseMean  log2FoldChange     lfcSE      stat    pvalue      padj  \\\n",
       "G_409  247.896269        0.426834  0.341703  1.249135  0.211616  0.999304   \n",
       "G_455  256.032597        0.422385  0.497494  0.849026  0.395867  0.999304   \n",
       "G_27   244.078791        0.394053  0.408895  0.963701  0.335196  0.999304   \n",
       "G_225  254.563331        0.375690  0.511622  0.734312  0.462758  0.999304   \n",
       "G_775  247.954304        0.348207  0.322448  1.079887  0.280193  0.999304   \n",
       "G_732  268.965494        0.307362  0.204365  1.503986  0.132585  0.999304   \n",
       "G_526  270.449262        0.294261  0.205304  1.433293  0.151774  0.999304   \n",
       "G_423  246.780815        0.285675  0.349865  0.816530  0.414197  0.999304   \n",
       "G_243  254.108202        0.283231  0.282700  1.001875  0.316404  0.999304   \n",
       "G_786  271.956177        0.279859  0.455698  0.614133  0.539127  0.999304   \n",
       "\n",
       "         rank  \n",
       "G_409  1000.0  \n",
       "G_455   999.0  \n",
       "G_27    998.0  \n",
       "G_225   997.0  \n",
       "G_775   996.0  \n",
       "G_732   995.0  \n",
       "G_526   994.0  \n",
       "G_423   993.0  \n",
       "G_243   992.0  \n",
       "G_786   991.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcc042c5",
   "metadata": {},
   "source": [
    "## Compare\n",
    "\n",
    "1. mean rank of specific genes - mean rank of generic genes for template experiment\n",
    "2. We will need to re-run this notebook for each template experiment and then plot the distribution of difference scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3c68482d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load pickled file\n",
    "with open(template_specific_gene_ids_filename, \"rb\") as specific_fh:\n",
    "    specific_gene_ids = pickle.load(specific_fh)\n",
    "\n",
    "with open(generic_gene_ids_filename, \"rb\") as generic_fh:\n",
    "    generic_gene_ids = pickle.load(generic_fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bc3907e3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get mean of specific gene ranks\n",
    "sophie_specific_mean = summary_gene_ranks_sorted.loc[specific_gene_ids, \"rank\"].mean()\n",
    "trad_specific_mean = trad_de_stats_sorted.loc[specific_gene_ids, \"rank\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8fe89cb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "514.0\n",
      "543.5\n"
     ]
    }
   ],
   "source": [
    "print(sophie_specific_mean)\n",
    "print(trad_specific_mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3067e55e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_77     892.0\n",
       "G_483    776.0\n",
       "G_455    927.0\n",
       "G_263    117.0\n",
       "G_549    261.0\n",
       "G_901    675.0\n",
       "G_751    384.0\n",
       "G_911    280.0\n",
       "G_381    372.0\n",
       "G_689    456.0\n",
       "Name: rank, dtype: float64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.loc[specific_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "735497d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_77      70.0\n",
       "G_483    905.0\n",
       "G_455    999.0\n",
       "G_263    553.0\n",
       "G_549    355.0\n",
       "G_901    202.0\n",
       "G_751    697.0\n",
       "G_911    610.0\n",
       "G_381    336.0\n",
       "G_689    708.0\n",
       "Name: rank, dtype: float64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.loc[specific_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "10b4c68d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get mean of generic gene ranks\n",
    "sophie_generic_mean = summary_gene_ranks_sorted.loc[generic_gene_ids, \"rank\"].mean()\n",
    "trad_generic_mean = trad_de_stats_sorted.loc[generic_gene_ids, \"rank\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "491c6519",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "466.43\n",
      "509.89\n"
     ]
    }
   ],
   "source": [
    "print(sophie_generic_mean)\n",
    "print(trad_generic_mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d50a8569",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_668    952.0\n",
       "G_229    928.0\n",
       "G_341    469.0\n",
       "G_15     119.0\n",
       "G_314    606.0\n",
       "G_925     48.0\n",
       "G_57     674.0\n",
       "G_198    815.0\n",
       "G_87     970.0\n",
       "G_47     653.0\n",
       "G_567    737.0\n",
       "G_86     514.0\n",
       "G_75     982.0\n",
       "G_246     67.0\n",
       "G_974    250.0\n",
       "G_659    596.0\n",
       "G_322     87.0\n",
       "G_495     34.0\n",
       "G_203    487.0\n",
       "G_458    344.0\n",
       "G_0      124.0\n",
       "G_192    411.0\n",
       "G_954    458.0\n",
       "G_296    353.0\n",
       "G_490    406.0\n",
       "G_62     316.0\n",
       "G_843    789.0\n",
       "G_480    870.0\n",
       "G_810    265.0\n",
       "G_872    203.0\n",
       "         ...  \n",
       "G_200    489.0\n",
       "G_861    235.0\n",
       "G_918    122.0\n",
       "G_806    253.0\n",
       "G_541    678.0\n",
       "G_646    491.0\n",
       "G_466    578.0\n",
       "G_811    972.0\n",
       "G_213    447.0\n",
       "G_821     27.0\n",
       "G_238    697.0\n",
       "G_621    597.0\n",
       "G_76     695.0\n",
       "G_180    317.0\n",
       "G_570    692.0\n",
       "G_261    284.0\n",
       "G_558    479.0\n",
       "G_574    365.0\n",
       "G_698    595.0\n",
       "G_662    733.0\n",
       "G_712    292.0\n",
       "G_794    435.0\n",
       "G_818    562.0\n",
       "G_516    884.0\n",
       "G_329    502.0\n",
       "G_728    817.0\n",
       "G_210    318.0\n",
       "G_840    193.0\n",
       "G_223    552.0\n",
       "G_823    925.0\n",
       "Name: rank, Length: 100, dtype: float64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_gene_ranks_sorted.loc[generic_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "839f4a81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "G_668     44.0\n",
       "G_229    841.0\n",
       "G_341    719.0\n",
       "G_15     443.0\n",
       "G_314    887.0\n",
       "G_925    502.0\n",
       "G_57     145.0\n",
       "G_198    874.0\n",
       "G_87     979.0\n",
       "G_47     785.0\n",
       "G_567    925.0\n",
       "G_86     302.0\n",
       "G_75     985.0\n",
       "G_246    470.0\n",
       "G_974    401.0\n",
       "G_659    188.0\n",
       "G_322    554.0\n",
       "G_495    512.0\n",
       "G_203    712.0\n",
       "G_458    748.0\n",
       "G_0      656.0\n",
       "G_192    670.0\n",
       "G_954    804.0\n",
       "G_296    263.0\n",
       "G_490    179.0\n",
       "G_62     737.0\n",
       "G_843    913.0\n",
       "G_480     62.0\n",
       "G_810    630.0\n",
       "G_872    340.0\n",
       "         ...  \n",
       "G_200    270.0\n",
       "G_861    345.0\n",
       "G_918    406.0\n",
       "G_806    344.0\n",
       "G_541    826.0\n",
       "G_646     94.0\n",
       "G_466    131.0\n",
       "G_811    975.0\n",
       "G_213    726.0\n",
       "G_821    516.0\n",
       "G_238    878.0\n",
       "G_621    694.0\n",
       "G_76     919.0\n",
       "G_180    372.0\n",
       "G_570    149.0\n",
       "G_261    284.0\n",
       "G_558    206.0\n",
       "G_574    713.0\n",
       "G_698    102.0\n",
       "G_662     56.0\n",
       "G_712    310.0\n",
       "G_794    747.0\n",
       "G_818    226.0\n",
       "G_516     14.0\n",
       "G_329    609.0\n",
       "G_728    888.0\n",
       "G_210    332.0\n",
       "G_840    381.0\n",
       "G_223    820.0\n",
       "G_823     35.0\n",
       "Name: rank, Length: 100, dtype: float64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trad_de_stats_sorted.loc[generic_gene_ids, \"rank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8c08e343",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sophie difference:  47.56999999999999\n",
      "traditional difference:  33.610000000000014\n"
     ]
    }
   ],
   "source": [
    "# Difference\n",
    "diff_sophie = sophie_specific_mean - sophie_generic_mean\n",
    "diff_trad = trad_specific_mean - trad_generic_mean\n",
    "\n",
    "print(\"sophie difference: \", diff_sophie)\n",
    "print(\"traditional difference: \", diff_trad)"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:generic_expression] *",
   "language": "python",
   "name": "conda-env-generic_expression-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
